{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# KGB Genome Browser v0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "ename": "SyntaxError",
     "evalue": "from __future__ imports must occur at the beginning of the file (<string>, line 79)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<string>\"\u001b[0;36m, line \u001b[0;32m79\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m from __future__ imports must occur at the beginning of the file\n"
     ]
    }
   ],
   "source": [
    "###############################################################################\n",
    "# KGB user input vars  (Preferably implement as separate upstream cell)\n",
    "###############################################################################        \n",
    "\n",
    "\n",
    "################################################################################################\n",
    "# Config #1: local notebook with NCBI annotated genbank format genomes and gene tree           #\n",
    "#            Domain Annotations: NO (KBase domain annotations require KBase genome annotation) #\n",
    "#            Gene Tree: YES                                                                    #\n",
    "#            Search Terms: YES                                                                 #\n",
    "################################################################################################\n",
    "                                                                                             \n",
    "%pylab notebook\n",
    "KBase_backend = False\n",
    "\n",
    "GenomeSet_names = [\"Gsulf\", \"DvulH\", \"DdesulfG20\", \"EcoliK12\", \"Bsub\", \"DaudaxMP104C\"]\n",
    "ContigSet_names = []\n",
    "PivotFeatures_IDs = [\"GSU2863\", \"DVU2928\", \"Dde_2997\", \"b3987\", \"BSU01070\", \"Daud0216\"]\n",
    "PrimaryAnchor_leafId = \"Gsulf rpoB\"\n",
    "PrimaryAnchor_locusTag = \"GSU2863\"\n",
    "\n",
    "genome_annotation_system = 'NCBI'\n",
    "genome_data_format = \"Genbank\"\n",
    "# if relative path to scaffolds is e.g.: ./data_example/NCBI_genome_contigs/Gsulf/scaffolds/scaf_1.gbk\n",
    "gbk_ext = \"gbk\"\n",
    "genome_data_base_path = \"./data_example/NCBI_genome_contigs\"\n",
    "genome_data_extra_subpath = \"/scaffolds\"\n",
    "\n",
    "domain_data_exists = False  # NCBI annotated genome\n",
    "domain_data_format = None\n",
    "domain_data_base_path = None\n",
    "domain_data_extra_subpath = None\n",
    "domain_family_desc_base_path = None\n",
    "\n",
    "tree_data_format = 'newick'\n",
    "# if relative path to tree is e.g.: ./data_example/trees/rpoB_tree-names.newick\n",
    "tree_data_base_path = './data_example/trees'\n",
    "tree_data_file = 'rpoB_tree-names.newick'\n",
    "\n",
    "Search_Terms = ['dna-directed polymerase',\n",
    "                '16S',\n",
    "                'DNA and methyltransferase',\n",
    "                '1.10.3.-',\n",
    "                'fucI',\n",
    "                'sulfate adenylyl transferase']\n",
    "\n",
    "from urllib.request import urlopen\n",
    "import_code = urlopen('https://raw.githubusercontent.com/dcchivian/KGB/master/KGB.py')\n",
    "exec(import_code.read())\n",
    "\n",
    "\"\"\"\n",
    "##########################################################################\n",
    "# Config #1: local notebook with KBase annotated genbank format plasmids #\n",
    "#            as separate contig gbk files in a directory for each sample #\n",
    "#            Domain Annotations: YES                                     #\n",
    "#            Gene Tree: NO                                               #\n",
    "##########################################################################\n",
    "\n",
    "%pylab notebook\n",
    "KBase_backend = False\n",
    "\n",
    "GenomeSet_names = [\"GW456\", \"GW460\"]\n",
    "ContigSet_names = []\n",
    "PivotFeatures_IDs = []\n",
    "PrimaryAnchor_leafId = \"GW456.scaffold_130\"\n",
    "PrimaryAnchor_locusTag = \"GW456_SEED_annot.CDS.1\"\n",
    "\n",
    "genome_annotation_system = 'KBase'\n",
    "genome_data_format = \"Genbank\"\n",
    "# if relative path to scaffolds is e.g.: ../data/plasmid_contigs/GW456_old/scaffolds/scaffold_130.gbk\n",
    "gbk_ext = \"gbk\"\n",
    "genome_data_base_path = \"../data/plasmid_contigs\"\n",
    "genome_data_extra_subpath = \"_old/scaffolds\"\n",
    "#genome_data_extra_subpath = \"_circular_2KbpMIN/scaffolds\"\n",
    "\n",
    "domain_data_exists = True\n",
    "domain_data_format = \"KBase_domains\"\n",
    "# if relative path to domain annotations is e.g.: ../data/plasmid_contigs/GW456_old/domains/GW456_Domain_annot.json\n",
    "domain_data_base_path = \"../data/plasmid_contigs\"\n",
    "domain_data_extra_subpath = \"_old/domains\"\n",
    "domain_family_desc_base_path = \"../data/domain_desc\"\n",
    "\n",
    "tree_data_format = 'newick'\n",
    "tree_data_base_path = None\n",
    "tree_data_file = None\n",
    "\n",
    "#Search_Terms = []\n",
    "Search_Terms = ['pil or conj or mobil or partition or t-dna or vir or plasmid',\n",
    "                'toxin',\n",
    "                'tra',\n",
    "                'rep'\n",
    "               ]\n",
    "\"\"\"\n",
    "\"\"\"\n",
    "# KBase genomes in KBase context\n",
    "#\n",
    "#KBase_backend = True\n",
    "#genome_data_format = \"KBase\"\n",
    "#GenomeSet_ref = '16750/58/1'\n",
    "#GenomeSet_refs = []\n",
    "#GenomeSet_refs = ['16750/7/1',\n",
    "#                   '16750/8/2',\n",
    "#                   '16750/9/1',\n",
    "#                   '16750/10/1',\n",
    "#                   '16750/11/1'\n",
    "#                   ]\n",
    "#GenomeSet_names = []\n",
    "#PivotFeatures_IDs = []\n",
    "#Search_Terms = []\n",
    "#domain_data_exists = True\n",
    "#domain_data_format = \"KBase_domains\"\n",
    "#domain_data_base_path = None\n",
    "#domain_family_desc_base_path = None\n",
    "#tree_data_format = 'newick'\n",
    "#tree_data_base_path = None\n",
    "#tree_data_file = None\n",
    "#PrimaryAnchor_leafId = None\n",
    "#PrimaryAnchor_locusTag = None\n",
    "\n",
    "\n",
    "# KBase genomes\n",
    "#\n",
    "#ref='ReferenceGenomeAnnotations/kb|g.3899'\n",
    "#ref='ReferenceGenomeAnnotations/kb|g.2624'\n",
    "#ref='ReferenceGenomeAnnotations/kb|g.2424'\n",
    "#GenomeSet_names = ['ReferenceGenomeAnnotations/kb|g.2424']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# KBase Genbank genomes\n",
    "#\n",
    "#GenomeSet_names = [\"Desulforudis_audaxviator\",\n",
    "#                   \"Pelotomaculum_thermopropionicum\",\n",
    "#                   \"Desulfotomaculum_reducens\",\n",
    "#                   \"Bacillus_subtilis\",\n",
    "#                   \"Desulfovibrio_desulfuricans\",\n",
    "#                   \"Desulfovibrio_vulgaris\",\n",
    "#                   \"Geobacter_sulfurreducens\",\n",
    "#                   \"Escherichia_coli\"]\n",
    "##                   \"Desulforudis_audaxviator\",\n",
    "##                   \"Bacillus_subtilis\",\n",
    "#                   #\"Moorella_thermoacetica\",\n",
    "#                   #\"Thermoanaerobacter_tengcongensis\",\n",
    "#\n",
    "#ContigSet_names = []\n",
    "#PivotFeatures_IDs = [\"kb|g.1713.peg.340\",\n",
    "#                         \"kb|g.1713.peg.368\",\n",
    "#                         \"kb|g.1106.peg.1477\",\n",
    "#                         \"kb|g.996.peg.3228\",\n",
    "#                         \"kb|g.423.peg.584\",\n",
    "#                         \"kb|g.423.peg.176\",\n",
    "#                         \"kb|g.357.peg.1850\",\n",
    "#                         \"kb|g.3562.peg.757\",\n",
    "#                         \"kb|g.483.peg.109\",\n",
    "#                         \"kb|g.26833.CDS.2562\"]\n",
    "#PrimaryAnchor_leafId = \"Daudaxviator-Sat\"\n",
    "#PrimaryAnchor_locusTag = \"kb|g.1713.peg.340\"\n",
    "#\n",
    "#genome_annotation_system = 'KBase'\n",
    "#genome_data_format = \"Genbank\"\n",
    "#genome_data_base_path = \"./data/KBase_genome_contigs\"\n",
    "#genome_data_extra_subpath = \"\"\n",
    "##genome_data_extra_subpath = \"/scaffolds\"\n",
    "#\n",
    "#domain_data_format = \"KBase_domains\"\n",
    "#domain_data_base_path = None\n",
    "#domain_data_base_path = \"./data/KBase_genome_contigs\"\n",
    "#domain_data_extra_subpath = \"\"\n",
    "##domain_data_extra_subpath = \"/domains\"\n",
    "#domain_family_desc_base_path = \"./data/domain_desc\"\n",
    "#\n",
    "#tree_data_format = 'newick'\n",
    "#tree_data_base_path = None\n",
    "#tree_data_file = None\n",
    "##tree_data_base_path = './data/trees'\n",
    "##tree_data_file = 'rpoB_tree-names.newick'\n",
    "#\n",
    "##Search_Terms = []\n",
    "#Search_Terms = ['dna-directed polymerase',\n",
    "#                '16S',\n",
    "#                'DNA and methyltransferase',\n",
    "#                '1.10.3.-',\n",
    "##                'polymerase or dna-directed',\n",
    "#                'fucI',\n",
    "#                'sulfate adenylyl transferase']\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
