{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# KGB Genome Browser v0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "reading Gsulf/c:scaf_1 ...\n",
      "reading DvulH/c:scaf_1 ...\n",
      "reading DdesulfG20/c:scaf_1 ...\n",
      "reading EcoliK12/c:scaf_1 ...\n",
      "reading Bsub/c:scaf_1 ...\n",
      "reading DaudaxMP104C/c:scaf_1 ...\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "/* Put everything inside the global mpl namespace */\n",
       "window.mpl = {};\n",
       "\n",
       "mpl.get_websocket_type = function() {\n",
       "    if (typeof(WebSocket) !== 'undefined') {\n",
       "        return WebSocket;\n",
       "    } else if (typeof(MozWebSocket) !== 'undefined') {\n",
       "        return MozWebSocket;\n",
       "    } else {\n",
       "        alert('Your browser does not have WebSocket support.' +\n",
       "              'Please try Chrome, Safari or Firefox â‰¥ 6. ' +\n",
       "              'Firefox 4 and 5 are also supported but you ' +\n",
       "              'have to enable WebSockets in about:config.');\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure = function(figure_id, websocket, ondownload, parent_element) {\n",
       "    this.id = figure_id;\n",
       "\n",
       "    this.ws = websocket;\n",
       "\n",
       "    this.supports_binary = (this.ws.binaryType != undefined);\n",
       "\n",
       "    if (!this.supports_binary) {\n",
       "        var warnings = document.getElementById(\"mpl-warnings\");\n",
       "        if (warnings) {\n",
       "            warnings.style.display = 'block';\n",
       "            warnings.textContent = (\n",
       "                \"This browser does not support binary websocket messages. \" +\n",
       "                    \"Performance may be slow.\");\n",
       "        }\n",
       "    }\n",
       "\n",
       "    this.imageObj = new Image();\n",
       "\n",
       "    this.context = undefined;\n",
       "    this.message = undefined;\n",
       "    this.canvas = undefined;\n",
       "    this.rubberband_canvas = undefined;\n",
       "    this.rubberband_context = undefined;\n",
       "    this.format_dropdown = undefined;\n",
       "\n",
       "    this.image_mode = 'full';\n",
       "\n",
       "    this.root = $('<div/>');\n",
       "    this._root_extra_style(this.root)\n",
       "    this.root.attr('style', 'display: inline-block');\n",
       "\n",
       "    $(parent_element).append(this.root);\n",
       "\n",
       "    this._init_header(this);\n",
       "    this._init_canvas(this);\n",
       "    this._init_toolbar(this);\n",
       "\n",
       "    var fig = this;\n",
       "\n",
       "    this.waiting = false;\n",
       "\n",
       "    this.ws.onopen =  function () {\n",
       "            fig.send_message(\"supports_binary\", {value: fig.supports_binary});\n",
       "            fig.send_message(\"send_image_mode\", {});\n",
       "            fig.send_message(\"refresh\", {});\n",
       "        }\n",
       "\n",
       "    this.imageObj.onload = function() {\n",
       "            if (fig.image_mode == 'full') {\n",
       "                // Full images could contain transparency (where diff images\n",
       "                // almost always do), so we need to clear the canvas so that\n",
       "                // there is no ghosting.\n",
       "                fig.context.clearRect(0, 0, fig.canvas.width, fig.canvas.height);\n",
       "            }\n",
       "            fig.context.drawImage(fig.imageObj, 0, 0);\n",
       "            fig.waiting = false;\n",
       "        };\n",
       "\n",
       "    this.imageObj.onunload = function() {\n",
       "        this.ws.close();\n",
       "    }\n",
       "\n",
       "    this.ws.onmessage = this._make_on_message_function(this);\n",
       "\n",
       "    this.ondownload = ondownload;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_header = function() {\n",
       "    var titlebar = $(\n",
       "        '<div class=\"ui-dialog-titlebar ui-widget-header ui-corner-all ' +\n",
       "        'ui-helper-clearfix\"/>');\n",
       "    var titletext = $(\n",
       "        '<div class=\"ui-dialog-title\" style=\"width: 100%; ' +\n",
       "        'text-align: center; padding: 3px;\"/>');\n",
       "    titlebar.append(titletext)\n",
       "    this.root.append(titlebar);\n",
       "    this.header = titletext[0];\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._root_extra_style = function(canvas_div) {\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_canvas = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var canvas_div = $('<div/>');\n",
       "\n",
       "    canvas_div.attr('style', 'position: relative; clear: both; outline: 0');\n",
       "\n",
       "    function canvas_keyboard_event(event) {\n",
       "        return fig.key_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    canvas_div.keydown('key_press', canvas_keyboard_event);\n",
       "    canvas_div.keyup('key_release', canvas_keyboard_event);\n",
       "    this.canvas_div = canvas_div\n",
       "    this._canvas_extra_style(canvas_div)\n",
       "    this.root.append(canvas_div);\n",
       "\n",
       "    var canvas = $('<canvas/>');\n",
       "    canvas.addClass('mpl-canvas');\n",
       "    canvas.attr('style', \"left: 0; top: 0; z-index: 0; outline: 0\")\n",
       "\n",
       "    this.canvas = canvas[0];\n",
       "    this.context = canvas[0].getContext(\"2d\");\n",
       "\n",
       "    var rubberband = $('<canvas/>');\n",
       "    rubberband.attr('style', \"position: absolute; left: 0; top: 0; z-index: 1;\")\n",
       "\n",
       "    var pass_mouse_events = true;\n",
       "\n",
       "    canvas_div.resizable({\n",
       "        start: function(event, ui) {\n",
       "            pass_mouse_events = false;\n",
       "        },\n",
       "        resize: function(event, ui) {\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "        stop: function(event, ui) {\n",
       "            pass_mouse_events = true;\n",
       "            fig.request_resize(ui.size.width, ui.size.height);\n",
       "        },\n",
       "    });\n",
       "\n",
       "    function mouse_event_fn(event) {\n",
       "        if (pass_mouse_events)\n",
       "            return fig.mouse_event(event, event['data']);\n",
       "    }\n",
       "\n",
       "    rubberband.mousedown('button_press', mouse_event_fn);\n",
       "    rubberband.mouseup('button_release', mouse_event_fn);\n",
       "    // Throttle sequential mouse events to 1 every 20ms.\n",
       "    rubberband.mousemove('motion_notify', mouse_event_fn);\n",
       "\n",
       "    rubberband.mouseenter('figure_enter', mouse_event_fn);\n",
       "    rubberband.mouseleave('figure_leave', mouse_event_fn);\n",
       "\n",
       "    canvas_div.on(\"wheel\", function (event) {\n",
       "        event = event.originalEvent;\n",
       "        event['data'] = 'scroll'\n",
       "        if (event.deltaY < 0) {\n",
       "            event.step = 1;\n",
       "        } else {\n",
       "            event.step = -1;\n",
       "        }\n",
       "        mouse_event_fn(event);\n",
       "    });\n",
       "\n",
       "    canvas_div.append(canvas);\n",
       "    canvas_div.append(rubberband);\n",
       "\n",
       "    this.rubberband = rubberband;\n",
       "    this.rubberband_canvas = rubberband[0];\n",
       "    this.rubberband_context = rubberband[0].getContext(\"2d\");\n",
       "    this.rubberband_context.strokeStyle = \"#000000\";\n",
       "\n",
       "    this._resize_canvas = function(width, height) {\n",
       "        // Keep the size of the canvas, canvas container, and rubber band\n",
       "        // canvas in synch.\n",
       "        canvas_div.css('width', width)\n",
       "        canvas_div.css('height', height)\n",
       "\n",
       "        canvas.attr('width', width);\n",
       "        canvas.attr('height', height);\n",
       "\n",
       "        rubberband.attr('width', width);\n",
       "        rubberband.attr('height', height);\n",
       "    }\n",
       "\n",
       "    // Set the figure to an initial 600x600px, this will subsequently be updated\n",
       "    // upon first draw.\n",
       "    this._resize_canvas(600, 600);\n",
       "\n",
       "    // Disable right mouse context menu.\n",
       "    $(this.rubberband_canvas).bind(\"contextmenu\",function(e){\n",
       "        return false;\n",
       "    });\n",
       "\n",
       "    function set_focus () {\n",
       "        canvas.focus();\n",
       "        canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    window.setTimeout(set_focus, 100);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items) {\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) {\n",
       "            // put a spacer in here.\n",
       "            continue;\n",
       "        }\n",
       "        var button = $('<button/>');\n",
       "        button.addClass('ui-button ui-widget ui-state-default ui-corner-all ' +\n",
       "                        'ui-button-icon-only');\n",
       "        button.attr('role', 'button');\n",
       "        button.attr('aria-disabled', 'false');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "\n",
       "        var icon_img = $('<span/>');\n",
       "        icon_img.addClass('ui-button-icon-primary ui-icon');\n",
       "        icon_img.addClass(image);\n",
       "        icon_img.addClass('ui-corner-all');\n",
       "\n",
       "        var tooltip_span = $('<span/>');\n",
       "        tooltip_span.addClass('ui-button-text');\n",
       "        tooltip_span.html(tooltip);\n",
       "\n",
       "        button.append(icon_img);\n",
       "        button.append(tooltip_span);\n",
       "\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    var fmt_picker_span = $('<span/>');\n",
       "\n",
       "    var fmt_picker = $('<select/>');\n",
       "    fmt_picker.addClass('mpl-toolbar-option ui-widget ui-widget-content');\n",
       "    fmt_picker_span.append(fmt_picker);\n",
       "    nav_element.append(fmt_picker_span);\n",
       "    this.format_dropdown = fmt_picker[0];\n",
       "\n",
       "    for (var ind in mpl.extensions) {\n",
       "        var fmt = mpl.extensions[ind];\n",
       "        var option = $(\n",
       "            '<option/>', {selected: fmt === mpl.default_extension}).html(fmt);\n",
       "        fmt_picker.append(option)\n",
       "    }\n",
       "\n",
       "    // Add hover states to the ui-buttons\n",
       "    $( \".ui-button\" ).hover(\n",
       "        function() { $(this).addClass(\"ui-state-hover\");},\n",
       "        function() { $(this).removeClass(\"ui-state-hover\");}\n",
       "    );\n",
       "\n",
       "    var status_bar = $('<span class=\"mpl-message\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.request_resize = function(x_pixels, y_pixels) {\n",
       "    // Request matplotlib to resize the figure. Matplotlib will then trigger a resize in the client,\n",
       "    // which will in turn request a refresh of the image.\n",
       "    this.send_message('resize', {'width': x_pixels, 'height': y_pixels});\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_message = function(type, properties) {\n",
       "    properties['type'] = type;\n",
       "    properties['figure_id'] = this.id;\n",
       "    this.ws.send(JSON.stringify(properties));\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.send_draw_message = function() {\n",
       "    if (!this.waiting) {\n",
       "        this.waiting = true;\n",
       "        this.ws.send(JSON.stringify({type: \"draw\", figure_id: this.id}));\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    var format_dropdown = fig.format_dropdown;\n",
       "    var format = format_dropdown.options[format_dropdown.selectedIndex].value;\n",
       "    fig.ondownload(fig, format);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype.handle_resize = function(fig, msg) {\n",
       "    var size = msg['size'];\n",
       "    if (size[0] != fig.canvas.width || size[1] != fig.canvas.height) {\n",
       "        fig._resize_canvas(size[0], size[1]);\n",
       "        fig.send_message(\"refresh\", {});\n",
       "    };\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_rubberband = function(fig, msg) {\n",
       "    var x0 = msg['x0'];\n",
       "    var y0 = fig.canvas.height - msg['y0'];\n",
       "    var x1 = msg['x1'];\n",
       "    var y1 = fig.canvas.height - msg['y1'];\n",
       "    x0 = Math.floor(x0) + 0.5;\n",
       "    y0 = Math.floor(y0) + 0.5;\n",
       "    x1 = Math.floor(x1) + 0.5;\n",
       "    y1 = Math.floor(y1) + 0.5;\n",
       "    var min_x = Math.min(x0, x1);\n",
       "    var min_y = Math.min(y0, y1);\n",
       "    var width = Math.abs(x1 - x0);\n",
       "    var height = Math.abs(y1 - y0);\n",
       "\n",
       "    fig.rubberband_context.clearRect(\n",
       "        0, 0, fig.canvas.width, fig.canvas.height);\n",
       "\n",
       "    fig.rubberband_context.strokeRect(min_x, min_y, width, height);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_figure_label = function(fig, msg) {\n",
       "    // Updates the figure title.\n",
       "    fig.header.textContent = msg['label'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_cursor = function(fig, msg) {\n",
       "    var cursor = msg['cursor'];\n",
       "    switch(cursor)\n",
       "    {\n",
       "    case 0:\n",
       "        cursor = 'pointer';\n",
       "        break;\n",
       "    case 1:\n",
       "        cursor = 'default';\n",
       "        break;\n",
       "    case 2:\n",
       "        cursor = 'crosshair';\n",
       "        break;\n",
       "    case 3:\n",
       "        cursor = 'move';\n",
       "        break;\n",
       "    }\n",
       "    fig.rubberband_canvas.style.cursor = cursor;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_message = function(fig, msg) {\n",
       "    fig.message.textContent = msg['message'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_draw = function(fig, msg) {\n",
       "    // Request the server to send over a new figure.\n",
       "    fig.send_draw_message();\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_image_mode = function(fig, msg) {\n",
       "    fig.image_mode = msg['mode'];\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Called whenever the canvas gets updated.\n",
       "    this.send_message(\"ack\", {});\n",
       "}\n",
       "\n",
       "// A function to construct a web socket function for onmessage handling.\n",
       "// Called in the figure constructor.\n",
       "mpl.figure.prototype._make_on_message_function = function(fig) {\n",
       "    return function socket_on_message(evt) {\n",
       "        if (evt.data instanceof Blob) {\n",
       "            /* FIXME: We get \"Resource interpreted as Image but\n",
       "             * transferred with MIME type text/plain:\" errors on\n",
       "             * Chrome.  But how to set the MIME type?  It doesn't seem\n",
       "             * to be part of the websocket stream */\n",
       "            evt.data.type = \"image/png\";\n",
       "\n",
       "            /* Free the memory for the previous frames */\n",
       "            if (fig.imageObj.src) {\n",
       "                (window.URL || window.webkitURL).revokeObjectURL(\n",
       "                    fig.imageObj.src);\n",
       "            }\n",
       "\n",
       "            fig.imageObj.src = (window.URL || window.webkitURL).createObjectURL(\n",
       "                evt.data);\n",
       "            fig.updated_canvas_event();\n",
       "            return;\n",
       "        }\n",
       "        else if (typeof evt.data === 'string' && evt.data.slice(0, 21) == \"data:image/png;base64\") {\n",
       "            fig.imageObj.src = evt.data;\n",
       "            fig.updated_canvas_event();\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        var msg = JSON.parse(evt.data);\n",
       "        var msg_type = msg['type'];\n",
       "\n",
       "        // Call the  \"handle_{type}\" callback, which takes\n",
       "        // the figure and JSON message as its only arguments.\n",
       "        try {\n",
       "            var callback = fig[\"handle_\" + msg_type];\n",
       "        } catch (e) {\n",
       "            console.log(\"No handler for the '\" + msg_type + \"' message type: \", msg);\n",
       "            return;\n",
       "        }\n",
       "\n",
       "        if (callback) {\n",
       "            try {\n",
       "                // console.log(\"Handling '\" + msg_type + \"' message: \", msg);\n",
       "                callback(fig, msg);\n",
       "            } catch (e) {\n",
       "                console.log(\"Exception inside the 'handler_\" + msg_type + \"' callback:\", e, e.stack, msg);\n",
       "            }\n",
       "        }\n",
       "    };\n",
       "}\n",
       "\n",
       "// from http://stackoverflow.com/questions/1114465/getting-mouse-location-in-canvas\n",
       "mpl.findpos = function(e) {\n",
       "    //this section is from http://www.quirksmode.org/js/events_properties.html\n",
       "    var targ;\n",
       "    if (!e)\n",
       "        e = window.event;\n",
       "    if (e.target)\n",
       "        targ = e.target;\n",
       "    else if (e.srcElement)\n",
       "        targ = e.srcElement;\n",
       "    if (targ.nodeType == 3) // defeat Safari bug\n",
       "        targ = targ.parentNode;\n",
       "\n",
       "    // jQuery normalizes the pageX and pageY\n",
       "    // pageX,Y are the mouse positions relative to the document\n",
       "    // offset() returns the position of the element relative to the document\n",
       "    var x = e.pageX - $(targ).offset().left;\n",
       "    var y = e.pageY - $(targ).offset().top;\n",
       "\n",
       "    return {\"x\": x, \"y\": y};\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.mouse_event = function(event, name) {\n",
       "    var canvas_pos = mpl.findpos(event)\n",
       "\n",
       "    if (name === 'button_press')\n",
       "    {\n",
       "        this.canvas.focus();\n",
       "        this.canvas_div.focus();\n",
       "    }\n",
       "\n",
       "    var x = canvas_pos.x;\n",
       "    var y = canvas_pos.y;\n",
       "\n",
       "    this.send_message(name, {x: x, y: y, button: event.button,\n",
       "                             step: event.step});\n",
       "\n",
       "    /* This prevents the web browser from automatically changing to\n",
       "     * the text insertion cursor when the button is pressed.  We want\n",
       "     * to control all of the cursor setting manually through the\n",
       "     * 'cursor' event from matplotlib */\n",
       "    event.preventDefault();\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    // Handle any extra behaviour associated with a key event\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.key_event = function(event, name) {\n",
       "\n",
       "    // Prevent repeat events\n",
       "    if (name == 'key_press')\n",
       "    {\n",
       "        if (event.which === this._key)\n",
       "            return;\n",
       "        else\n",
       "            this._key = event.which;\n",
       "    }\n",
       "    if (name == 'key_release')\n",
       "        this._key = null;\n",
       "\n",
       "    var value = '';\n",
       "    if (event.ctrlKey && event.which != 17)\n",
       "        value += \"ctrl+\";\n",
       "    if (event.altKey && event.which != 18)\n",
       "        value += \"alt+\";\n",
       "    if (event.shiftKey && event.which != 16)\n",
       "        value += \"shift+\";\n",
       "\n",
       "    value += 'k';\n",
       "    value += event.which.toString();\n",
       "\n",
       "    this._key_event_extra(event, name);\n",
       "\n",
       "    this.send_message(name, {key: value});\n",
       "    return false;\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onclick = function(name) {\n",
       "    if (name == 'download') {\n",
       "        this.handle_save(this, null);\n",
       "    } else {\n",
       "        this.send_message(\"toolbar_button\", {name: name});\n",
       "    }\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.toolbar_button_onmouseover = function(tooltip) {\n",
       "    this.message.textContent = tooltip;\n",
       "};\n",
       "mpl.toolbar_items = [[\"Home\", \"Reset original view\", \"fa fa-home icon-home\", \"home\"], [\"Back\", \"Back to  previous view\", \"fa fa-arrow-left icon-arrow-left\", \"back\"], [\"Forward\", \"Forward to next view\", \"fa fa-arrow-right icon-arrow-right\", \"forward\"], [\"\", \"\", \"\", \"\"], [\"Pan\", \"Pan axes with left mouse, zoom with right\", \"fa fa-arrows icon-move\", \"pan\"], [\"Zoom\", \"Zoom to rectangle\", \"fa fa-square-o icon-check-empty\", \"zoom\"], [\"\", \"\", \"\", \"\"], [\"Download\", \"Download plot\", \"fa fa-floppy-o icon-save\", \"download\"]];\n",
       "\n",
       "mpl.extensions = [\"eps\", \"jpeg\", \"pdf\", \"png\", \"ps\", \"raw\", \"svg\", \"tif\"];\n",
       "\n",
       "mpl.default_extension = \"png\";var comm_websocket_adapter = function(comm) {\n",
       "    // Create a \"websocket\"-like object which calls the given IPython comm\n",
       "    // object with the appropriate methods. Currently this is a non binary\n",
       "    // socket, so there is still some room for performance tuning.\n",
       "    var ws = {};\n",
       "\n",
       "    ws.close = function() {\n",
       "        comm.close()\n",
       "    };\n",
       "    ws.send = function(m) {\n",
       "        //console.log('sending', m);\n",
       "        comm.send(m);\n",
       "    };\n",
       "    // Register the callback with on_msg.\n",
       "    comm.on_msg(function(msg) {\n",
       "        //console.log('receiving', msg['content']['data'], msg);\n",
       "        // Pass the mpl event to the overriden (by mpl) onmessage function.\n",
       "        ws.onmessage(msg['content']['data'])\n",
       "    });\n",
       "    return ws;\n",
       "}\n",
       "\n",
       "mpl.mpl_figure_comm = function(comm, msg) {\n",
       "    // This is the function which gets called when the mpl process\n",
       "    // starts-up an IPython Comm through the \"matplotlib\" channel.\n",
       "\n",
       "    var id = msg.content.data.id;\n",
       "    // Get hold of the div created by the display call when the Comm\n",
       "    // socket was opened in Python.\n",
       "    var element = $(\"#\" + id);\n",
       "    var ws_proxy = comm_websocket_adapter(comm)\n",
       "\n",
       "    function ondownload(figure, format) {\n",
       "        window.open(figure.imageObj.src);\n",
       "    }\n",
       "\n",
       "    var fig = new mpl.figure(id, ws_proxy,\n",
       "                           ondownload,\n",
       "                           element.get(0));\n",
       "\n",
       "    // Call onopen now - mpl needs it, as it is assuming we've passed it a real\n",
       "    // web socket which is closed, not our websocket->open comm proxy.\n",
       "    ws_proxy.onopen();\n",
       "\n",
       "    fig.parent_element = element.get(0);\n",
       "    fig.cell_info = mpl.find_output_cell(\"<div id='\" + id + \"'></div>\");\n",
       "    if (!fig.cell_info) {\n",
       "        console.error(\"Failed to find cell for figure\", id, fig);\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    var output_index = fig.cell_info[2]\n",
       "    var cell = fig.cell_info[0];\n",
       "\n",
       "};\n",
       "\n",
       "mpl.figure.prototype.handle_close = function(fig, msg) {\n",
       "    // Update the output cell to use the data from the current canvas.\n",
       "    fig.push_to_output();\n",
       "    var dataURL = fig.canvas.toDataURL();\n",
       "    // Re-enable the keyboard manager in IPython - without this line, in FF,\n",
       "    // the notebook keyboard shortcuts fail.\n",
       "    IPython.keyboard_manager.enable()\n",
       "    $(fig.parent_element).html('<img src=\"' + dataURL + '\">');\n",
       "    fig.send_message('closing', {});\n",
       "    fig.ws.close()\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.push_to_output = function(remove_interactive) {\n",
       "    // Turn the data on the canvas into data in the output cell.\n",
       "    var dataURL = this.canvas.toDataURL();\n",
       "    this.cell_info[1]['text/html'] = '<img src=\"' + dataURL + '\">';\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.updated_canvas_event = function() {\n",
       "    // Tell IPython that the notebook contents must change.\n",
       "    IPython.notebook.set_dirty(true);\n",
       "    this.send_message(\"ack\", {});\n",
       "    var fig = this;\n",
       "    // Wait a second, then push the new image to the DOM so\n",
       "    // that it is saved nicely (might be nice to debounce this).\n",
       "    setTimeout(function () { fig.push_to_output() }, 1000);\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._init_toolbar = function() {\n",
       "    var fig = this;\n",
       "\n",
       "    var nav_element = $('<div/>')\n",
       "    nav_element.attr('style', 'width: 100%');\n",
       "    this.root.append(nav_element);\n",
       "\n",
       "    // Define a callback function for later on.\n",
       "    function toolbar_event(event) {\n",
       "        return fig.toolbar_button_onclick(event['data']);\n",
       "    }\n",
       "    function toolbar_mouse_event(event) {\n",
       "        return fig.toolbar_button_onmouseover(event['data']);\n",
       "    }\n",
       "\n",
       "    for(var toolbar_ind in mpl.toolbar_items){\n",
       "        var name = mpl.toolbar_items[toolbar_ind][0];\n",
       "        var tooltip = mpl.toolbar_items[toolbar_ind][1];\n",
       "        var image = mpl.toolbar_items[toolbar_ind][2];\n",
       "        var method_name = mpl.toolbar_items[toolbar_ind][3];\n",
       "\n",
       "        if (!name) { continue; };\n",
       "\n",
       "        var button = $('<button class=\"btn btn-default\" href=\"#\" title=\"' + name + '\"><i class=\"fa ' + image + ' fa-lg\"></i></button>');\n",
       "        button.click(method_name, toolbar_event);\n",
       "        button.mouseover(tooltip, toolbar_mouse_event);\n",
       "        nav_element.append(button);\n",
       "    }\n",
       "\n",
       "    // Add the status bar.\n",
       "    var status_bar = $('<span class=\"mpl-message\" style=\"text-align:right; float: right;\"/>');\n",
       "    nav_element.append(status_bar);\n",
       "    this.message = status_bar[0];\n",
       "\n",
       "    // Add the close button to the window.\n",
       "    var buttongrp = $('<div class=\"btn-group inline pull-right\"></div>');\n",
       "    var button = $('<button class=\"btn btn-mini btn-danger\" href=\"#\" title=\"Close figure\"><i class=\"fa fa-times icon-remove icon-large\"></i></button>');\n",
       "    button.click(function (evt) { fig.handle_close(fig, {}); } );\n",
       "    button.mouseover('Close figure', toolbar_mouse_event);\n",
       "    buttongrp.append(button);\n",
       "    var titlebar = this.root.find($('.ui-dialog-titlebar'));\n",
       "    titlebar.prepend(buttongrp);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.figure.prototype._canvas_extra_style = function(el){\n",
       "    // this is important to make the div 'focusable\n",
       "    el.attr('tabindex', 0)\n",
       "    // reach out to IPython and tell the keyboard manager to turn it's self\n",
       "    // off when our div gets focus\n",
       "\n",
       "    // location in version 3\n",
       "    if (IPython.notebook.keyboard_manager) {\n",
       "        IPython.notebook.keyboard_manager.register_events(el);\n",
       "    }\n",
       "    else {\n",
       "        // location in version 2\n",
       "        IPython.keyboard_manager.register_events(el);\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "mpl.figure.prototype._key_event_extra = function(event, name) {\n",
       "    var manager = IPython.notebook.keyboard_manager;\n",
       "    if (!manager)\n",
       "        manager = IPython.keyboard_manager;\n",
       "\n",
       "    // Check for shift+enter\n",
       "    if (event.shiftKey && event.which == 13) {\n",
       "        this.canvas_div.blur();\n",
       "        event.shiftKey = false;\n",
       "        // Send a \"J\" for go to next cell\n",
       "        event.which = 74;\n",
       "        event.keyCode = 74;\n",
       "        manager.command_mode();\n",
       "        manager.handle_keydown(event);\n",
       "    }\n",
       "}\n",
       "\n",
       "mpl.figure.prototype.handle_save = function(fig, msg) {\n",
       "    fig.ondownload(fig, null);\n",
       "}\n",
       "\n",
       "\n",
       "mpl.find_output_cell = function(html_output) {\n",
       "    // Return the cell and output element which can be found *uniquely* in the notebook.\n",
       "    // Note - this is a bit hacky, but it is done because the \"notebook_saving.Notebook\"\n",
       "    // IPython event is triggered only after the cells have been serialised, which for\n",
       "    // our purposes (turning an active figure into a static one), is too late.\n",
       "    var cells = IPython.notebook.get_cells();\n",
       "    var ncells = cells.length;\n",
       "    for (var i=0; i<ncells; i++) {\n",
       "        var cell = cells[i];\n",
       "        if (cell.cell_type === 'code'){\n",
       "            for (var j=0; j<cell.output_area.outputs.length; j++) {\n",
       "                var data = cell.output_area.outputs[j];\n",
       "                if (data.data) {\n",
       "                    // IPython >= 3 moved mimebundle to data attribute of output\n",
       "                    data = data.data;\n",
       "                }\n",
       "                if (data['text/html'] == html_output) {\n",
       "                    return [cell, data, j];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "}\n",
       "\n",
       "// Register the function which deals with the matplotlib target/channel.\n",
       "// The kernel may be null if the page has been refreshed.\n",
       "if (IPython.notebook.kernel != null) {\n",
       "    IPython.notebook.kernel.comm_manager.register_target('matplotlib', mpl.mpl_figure_comm);\n",
       "}\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<img src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAA3AAAAJYCAYAAAAjYgQ4AAAgAElEQVR4XuydBXhUxxbH/3GFhAR3iru7u2uB4tBSCqWFKi1tqQu0lPIKhbYUd3cr7sHd3S0JEPfkfWfSu2w2u9m7dyV7d8983/teyd478hu5859zZsYFHJgAE2ACTIAJMAEmwASYABNgAkxAFQRcVJFLziQTYAJMgAkwASbABJgAE2ACTIAJgAUcNwImwASYABNgAkyACTABJsAEmIBKCLCAU0lFcTaZABNgAkyACTABJsAEmAATYAIs4LgNMAEmwASYABNgAkyACTABJsAEVEKABZxKKoqzyQSYABNgAkyACTABJsAEmAATYAHHbYAJMAEmwASYABNgAkyACTABJqASAizgVFJRnE0mwASYABNgAkyACTABJsAEmAALOG4DTIAJMAEmwASYABNgAkyACTABlRBgAaeSiuJsMgEmwASYABNgAkyACTABJsAEWMBxG2ACTIAJMAEmwASYABNgAkyACaiEAAs4lVQUZ5MJMAEmwASYABNgAkyACTABJsACjtsAE2ACTIAJMAEmwASYABNgAkxAJQRYwKmkojibTIAJMAEmwASYABNgAkyACTABxQIuLS0tjfExASaQNQEXFxfuY9xImAATYAJMgAkwASbABDIRUDpP5MklNyYmYEUCSjsmZYkXSaxYMRw1E2ACTIAJMAEmwASymYDSeSILuGyuOE7esQko7Zgs4By7XXDpmAATYAJMgAkwASagdJ7IAo7bDhOwIgGlHZMFnBUrhaNmAkyACTABJsAEmIAdEFA6T2QBZweVx1lwXAJKOyYLOMdtE1wyJsAEmAATYAJMwPEJ0HEhxo5CUDpPtLiA087svYdPcP7KTYSGv0B8QiJS01Idv7a4hA5NwM3VDT7enihUIC8qlX0FeYICsyyv0o4pR8Clpqbi9v3HuHbrHqJjYpGQmET75hyaPxeOCTABJsAEmAATYAL2SMDD3R0+3l4omD8PypcqBl8fb6PZVDpPtKiAk8TbnfuPMW/lZjyPiDKacX6ACaiZQNFC+fB6704IzOmP1NQ0uLpm7FJKO2ZWAi42Lh7rtu3HhSs3ERMXr2Z8nHcmwASYABNgAkyACTgcAVcXFxQvUgBtm9ZFmVeKGiyf0nmiRQUc5e7mnQeYsXidsAZwYALOQCAoMCfeHdITuQJyIDwcCA5+WWqlHdOQgKNFkRmL1uJx6DNnQMtlZAJMgAkwASbABJiAagm4ubqiT5dWqFW1vN4yKJ0nWlzATZu3Ctdv31ctaM44E1BCoEWDmujcuhGePwfu3QOqVEmPRWnHNCTgfp6+gMWbkgrid5gAE2ACTIAJMAEmkE0ERg3piVeKFUJEBBAQYP5Cv0UF3IvIaHz3v9m8DyebGgcnm30EyIXyy/ffAJnM//gDGDECcHe3rIC7de8hpsxekX2F5JSZABNgAkyACTABJsAETCZQu2p59OvWBvHxwLNnQMGC5i30W1TAnTh7GQvX/GtyofgFJuAIBD57ZxDy5s6Ff/4BGjcGypWzrIBbvnEnQk6cdwRUXAYmwASYABNgAkyACTgNAU8PD3z38TB4eXrgyBGgbl07EnD7jpzGmq17M1UGbeBr0bAWPv1pWobfmtWvIdzO/t1zGNv2HdX8VrhAXjSvXwOvFCsIf18fJCYlC7exc5dvIOTEOc3+OlKzfbq21ryXlpoKsgJevHYLW3YfRlx8gtGG4eHuhoa1q6J6xTLIExwId3c3RERG49bdhwg5eQFk9XC2YAmuzsaMyjv6jV4oUaQgli0DEhKAQYMsK+B++XMRHj0Nc0a0mjLTwPftR2/i1PmrWLZhp9VYFMyXW4jx0xeuWS0NjpgJMAHlBL5873UkJScjKSlZE8miNf/iadhztG5SR3zTU1JTQaf13n34BBu2HxCnYas1kHeHJculb4z76K2+mDpnhZhzWSO4urqidePaom7oVPKUlFQ8exGJf/cewcMn9v9tozY3c8mGDN/hdwa/ir2HT4kT160ZShYrhG5tm2DSjCXWTEZ23NRWKLi7uYlvpVR/1P8WrN4qOx5ne/D9oa+hWOH8WLMG6NQJ8PBQPk+0qAVud8hJrN+234CAq4lPf5qu+c2QeKtTrQJ6d24pBNTR0xcR9iwCXl4eKF2iCOpVryj+RifwUZCExj+L1oFO5nNzc0XBfHnQvkV90EmY/yxel2XboKM+3x7YXQi3A8fO4sbtB0hMSkKe4FyoVaWc8FX9fMKfTncgi7lcna1DSuWlgbxU8cLYsAE4ehT4/nvlHZPiTNO5E+Cb32YhIiraWfGKctetXlH0zQL5cuO7ybOsNtGgPkDXRMxZvsmpeXPhmYC9Ehg3eghmLd2YaVGrX9fW8Pb2wuK12zSCrWr5Urj36KkQC2oNli5Xdoxx/bu3haeHO5as266pG5rb+fl44/RF+18s09fmRg7qIQTchau3rNq07E3ASYWlw9s+Ht4PX/zyd6by06JDKl9tlIHLsH5dUaF0caxfD5QsCVSsqHyeaEMB99ICZ0i85c8TjI+G98XJc1dEB9cN1MlJVJElTlvAff3rP4iOjdM8TgdKdGjZAJ+N/1Os0BkKA3u0Q4UyJfD7rOV4HBqe6bEalcri/JUbVpskWrW3mxG5JOCUcJUuLFTbfWR0ShCt1poTJAG3aROwYwcwebLyjqlPwH05cUaGdm5OXtX6Llk5t+09iga1KosVT1rQofZKoi4qJg4F8gYjOTlFXGNCkzX66L3aoRlu3HkgrKN0zcPitdtx/9FTgYDeo/GIAlnvl2/YKVbsP3yrL7y9PBH+PELctbdq8261IuN8MwGHJECT6dnLNmaw3OTOFYCPR/QXizuxMjxw1ALGWLmaN6ghxkGaK5MlZOWm3UhITBTHl+fLHQQPD3dQHJHRsZi7YhPoe/fBsD7ivirtMW7SV6Px+YS/xLvFCxdAz47NxSXEZMEsUiCv8LCisZSsaDUqlxVjLQUS0i8is742KndQID4e3he0EGnIEkpjcbUKpcU4HR0Th+Ubd4l4DZWDvKwof51aNUS5ksVEXugQPVrkp3G8b9fWSE5JQXBgAHIHBeDa7fsIOX5OPE/C49yVmxqjQw4/X3Rv1xR0qjTxorkfeXJpB31tThJwZDTo1akFqJwUDhw9g5CT6VseyHJ3/OwllC5eBIEBObB9/1FhfaxXoyJy+vth486DGm+PsiWLomOLBiBrJZVvxaZdwqqsK+D0fbsio2PEe1SOMiWKCMPGrfuPUDh/Hkyfv1rkjRYCPD09BDd9ZTS1T+gKOCorecjQYnZo+HMxl+/QooH4NxlZ6F5o2g5CbYA8arq2bYKCeXML77c7D+hbu0fUnaOG13t3RJXypUDzxJw507fbKD3szuYCjgaaTq0yu01SZZHlrWblsqKDy3F/NCQ0Gtaqgh4dmmVpPQvMmQPj3huCvSEnsWHHQdltpUndamhQszKCcuUUA8yRUxewbe8RSNcnS3n6bcaS/+5+KCKeI/dS+p92eKVoQdFRCxfMKwbCKzfuioFH28pCA+rmXYdAlwPWr1lJ/P/J81dEIy9eOL9o/PnzBIlBmzoKdQ7tYCy/+gpuClfqrJeu3xYDDKVFg9PP0xYg7NkLtGxUWwxQATn8xOT48KkL2HnguEiSVmZ+/HQENu44iIPHz4q/Ud33695WPENlpkCT7ndf74kJf8xH6LMXwvRMzArlzyMGA4r32OlL2HkwPV4KRQvmQ4cW9cWzgIu46Jo+PNr3EhLXLbtC4OHhhjrVaBD1xSc/TjNLxEkCbvNmYOtWYMoU5R2TyqFrgXN2AUcTkeEDuomDkiqWKYGWDWthypwVYuJCH6yJfy0SddyxZQP4enuLDx999MjK/vvsFbj38InoQ5XLlcSMRetAC0b026QZi8XEplWj2ihRtKCw3GfH6rTsQYgfZAJMQEyKdV0oyRJCIuDXvxc7FCESNa0a19ZbrnKliqFrm8b438zlQniRiKA7SWnRiYQPjWXkdkdzKlq0prkCfS/1jXGSgEtOTsYXo4eAXFJJsNE4OnLwq5g2dyUePQ0HCZmvJs1ESkqKcKGj+Q/9d1YhqzLQe7RgXqp4IazYuEvERwKlWsUymLlkfZbloMU8ivuvhWvFAXpv9uks8rzr0Akh4PIEBYJOR6c4x44cKMpPC3yUb5oD0m80fxnev5sQVjfvPhTzkzf7dcGRkxdw5tJ1TbH0tTkSRcSJ8h/2/AU27wqBn68PyMVw7vJNQvwSr7OXbwixGJwrAJ+8PQDb9x3FjgPHhDAe1r8rvvr1H7Fl6NN3BgrOtG2I4iSx/POfCzMIuKy+XQ1rV0Hlsq/g70XrQBP8t/p3FaKOBFz3dk0QGRWrmS+RgJcz186qXnUFHJX16s27QnxToO+qiwuwff8x8W9yAyaxvHrLHvTu1ELwPn72svjttc4t8STsOfaEnHSo/qtdGG0BRwfdtW2rfJ5oUwG3ZXeI3j1vUuE+HzVY7D+jDiUnSELj299mITomVkzoC+TNjf492opVJZqkGQqSWPhz/mrZ1x7QINmgVhXsPHBMNDoSEe2b18Pew6dBZaMg5elp2DMcOXVRrPRXr1QG9WpUwt8L14qGTYH2+b33Ri/cuvcIe0JOwcfbEx1bNhSD4MS/FgtXTgo0oBITul/v6JlLKFIwr1jNOHT8HEgA7th/TLh4km80Dd7a/tFy8quPjylcqbO6ubnheUQkdh86KfJ998ET0UlJ0JEYu3H7PkqVKIKWDWtiz+FTYi8CBRpY4uMTMX/VFvFvScA/eByqOW2RPlqNalcRop5Wa756/w3B7MCxM2LvQ97gXAjI6Y+te9JXyki8keC7evOe2C9Jq0z0ESNryvhpCzQrO8Q1MioGDx49xcHj50TboYuxJSEup/3pPsMCTgk1+e90adMYNLGgDyR9YL/64A1Mn7daCHXaU0H3T1Igcde4TjX8tXCN+Oj16tgCE6YvEL/Rvo8hvTripz/moXGdqqIPL12/Q/xGHzPaX0dCngWc/HrhJ5lAdhDQ585mTCRkRz4tkWZW5aJxMT4+QXOOgPYYR98+fz9fjQcBjXnkfk6eBlkJOLJCDe3TCT9MmavJ/ufvDsbS9dvF9pb33nwNLyKixKIznTkQERVjtJi6ZSAhM6RXB2Htun3vEeiAh6KF8mkEBYkOEmQkxrMqB43nZEmSRAC5vjeqXVWM/326tAbNxUjMUaDJ8+XrdzSWMcl6Rou8P419G4+09uGRlYoWh0lkScGQCyUtzr/WpRUm/b1EY4mkORl5htEcjd6bv3KLEHMUfvhkOH6fuUwsSlP49ctR+OLnv4WAbVqvuhBbUvjxk+H4efpCsdVH2gNn6Nv16Y/TMKR3R+GlJvEgSyltP6I4afGyS+tGOHn+qhC50nzUaOVl8YA+Abdw9b+4ff+ReOv9N18T8y9pryoJ5/AXkUKY04EeUdGxmnkZtQWyoJIF2VGDtoCjMnbsqAIB16Zp+nErZFpdaGCD48+fv4Nzl65nOsmSJmtSoAm25J6ne9iG9AxNykm8abtV6jaGFg1rCsE0/o/5wlqkHbTTk/x3gwNz4rNRg7Fmy16NxYjeIZFC1h7JLUDK09qte7H/6BkRLeWeJpuXrt3WrEpQJdJg9cPvczVWH5pMkuvWmi17xJ48CiQ0qDy//bNUk0W6S6J40YKYPGMJ7j8OFX+nTkr+5T9NmSs6h9z86uskpnClFSkSVt/9b45GdJKr6zcfvYn9R05j/X9ijdKhwYdWy76eNFMM0i0b1UKTOtXw9W8zRTboFEey5tEKEu09TEpOwYgB3RATGy82xdJK1fvD+mDin4v0urxSHDQg04fgf7OWaYpGqz200rb2332aUxyJK5n3J0xLn9hbIrCAswRF/XHQx/ybD4eKBQ5yPaFAgousuo+fhmfYr1ahdAk0q5/+EdR1O6GVyzf7dhYTE1oYKFwgn5iUUKCPDH1QWMBZrx45ZiZgKQL6JtPGXA0tlbat48mqXDQhp0VcOgiEgq6Ao3GSvn0UtMc8UwXcZ+8OwrL1O8TiNc1pihcpKARH/ZqVsWDVVqMHvhlyoZTyQW5zV27exeGTFzLhJQFnqBwkAmn/2bEzl8R76QKuirDIkYB7+CRU4/00uGcHITilZ0cM6C7mcyREf/p0uDinIautFMYEHHleSZ4+QsDFxAkBqPve9x8Pw6QZSzVi79dx7woLHHmAyBFwhr5dJOAG6whabQFHfEjQlytZVLjxkSvtP0vWm9Wc9Qk47b2pJOA27zyEq7fuZUqHvrdTZi1H2PMIs/KgppdVKeBosn7zzkNxsuSsJRtER9UNP38+Eucu3cgg4KhxjHvvdc2joWHPNavpktAgy1ZsbBxcXF2Fn3ObJnWES+KU2cuFCNAXDAk4Mrlr35a+YNUW4ZtMroC9OrUE7Qsjv2Ip5M0dhDFv98cfc1aKAUzK04Rp8zO4M749sIeYfEpWAurAtGmXXCG1g2Til6xSJDR2HTyOTTvTXQop0I3ulcqVxDitTaNkeaJVMTK90wBrLL/Sc/rYmMKVBBxZGbUPe6ANmkP7dsH/Zi4TbmtSKFYoP0YP7Q06dObyjTsoUaQA3n29F4hVQkISvv5wKH6cMleIWFqtopUYWn0ii92hE+fEBJvSexL6TIhj+j0qJlYTP50oOn7s29i48xD2HT6VoWiULrl+0AeIAnGlZ6QDcSwxALCAswRF/XHQx4ZEmfY9eGR9fWfIq8IiR21OaoNyBZyuGwq5ZNIHlFYGaaWSVjm1V0KtVzqOmQkwAVMJ6NuPRHHQPUu0qKh9UAaNH/SdUvUhJgbKRQuUNKGnRUsScrRvjVwoyUUtK+Gjb4zTuFCmpOCLUYPFYjvNJ7RdKO8/ChXfYtpvRYHcMmnuIy06Z1WPA7q3hbu7u1g0k/bB0YIt7dc6c/G6GOP/XLBGLPDSop20PSSrctCWlqoVSuPvRWtpdR9v9Oks5gbkhkfzOfLokbavGBJwZLGixWLy7pGEMO1NI9uBtnUxKwFHnlZ06B55Y5EL5YfD+mDuis1iDiRXwNEcdqyWCyV5lpAXEm0P0F6MzOrbRTzJC2Xm4vVk2sGwvl2Eh5G0By782QvhaUTfT9pTPm7iDFO7XobnjQk4aWvC3OUbxXyctgGRhfdJ2DPhQkleUuRuSYYZEum+Pt7Cg85RgyoFHF0jQIKDLCQF8gWLTkqbPrUDmehpUPhj7krNn6kTF8wbLBoidWISaGROpmBor5ZkyVq9eU8Ga5l2WoZcKKkxkgUpZw5/DO3bGQtWbhFCizpR++b19beptDQsWrNN7E0zlCdyFyTTsTQhpBWXf/cdFX7Q2oFEAIlPMRj9JzQ27TiocQGgv5FLGG101XZvoFW3j4b3w4yFa4U4lptffQUyhSsNTCTGtE3eEtvv/zdb7FGTgiTG6aNAllha/aF9cOu27RODOe1tozK98Von4f994epNjH6jN36evkD4qFOgzbhkzS3zShExENx7+BTrt+8XHxnaa/fVB0MN9vtrN++KVTkN150HsetgumuFJQILOEtQ1B8HfYQuXr+Ng/9ZpqWn6CN54twV4U78UsAVF/tgjFngKA5qq80b1BTRkUsQba6m/XBenp4Y3r+r2OxNH3U+xMR6dcsxMwElBOjbQwdUaF8jQJYmcvETx+1XKiNcs2iCSO5itN/aIa4R0FMu2qJRp9p/h5g8DsXKzXSISZJm+4C0UEmTe/JkIbdxfWPcpC9H4fOf/xbbMWiB9dUO6YeY0HeW9tzPW7kFsXFxwg2dxkaadNO+exJklJ6xoLlG4L8yxMUliMOndh06LrZe0KIZnTRMgQ4yoW0otK9R2gahrxxkDezUulHGQ0z+3SdOPyQL3IPHTzXeUCTgaF4huRdKFjgScCS6urVpjIL584j0ExOTxPeAFn6lkNUplDSf7dmxhXB1pLD/6GmNNVH3PbI8/aZlgZs47t1040B8gpjbSYeYkLGA5lZPwzMfYmLo20UeZFRvpUoUBvG99+iJOCiFvo+0SFmzSlnhxUL1SvNP7T1+xupP3+80r6O5p2RQ0C2rtI2FFgwk7znag0lzQDqRlA6UKVmssPiN+it5bZEAd9SgWgFH98CRwn53SE+xb2n6vFUZTpBK3wNVDt/+NlPvCVK0mkJWHGk/iyGhQZaYCZ+/g0PHz2aycEmNIjCnv7DmaO/J0m4wktiQBBy5/lGnmDp7hdiHoxtoZY86n1wBRx34jD4L3DsDxYoRuSRQoBUxJQJObn71dRJTuFJnvXT9ToYJrsYC989ScXSzFHQtcPR3OkSCBvD4hAS4u6WvzJELQdUKpcQpgzQRJ9cC3UCDFFlLOjSvL3z6yYUVSBN+7CTKzuo5kjg+MUnjLquPq7kDBgs4cwny+0yACTABJmAPBGgrgrQXv0jBfOl74n6fK0QzB/smINUdCWayepKIozMKOGQ/AVULOMJHPrijX+8l7nebOmelZlJNR4CT+9ypc1ewWM81AnIFHA025HdLp0NK5nB91Tbw1XbC1EzXCGivstCzugKO/Lc/e2egMIlL1xiYIn50LXC0glWscMY9cJIVjVwfJEuDUgEnN7+mlEEfV30Cjkzg3344FPuOntEcWELpkE84+ct/81v6HjgKtFJKLhAJCYniZCTyTScrG7mD0ooWbXClE6MMBao/cpmgky9plYoWByjuWUs3ZNlTWcBl/0DGOWACTIAJMAH7JEALubSYSsEZLCP2WQvKcvXe0N7C44u8lG7efYBVW/YaPSVUWUr8lqkEVC/gJIE06vVewmw6dc4Kjasdmc97dWwu7l6iI/rJF5ZOpqHTJenAEDo5UDqgQrIU0bHfZComkzudbNS6cR3kzOGLX/9aLA70MBTIGkgunfQObWQlNwsym+fw90XV8qWFFWjmkg3iYA0KnVs3Qv0alYRJn05Nop28uXMFomLZEuI5GuTkWuBIpNC+LHL3oFMspVMoaXXr178Wae6eUyrgjOWX9iEa2qxrCleyYl68djuTixmxosGfTmAi90ba7ExutOSXTq4sUiDXt3eG9BT//OH3OZoNwD99OgJeXp4ZDnShvU20t48ENFk8SSiSf7Wfr7fYO0d+3SQyyRJ25cYd4VoXExsn3AfoDpKrt+4KP3sKLOBMHXb4eSbABJgAE2ACTIAJMAGlBFQq4GqKE360A22ipCPfY2PjhYiL+e9wEPLRJtc5urTb389HiCo6aY7u0aCj4ROT0l0YM52WmJYm9tCR5YYuYKRNksYCuVs2qlMV1SqUQd7cucRmTxKJJObo4AzdfXokIOieOTq8hFwp6RjYi1dvaY7wFXnq0kqctKh9CuZb/bqKiwq1D0WgO846tXx5DxztJaN7QrQ3zeoTGrRJmS6tzLQH7q2+4vRN7QNijOVXHx9TuOqzwElx0sE1JHhpb9rzyGhxp4r2fW30HF1BQGKNroH4/vc5muzQXS7lSxcXRwhL1lG6z4X2IdKx8WTFpaOTyVea7ozTFuok9ts1qys2/Xp4eIh79cS9MAePaw6WYQFnrGfw70yACTABJsAEmAATYAKWIqA6AWepgnM8TMCeCfAeONvXDlnwExPS3XHNCZ5eXmJTNwcmwASYABNgAkzANAL8LZbHiwWcPE78FBOwKQEWcDbFLRJLiI/HnxPGmp3w22MnwMvb2+x4OAImwASYABNgAs5GgL/F8mqcBZw8TvwUE7ApARZw1sdNq3x0n6IUEhPiMWPiOGzfvh3+/v4mZyA6OhqtW7fGW2N+gKcXCziTAfILTMDBCJBbvxqs8bpjoYNVAxdHZQSs9S1WS3+UW10s4OSS4ueYgA0JsICzPmzad3p8/za4iCNrIPahHju4GyEhIYoFXP369VG7YXNxwSwHJsAEnJdAGlxQq3EbVYwFumOh89Yal9weCFjjW6ym/ii3DljAySXFzzEBGxJgAWd92PSROLH/X9Sr3wBu7u5ISIjH1Ik/mi3gRo35Al5sgbN+BXIKTMBOCaQkJ+NwyCHUbNxWNQJOeyy0U6ycLSchYOlvsdr6o9xqZgEnlxQ/xwRsSIAFnPVhSwKuYeMmYpJFfveTJ3xrtoD7YOzXvAfO+tXHKTABuyVAY8vB/ftUJ+CksdBuwXLGnIKApb/FauuPciuZBZxcUvwcE7AhARZw1ofNAs76jDkFJuCMBNQ2YdQdC52xzrjM9kOABZy8umABJ48TP8UEbEqABVw67lwBOTDuvdfx05S5Ge7ns0RlsICzBEWOgwk4LgFXV1dN4VJTU2UX1FkEnDYfgmMKI9kwbfig0vq2YRZNTkqNZXIkAVemRBEMH9gdH303xeS6M/YCCzhjhPh3JpANBBxJwBXIG4xWjWuLy9C9PD0RExuH2/cfY/ehE3jwODRLuizgsqHxcZJMgAmAJr71a1WFp4cHEpOSEHL8jGyB4gwCTpsPNRdTGdlbEzOnvu2tLFJ+1FomWws4mm+9UqwQVm7chZCT5zXV6eXpgW8+fBOenh6KF5FZwNlr7+B8MQErEXAUAUei7a3+3RBy4hz2hJzCi8goeHt5onK5ksgTHIjNu0KsKuDEddouLqBjsnUDW+Cs1Hg5WibgAARo8tu0fi3EJySKMWtvyHEWcFr1KvGJi08Qf/Xx9jKJkb01EXPq297Koi3glLbh7CyTrQXcyEE94O/ng+TkFPz2z1JN0RvWqoIGtSojf54g/DR1niIvIBZw2dmSOG0mkA0EHEXAffbOINy5/wiL1203SLF21fJoVr+GcJd8HhElLHPHz14Wz+uzwMl5ftn6HWhSt5oQidPmrsLdh09sJuBGfjAWXl5e2dBqOEkmwAQsRYAm9A3rVNcIuGNnLupdCNKXnjNY4Oh+u9pVKyAuPv4/AeeNg0dPyRa5lqonS8WjW99qLou2gNNuw2opU0JCAqZPnmCxA8WM9UcScHcePEatKuUwZ9kmzXxhzIj+YvG5e/tmGgtchdLF0a5ZPeQOCkRUdCxCTuo7IT4AACAASURBVKYvTkuhcIG86NmhOfLlCcLTsOc4ef4KurRpnMGFktKhOU9QYE6xqL1t31GcvnDN5KbMLpQmI+MXmID1CTiCgKMB7rN3B+GvBWtw7dY9vdCqlC+FPl1aYfayjbh++z5KlyiCN17rhEVr/sX5KzczCTi5z9+++xDzV21BRFQM3FxdkaJn/4q1LHDWbx2cAhNgArYm8NHn38LD01NWssYmjLIiseFDSg4xSUpMxKSfvrZhLjkpZyNg7p2s0onQxvojCbhb9x4iNTVNzDmWrt+BEkUKYkCPtpg2bxW+GD1ECDhfXx+Mfr0nFq7+F2cvXUeh/HkwtG8Xsei878hpsUVk3OjBOHj8nBBluYMC8GbfLggOzImPvp8qqo8WoEkAzlm2Efcfh6JEkQIY1q8r/lm8DrfuPTKpilnAmYSLH2YCtiHgCAKOBqZ3X++FCdPmIzT8hV5ww/t3w6PQcKzftl/ze7e2TZA3OBdmLF6XScDJff7vhWtx9ebdLCuLBZxt2jKnwgQcgQALuIy1yALOEVq1fZfBlgLu5t2Hwtr26ciB+O5/s4UV7XHoM5w4dzn9ILWp89CiYU34+fhg7opNGnDk6dOgZmVMmL4ANSqXRdc2jfH1pJma3xvVriIseNIhJmTV23/0NA6fvKB5pnfnluK/l2/YaVKFsIAzCRc/zARsQ8ARBJxkgft7wRpcNWCB+/TtAThw7CwOHj+rAdu4TlUxIP7858JMAk7u8+OnzkPY8wgWcLZprpwKE3B4AizgWMA5fCO3swLaWsBt3XNYeAA9ehoutmD8OHUePNzdNAKuR7umeBwajg07DmpIVSxTAgNfbY+x46cLgVe1fGlMnvlyH12F0iUwtG9njYD7+fN3hJux9omt5L57484DzFyy3qQaYAFnEi5+mAnYhoAjCDgiNfadgbh7/7HBPXBv9e+KJ6HPsE7HApcnOJdwKdDdA2fq81nVFlvgbNOWORUm4AgEWMCxgHOEdqymMmSHgCtbsije6tcVZy5eF9swNHMQssA1qCkOO5mzPKMFrn7NSvh5+kKNBe6bSTMhHZuma4EbN3oItuwOwYlzV8yuChZwZiPkCJiA5Qk4ioBLP4Wyq/AJ33f4tNiwS37ilcu9AhJpDx+H4rUurTBr6QaQCwM9P7QP7YHbpncPXNXypUx6ngWc5dsmx8gEnJEACzgWcM7Y7rOzzLYScDTfunn3AbbsPiyKW6p4YXEASWR0TIZFZD9fH4x6vSfmr9oq5icF8+XGm307Y/ehkxn2wJFX0Y4Dx8XeN7EHLihAY4EjD6NGdapi4aqtuP/oKVzd3EQ8aampYk+cKYEFnCm0+FkmYCMCjiLgCJd0DxwNinSnEt0DR5t1aePvwydhqFu9IprVq47AAH88exGFPSEncezMJUFarH7RBmKtI3xNfd5QlbEFzkaNmZNhAg5AgAUcCzgHaMaqKoKtBBwdYkILyORCqRt05yDkMtm2aV1xCmV0TFz6KZSHTmosbkUK5sOrHZohX+4gPAl7hlPnr6JL60aaQ0wo/hqVyqJZ/eoIzhUgXCnJZZPEIx2kYkpQhYCjCWDLRukXAft4eyI2LgGh4c/FUePSRM+UQvOz6QRKeidiUO5IfH0/t/i3n2sqBueJQGGvZJyO8cLy8Jx2hcre82dJWI4k4CzJxZJxWUvA8TUClqwljosJZA8B6Vj5hMRE4TXA1whkrAfpGgG6J48C3ZWnlmPq9bUo3fpWc1mk8qm1TLa+RiB7RhjzU7V7AUer9nTEJp0Os/fwKXFPFDXKV4oWFJsM6fhxZwhVfOPRLGccCngmIzkNeJHshtOxXtgX6YukNHFdsclBV8C1CohBUa8kzH4amCEuEnVnY71wKsbb5DQs+YKh/OmmkcMtBb2Do1DYMxk53VIx4UEQQpPdLZkVq8fFAs7qiGEtAScdXWz9EnAKTIAJWIsAzTPq16oqvAYSk5IQcvyM7DvOjB1bbq08K41XyTUC2nwoXVMZKc2rtd4zp76tlSdz41VrmWx9kbe5nLPrfbsXcHSP1K27D8W9DFmF8qWKp5s1gwOFi9aBo2ew/+gZ8Yq0CXHJ2m3i8rzgXDnx+OkzLFm/Xfi5UnB3c0O75vVQrUJpeHt74fHT9KPNpQuA6e6G9s3rY8eBY2jZsBb8fL1x+sJ1rNm6B51bN0LVCqWRkpKCf/ccQcjJ85qsFi2UD51bNUL+vMFITEwSFsN/9x6RfSEoRdQkZyxaB8Rg9bMcuBjriYQ0V+T1SEaDHHE4HOWDx0nKxImugHstOBLRKa7Y9MJfk39XpOHbImH48X4w4tNcs6udinT15U9fhvxdU1HFLwH3EtzxfoHnLOAAam/SnlqB7MuJMxAdG5et9ZndibOAy+4a4PSZgH0ToAmwFLRPjTOWa2cQcMRAmw/92xRGxhhmx+9K6zs78io3TTWWiQWcvNq1awEXcvwcxr47CH/OXy0u+TUUyEpHBaGTYei5/HmC8Fb/bti446C4BV0ScNdu3sXCNdsQF5+AAd3bitNk6JI+Cj3aN8UrRQuJy/XIyte4bjW0a1YXE6YtEJcBk4B7rXNLYQXcvCsEATn98eGwPuIm9k27DokNjXTJ8MBX2+GH3+ciIioaeYID8eGwvli2fgfOXLyGgJw5xIZHunGdhKCc4O2Siq+LhGNpWA6cidVvASvimYRXg6OQ1yNFWOeuxHliUVgAcrmnYFyhcHxyJw9SkG6lG5gnAk8T3fFvhF8GF8r+uSNQzS9BPJOc5oLFYTlxLtYLpbwT0TYwBtMe5xK/1fKLQ/OAWAS5pyIqxRWrwv1xJd4rU1HK+SSgc65o8VxiqguOx3hjw/N0YVjUMwldgqKR3yMZKWku2Bflg50RfghyTxEiraBnsniOyrHyWQ7Ep7rCUP6MMZxU7CkLOBZwepsJCzhjvYd/ZwJMQAkBZxFwStjwO0zAGAEWcMYIpf9u1wLu3KXrmS4Cpv1w7w7pKTLv7u6OvxeuQfMGNcUmwM27DmlK3apxbZQqVhh/LVyjEXB/zFmp2SRYoXRxDOrZQdzdQNJm/GcjsWDVFly4eksTx8fD+wkBuOvgCSHgenZsIZ6XjBkEz8PdXVw4LIWfPh2BRWv+FfF0b9cUdGrNwtVbNb/TRX/tmtbDT3/Mk1VDJITezBuBMXfyIO0/Eab74nv5n+F8nJcQQW5IE3vY7iR46BdwuSPwJMkd23QEHMXZJzgSkSmu2KxlgeuaKwoRKW7YE+mLSr4J6BUUhdmhASL+ALcUeLmm4WmSO2r4xaNFQCx+fRgksvdt4TCse+6PkzHe8HBJQwGPZNxNTH/n00LPsDr85W/5PFLEbyTgcrun4Hq8B7xd0zAkT4SwLpLl0VD+jEFkAZdOiC1wmVsKCzhjvYd/ZwJMQAkBFnBKqPE7TCCdAAs4eS3BrgWcZIHTdxEwbaD99ctRmD5vFXp1bIHAgBzChVEKrq4u4kS7iX8tynSXFD1DB6KMHPwqPv5uihBZ3348DL9MXyhOjZHCkF4dEBkdi9Vb9ggBRy6UdEO7FPp3a4OU1NQM7p3ffvimuNOKhB/dJVGyeGGx10YKlG8XF+CzCX/JqqGafvHokitac9AIvfRG3hco6ZUEN5c0rHmWA7X84/E0yQ3bXvgJsSUFvRY4IwKOrGraLpSfFQrHrKcBQqS9le+FsIrtjfQ1mvcvC4fhaLQ3DkT6Iib1pRtKy4AYFPdKwiydfXb6Iqzkk4D2uaIx8WGw+JkEpm7+jGWEBVw6IRZwmVsKCzhjvYd/ZwJMQAkBFnBKqPE7TCCdAAs4eS3BrgUc7UFL3wP3CEvXb89QIlcXF0z8T8DR3rdrt+9j+76jekutexkwPaQt4OjfEz4fifkr9Vjgzl3BrkPpFjj9Ai4tQ960BVyfLq3E0aLkQqk0ZGWBG5nvuThY5Gq8J9oFRqOsT6LYw0bWsqPRPooscNoCKbd7Mobne4EfH6SfUvlpwXBsfO6PC3GZXSZ1y1fIMwmtA2KFC2ZYcrq4vBjnhVeDopCSBqx9nm5V0w60d617cBRe8UoSlj2yjManuuDb/07JZAFH0l9ZYAFnWMDVq98Abu7uSEiIx9SJP8Lco4tHjfkCXl7Ze+CPslbCbzEBJmAJAinJyTgccgg1G7cVnkL2HqTFLGkstPf8cv4cm4Clv8Vq649ya9fuBZzuKZQvIqLg4uoqBNiIgd0xbe5K+Pp4o3fnlkKA3bzzQJQ9b+4g+Pp4ibsdjAk4Elk92jcTJ1vSqZYRkdHioj3dPXCZBFz3tkhJIQvcS3GpLeCKFMiLtwf1EAKO9silpqWJi/3oUr8rN+7KqiPaA/dNkXAs0bMHThJwIdE+mrhKeiViRP4XmPAgWIif74qE4Yt7ucU+Mgrv5H+Oa3GeBl0otQVc05y01y1FWPkovJX3hRCLJBDlhzRU90sQ1rNx9/KgcY5YvOKdhJl6LHC0/83TNQ0rw3MgLtUVZIHrFRylsT6ygGMBJ7/dGX+SJi3H92+Dy383uNC/jx3cbbaAq92wuSombcYJ8RNMgAkoJUBLkLUat1HFWKA7FiotM7/HBCxBwBrfYjX1R7kM7V7AUUEK5M2NVo1qoWTxQvD28kJsXDxCw1/gyKkLOH3hqhBGZUsWRZsmdZEvd/phG/T7zoPHhXDSdxmwsMAN6oGPv58qpm90CmV7OoWyYhlxnwntqVu/fT/uPngi4lPiQknvFS2YT1juChXIAzdXVzx7EYmDx8/i8MkLcusIJKTI9XB1eA5ciqNTKF3EXjHaG0fujElpwOU4L0SnuqKgZ5I4eXH8/WA8T3HDV4XDhODaH+mDij6JGJQ3Ajte+OkVcH1zRyIy+aUL5dv5not9dSTaKFT2TRAWtNlPA3A30R0BbqnwdEnLdEQ/nVxJe+LIUkdCrLxPgrhf7ou7eZDDLRWfFHqGVeE5cCrGS+yPk/bADcoTIQ48WRaeQzw3OE+kKKd0T51u/rIC6O6SbsGbUDRU7MsLTXYTh7OoJfA1AtavKdrLqu12nZgQjxkTx2H79u3w9395EqvcnERHR6N169Z4a8wP8GQLnFxs/BwTcFgCbm5uoG0T9h50x0J7zy/nz7EJWOtbrJb+KLd2VSHg5BbGkZ+r6huPpjnjxAmNJNjoHriTMV44GOWDnsHp7pMkpugQkp0R6S6UFMp4J6JncBT83VKFYPJ1TcOjxJeHmAzMHYlvtFwUpUNMvFxS8WXhcHx1LzdStQ5PqeMfh+Y5YxHonirSIiFGAo8EG93T9svDYHGQyht5I1BM7NMDwpPdsPl5ugslBfo77eujUyjpDru9kT7YHemHfB7J6J87Enk8UhCW5CZOrqS09OXPWF3T3jfd8P39YLzQ2iNoLI7s/J0FnO3pk9/9nxPGmp3w22MnwMubXSjNBskRMAEmwASYgNMR4G+xvCpnASePk9M9RYKRXB/nhgY4XdntocAs4GxfC7QKnZiQfpWGOcHTy0sVq+7mlJHfZQJMgAkwASZgDQL8LZZHlQWcPE5O9xRZ7shVk64L4GB7AizgbM+cU2QCTIAJMAEmwASYgBoIsIBTQy1xHjMRoANVSngnZfr78WhvrPrv0BU1Y2MBp+ba47wzASbABJgAE2ACTMB6BFjAWY8tx8wEFBNgAacYHb/ooATYrcZBK5aLxQSYgGICPC4qRqf6F1nAqb4KuQCOSIAFnCPWKpfJHAK8sd0cevwuE2ACjkiAx0VHrFV5ZWIBJ48TP8UEbEqABZxNcXNiKiAgTVTMveaBTwlVQWVzFpkAE5BFgMdFWZgc8iEWcA5ZrVwotRNgAWf9GrSE64m+EyctES+VXjdua8arfR+e9ckrS0G6GygkJETxPX3169fne/qU4ee3FBBQy71TfA+cgsq1k1d4XJRfEWrpj3JLxAJOLil+jgnYkAALOOvDtoTriT5rjiXipdLrxm2teJOTk3F8/za4IM360M1IgfJ57OBumCvgajdsDnd3dzNywq8yAeME0uCCWo3bqKKtqWUMME7d+Z7gcVFenaupP8orEcACTi4pfo4J2JAACzjrw7aEIHIUAXdi/7+oV78B3OxY2CQkxGPqxB/NFnCjxnwBLy++aN36Pcx5U0hJTsbhkEOo2bitagScGsYA521RhkvO46LxVqG2/mi8ROlPsICTS8rIc2VKFMHwgd3x0XdTLBQjR+PMBFjAWb/2WcClM6YVXJq8NWzcxK4nm1Rfkyd8a7aA+2Ds1/DyZgFn/R7mvClQnzq4f5/qBJy9jwHO26KyEHA8LhptFmrrj0YL9N8DTiXgaFL8SrFCWLlxF0JOntcw8vL0wDcfvglPTw/8NGUuwl9EyuWnec6aAu7LwmFYHpYDV+K9TM4Xv6BOAizgrF9vLOBYwFm/lXEKzkhAbRNGtSziOGNbMlZmXtgyRih9kVJNCyrGS5T+hFMJuJGDesDfzwfJySn47Z+lGkYNa1VBg1qVkT9PEH6aOs/uBNy4QmFYEa5fwLkiDalwkVvf/JxKCDiqgAvMmQOfjhyAX/5ciOcRUQZrY/zYt/HP4nW4efeh1WqMBZx5As7V1TVD3aSmpppcV6bEYcuJina+lJRLG4Sl4jKFlbGKsFZcjsZKad2pbcKoVMBZsh0ZarO2SENKW2l9G+tvur9bsky2HBezKqcly2TptqC2/ii3PTmdgLvz4DFqVSmHOcs24e7DJ4LTmBH9EXLiHLq3b6axwFUoXRztmtVD7qBAREXHIuTkOewJOaXhWrhAXvTs0Bz58gThadhznDx/BV3aNM7gQknpNKtfA0GBOfEiMgrb9h3F6QvX5NaNeG5g7ghU80tAchqEUNsX6YNWAbFY9cwfTXLGIcAtFZ/dzYMinknoGhSN/B7JiEpxxeYX/jgXm26xc0Ma2gbGoIZfPLxc03AlzhMrn+VAfGrGCZhJGeOHrUrAHgQc5aF44fxITkkFnVJGgmvv4VM4evqixcpeu2p5tG9eH9/9b7bF4pQbEQs45QKOPtb1a1WFp4eHiCQxKQkhx8/AlAm8qXHYaqKinS8l5dIVbxInc+IylZWxiZa5dac94bVE+Sg+S3G3FCtz8qO2CaMSAWcpzrZqq8a+C+bUt7G4DY0JSsdO7fhsNS5mdz2Z096M9cdXihbEsH5dMxWR0nR3d8OiNf/i5LkrplSzTZ51OgF3695DpKamIVdADixdvwMlihTEgB5tMW3eKnwxeogQcL6+Phj9ek8sXP0vzl66jkL582Bo3y7YfegE9h05DS9PT4wbPRgHj58Toix3UADe7NsFwYE58dH3U0XF0cSUBOCcZRtx/3EoShQpIBoIWRVu3XtkUuXqWuAmFXuKG/EemBcagPhUF/i5pWJMwWdYGpYTF+I8UdQzGW/le4Epj3PhaZI7uuSKQj6PFCwOy4mENBf0DIqCm0saFoUFmJQPfth2BOxBwJHFmixgW/ccFgWvUbks+ndvi+nzVuHGnQcWgeFsAq5r16746quvUL58eURFRWHt2rX45JNPEBERkYmnLU+hNHUPHH3Ymtavhbj4BJFvH28v7A05brKAMyUOW01UpLLFJyTC28vT5HLpTtaojObGZQne2qLLFO7GJm6WKJ8k4CwRl6VYmdMOjE0YLTJ4WjASpQLOUu3IUFEsVZdyUJlT33Lit0b/ozhtNS7KGQfM+R4YY2hOW1DSH8mvjebs9G37Y94q2ONVO04n4GhCSta2T0cOFKv+ZEV7HPoMJ85dxrj3XhculC0a1oSfjw/mrtikaVNN6lZDg5qVMWH6AjGR7dqmMb6eNFPze6PaVYQFTzrEhKx6+4+exuGTFzTP9O7cUvz38g07jbXVDL/rE3AzngRo9sS1yBmDQl7JWBD6UpC9FhyJZ8lu2B7hiwlFw/Dbo3QxRyHQLQVfFA7HmDt5AHa/NKkubPWwPQo4Kvv3Hw/DjgPHERr+3KCF2s3VFd3aNUWVciXh6emOmNh47Ak5iQPHzoqFE9HPpsxFQE5/DB/QHe5urkhMSk7vGxt34tT5q5j01Wj8OX81rt++L/5uzCJOz6/ZsgfVK5VFwXy58exFJFZu2pXlYompFrgCBQrg0aOMiy+mnELZuXNnIdhWrVqFWbNmoXjx4hg/fjwuX76Mhg0bCiundpAj4PLly4cnT9I9CeQG3XiVTN5cXFxQu2oFxMXHi2R9vL1x8OgpkwVcwzrVZceRkJCA6ZMnmH2IycgPxsLLy/B+YpooUL4k0WVqubTrwVJxSfGYw1t7AmkKd2MTN0dlpVt3x85czNRHDbFRMmGU23+t8Vx2jQHGymLJdi83LUv0ezl9xhJ9mdKx1bhoyzLpS0tfW5DbJ5X0xx7tm4l5x+SZyxATGydmys0a1EC96pWQw98XoeEvsGV3CC7fuKPJrpx5ytqte1G9UhkxTwl7FiEMRfny5BKeSDn9/XDlxl0sWbdNMyfKirtTCjiyKLzxWic8ehoOEmY/Tp0HD3c3jYDr0a4pHoeGY8OOgxp2FcuUwMBX22Ps+OlC4FUtXxqTZ77cR1ehdAkM7dtZI+B+/vwdMZnRdimiBkiWi5lL1hsbSzL8rk/ATXgQhNDkdEH2alAU6vjHITnt5V44VxfgeLQ3tr7ww3dFwoSlTju4u6ThxwfBiExxMykv/LBtCNiLgCOL9Zbdh+Hq4iIWLvp0bS0scG8P7G7QQl2vRkXQvlISYLHxCfD39RFi7cHj0AwCjg4LMmSB0xZwRQrmy9IiTjVCzz98HIq5yzfhWUSUWGCpXK4kvv99jsEKkyPgypUrh9deew29e/fGzZs3QSJMO5gi4E6cOCFW8erUqaOJgixyq1evRpcuXbBp08sFI3pAjoCbNm0a2rdvj+XLl2PZsmU4deqlm7ehgltCwCUlJmLST1/bpjPopGLuPXDZkmlOVPUEPvr8W3h4esoqh5IJo6yIrfSQEgGXnWOAlTCoOlpnHBfl9klT+2PjOlWFoPp91nI8CXsm2kWz+tXRpG51MX9/9CQMVSuWRv9ubfC/WcvF3EbuPOXBo6eYs3wTIqJiMKB7WxQumFfoAlqApi0Jo4f2xpGTF7Dz4HGj7dFpBVzZkkXxVr+uOHPxOuav2vJyYkkWuAY1xWEnBFkKJPTq16yEn6cv1Fjgvpk0U3P1ra4FbtzoIUKdn7CA36wxAUcWuLweKVganjNThdPlvOOLhmLCg2C8YLFmtEPYywP2IOAoD0UL5RcnONFCxLMXUcKqXKJowSwt1LT3s3WTOli2fgdu33+cYRFD2wInV8D16tQiy/QkAbd4zb+a/kYHEo15ewC++vUfsXqmLxgScKVKlRKCjf6XP39+IbBIIO3ZsydTNHIFnJ+fn3CTHDt2LH799dcM8URHR2Pq1Kn47LPPshSHhvJbt25d9OrVCz179kRiYqIQcpTfc+fO6S03Czh76eWcDzURkDtZpDKZOmHMbg4s4LK7BsxPnwWcYYam9MfypYpjSO+OmLV0A67evKuJ9LN3BuHAsTPYf/SM5m9kCIqMjsHKTbshd56ivZ+uUtlX8PprnYQ3X3RMrIi3e7umCMzpn0F/GCqZUwk4mpDevPtAWBQolCpeWBxAQhWgPbH08/XBqNd7Yv6qrTh/5aYwdb7ZtzN2HzqZYQ8cuYSROxntfRN74IICNBY4UvCN6lTFwlVbcf/RU7i6uYl40lJTxZ44U8J7+Z/heIw3Dkb5itdoD5y2BS7ALQUfFnyOZWE5cDnOU5h6C3kmIz7NRbhNds0VhSD3VKx85o+oFDf4u6aimFcSLsTxtQSm1IMtn7UHAae7B04q/7C+XbK0UFP7a1q/OqpWKI18uYOEG+OWXYdEu1ci4IylJwk4bZdL3XTkCLhXX31ViKjChQsL0bZixQrs3r07y2qXK+AKFSqEu3fvYvDgwVi4cGGGOMmyt23bNowYMUKRgNN+icQcCU8SdCQMf//9d/z9999Zxqu2yZszTlRsOfZwWvoJsIDLyIUtcPbVU5xxXJTbJ+UKuAJ5c4u5/4btBzJcNUY1PeGzkZi3YjMuXb+tqXg6uDBvcC5hlaP9co+fhhn03NM3T9F3/Vinlg1BhyT+tXCN0QbmVALO0ISUKIkJHx1i8t81AuQy2bZpXXEKZXRMXPoplIdOaixuZC59tUMzMUElEyvt2+nSupHmEBOKs0alssLsGpwrQFghyGWTxCO5pZkSKvokoEdwFLxd0rA/ygetA2IzCDiKq7BnEjrnikZBz/S9RA8T3bH2WQ48SnIXp1DSyZU1/eORwy1VnFJ5KsYLW174m5INftaGBOxZwPXq2CJLC7U2Jg93d7RvUR/VKpQWe051hRVZ6zq0aJDpFEptF0o56enumVMi4EaPHo1vv/0Wp0+fFlYsEnFPnz61iICTLHCffvopJk2alCHOmJgYTJkyRbEFTjuywMBAkFsmWeOaNWuGP/74w2i8LOBs2LE5KdUSkDtZpALKnTDaCwy1jQH2ws2e8sECznBtyOmPOfx88cGwPjh1/koGESbFOvadgTh47GwmCxy5Qq7avBtK5iks4OypB3FemICFCNiDgNO1WEtFK1owX5YWarJsJyQk4uGTMLHpv03TuuLqjh+mzM0k4Mq+UhRv9OmMb3+bKfbLSUFbkBlLT9/KlhIBR/F4eHiIPWW0761Dhw5CzK1cuVKIOd0DTOh5uRY4evb48eNiDxxZyaRAe9/WrFkj9tZt3rw5Q+uRsweOXggKCkK3bt2EaKPDUHbu3ClcKNevX4/Y2HS3DO3ALpQW6qQcjVMRYAGXsbrZAmdfzZ8FnHIBR2dgvDOkJyIiow26LtKVYLSViqxtj5+Go0r5UujfvY3YJ0feRUrmKSzg7KsPcW6YgEUI2IOAy8pirc9CTS7GFKpVLI3WjeuI+w9TUlPF4SLrtx8QrsS6lm46zZCu8SjzSlFxUAr5kp+6kPkUSmMWcb0WOC2Lur5KMXaIiaenJzp27CjEHIm6pUuXSfZc5QAAIABJREFUYvjw4VmKIfrRULydOnUSp1CSIJw7dy6KFi2KCRMm4NKlS0J4GRNa+uL98ssvxTUEu3btEqKN4ieLXlaBBZxFuihH4mQEWMCxgLPnJs8CTrmAowXmvt3aICkpWe9Js8fPXsbqzbvRvGFNcQolnZFBp1DSgYjaLpWmzlOEgBvQLYPnXseWDVBEuFCuNdrcnMqF0igNfoAJ2AkBexBwdoLCatkwJuC0E/b29kbVqlVx5MgRxQKOXiSLG90DV7FiRURGRgrBNWbMGPHfSgRctWrVcOPGDXGnnNxgCQEnXSNAR25ToPvSTD1uX/uYdjlx2Oq4bClfCYmJ4s5PU8ulXQ+WistUVlm1BWvE5YisdOtO7pHlxF6Oy5bc/mqL55S4UFpiDDBWNku2VblpWaIt26r/UTq2GhdtWSZ9aelrC3L7pNr6o7G2Kv3OAk4uKX6OCdiQAAs468M2RcAZyo0pLpSmlkiuC6W58SqZvNHHtH6tquLYYwqJSUkIOX7G5HvgTInDVhfWapdNSbl0BZxURnPisgRvKV/Wisuc8lHeLMXdUuUzJz9qmzBm1xhgbOyyVF0aS8eS7c9YWpYuk63GRWMCzpSx3BgjQwJOaRpq649y+bCAk0uKn2MCNiTAAs76sFnApTNWMnmTJjzataR956Xc2qPJjNw4bDlR0c6XknLpijjp3+bEZQorY/ytFZc55dNtU+bEZanyKW0HapswZucYYMu2akpa5rQ/U9KhZ81Jy5bjojERJ3csN8bH0O9K+7Xa+qNcPizg5JLi55iADQmwgLM+bBZw5gk469dQxhTsZaJi63JzeuojoLYJo1IBp76acbwc87hovE7V1h+Nlyj9CRZwcknxc0zAhgRYwFkfNgs4FnDWb2WcgjMSUNuEkQWcelspCzjjdae2/mi8RCzg5DLi55iAzQmwgLM+chZwGQVcvfoN4Obubn3wClNISIjH1Ik/wtzT1kaN+QJeXt4Kc8GvMQHjBFKSk3E45BBqNm4LdzvuU1JJJAFn72OAcfLO9wSPi8brXG390XiJWMDJZcTPMQGbE2ABZ33kLOBeCrjj+7fBBWnWh25GCjTJPHZwt9kCrnbD5qqYVJuBil+1AwJpcEGtxm1U0daob6lhDLCDarW7LPC4KK9K1NQf5ZWIXSjlcuLnmIBNCbCAsz5uumQ8MeHl5eFKUvT08gIdp60dLBEvxacbtzXjpQvG7T0kJsRjxsRxZgu4t8b8AE+2wNl7dTtE/tzc3DKND/ZYMBpb1DAG2CO77M4Tj4vya0At/VFuiXgPnFxS/BwTsCEBFnA2hM1JqYKAZDHdvn07/P39Tc5zdHQ0WrduDX1XP5gcGb/ABJgAE7ADAjwu2kElZFMWWMBlE3hOlglkRYAFHLcPJpCRgCVcXilGFnDcspgAE3AUAjwuOkpNml4OFnCmM+M3mIDVCbCAszpiTkBlBKzlQqoyDJxdJsAEmICGAI+LztsYWMA5b91zye2YAAs4O64czhoTYAJMgAkwASbABLKRAAu4bITPSTMBQwRYwHHbYAJMgAkwASbABJgAE9BHgAUctwsmYIcEWMDZYaU4aZZyBwWiX7fW8PPxQVxCApas3Y4nYc8y0ahQugS6tG4EF1cXPHoSjiXrtiEhMUk8JzcOKdK+XVujVtXy+HzCX0hITMyQlty4WjSsidpVyiM5JQXJySlYvXUv7j18oomre7umqFimBHIF5sSvfy3Go6dhmcrk6eGO13t3QuECeeDq6oovfvlbbyuQE1eugBwYN3oIHj55mc6c5Zvw7EVkhjjlxEUvNG9QA7WqlIeriwuehj/HknXbEZ/wkhWduDa4Z3vkyx2EpORkRMfEYsWm3Qh/HpEhvaDAnBjSq4M4LZHK+DTsGZZt2KkorrIli6JTy4aa+P39fBEVHYPf/lmaIc0RA7qBfiP3L2ojq7fsycCFHpbDXk482gkbaldy4zHWppxpiJDbTs1hIrcNm5OG7rt1qlXAa11aYfbSDbhw9ZYlo9bEVb5UcbRvXk/T53YfOoHjZy+bnVZWddK2aV3UqFRWjIcxsXGYPn+1WekZ6jN9urRGiSIFxJhDfXstjbuPnipKi+q/a5vGoHGFxvCHj0OxaO02TVylihfG2wO7Y922/dh35LSiNNT8Egs4Ndce591hCbCAc9iqVV3BRg7qgaOnL4oJRpXypUCT2P/NXJahHJ4eHvhi9GD8MWclQp+9QI/2TZGUlIwNOw6K5+TEIUVYuVxJlC9VDHWrV8TnP/+dScDJiatgvtwY2qczJkxbICYSNSqXRZM61fC/WS/zXaJIQTx7EYFRr/fCrKUb9Qo4N1dXFC9SAHHxCaA+aUjAyYmLBNzHw/sZjEMqv5y4yrxSFN3bNsHkmUuRmJSMVo1rI6e/nxBCUqDJT+nihXH5xh3xp4a1q6Bq+VKZJm5URri4aI6R79a2iXh+7b/7TI5Lt3G/2aczrt2+j72HT2X4ycvTU1Ovlcq+gnbN6uHXvxdneEYOeznxyGlXcuKR06ZU17nNyLCcdmpG9OJVuW3Y3HSk96mPDuzRTvxz58HjVhNwP3wyXIyVj0PDQWmOfWcQvpz4t+jL5gRDddKkbjW8UrQg5q/cgtS0NPj7+iA6Ns6cpGCoz9Ci2MWrt8StohVKF0eP9s3ww5S5itKisYhu6VmzNX0sokUfWoii4O3liREDuiMyOgbXb99nAQegY0caynXuNZJJPuNlSDJfosfSaBlOJ+wOOYn12/abEAs/ygQchwALOMepSzWXhD70n48ajC9+/ktzzfe3H76JKbOXI1zLclS1QinUqVYR/yxeJ4qbN3cu8XH97n+zxWRBThzSB3pY3y6YNm8Vxn/2diYLnNy4aLI9rF9XTPp7sZioNKpdBbRaO3fF5kzVQVYxQwJOe2InR3xlFZdcASelmVVczepXR57gXFixcZd4vHD+PBg5+FV8/vNfBptbkQJ5MbhXhywnU/Tt792phbC+0aq2oSAnLhKUJOq/mzwbMXHxBuOqXbU8aII5acYSvc/I5ZZVPDTxy6pdaSdsKB5T2pSa+7ypeZfTf0yN05x2pzQtmsAOH9AdG3ccQJc2jcWig7UscN+PeQtzlm3EzbsPUSAvjVVd8MPvc4S4skTQrZOv3n8D0+etQpiO9d0SaVEchvqMn483vvnoTXzywx+a74fcNMkC/82Hb+Kb32YhMSndk0M79O/WBqcuXBOLig+fhLKAYwEnt2nxc0zAugRYwFmXL8cuj0DhAnkxoEdbYcmSwvtDX8OGHQdw484Dzd+EoAjKhRWb0gWFh7u7EGBjvp+KQjLjoPfeeK2TmDhR3JO+Gp1JwMnND8VFLoZtm9ZDbFy8cBuiFW9ardUNciagckWEMQFHQvbRfy6U567cxI59Rw1ObLKKq2SxQujTpRV+n71CrEjThLNpvepCaGu7UepOeEhIaVvWpN/JdfLDYX0QGJBDuCnNWrpB4/6qr6XQ5MlQXNLzrRrVRqH8eTBvZWbRTM/069ZGiGoKMxatxePQzG659Jsx9nLiMdau5OZHbpuS17sc4yk5/cdSJZXT7pSm1ax+DXh5euDfvUeEx4A1BVzpEkUw6NV2wuLm4+2F2cs2CiuSpYJ2nZCl7IdP3sKmnYdAC20U9oacwumL18xOzljfIzfR/HmCQa7ipgYStkP7dMKpC1dRpkRR4Umxdc9hwYk8CcqXLoGl67eDXDYfPH6K/UfPmJqE6p9nF0rVVyEXwBEJsIBzxFpVX5nkCiYSD3nJImSGgCOXyUL5c2P1lr0CFAk4XUEiNz+0r4uE55xlmxAVEyvcB2tULIOpc1dmm4Ajl0By+yHhQ5M22p925eZd7D50Um/DMDYxblCrMupWq4jUtFScu3wDHVs2xGfj/9S7Wk1iityZaN8LiVlDgYTcq+2bIez5C4P5khvX5+8OFi6dkgunoTRrVSmH6hXL4J8l6/U+YkzASS8ZikdOu9JO2FA8prQp9fV05Tk21k6Vx5zxTbntTkl6JDJ6d24hFnnICkYCjvZUnb9yU0l0Wb5De1bfHtQDm3eF4Na9hyBr9tC+nfHLn4vEYpMlgnad+Hp74ftPhmPr7hBs339MLIiMfqM3Zixap9dtXEn6+vpMzcpl0bpJHUyds1LsuTM1kFfBB2/1xeI1/+LEuSsgCzjtd5s0Yyne7NsZf8xdKRaraF/rg8dsgSO+7EJpaivj55mAFQiwgLMCVI7SZAJyXRZpRZQmyjP+c6GkgzOGD+hmkgtl/+5tQZal1NRUkU86XORFRJSwBkkHf8jND62m5wkO1LgYkjvO+M9GCougrpuSnAmoXBEhJy6pEki00N48Kp++YEpcxQrlF+6R5LKqG4hFtYql8ef8NZn2E+pLl+Lq3bklJv61SHFcVI9Un/ryoy/Nnz8fKVylaK+hbpDLnt7TF4+cdqWbpr54TGlTJnc0Fb9gSjtVWkxT27Cp6TSoWRltmtYRB2VQyOHvh4SERGzZE4KQE+dNjS7L50mwDejRDuOnzdc89/6brwkL2bVb9yySlm6djB87QgjE5xFRIn5aPLp84y6OnLpgkfR0+x6NN+2b1RcLRhFR0YrSIPfLbz8eho+/n5qB0/Z9R8X4RHusKfj5+oj9uwePn8WW3YcVpaXWl9gCp9aa43w7NAEWcA5dvaoqHK1GHztzSfyPhFpzPYeYkOvRF6OGYOrcFQgNp0NMmiEpKSnDISbG4tCFos+Fkp6Rkx86CIXcd+iwFXJToglFu6b1MGH6S1dQKT2a7JALk/bpkEpFRFZx0USDBAoJVDqcYUD3tuIQA3LZ0heM5Sunvy8io2OFuyq5GpFL5sFjZzNERZZRWgmniZQh18rAnDnECjm5KNE+oM5tGiOHr0+G094oUjlxSYmTqxvtkSSXJ91AVkg69EZyZ6VDTKi9GBJ7hgScqfFI+dBtV3LjMaVNqaqDm5lZY+3UzOhNanfmpiW9b00XStqP+cWoQZj8zzJxemzuXAEgATfxr8WKxY5uuXXrpFenFsJKdej4OZBF7qPh/TB3+SbFp0Nm1WeEeGteH3/OX40XkcrEm1Se4f27Ye+RU7h8/Q7IAv6B4LRIjHtSYBfKUtj0n4cqW+AsNQJwPEzADAJWF3C//qM5zcmMbPKrTkAgT1Ag+mpdI7B03XaxX4lODqTVVWmFmk4f69yqkTiKno7kX7z25TUCcuPQxjnpy1F6T6GUG1eHFg1QpVxJ4TIoHVVPkxgp0KSGjvPO4e8rXJdI4Iz/Y36mco0Z0R9+vt7iBLTIqBhcv3UPi9dtz1DzcuIiAUDM6MwuV1cXsdq+btsBjcVRSb7ovDESg8fPXBLuUdohIIc/vvrgDYQ/i9BY3sjC8Pvs5RnKSNc/dGhRX7xKh5jcf/RU7JPTtobJjYvioMndNx8Oxc/TF2pW/LXzRYKRri3w8HAXLGgPH3GgNqPbprJib0o8WbUrU+Ix1qacYDgw2n8sySCrdmfJdHTjsqaAo7TI+k4nx1L7pz63Y/8xsdfL3GBoHCLR1qdrawTnChBJ0ELPoRPnFCeXVZ+ZOO5dMU5qu4OSmIvVY103lgESbbTXlxa/iBUtdpG7uHZwZgFHe3vpu2J3Am5PyMksT8EyVvH8OxNQMwFrC7gfp8y12olUaubOeWcCTIAJMAEmwASYgL0ToIUGOhBn1SrA29uO9sDR6oB0RLK9Q+T8MQFLE6AT4YoUzIc1a4Ddu4EpU5Tf70F5072qY+qcFeIIYw5MgAkwASbABJgAE2AC6iIw9p2BoP3mc+cCefMCHToonyda9B64Kzfu4K+Fa9VFk3PLBCxAgFwqvvvoTeGytWgRcOSI5QXcviOnNJdjWiDLHAUTYAJMgAkwASbABJiADQgUyBuMT94eIFL66iugaVOgZUs7EXApqan4etJMRceP2oAdJ8EErEagTIki4phhOi3vu29dEBMDTJyovGNSRnUtcHS58TeTZoL6GQcmwASYABNgAkyACTABdRDo0roRmjeoiadPgY8/BkaMABo0UD5PtKgFjhCu2brXKW9XV0fz4VxaiwCd4FaranncvAlMmgQUKACMG6e8Y+oKOGnj9OZdhzIdfGCtMnG8TIAJMAEmwASYABNgAuYRILfJd1/vCbpWZ+tWCE+t8eOBwoWVzxMtLuDoWON5KzbjwtVb5pWW32YCKiFAl5V2bNlA5HbdOmDlSqBfP6B9e+UdU1fAaaOgw4LWb9uPNJXw4WwyASbABJgAE2ACTMAZCbxStCCG9ukMXx9vxMcD33wD+Piku1HSqcQutAdHQVD0UlaTS/otJSVVXNB39tJ13Lr7MNMlrAryya8wAbsiQJcMly1ZDDUqlRX3VVG4dAlYvhy4cQP45Rcgf37lHVNfH0tLS+/sFOj47vNXboqFkrsPnojjejkwASbABJgAE2ACTIAJZC8Buk6jQpniqFi6BMqWKgZ3NzdERAALFwKHDqW7TzZunJ5HuxFwtD3H1fUluJi4eDx7HiHu60lN5Ulm9jYpTt1cAm5uruLOpLy5g0AiTgqnT6db30i8VawIfPKJeSsrhhZJHj9OF4baISExETGx8eLeKO5j5tYwv88EmAATYAJMgAkwAdMIkCHNw90d3t6eyOHnm+HlJ0+ABQuAU6eAMmXS98D5+dmZgKPsvHgBPHsGFC+efvkpBybgiASSkyH2vF28CJw8Cdy6BZQqBbz3HhAYaF7HNCTg6O+U5oULQOXK1McckSyXiQkwASbABJgAE2AC6iUQHQ1cuwZcuQKcPQvcvg3Urg289dZL8UalsxsLnISaTlmZPRvIkQPImTN9QuvpKVkl1FshnHPnJkAW5rg4CFP45cvA8+cAdVJybaxWDRg+HPD3f8lIacfMSsDRb3fuAEuXAo8epW+ClfqZm9tLN0vnrikuPRNgAkyACTABJsAEbEeAFvZpjhgeni7YoqLoRHEgICDdZbJ793QtpB2UzhMVm8d0jzjXh4cmu/fvp1sojh5Nt8rFxqYXhgMTUDMBEkq+vkC+fEDdukD58umXMupuRVXaMY0JOIldZGT6Cs+ZM8DDhxDXFyQlcR9Tc9vivDMBJsAEmAATYALqJEDzQDqkhBbWK1RInx8WKwa4v9x1k70CTp1YOddMgAkwASbABJgAE2ACTIAJMAH1ElBsgVNvkTnnTIAJMAEmwASYABNgAkyACTABdRJgAafOeuNcMwEmwASYABNgAkyACTABJuCEBFjAOWGlc5GZABNgAkyACTABJsAEmAATUCcBFnDqrDfONRNgAkyACTABJsAEmAATYAJOSIAFnBNWOheZCTABJsAEmAATYAJMgAkwAXUSYAGnznrjXDMBJsAEmAATYAJMgAkwASbghARYwDlhpXORmQATYAJMgAkwASbABJgAE1AnARZw6qw3zjUTYAJMgAkwASbABJgAE2ACTkiABZwTVjoXmQkwASbABJgAE2ACTIAJMAF1EmABp85641wzASbABJgAE2ACTIAJMAEm4IQEWMA5YaVzkZkAE2ACTIAJMAEmwASYABNQJwEWcOqsN841E2ACTIAJMAEmwASYABNgAk5IgAWcE1Y6F5kJMAEmwASYABNgAkyACTABdRJgAafOeuNcMwEmwASYABNgAkyACTABJuCEBBQLuLS0tDQn5MVFZgImEXBxceE+ZhIxfpgJMAEmwASYABNgAs5BQOk8kSeXztE+uJTZREBpx6Ts8iJJNlUaJ8sEmAATYAJMgAkwARsQUDpPZAFng8rhJJyXgNKOyQLOedsMl5wJMAEmwASYABNwDgJK54ks4JyjfXAps4mA0o7JAi6bKoyTZQIOQIB2OCQmJjpASbgITMA0Ap6enjBj54JpifHTTMACBJTOE1nAWQA+R8EEDBFQ2jFZwHGbYgJMQCmBhIQETJgwQenr/B4TUC2BUaNGIVeuXCziVFuDzpdxpfNEFnDO11a4xDYkoLRjsoCzYSVxUkzAwQhIAm7oyA/g6eXlYKXj4jCBzAQSExIwa/pkbN26FTt27ICfnx9jYgKqIKB0nsgCThXVy5lUKwGlHZMFnFprnPPNBLKfgCTg3v5gLLxYwGV/hXAOrE6A2vyfkydg5cqVOHz4MAs4qxPnBCxFQOk8kQWcpWqA42ECeggo7Zgs4Lg5MQEmoJQACzil5Pg9tRJgAafWmuN8K50nsoDjtsMErEhAacdkAWfFSuGomYCDE2AB5+AVzMXLRIAFHDcKtRJQOk9kAafWGud8q4KA0o7JAk4V1cuZZAJ2SYAFnF1WC2fKigRYwFkRLkdtVQJK54ks4KxaLRy5sxNQ2jFZwDl7y+HyMwHlBFjAKWfHb6qTAAs4ddYb5xp0YqoiLaboJZ5ccpNjAvIIKO2Y3Mfk8eWnmAATyEyABRy3CnslYM4dhVnd8cYCzl5rnPNljIDSeSILOGNk+XcmYAYBpR2TBZwZ0PlVJuDkBFjAOXkDsOPiS0JLSRazOlWVBZwSovyOPRBQOk9kAWcPtcd5cFgCSjsmCziHbRJcMCZgdQIs4KyOmBNQSEASWh988IHsKy7oncmTJ4MFnELo/JpdE1A6T2QBZ9fVyplTOwGlHZMFnNprnvPPBLKPAAu47GPPKWdNQBJwY8fKv6NQTntmCxy3PLUSUDpPZAGn1hrnfKuCgNKOyQJOFdXLmWQCdklAzoTXLjPOmXJ4AizgHL6KuYAmElA6T2QBZyJofpwJmEJAacdkAWcKZX6WCTABbQIs4Lg92CsBXQEXFhaGBQsWYMiQIciVKxcOHTqEW7duoV+/fnQ6nyiGnPbMFjh7rXHOlzECSueJLOCMkeXfmYAZBJR2TBZwZkDnV5mAkxOQM+F1ckRc/GwioM8Cd/78eYSEhKB169ZYt24dhg0bBl9fX00O5bRnFnDZVKGcrNkElM4TWcCZjZ4jYAKGCSjtmCzguFUxASaglICcCa9u3BfOn8PkX8bj4IG9iIyIQGBgLpQqUxa9+w5A/0FDkJqaiskTJ2Dpovl4/Ogh3D08UOKVkhj1/sfo3rO3iC6XjyvWbt6Bps1bZIi+cpniGPPZlxj0+lDx9/PnzmLM++/i7OmTyBkQgMFvvIWx477WvFOvekXcv3dX829KOy4uDguXr0HHzl0RHx+P4W8MxPmzZ3Dr5g18+MnnGPfN9xnSnPDDt1i2eAHCw8Pg4eGBatVr4psff0blKlWVYuX3LEDAkAvlhg0bcObMGQwePBhFihTJkJKc9swCzgKVw1FkCwGl80QWcNlSXZyosxBQ2jFZwDlLC+FyMgHLE5Az4dVOdd+e3ejdrQNeHzYCI955D8WKF0dSUhJCDh7A39OnYNHyNfjfrz9j/pyZQkRVqFhJiKgTx44iMSEBzVu1zlLAVSlbQgi4gUPeQFRUFGpVLoMBg97AJ198hRvXr6FXl/Z45/2PMHLU+3ph/DVtCn6d8AMu3rgPuguMyjdrxp+oXKUavvvqMzRr0RpffP1dhnevX7uKPHnzISAgAMnJyaA4pvz2C67cfqRxzbM8eY7RGAF9Ao4E+pw5cxAREYG2bduiYsWKLOCMgeTfHYaA0nkiCziHaQJcEHskoLRjsoCzx9rkPDEBdRAwVcDVrFQG9Ro0wrQZsw0WsM+rXVCseAn8POl3g88YssBpC7jFC+bh6y8+EULK1dVVxEXi6u9pU3Dq4nW9cdepWh4dO3fD1z+Mz/R7pzbNUb9h40wCTvtBSfCN+/QjXL/3FEHBweqoSAfMpT4Bt23bNmFhbdy4MebNmyescEFBQZrSy2nPbIFzwMbiJEVSOk9kAeckDYSLmT0ElHZMFnDZU1+cKhNwBAJyJrxSOclSVbtKOb2uj9os/vj9N/zy03fCZZLEXtXqNZAzZ84MuOQIuM/GfIBrVy5j5fotmnePhBxCuxaNcC80Ev7+/hni3Lt7F3p0aoNTF2+gaLFiJgm4f7dswluvDxAuoXQgxsjRH+CHCb86QhWrtgy6Au7q1avYuXOn2Pfm7u6Oixcv4sCBAxg6dCjc3NxEOeW0ZxZwqm0STp9xpfNEFnBO33QYgDUJKO2YLOCsWSscNxNwbAJyJrwSgcOHDqJ9y8Y4euYSSpcpK/5M++E6tGqSPnmOj8fqjdvQoFFjrFq+FCuXL8GxIyF48fw5GjVphp9/m4Ky5cqLZ+UIuHeHD0VsbAxmL1iqqYQrly+B9r2Ri2SBggUzVM6gvj3FBH7Z6g16K02OBY7yunjhPBQqVBhde/R07Mq389LxNQJ2XkGcPZsTUDpPZAFn86riBJ2JgNKOyQLOmVoJl5UJWJaAKQJOssCt2bgNzVq2ypCRlJQU5Pb3wMZte9CwcbqgkwKJLjqI5M7tWzhz+ab4c54cnli6egNatm6b4dkKJQvjq+/Go0//gSAL3NXLl7Bqw1bNM4YscI8ePkSVssWxeOU6tG7bXrGA+288RbF8gdi8cz8qVa5iWeAcm2wCkoD74IMP4OXlJes9emfy5Ml4+wPDl3+zBU4WSn7IDgkonSeygLPDyuQsOQ4BpR2TBZzjtAEuCROwNQFTBBzljfbA1a3fENP/mZMhq3T4B4kyfQKOHtyyaQP69eyKW4+eITAwULhivjniHQwfOUoTDx1aUjx/LmzesQ916zfAkoXz8dXnY3D51kONi9yff/yOf/78AycvXMuQ/vjvv8HyJQsN7o2jh+VY4Og5KkvRvAH4a/YCdOnWw9ZVwun9R0ASWkqAsIBTQo3fsXcCSueJLODsvWY5f6omoLRjsoBTdbVz5plAthIwVcDRPrPXunfUnEJZpGhRkPXt0P596NaxtRBwp04eR5my5VC7Tj3kCgrCvbt38cG7w/Hw4QMcOn5WlPf3Sb9g5t/TMX/JSrFHjlwXv/j0I5w5fRL7Dp8Ue5yio6M1p1B+/Nk4cQrla906YuR7H2Y4hZIEV+UyxfD2u+9j9IdjMvGkMqalpaF7xzao36CRONGSDkU+121OAAAgAElEQVShUyopkCjs2bsv8uTNi7DQUHz/9RdYt2Yljp6+hLz58mVr/Thz4lRniYmJihBQ3UqXe+tGwBY4RUj5JTsgoHSe+H/2rgJaqqqNblokBARFJES6u7tBulM6pPvR3d3dINLd3d0disIvLSHd8K/9Pe44b97kfTPzJs63lgth7j33nH1yn68UgfOAzlNV8F0E9E5MReB8d0yolikEXI2AowSO9WFuNuM8cDFjxkKSZMnxc/1GkueN0SMXLZiLK1cuSeqAr76KgXwFC6NHnwEg4aMwHDyJ069zZ+HWzb8RLVp05MlXAH0GDsX38eMbmk0fu05tW+LMqROIFj06GjZpji49egeBZc2qFfil4c/iF0fCaCrMLcdccTzQkxRQ6JO3bstO+f/qlcri1IljePH8uXwjc9bsCOjeGxkzZXY1/Kr8UEBAEbhQAF190ikI6D0nKgLnFPhVIQoB8wjonZiKwKkRpRBQCOhFQA+B0/st9Z5CwBMQUATOE3pB1UEPAnrPiYrA6UFbvaMQsBMBvRNTETg7AVaPKQQUAsEQUARODQp/Q0AROH/rcd9pr95zoiJwvjMGVEs8EAG9E1MROA/sTFUlhYCXIKAInJd0lKqm0xBQBM5pUKqC3IyA3nOiInBu7ij1Of9CQO/EVATOv8aJaq1CwJkIKALnTDRVWd6AgCJw3tBLqo7mENB7TlQETo0nhYALEdA7MRWBc2GnqKIVAj6OgCJwPt7BqnnBEFAETg0Kb0VA7zlREThv7XFVb69AQO/EVATOK7pXVVIh4JEIKALnkd2iKuVCBBSBcyG4qmiXIqD3nKgInEu7RRXu7wjonZiKwPn7yFHtVwjoR0AROP3YqTe9EwFF4Lyz31StwVQouriYrpfU4VINOYWAfQjonZhqjtmHr3pKIaAQCI6AInBqVPgbAorA+VuP+0579Z4TFYHznTGgWuKBCOidmIrAeWBnqiopBLwEAUXgvKSjVDWdhoAicE6DUhXkZgT0nhMVgXNzR6nP+RcCeiemInD+NU5UaxUCzkRAEThnoqnK8gYEFIHzhl5SdTSHgN5zoiJwajwpBFyIgN6JqQicCztFFa0Q8HEEFIHz8Q5WzQuGgCJwalB4KwJ6z4mKwHlrj6t6ewUCeiempxC4d+/eYfDgwVi8eDHChQuHiBEjIlGiROjbty9Sp06Njh07YufOnQgfPjz4bOPGjdG+fXvs3r1b/jx16pShn65fv45MmTLh8ePH8m9HjhxB06ZN8fr1a8SPHx8LFixAvHjx8PDhQxQtWtTw3suXL/Hnn3/in3/+QYwYMXD//n3UrVtX/i1SpEiYPHky8uXLZ3ief584caLUKWzYsPIdPqfEexDwxHHHeTB//nz8/vvvWLlyJcqXLx9kjDZq1AjHjx+XMcdnK1euHARwd7Zp1qxZmDdvHpq372rX2I8eOSye3vsXiB7d5YMk3N49CPPmNd4XKyHfCnP9OqLlzISndwPXBbvlyRNEmj4Fbzp3NbwSuXkTvK1ZBx/yF7C7GOMHIzeujw8ZM+Ftq7a63rfnpYgTxuJdtZr49O23Nh+P1L8PPiZPgXc1atl81t8fcDaBc+d8tXfvs7YGVapUCX/99ZdhGJw9exZr1qxBmTJl3DY0vA0zruUtWrSQs8X79+/Ru3dvVKtWzW14aR/Se05UBM7tXaU+6E8I6J2YxOjTp0+fQhurOnXqgARqzpw5+Oqrr6Q6O3bsEJL1v//9T8jRkiVL5ND69u1b/PHHH0LsbBG4jx8/Inny5OBBs0CBAhg1apSUtXTp0mBN5m979+6VzYjSsGFD/PDDD7LY8sBcsWJFkBySYPKZESNGYNOmTYgWLZrUM2bMmFI/Jd6DgCeOu2PHjiF27Ngy/ng5Ua5cOQOg/fv3lzE4e/Zs+TNHjhy4dOkSYsWKZXjGnW1iXVOlSuUYgSOB+jzHXTlSIg3oizBPn+D1iDHyGb0ETu971toWuUmDQALXso1+CD58AMKFs/h+tBSJ8WLZanxMn0H/N9SbwRBwNoFz53y1d++ztgYZA3LixAmULFkSd+7ckYtMd4m3YZYnTx7w4o1r+oMHD5A1a1YcPHhQLpLdKXrPiYrAubOX1Lf8DgG9E9MTCBxvp6gxu3nzpmi+TIWHWJK2SZMmBfvNFoHjRkQtGg+5lGfPnuGbb77BkydPRMtnLDyIDhs2zHBgJjG7du2aPE/hYXnIkCEoXLgw8ubNK9pBYw2e3w06L2+wp447DdZChQoFI3Bp06YV8pY9e3Z5rHr16ihevLgcDijubhM12h06dHCIwL3pNwjh161G2H/+wevuvfGubn2EX7kcEWfPwMv1WwKb/+EDoqX8ES/Wb0GYu3cRuX0rfMiUBeFOn8SnSJHwauosAzGJOHoEIv46DwgbFh/Spser8ZMR9q8/EaV8KSnnU7zv8a5CZbyt9TOi5ciINy3bIvym9Qjz5Alejx6P9yVKySfDHT+GL3p2BZ49RZgPH/A6oDveV6qCL8uWRPid2/ExbTp8ihABL/YfRZRiBfGmTXu8L1seePIEkbt0RLhjR4RQsZ6vps1CuJ078EX/Xgjz+jXw9i3etOmAd/UbyreEwGXIaFYDRy3lmy49EGHzBuDFC7zp0cegGZPfuvdG+C0b8aFAIbxp1wmRWzVD2Gt/8CYOb5q3xrvGTRFpUH9EGjYIH39IDESOjJcz5+FjylSI1K8Xwu/ZhTBv3+JDsuR4NXEaECMGjDWCJL5hr1xCmFevEPbPa/j0bVy8WLQciBnTy2e8c6rvTALn7vlq795nbQ0yRrF58+aIHDkyRo8e7Rxw7SjFGzHjWYL1jhs3rrSQVhUFCxaU9d2dovecqAicO3tJfcvvENA7MQlUaGvgqA2jycbp06fN9tvFixflli9q1KhCnIoUKYKqVauKtssWgVuxYgVmzJiBzZs3G8r+9ttvRQtH7ZomvA2jKdqtW7ekXGrUvv/+ezG71ISH5VKlSqF+/fr4+uuv0alTJ2zYsAHc0EkSW7du7Xfjzpsb7InjzhhPcwQuevTochDgGKZ06dIFX3zxBfr16yd/d3ebSOCaNGniEIF7PWwU3rZpj7BXryBqnmyBJpWfPiFa6qRC2D4mS47wa1Yh4rTJeLlxG8Lt2Y0oJQvjxaYd+FCwEMKvWIYv+vfG8zOXEH7LJnzRpSOe7z0sZplftGwmJOr1+MmINLBfoAZueODhUjRpqX7Ey8Ur8L58RYTftgVfdGyL52cvA//+i6j8xuqN+BQ3LsI8eICoubIElvv2rRA/Y9PLKMULBRK4MuWEjH2KEgWvx04M/M6DB/gUO7aUKaaiYcMizKNHiJozM57vPohP8eLZJnDdeuFN734I89dfiJonK54fPoVPCRNCCFzv/njTrad8K3Kd6vj4Y1K86T8IYf75R+r8ctFyfMiWHaKBW74GH9Oll2cjDRsMfPxoeDfS4AEIc/+e1NtYI0gCR0L87PBJIW2Rf64pZNnYhNSb531I6+5MAufu+WrP3mdrDdJ+f/XqlWiQ9u/fjzRp0oQUVrvf90bMSNZKly6Nzp07i0sGL4Nr166NsWPH2t1uZzyo95yoCJwz0FdlKAQsIKB3YrI4TyNw1HpVqVIF3CBy584tGgfavHOjINGiHX/SpEmxfv167NmzB+3atQvmA5c5c2Y8evQI9hI4ajDixImDoUOHCsK2CBzNPGvUqIGpU6fKd2ieSe0dF2kl3oGA6UHAE8adrcOTowTO1W3SQ+Ce3biLT5+12tG+i4XnJ84LqaEmLeytm3g9ahyilCqKNy1ai4aLBO7LZg3x7PKfBniix42JZ8fOItLEsfgU/Su86dFbfgt75jS+rFUVzy/8DrMmlJlS4+njl4HlPHmC6N/HxtPn7xB+80Z8Wa9WoMbqs4R5/BgvZ8zFx0Q/BPOdMyZw0RJ8g+f7juKT0YWQ1OXaH/iiewDC/vE7QD/Za3/g5aIVeF+suE0C9+zqDXxKkEBq8mW1iqJBfFerjhC4Z9duCl6UaN/HxvNDJ4XcUb7o1A6fYsfBm649ghG4KHmzI8zTp6KRE3n7Vtr7ctX6YAQuzD/3hQRTIk4aj3Dnz+HVlBneMbFdXEtXEjhXz1dzBM5077O1Bmm/cx+mVczhw4ddjHjQ4r1x3b5x44b48bN/6dvPSzdeENNtw52i95yoCJw7e0l9y+8Q0DsxCVRoEzhLJhEMjrB69WqsWrUqSH/eu3cP3333nRCn27dvi3aOfnKa0F+NNvKXL1+GORNKErWnT58aTCifP38uN4l8j/5ymlDjxwVX03YYm1CmS5cOEyZMEDMISkBAgARxGDBggN+NPW9tsKeOOw1PSyaU9OfkWKTQEZ7jn74VFHe3iYcQHkz0BjGJFj8Onh88EUhAHj9GtMxp8HL5WiFhQtjChDFL4IT4aQQuajS86dlH2m+TwBkHMXn+HNHjRMfTVx8RftMG0VC92H0g2HA25wMXjMDtPYJPif8jfyyEJPTdT2XwtnU7KZPasdc9+hi0dtZMKJ9duW4gZV9WrYB3lariXc3aQuCMg8AIgSN+iRLJN4TAxfkGb7p0N0vgXvcbjA9F/gvcpDU2iAaOmssn/xp8ByNOmYhwJ0/g1Yw53jrVnVpvZxI4d89Xe/c+a2uQ9hv3PmqRqIF3p3gzZhpOXLNp0dOgQQN3Qge950RF4NzaTepj/oaA3olJnEKbwLEO3AhorkhtmxbEhLd727dvFx8batxI2ig0W6QZIyM6UTOXMmVKiQZJ80YGLWnWrBm+/PJLjBs3Tv5OUjZz5kwhWyNHjsTRo0eDBDHRIukxgImxcHGlmWWfPn2ECDKICW/SGMSEvnD//vuvaN2oKaQGjuZsphEB/W0celt7PXHcGR+Q6CNhHIWSppIMXsJgP4wElzNnzmBBTNzZJl56OBzExCgKZRACRwLSujkibNqAN7+0xNtOXQQKgwnl5p34UKCg+MuJCeXpiwi/dfN/JpTRosn7YkI5diIijh8TqDmaPlvKCUbEjAicgTzOWoAPhYvI8ySDH1OnQZhnzxAtcTw8ffAMiBBBfgtiQvlLY/HLExPKMGEMJpRRcmfFG/rRVaiEcPv2ihnoyyUr7SJw9HsjKWWdxYTys5bNlMBF/rlGoAllv4GBJpS5swaaUGbNhqjZMuDV6An4kC+/1JkENdzB/WJCKlq4ly8R9sZ1fEyVOohG0FRzqQhc0FXNmQTOU/c+a2sQf2MQMfqtM3gJLzrdLe5c4+w9L1jDjBGtGZiK7hlbtmyRM8qVK1fsitzrTGz1nhMVgXNmL6iyFAImCOidmJ5C4EjEBg0aJGkEGM2KER3pcE3NFjcLOkmT4DHwCB2CaepI80oKwxiT5NHskYSNh1qaJmgbC008uGDyfWoMaPrBPzVhhCimGahXr14QVLno/vzzz3JQpnaNJJFEjcJNnGWS2IUJE0ZMPhnURIl3IeCJ427gwIGYNm2aRCvjGGaQAKbJoN8lI7VS20bixIsEzhmOPWNxZ5t4+cGUB3Zr4L4MF+hL9jmNgJgfHjhu0DaFPXkCUQvnxbM/b+HT58iaJHCRO7QOGsRkysz/gpiMGYmIC+YGBjFJl0GCmCBaNCE/UWpUEv86LYhJ1JyZ8OzOo0C4SOC++QpPX36Qv4Y9fQqRu3ZCmEcPgXfv8DFhIrxcthqIGBGRWzRFuAP78ClatMAgJkY+cHj2DJE7t0e4o4eF4L3Pmh2vJ01DuB3bEbltCyBKFHxIn1FMKd90DPiPwFmIQil+bt16IcLGdYFBTHr2xbvqNaWO0U3wC3P/PiK3aR5opskgJi3a4F2jQI1IhDmzEGn0cODLLwODmKROI8FNIqxdJURT1rFOXaXsYBo4I9/BiFMnBWrgPhNh75rhzq+tswmcO+ervXuftTWIiPbo0UOsX3iRFBribZhxneSZhWs2zx603mEUbXeL3nOiInDu7in1Pb9CQO/EJEieoIHzq85SjVUI+AgCzk7kHXHMSIS7eiWIv5UQuID2eH7kv1yPPgKf2Wa4M1eeL+PoqrY5m8C5qp6qXIWAKQJ6z4mKwKmxpBBwIQJ6J6YicC7sFFW0QsDHEXAmgYuaKY2YP75Yu9kQpIPwMSG3EDhGRfQDMdWy+UGTvaqJisB5VXepyhohoPecqAicGkYKARcioHdiKgLnwk5RRSsEfBwBZxI4H4dKNc9HEFAEzkc60g+bofecqAicHw4W1WT3IaB3YnoDgWOQEPq+MZk3/YFMhf/OyJS0i6efEJ8195zpe+/fv5f3+H6ECBHE747v8v/1Cv0C6DP34UOgX40Sz0OAEUcZEIdjhE7lloRjg/6Xxn2ZLFkyMJR/lChRLL7H5+/evSvj6tOnTy4FgP6iHPMcs6aJ6c19mPViZFVr45OpARh5lYGA6N9pTvj+uXPn8PjxYzD4j70+cC4FQxWuEHADAhqBW7hwoaS24b6hZ8/g2sDoyVyP7H3/yZMn4uvta8K1i75hltYbR9rLPGvch+kP725h+iKuyUwTYEnox0yfep4/+Kc7Re85URE4d/aS+pbfIaB3YhIoT/aBY+AREqILFy5In5oGbOBiyMiQ2oF0zZo1SJAggSTWtibchPkeD9kUJvDmZsz0AylSpHB4/HAzZnhjS8nIHS5QveAyBNjXHDetWrUyRDw1/RhJzsmTJ4MRsKtXr8o7lpK2m44rlzXic8GMhMqorJwXWbJksfo5klEGQ7FFKv/++2+ZF23bthUSZyqckwcOHJCDJOcdMVEEztU9rcr3FAQ0AsfAFAwFz0BX6dMHJkt3RHgxSRLI6ITFixc3O9dMy+P+liRJEkc+4zXPkrwxojQDfegREjZeKjG6Y2jJiRMnkD9/fov5YLl2MorwoUOHhLCy390pes+JisC5s5fUt/wOAb0Tk0B5KoHjRknyxpuqGTNmgLdbxpEieXjlYdz4QLpr1y651WJUSUvy4sULufk01kJQk7Bx40ZJT+AogePGwYMxD/1KPB8BHpyYI7Br167BCJytQwAjQz579kxSRpiKuXHlajRIRFeuXImqVataJHBsEyO1kmjZIyyTh6BevXoFO1RSA8CLDraVogicPYiqZ3wJAY3AMeUNL064J+klcExvQ203yUvhwoVFm25Nli9fLiTHV4VrFfdfRzVTPCMw2jQ1b6EpvGjmRVq5cuWCVYOXYtQO8k+mSCpbtizKlCnj1urqPScqAufWblIf8zcE9E5MTyVw2mLHhZmyYsUKyTfFxMUkbMwBxzD/pmKLwJGokfSZil4Cx3py46BpmhLvQMASgeNYO3LkiNnxobXMEoGzNK5cjYgtAkezYrbJkYONJQJHk9KDBw+KqbImisC5uodV+Z6GgCsInNbG7NmzS+5RS+LrBI7t5pry448/Sroge4TrOS+VuAaHtlgicNxbNPLG/1+0aBGKFCmiNHCh3WHq+woBT0DAlwgcSRE1b8YHxXXr1onpCG88eRil9s2cWCJwJH0kfCR+5kQPgaMWYt++fXj69KknDAFVBzsRMEfg7D0EmBI4W+PKzirpfswageNvPNhYmiuWPmqOwFkyKVUETnfXqRe9FAFXEjiuJ9RAZciQwaw/mD8QOA4LauLixo0ruWCtCfdtrnFcvz1BzBE4rpE8z9B8UiOoy5YtE1NLZULpCb2m6qAQCGUEnEnguHjyEMc/6QNDp1xNuFCSYNHEwdhRl7dKfIdBIRjgQa8zMkkbFzvNN037Lglc8uTJkSlTJvmOJTFH4NiOmzdvWiVajhI4+v9w4+BmrsS7EDAlcI4cAowJHMcVtbk0KwwtsUTg2CZeLmiHBkfqZ0zgOMet+ZUoAucIsupZX0DAlQROw4fkJW/evMGCLPkLgdNwoBYuYcKEZs8TvMilL5knBQwzJXDcI+hrr5mcO0LguHZzrJkGZdPOWjxj8aylBeIi+ed3+E2u2+aCWuk9JyoTSl9YuVQbPBYBvROTDTL2gdNuizRiwkUiceLEErGPJlTUYPHfuFjwdoz/8cBHx1zNF43kju846ozMhYnkzRwp2rlzp2xotiLtmRI4lskF1NYNnSMEjkEeaJYWGlGuPHYAelHFjAkcNzxHDgEagevYsaNd48rVsJgjcCSVNOvVe7DRCFy3bt2EvJkzOdbapQicq3tYle9pCLiDwLHNUaNGRdGiRYPsef5G4IgDI3Qy+q8xUfHUgGHGBI7nIZ4VTC10uGba0sBxn6FbhnbWom8kIydrQd20swexoWUSseEZzPhym8TX1AxV7zlREThPW4VUfXwKAb0T05TAcdGgdsk4MAhJExcDRrEzlaRJk0pAEGONGRcdRtaic7a9Yo28sQyWb4u88TljAsfFjuTN2BTTUn3sIXDE5NKlSzh//ry9zVLPeSACGoGrWbOmBDNxRLixchzwYGXPuHKkbD3PGhM4BvlhoJIzZ87oKcrwDsukySQvTGxpFxWBCxHU6mUvRMBdBI7Q0PqlWLFihoO4PxI4bYgweAuJiicHDNMIHAOUUENozi/PFoHjWcc04JR2kc4yja0q+O9MY8FLc57djM9txIpxA4ytofSeEz2WwHEz19isxnY9bU2xlo/H0+qq6hM6COidmKYEjmTM9MaI84Kh+XmbZLxA8N9J7PiOaWhy5srib/bKnTt3LOa3Ydn2mmRqBK5JkyZCOO01cbSHwHGBZM4rJd6NANd8jmWa5DoqJHAZM2aUXHCeIMYEjvON4z+kwjJJTpkuwZYoAmcLIfW7ryHgTgJH7GjJUr58eSFz/kzgSGy4JjFUv6eKRuAYoITWRObEFoGjVYixRRPL4PmHaQd4TjK1rKAZJS2kuDeZSurUqYOY4eo9J3osgevbt68cQCmM/kPgPE2GDRuGOHHieFq1VH08CAG9E9OUwGmBPowJGe2pSeBotmAqPARz/pjeCtG00pExS3LExc+c1u7YsWMyN+0pz1gDx1xVXPTtec8eAqfl7aFpi7UE0B40LFRVzCDAfuRYY6QzHggcMfXlJkntVOnSpW062LsDfGMClzZtWskrxc3ckTaZ1pNlMtokw2HzdteaKALnjl5W3/AkBNxJ4PgtapwYlr5gwYJ+S+BIXLZu3YqSJUvKOYTrkr2Xuu4cO8YauC1btiB+/PjBPm+LwPHyjBo400txWjtRA2dsJkkMaF5JDRx/M36HhN805YTec6JHEziaWVHot2OJNbtzEBh/ix2kCFxooe8939U7MU0JHBcAEjLmuqLQxprkiQuEFipd01TzRogLKTcZXnxoJmXWHI+tIUrCRSJoSuK4YDFvClMIJEqUyGqnGBM41odJjtkGW+/ZQ+D4YfoD0X6dBNU4uIv3jBRVU82EskGDBti8ebNsgBwj9ggJHKM6pkmTRkiStZDf9pQX0mdMfeB4EbJ06VKH2mSOwNG0lCaU3A+tmUIrAhfSHlTvexsC7iJw2kVT1qxZRQPHPdgfNXAMv8+ATMyTx6iNXJNWr14t5MjTLlKNfeAYOI2Xz/TfMxZbBI7P0nSd738+n4n/G/cp0yBv1L7xbMMzmWayyf8nLtybeJlnwid0cTFdL5keLl0x0amBUwTOFciqMt2JgLMInLZgUHNFMke/M+NFkr5qXER44DUmMDRD5jtcPPiO3tsxmihygTPVmnHBpmlB7ty55aLFkpgGMWEbWCZv8KihsCT2Eji+zwP8kiVLJNKT6QLpzj5X39KHgHEQE45TEh6OZZqq2xItiElAQIAcKkjo06VLZ+s1l/1uLogJTaA5PjlH7WmTOQKnJfLmTfCOHTvksGRuTisC57KuVQV7KALuIHA8wHPuUeNUqFAhwx7sTwSOZwoGUeJ/FStWBHPkaWuQMy6qXDG8TKNQ0lqDfvPGF332EDjWTc9Zi2cwlk+yb47c6j0negWBa9OmjdyseppwI9Z7IPa0tqj6uAYBvRPTHZckjrSYhGvPnj3yCgOhaMI0AjR540GS2j7myTEnlvLAcRPgbRhvM80tbI4QOH6XmsJVq1YJqfQUfyhHcPbnZ03TCPBAtnbtWtEwx4gRwyo0pnngGNDm6NGjFseVq3G2lEaAbVqzZo1cNthqkzUCRwLIG++VK1ciXrx4wUwzFYFzdQ+r8j0NAVcSOO5/jPbMC8fq1auLv62x+AuB4yUxowMTizp16gTTYhETXlQ5cvnmjnFkLg8c/40XfVw/KfYSOFfUV+850SsIXPv27ZE+fXpX4KbKVAi4FAG9E9PTCNzn+sitPzV5sWLFEty0RN40o7x48aL46GTLli3YgdISgWMZNA2lyRxJnGlES0cJnLYQ0y6fJh72+Nm5dACowu1GwFwib972cuxwbFkzGTQlcNq42rRpk4xHeyKl2l1ROx60lsibbeI8ojmkIxFhzSXyZvoQavVIBo3bqAicHZ2kHvEpBFxF4DhfNVcEkhZzlib+QOC4PnPdolVE7dq1xXzQkoTkosoVg9IcgeN3jh8/Lm4pPCcoAudE5I1NKBWBcyKwqii3IuBLBE4jcSRHtPHmodGYwPF32ofTv41kjOYCmlgjcHyGKRJojsnoTMY5UvQQOO2b1MAwqTd9ApWm3K3DXtfHzBE4raDTp0+DOQct9aU5AmdtXOmqoAMvWSNwWjEMgsA2WTKDNP2cOQLHZ54/fy4+ONQScF5SHCVwMSOHxY17/4a61nr/3j148+Y1ihQr4QDa6lGFAMTne8qYoZg9ezaqVKmCevXq6br45zo0c+ZMuVyhuRzNlulTTtJi6ULQ1wkcTUd5ycqgadRA2hMJl8SX6xtNFR25qHLFWLZE4PgtmtzzfMBzh608cK6oG8vUe05UGjhX9YgqVyEQgon5mSx98kQQuTAzkhMXcS7QjHhJDZwmJF00Y0yRIoVhobdF4PguD6g0L2MgEm3BDwmBY5mMjEUzPB78Pc2x2hP7NjTrZI3AsV40GeS44s2vaTRHSwROG1e8HOCYctdBwh4CZ6tN9hI4PkfTpvXr1xN7t2MAACAASURBVItpE7Xj3krghgzoi6dPn2DIiDGhORTVt70QAWcTOJI2mmJzH6tataok8LYkvkzg6NfGS1vjoC2ODA9evlnz13WkLL3PWiNwvPjavn27+CVzDaVvI4OyuFMUgXMn2upbCgE7EdA7MT2ZwLFuJHEbN26UBY9R8WhaYiw8jJM48SDJjZC3cNwAmzZtahU53njypo/EjQdRHkyPHDmC+vXry0aqRzTHahI45RenB0H3vMO+J+Hu2rWrxRtezWSQmy77UtOs8oaYpKlLly5mK2s6rlwdqZTjltpfHvwY9t+aGLfJ2s02y2QOxV69epkNgqIFBqJPKbGhv07z9l2DaMIt1cNYA3fqxHF06dhG/EgZEGjw8DHIkSu3vLpl0wYMG9TPEBhp3KTpyJw1G0w1eD9+Hxt7Dp1E/AQJENC+Nfbu3okIESOK+dWWXQfM1unsmdOoWr6UrBnfxfse5SpURoeAbqhWoTQePXqI169eIW26DBg3ZYah/YP798HyJb8hRsyYKFykOJYuXoizV8zneXLPKFZfCS0EnEngRo4cKX6q+fLlk9QktqLhLlq0SAJn+VrwLK6rXE9KlCgh0Sb1XoJq/rrExxoRdtXYYdASrsNM+2BOeJ6h5RAvkCtXroxSpUq5qipmy9V7TlQaOLd2k/qYvyGgd2J6OoFj/XjQmjt3rhymatasGaxr+Ttv7hj8hIdLmkfSrMWW8Fk6SnNB5cLKQzrf00vg+D0tAiAPwHo3IVv1Vr+HHAFu8G3btrVqokOTQZq60LRJI3AcJ3RG57uWhOPq8OHDcunA510t/AY107YIHOthrk3m6se51rFjR6tRLKk1ID4c544SOBK2zKmTYsLUWShUtBgOHzyAerWq4NTFa7h96yZKFcmHTTv2IWmy5GJeRtJMsmhK4JLEj4PdB08I8WpavzaOnL4ozaG/CQ9wlkyahw7sJxq4wcNHG5r/+NEjxPzsc9uhdXMkTPQD2nXqImSyX8+u2Lb3sJiOtmzaEPv37saZy3/KuyR+3fsMQMZMmV3d1ap8D0BAI3BTp06V/YiXfnpiJ/Dycfz48ciRIwcKFChgl/n9b7/9hjNnzvjc3sILF0aazJw55HOIF1XMh8kLVXe7NPAswr4kGbckvCDj2EmSJAl++uknt45ovedEReDc2k3qY/6GgN6J6Q0EjnXkIY6HT0sR9Xhopnkb/+RmoAU/sWccPHr0SMqn0D49pLebDOXLW1Ulno0A8+rYSnjNzZZaWmPhocCeoDXG48rVSJBYaH5ptr5lrk3m3mEUWFuXEPfv38eUKVMcJnB//+8GalYuF0SLlTd7RowcOwnnzp7GyePHMGXm3GDVskTgvooRA4VyZ0WOXHmQr0AhlChVGrG+/toiFCRwT578azCh5LoxqG8vbN2yUdaCp0+eIGeuPJg5/zd0D+iAGDFiIqB7LymPZLNpgzpKA2droPno7xqBY1AfXv7RXFpPug5evHBt4Tpkr/B5Ld+qve94w3PUPNKCxllCcqzlsnVWmfaWQ598W757PCMwkrW7LXX0nhMVgbO399VzCgEdCOidmN5C4HRAol5RCCgEXIwAD7NDhw51CoHLlyOTgcAdP3oEU2fNC1b72FEj4I+/74spIyV+7Gg4dPICEiRMKOaWB/buwb49u7Bo4Txs3LYXiZMkMYuAKYFb8tuvmD19Clas3yKau6mTxks5C5euQo8uHRE9+lfo0qO3InAuHk/eULxG4OiPRk27vRcn3tA2VUffRkDvOVERON8eF6p1oYyA3ompCFwod5z6vELAixHQS+DEhDJNMkycOgsFixTFkUMHUbdmZTGhvHvnNkoWzosN2/YgWfIUonHgjTpvq7NnSIVBw0ejWIlSWLt6JerVrIKzV66LBoTaQppAUptWslBetA/ohpI/lTGL7qTxY3Dx/DlMmj5bfp8+ZSJ27diGRcvXyM196WIFxITy1yUrxYSyf+/u2Lr7oHyn9S+NxddO+cB58cANQdUVgQsBeOrVUEVA7zlREbhQ7Tb1cV9HQO/E9HcC98MPP0gABWOzyV9//RVp0qSxa8jwcDl48GAsXrxYzPGYIytRokRgehL64tGPiBE0adbJZxs3bgymK9m9e7f8yRDvmjAHUKZMmQwmewyqwmAsNLVgCPgFCxaI/xVTIRQtWtTwHv2DmIuOtv80MaVZW926deXfaM4xefJkcZKnMHhHixYt5FmaivXu3TtIZE+7Gq0eCoKAJ44Bjsn58+dLfzMJd/ny5Q11ZvjzMWPGSH44BlGw5s+nvWSpjT169BAzMnt94GJ9GQ7X7z4WMnb65Al06dAGz188l/k3aNhoQxCTbVs2YXD/3jJnOK/GTpyGTFmyYvvWzfJO1GjRULxkacyZORW7DhwXH7i2LZrI8x8/fEDO3HkxctwkiyayN65fx881KgnZYxCTZi3boG6Nyrh182/Ejh0HGTJnwc2//ycEjjKoX2+sWLpINH+58+bHnl07sO9I4NxVPnD+tSBoBG7EiBESMZF+afasy9bmJBHkOj1x4kTZK3gZwfXfOEWOMcretuZwHeJeRtcC7mf0ESN+7vZRC42R6kl99f79e5ou0L9jDgBGjHpF12gA7Ziu7jM+XwKYBSAr48gB6K4IXGiMHPVNv0FAETh9Xc1kqYwIpccJnV9kVEwSqDlz5hjs3hnKmCSLicO5CdNXghsyTbwY3ITEzhaBo38E0ybMmjVLnKJHjRolZS1dujRYQ/nb3r17pR2Uhg0bgsSU5IwJROkcTnLIg3CePHnQqFEjeYY+gzyAMCk6DyBK9CHgiWOAEd3ow8Z+5kWBcVS0s2fPykXDkCFDJPBJmzZtbDbcUht5WXD16lW7CZzND3noA/S/1aLa0Sfu7Zs3QhCV+B8Cxho4BjFh+Hp71mVrc5JrNwnNpk2bxA+b+wd9wiz5oHrbmlOhQgW5dGzVqpXk0cuWLRuGDRvm9iiMoTFaPamv1q5dq3GxsgA2fCZojLgyEUDiz/jQVvwHHiU+/3lEEbjQGDnqm36DgL8QOJpSMeoXI+DR8ZkO5N27d0e7du0M2iz+VrZsWcnnRZk2bRpGjx4tBzBuJH369DFEBySBY+6uDBkyODxWeKtIjRmTipsLrsKDM0nbpEnBD3q2CBw3e2rRmBaBQrMu5q1juGUevo0lVapUshlqh3QeABjOmM9TGOWMh3WGZ+ZvrHfcuHHlN2pmChYsKId8JY4j4KljQGsJcw2ZEjjttwYNGiBjxow2NXDW2uioCaXjCHvGG3WqV8L/blzHm9evkSpNWoyZMNUQsdIzaqhq4S4EjAnctm3bwD3EnnXZ2pxkihxabRhbVlhqjzeuOZUqVZIIkz179gSDO/EikVYrevZdd/WzM77jaX1Vvnx5c1wsNoDbAL74TOjOfyZvRz9jsEQROGeMBlWGQsACAv5C4JhgmYSMOdwoNMngDaixOaIxgeP/M1kmnyGhIXkbMGCAgcCZmlDSpIMaKZpV2hLeutIshmWbk4sXL6JkyZJCHLlBFylSRPJ18VbVFoFbsWIFZsyYYWgnyydZpRaOddaEdWU+mVu3bkm5vLllMnGaqWhSvXp1uekk8SVZo/lK586dxcSS5K527doYO3asreaq380g4IljwLiaziBw1troyQSO+d5aNm0QrNdq/VwfzVtZTgOhBrpCwBoCpj5wP/74o8112dacZCTKTp06iTkyy+flXevWrc1WwxvXHFqj8FKVVh+MpEnrEObh9HXxtL4KFy6cOS42AEBaABU/98dTAMkA3Pv892GKwPn6SFXtC1UE/IXAUatGElKmTBkxLWQeFZoJWiJwzLPDvDk0RaSQ6CRIkCCIBk6vCaXp4kytV5UqVSTgQu7cuUFfIy3RMokWfdiSJk0q+cGYs85Ya8i60cyRt5S8obSXwNEckiHtGQmQYovA3bhxQ/zyWFf66pGokvDRDFOJ4wh44hiwdVjUfrdXA2etjbxJT5gwoc+bUDo+MtQbvoqAPQTOdF22NScZdr5GjRqSH4zrP/c2WlWYyyfmjWtOly5dJLUP/6T/NS+WeGloj8bRm8eRp/WVmXNiHQA9AdBJ/h9F4Lx5tKm6ey0C/kLg2EH0OaOf2fbt28Xvi0FHGJiDvj0UEjpqukj2JkyYIKaVJFPOJnCWzCPmzZsnZpnUFhrLvXv38N1338kGffv2bdHO8WZSE9ab9vIMLmHOhJJEjYnCNRNK+uXQd43v0V9OE2r8SNCosaMYm1CaDnDWgRo6HuaVOI6Ap44BrSW2NHA0AbblA2etjWvXrhX/UXuDmDiOsHpDIeBZCJiaUNIiwp512dqcTJcunexVvJykBAQESAATWouYijeuOaam+2wf97GBAwd6Vuc6uTae1lcm58TqAPoDKALgplHTaULZCMCRz/+2VGngnDwwVHEKAWME/IXAUYNGfzPm3qF/WbJkyYTAUSNH0sLADYyqx4MlCdy5c+eEKNHMkQSof//+4mvAICGUkPjA8X2aH9JckQRRS95JnzeSyw4dOojGjaSNQvMYmjHyBpKauZQpU0rUMZo3sj7NmjWTMOXjxo2Tv5OUzZw5UzZ1Rgs8evRoEGd5ahVJFhnAxFhIxniooLkoiSCDmFDzxiAmDDqhJWjesmWLfPPKlSsWo52pWWYbAU8cA1qtNf9G4yiU2m8ciyRw9kShtNRGRlh1hMAx19vmnfvN5s4qmDsrBg4dhbz5C0gVGdr/5YsXaNepi8VOmDF1Ep49fYoOAd1sd5R6QiHgBARsBTGxtC5bm5P0UaY7ALVutOCgBo7aKprHmxNvW3Po60aLE+5NL168kD2NJqO8PPR18aS+MjonVgNA9kzy9rdJH/T5HLyEt7oMbHJYEThfH6WqfaGKgL8QOPq+devWTcJ/Mww+yQlvKQcNGiRmktQ6kRDNnTtXfLwoU6ZMkbDpDF1eokQJCddMO3yNwPGm0ziNAE07uIHaIyRi/DYdshn+mZHD6GvHG0ZGnGTwFBI83jbyFpKmjjSvpFBjSJJHs0cStpw5c4opoxbtjkliSbD4Ps0caYLJPzWh+RpDM9erVy9IVUnSfv75ZyGwbBtJotYeYsQ6kMyxLN76MiqmEv0IeOIY4M02fUXpc8LxxPFNTTR9bTg3evXqJXOAgYD4O816rQUUsNRGjl9eVjhDA1coTzYhcHny5ZdDbK7MaXHo5Pkgc9O0l1gv5obbc/ikzG8lCgFXI2CcRoBRXJlGwJ512dqcZJlc63nhRj9smuLzotGSeNuawwtURqCk1QgvXnmhRNLqD+JJfWV0TnwL4A6AR0Z9QDLHvzONAE2WmEbgA4AeisD5w0hVbQw1BPyFwOkB2DgEOLVbW7duFW2YEoWAQiBkCDgaxCRm5LC4ce9fIVtM3t2xbQvJ25YpSzacPXMKQ0eOEwK3eOEC7N65HVNnzZMKHjtyGJ3bt8KHDx/w4f17NP6lJRo2+UV+69imBdKkS2/4e8hapN5WCFhHQCXyViPEWxHQe05UBM5be1zV2ysQ0Dsx2bhPVGf5sPDm78CBA2K2yJtSaiaMIzn6cNNV0xQCLkVAL4Fj8JxMqZJg8sx5KFCoMHZt34ZKZUtg/dbdQuBaNWuEjJmzoHGzFlL/WlUroHLVGqhcrYb8neZmWuoOkr1NG9Zi3m/L5DeaaS5fswnffk6V4VIAVOF+h4AicH7X5T7TYL3nREXgfGYIqIZ4IgJ6J6Y/EDhP7C9VJ4WALyCgl8Axp1rtqhVw5nKgmTMlU+qkmDhtthC4ymVLom7DJihfMdAHaMrEcZg3azoqVa2B/AULI2fuPIb3dm7biiED+2DbnkO+AKlqg4cjoAich3eQqp5FBPSeExWBU4NKIeBCBPROTEXgXNgpqmiFgI8j4EwClzlNMkyYOksIHBNnly1fCdVrMcp1oPzvxg3s3rENi39bgNRp0mLkuEny7xvXr8WUCWOxbstOH0dbNc8TEFAEzhN6QdVBDwJ6z4mKwOlBW72jELATAb0TUxE4OwFWjykEFALBENBL4GhCmTl1UkyZNR/5ChTE7h3bUbFMcYMJ5ZABfcXfrWffwDDqv1+9gmTJU8j/b9+6GQP69MCeQyfk72NHDsP9+/cwePho1UMKAZcjoAicyyFWH3ARAnrPiYrAuahDVLEKASKgd2IqAqfGj0JAIaAXAUcJXKwvw+H63ccSxOTo4UMSxIRELfPnICZDRowVDdzFC+fRqllD7Nx/VKoW0KEN9u3eiQgRI0oE1QFDRhrSDZQtURg9+w5EjlyB0V2VD5ze3lTv2YOAInD2oKSe8UQE9J4TFYHzxN5UdfIZBPROTEXgfGYIqIYoBNyOgKMEzpEKVq9YBl179kWmLIxmbV4uX7qIDq2bY+P2PY4UrZ5VCOhGQBE43dCpF0MZAb3nREXgQrnj1Od9GwG9E1MRON8eF6p1CgFXIuBKAnf9zz9x5collChV2mITGMAkfsKESJ4ipSubqcpWCBgQUARODQZvRUDvOVEROG/tcVVvr0BA78RUBM4ruldVUiHgkQg4SuDSJf8B9H/7InJkQ3umz/kVqVKnCda+v65dQ5+eXXDm1EnEjBVLTCfrNWyKug0aeSQWqlL+gYAicP7Rz77YSr3nREXgfHE0qDZ5DAJ6J6YicB7ThaoiCgGvQ8BRApc+RWL8tnwN0qZLb7Wt9+7eRf6cmdCjz0ADYWPut1XLlqBBk2Zeh5OqsO8goAic7/Slv7VE7zlRETh/GymqvW5FQO/EVATOrd2kPqYQ8CkE9BC4hctWI136DFZxGNi3F/689jtmL1jsU3ipxng/AorAeX8f+msL9J4TFYHz1xGj2u0WBPROTEXg3NI96iMKAZ9EwFECZ2pCGSZMGGzdfVDMKo2lWoXSKFS0OJq3amsXbm2aN0GpMuVQqnRZu55XDykE9CKgCJxe5NR7oY2A3nOiInCh3XPq+z6NgN6JqQicTw8L1TiFgEsRcJTA2WtCSQJXsEgxtGjdzqX1V4UrBBxFQBE4RxFTz3sKAnrPiYrAeUoPqnr4JAJ6J6YicD45HFSjFAJuQcBVBG5Qv9744/crmPPrEre0Q31EIWAvAorA2YuUes7TENB7TlQEztN6UtXHpxDQOzEVgfOpYaAaoxBwKwJ6CJw9PnAMYpIvR0b0GTAUtevWlzYxiMnq5UtRv3FTt7ZRfUwhYIyAInBqPHgrAnrPiYrAeWuPq3p7BQJ6J6YicF7RvaqSCgGPREAPgYsUKVKQNAJDRoxF3vwFYOrH9ue1P9C3Z1dJIxA1WjREiBABTX5pJYRu4/q12LxhHcZPmSG4KB84jxwePlkpReB8slv9olF6z4mKwPnF8FCNDC0E9E5MReBCq8fUdxUC3o+AowTO+1usWuDvCCgC5+8jwHvbr/ecqAic9/a5qrkXIKB3YioC5wWdq6qoEPBQBBSB89COUdVyGQKKwLkMWlWwixHQe05UBM7FHaOK928E9E5MbyBwP/zwg4QZ538vXrxAmjRp0KVLF+zbtw/Hjh3DsmXLgnR+27aBocc7dOiADBkyiO+MsYQNG1b+LXr06Pj9999Rr149PHz4EF999RXmzp2L1KlTy+MFCxbE//73P/l3Cp9r1y4wKl6lSpXw119/GYo9e/Ys1qxZgzJlyqBly5Y4ePCg4bfLly9jxIgRaNWqlV2D1Nvae+/ePTRv3hzXrl3Du3fv0KxZM2h9YFeD7XjIUzCpX7++oW0cO02bNpWx9Pr1a5QuXVr6maHxN2zYgN69e+PChQuCzZgxY+xopXc8YtwXb9++RfXq1dG8fVfQNFKJQsDXETBH4Fy159i73hDzFStWoF+/fvj06ZN0wfr165EoUSJf7w672me859v1AoDdu3ejffv2OHXqlMVXdu3ahSJFimD+/PmoU6eOvUWH2nN6z4mKwIVal6kP+wMCeicmsfmkrfgeClTixImFHKVPn15quGrVKjRs2FD+7aeffhKSFStWLPmNB8rvv/8eO3fuRLRo0ZApUyY8fvw4SMuMF/PChQuDm2TdunVlAxw2bBiOHj0qzxcqVEgW8HLlyllF5sSJEyhZsiTu3LmD8OHDB3mW5Ib1v379Or755hu7EPa29tauXRtJkiRB//798fLlS+TJkwczZsxA1qxZ7WqvPQ95IiYVKlRA0aJFhZjzUJctWzYZP6VKlZKLgVevXsnlwvPnz32KwBn3hdLA2TN61TO+hIApgaNvpjv2HGvrDUlGrVq1QEIRN25cuejkPhc5cmRfgl53W4gFzwHaZaw9BdlD4Eja7t+/LxeXxN7TRe85URE4T+9ZVT+vRkDvxPRGAsc6d+vWzaDxIdFq06aN9N/SpUsxcuRIIWEkTdYIHLUmyZIlk4WdCzx5bLx48XDgwAH8+OOPQuCocStfvrzVsUENCzfK0aNHB3uOB/ojR45g5cqVdo8vU7Li6e1Nly6dELacOXNKG7Xb6HHjxtndZlsPeiIm1MJmzpwZPXv2xKNHj4S4Ll68WLS+mvBGnBo6X9LAhYTADRnQFx0CuoWatu79+/cYOXQQVi5bLJct4cKHR5as2dFv8HD8+/gxMqVOgjTp0uPDhw94/+4dcuXJhy49+iDe999Ll144fw5dO7bF48eP8PHDBwnGMmn6HKRKncbWEFa/+wgCpgSO2nZ37DnW1hsSCVqMNG7c2EdQdm4zuL8PGjQIq1evxj///CPWEby4Xb58uexdW7ZskQ9y3nPv59/v3r2L1q1byxp/8uRJWbNmzZplWN+5rvPZS5cuyeUyrW54kenJovecqAicJ/eqqpvXI6B3YnorgaMWjgdnEqRevXoZzByoCatYsaKY8dkicNSSUHtEE0dNcuTIIWVyMySBu337tizcNKscMmSIaNOMhVoWkr79+/eLaaeppEyZEqNGjRLzOnvFHFnx5PZyI6S2c/z48Xjw4AHy5s2LVKlSyWbpLPFETKj5LVu2rLSZlwA8FHTt2jVIkxWBCzoCYkYOi+t3HbsJd9YYYjm/NKqHJ0/+xdRZ8w238WtWLkeGTFnE9DV/zky4cTdQY89b9RFDBuK3BXNx6OR5GeO5MqdFz36DULps4KXO7Vu3EDFiRMSOE8eZ1VRleTACpgSucuXKbtlzrK03WbJkEWsUuhU8ffpUTPn79u0rF5NKIDhwH6ZFzZUrV8RaggSMl7ZJkyYVwpY8eXKx7pk8eTK2bdsmJpS00NmxY4ecBWhNwTWehI3C57jv//bbb+jYsaO4eJAkerLoPScqAufJvarq5vUI6J2Y3krgqNEicTt37hwSJEgg9v5x4sQRonXr1i05bN24cQMZM2Y0a0L55MkTXL161SqBu3nzJuLHjy9jY9KkSbJg06fJWBYsWCC/HT58ONgY4mZao0YNsBweDu0Vc2TFk9tL/8FOnToJiaaZKG8lectJk1RniSdiQj9Mmu7yT7aXm/zYsWPFrFITReD+GwHtW/2CubOmI1WatKL9+m3ZGhTOmw1nLv9lMPVqXLcW8uQrgAZNmoFkr2OXHti6eQNevnghmrCqNWpJgSePH5MUA8+ePZVb844B3VG+UhWrw41pCfJmy4ALf/yNmJ9Nro1fuHH9ehACp/1WvGBuVKtZB42btUCiuDGxav1WZM6azVlDW5XjZQhoBG7hwoWy79BU3B17jrX1hhog+rtRo8T5QLN/XmTSH1tJIIGjRk1zY+C6ff78ebl8pd8y92hajHDtptaNVjckcHTV+PPPPw0QxowZU84cPBeQNA8dOhTFihWTsnh5/Pfffzu017u7b/SeE+0/vZi0yNX+Obyl4EGPQnau+dm4G1j1PYVASBDQOzG9lcDRhJJBRGiy1r17d/Ez4uJMrdq8efMESvpj0eadppLhwoWTf6O2JGHChPIbbdetmVCa9gfNJKmR4yKuCTV11OI1adIkWPdRM0XfCEdv5SyZC3p6ezUAfvnlF3z77bfiUO8s8URMeEnA8UafE0pAQIBoYwYOHGhotiJwQUcASdmNe4EBhChN69cRwlavURPcv3cPebKlF0L35ZdfCoHr3K0Xuvfuh+t//YVCebJi7+FTiBY9OsqVLIxlqzfi27hx8fDBAxTIlQXb9x5G3O++Q7UKpdG9zwBkzJQ5yMdXLV+KUcMHY//R02aHpSUC17VTOzGnHDluEiaOG40h/Xsja/acyJY9J8pVqoL0GTI6a5ircrwAAY3ATZkyRYIY8fzojj3H2npDSwBqArnnUHjZeOjQIfCCUUkggdMClxEPXvbSd51nAVpP0Hpm7dq1qFq1qhA2XriaI3AkfiRwvLDLnj27EEDtcpYEkQTaEWsbd/eN3nOiInDu7in1Pb9CQO/E9BYCR3M8zbeIwUu4UW3dulVMIf744w/xv4oRIwZmz56N/PnzG/qe/1+zZk2JBEgZMGAAjh8/LgFQKNSasCxGmOTiO3z4cPGf4y0myR6JCIXaJGqZjCNP8rv0sWPwkqhRowYZbzRjIXk7ffq0w3bxJCve1F76f/FwQWd+auFoysN2a9g5YyJ6IiYcj/SRbNCggQQNIJnnGGFURk14QUhtr6/5wGnj09EgJqYEjpq0Ni2aCKmiueKjRw8xZERgxE4+e+7qDcRPkED+XrtaRZQtXwmxvv4ajevVQqIf/jNnpv/a5Blzka9AQYvDTS+BC+jQBp8+fsSIsROl7H/u38fe3TtxcP9eLPltASZMnYWKVao5Y5irMrwAAY3AUfvy3XffSeRid+w51tabRYsWCQGhOR/3LpI51ommfUqsEzjiw/MBfRmpsaSmk6KZUDIgGtd2ng9oQnnx4kXR0nHPGzx4sAHeqVOnypnEEX93d/eN3nOiInDu7in1Pb9CQO/E9BYCRz804zQC9DXSgmawDVxgqR2jWaSx8HaUAU7oP0B79xQpUoivlkYu+DwJnJZGYM6cOXIbpx3IuVnz9o43dgxSwoAdmvTo0UO+yXdMZfr06eCmqicyvMsNSgAAIABJREFUFcmKN7V38+bNgjHN4qhZoUM//eCcKZ6ICUkqI1BS+8vopzS7oZ8khX4THFck8hx31ATzxp6+Kd4uxn3haBoBUwJHLGii2LPvILRoUh9rN+/Aj0mSCkR89uyV60iQMKH8vVbVCqhQqSpixIyJUUMHYcvuAw5BqdeEsliBXKhZpx4aNvkl2PfGjhyGI4cPYtHywAshJb6PgEbgeCkTO3ZsuUB0x55jbb3hGkMLAJIQWpuQvNEk0DQqsu/3jvkWEhNq2jTNP611eJFLDRyF2jjuWTSF1SJa79mzR9Z3mkpqQUxmzpwpZwhezu7duzeI3zvPEEyxQg0ezwueKHrPiYrAeWJvqjr5DAJ6JyYBcLWZss+ArBqiEFAIBEHAUQ1cwm++woHj5wykjIWtXLYEvbp2Et+45Ws3GcongaPfW9eefUDzRppQ7jl0Um6+c2ZOg2mzFqBA4SLy/NkzpyUSJLXA1qRZw7p4/vyZaOtIqnnwXbt6JTJkzBwsiAnJ6ejhQ/DrvNmGICbr1qxCqdJl5WDMiJa/NKyLb7/7DoOGjVIjw08QUIm8fa+jeenI4CaMSOnLovecqAicL48K1bZQR0DvxFQELtS7TlVAIeC1CDhK4IYN6o9lixfiyyhRsHLdFoneSCKUNME3mDZ7AUqU+i9aq+YDt3njOgli0rVnX1SpXlOwOnP6lJA+mlzSPy1BwkRYuGy1+CBa8oHje/wWTTW1NAIfP34UH7y+g4bh8aNHyJwmKVKnTYcP799LFMrcefMHSSNAAnjy+FFEjBRJ0ghkyZYDQ0eNC2ZC7bUdqipuEwFF4GxC5FUP0OKGGjpaktCnzZdF7zlREThfHhWqbaGOgN6JqQhcqHedqoBCwGsRcJTAmWvoqRPH0bRBHRw7+186Dz5nztzSa4FSFfcZBBSB85mu9LuG6D0nKgLnd0NFNdidCOidmJ5O4JhnjXbpdMx2ttDW3ZatOgOU0I/JGcLIegw/bC2lACNZMeiFHvn666/FJ8OSEEOGS2ZUTncIzczoE6BFAHXlN4kZg85Qw6JHGGGU/cNAONb8Rnh445igeR3N7/xJ2I/Eh1hFiRJFmh5SAtf6l8bYtWMbJk6dhYJF/ku/wLJjfRlOcsZpfiv+hLVqq+cioBE4BgxhFGSuG0pCBwGu1UyfYG3N5jrN8P48S+gV+t/TX86RdEB6v+XK9/SeExWBc2WvqLL9HgG9E9PTCRxD9jLkv3ZgdHZHs1ySDEsL89KlSyW3nLOEh2A6QVtKsMpkoUyCrVd42GVePHPtefbsGRhRK0mSJHqLd/g9mqixvQzK4grh5kzidu/evRAVzzw+9IGg0zqJsDlhYBsGw3HFZUKIKu+mlzl+GH2SEe4YQtsZBM5NVVefUQg4DQFjAscw/YrAOQ1aXQVxPeYeQ5JlTthfDO7CRN0hEe6pKVOmdMuFZEjqae1dvedEReBc1SOqXIUAwAO7x86xkHSQqwkc68bbOy7u5kgVk7Vq6QtC0g7jd0k6uBGYC7jAW92Q5qKkH1DSpEmDtYcHcIY55mbnTuEGy+TeDD7hTCE5ZBRQ5vcJqfCGllHHOnfubJbA8RvUBPub1s0YV95gM51GpUqVFIEL6YBT73stAorAeV7XcY9hdFxz2nr216pVq5A2bdoQV5x7Dvdu7rHeKHrPiR57uFSJvL1xGKo6myKgd2KyHE+OQukOAvcZA9F8mZpiuILA8XvccEiyTDWLziBw2tjgRmPcntAicKwPNz4mvGb4ZmcITSWpDaN21hliicCRsDHhOxO3+ruYI3A0xx02bBiat+/qMi2rJdznzJgmESVbt++Ec2fP4Pcrl1Gp6n85+PLlyITNO/e7THtfNH9OMSGV+fz+PS5dvIADx88idZrAg+LMaZMxY8pEhAsfXi5Tdu4/Kge/LZs2YHD/3rh88QIaNm1uyHvHd1YsXYyxo4ZJeZRadRugVdsO8v/79uxG1fKlkCxFSvk77+y27j4YTPPQvHF9LF44P0jCdGMML5w/h18a1TX8E3PoEce/bj+Uf2ObenbpiJ3btyLSF18gbboMmD4nMCF0iyYNcPTwQXwROTKiRomKISPHIlOWrPLbtT9+R9sWTfHkyb948/o1ipcqjQFDRgSxBmAOvdxZ0yF7ztxYuHSVvEfNdud2rXD61AkJSFO6bAX0GRiYjsOc9OzaCZmzZAvS19pzxHbT+rUYO2may6arInAugzZEBXOPYXog0/yjziRwMtc/fBArFtPcryGqvJte1ntOVATOTR2kPuOfCOidmERLEbjAMcOFmZo4+vho4ioCx/K54TCfjLHJnjMJnKlpSWgSOA1PauFC6kvADZn5/eiH5iwxR+DYP9S66fVJdFbdPKUcUwJHfK5duybJg0ODwBnjsnD+XGxcv8ZACtyN2ZpVKzB8cH8cOHZGPr1h3RpMGD0Cy9ZuEs3zo4cPJX8diRyJDrFcvWIZXrx4HoTAHTl0UPLgxfnmG/G9LZgrC8ZPmYm8+QsIgese0B77jpyy2DymRNixdTMWzJ1lt/9g5/atpV7DRo2Tcrt1bi9rk/b3+/fu4Ztvv5XfNm1YJ5FC+TzJEonX2St/yW/M01ewcFE0bd5KSGChPNnQb9AwFCtRylBfJmOPFetriR6qEbgBvZlP8xamzJwrPqw1K5dDnXoNUb5SlWDtvH3rFiqVKY7Dpy5YxKBg7qyYvWCxIZ+gs8eCInDORtS55Zm6RTibwGl7N5O42/Khd27LQl6a3nOiInAhx16VoBCwiIDeickCFYH7D1aSHhKMmDFjyj+6ksB9xl6+RSLHG3VnEjiWz/bQx485rzyBwLFONB1NliyZRT9Aa9PcVX5opgTO2Ro+X1i6jAlcpkyZhERzTK1du9ZuAsfIkh279MDWzRskNQDzvFWtUUvg2bFtC/r37i6H+BgxYmL0hClIkTKVEJ4WjevjxcsXQiyooenRpz+GDOiLp0+foGNAdyE6T589RaIfEiN7jlwYNX5ykCiWjHTZpWMb0fbQV2bw8DHIkSu35JfLlyMjmrdqhy2b1uPpkycYNnp8ENJhT99VKVcKRUuUwi8t28jjJQvllbQHpoFZjMsaOrCfaKuGjBhj8RPVK5WV5OU169S1SeBItKpXKoN1W3YhQZzoFjVwxh+jBjVl4nhYv3U30qZLL/ikShwPF/+8ZVPD8PDBA3n33pPXMpfrVK8k+fQ6d+spKRlKFMqDWQsWI136DPLJ+XNm4eqVS0idJh02rFttIHBsY9XqtQwpIiaOG40D+/aYTY4+fPAAGR/de/eziNm4UcPx+PEj9B041J6uc/gZReAchsztLxi7RbiCwGkNcsaFpDvB0XtOVATOnb2kvuV3COidmIrABR8qNJujVow5YVxN4LSv81BJs4xFixaF2AfOtEWaaQmd7UPDB87SZDQ18bQ1aR8/fiwaMVeIMYHjDS7JCfOAKfkPAY3AMYgJDy48SBMjRwlc52695AB+/a+/JDn33sOnhFTlyJQaG7btkYTcyxb/hlHDBommpUvHtogb9zu079xVKkN/REbDJAEigRs8fDR+WzAvCCngc1oaApadOXVSTJg6C4WKFsPhgwdQr1YVnLp4DTTpy5jqR8xfvAJly1cUEtm1Y1tDSoM2zZugVJlykrzbktz8+29kz5ASF/74GzFjxZLHEsf7Gq3bdcLWTRvw5u0b1KhdF81atA5ShC0Cd/nSRfxUND8OHDuL7+LFEwJXs3JZJE6SVC5BatdtgEZNmxvKrFmlPFq0bo98BQranYKBOE+eMAa7DhyTcs6fO4taVcqjctUa2L1rO774IrIQ0QKFCgdr/sC+vXDp4nkDEfv7f/9Djcpl8ejhA9AsM6B7b0Ofsa8b/lwdm3bsw/Ili4JoS0nEL5w/K1ozatWrlv8Jz549xf6jp4N9s3ypomjZtgOKl/zJYn8c3L8PPbp0MLTJ2XNYEThnI+qa8riP0y2Cl5jO8oEzV1NLPueuaVXIStV7TlQELmS4q7cVAlYR0DsxWajSwJmHlqaUhw4dQsaMGd0y+qiBO3XqlMu+xw2NkRZJnDxBbEUP0+roDj80jcC1adNGCAJJr5KgCJDAHT58GAULFjT4NekhcOeu3kD8BAmkcJrUlS1fCV/FiIFJ40Zj/dZdho8mihsTh09ewLGjh9G7W2dUqVZTkm5Tq8W5YkyAzJlQagTu7//dELM8zdSPH8ibPSNGjZuMuN/FQ85MqXHncaAvJc1lk3wfGw+e20/eqRW6fOmCEBBNEn7zFSpVrYExE6eKNqp0scBk4caJyq0RuFs3b6JcycLoPWAIylesLMVS20kheaYpYdUKP6Fz156oULkq5s+eiXPnzmDEmAkSaEdLwUDNuzUpV7KIvN+wyS/y2OlTJ1Eod1ZMnTUf1WvVwdkzp1GxdDHpB5p1arLkt18xcuhAIWRMxk7p06MLYsaMhXaduuDBP/+gbIlCGDJiLAoULoKyJQqLOWWWbNlh2lfUAvbt2VW0btGjf4XMWbNh7+6d2HPoRLCqZ0ufEjPmLkTGzFksNuuP368K3leu33HJFFYEziWwuqRQ7jFMM8Ak3c4IYmKtko5eSLqkwTYK1XtOVAQuNHrLzd/8/fffsWPHDsMmo2049laDm3KdOnVc5nRubz288Tm9E5NtVQTOfI/zILR//37ky5fPbUNi165dKFSokEu+x/n4119/OV3DF5LK0mSL5njWonoxUMmff/4Zks/YfJcEjn1du3ZtRd4soEUCxwORsfO+HgJ39sp1JEiYUL5CvymaCEb/6itMHDvKLIGj9onmgQyqsWbVcgnwsXT1BoMJJU0QHSVwDHAycuwkIXD5c2bCjbuPpT7Pnz8X88PHr+wj8FwjMqVOijETpop2T5PcWdJh2OgJog2j9O4egIgRI6Fn3wGGZywRuDu3b6N8qSLoENAdNWr/bHHsjhkxFHfu3Mbw0ePRtH4dHNy/F2HDhZPnSVrjJ0iIhctWY+e2LVixLJBc9h803FBPasXyZE2HS3/dNkTvo1lk8kRxhcBqgY0L582OPgOGGrRwK5ctkSAsazbtwPfx4xvqFz92NJw4/zu+jRv3vzZHiIg2HQNEyxk1amAUWvr9vXr5UgKZrN64LVj72K4rly9h6qx5wX7LnzMzho0aj1x58spvJQrmwctXL0WDu23PIfk3ahFrV62AM5dds2YoAmdzOfWoB3ghx4BXIY3ubKtRvIjIkiWLR+eK03tOVATOVu/7wO/UVkyfPl1aQj8iThpHhBvG6NGjxTxGiWMI6J2Y/IoicMGxpnnY7t27JfpgzZo1HesMHU/zIMhcZCdOnED9+vV1lGD9FbZn3bp1YhqaP39+p5evp0AmLd+4caOEpc+bN/BAZk5oVrVy5UqJ2mkp14+e7xu/o2ngcufOLXmdGDFTSVAESOCmTp2KGjVqgA78FD0Ejn5vXXv2Ef8zmlDuOXRSAgdlz5jKYELJaIw0oTx44pz4wCX+MYn4WV29chklC+fFn7ceBCFw69aswvTJE7Buy05DpYOYUKZJZkgWzkAhdWtWNphQhoTA7dm5Ay2bNcT5328EAWv08CHi30atE3GjRqhtxy4GbRof1nz4jH3g7t65I5q3dp26otbP9YKUee/uXdGCEQdexlQtVwo/N2iM2nWDrxda260lQacJJInetNnzg3ynUpkSaN66nfgBkuQVyZddzDjjfvcdVi1fioF9ewp507So2st5smUQX8I69RqIL12Z4gXFjNQ4MiifNTV3ZVuYH5Pzjt+j5m7RirXik2cqjLCZK08+1G3QyOL0JMFcuXwJfl2y0iVTWBE4l8DqkkIZQZjmk+nSpQPXdlfJ5cuXZe+uW7duiPK4uqp+Wrl6z4mKwLm6ZzygfEXgQq8T9E5M1lgRuKD9xg2amjAeTmPHjo1cuXK5tGOp1eDiz4h+PBjTRM2ZwgMkNeNsD4OHuNqUxJ66s6379u1DsWLFULRoUZvJUR89eiQbMX0arB1K7fm2uWc0AteuXTscP35czNCYV0jJfwhwHDGxPcN0M7k9L+n0EDj6wG3euE6CmNC/qkr1wAsS4yAmNMVjEJPkKVKCGpmli35FhIgRRTtKs0FGKDT2gWPERhKa5y+eI2euPBLERDMj5Hg5ffIEunRoI7+TLA4aNtoQxKRArsy4fueR1IEaOJo/Pnr5Qf5uyweuSb3aSJIsuRBSY+Ea0q5lM5w6cUxu5MtVrIJuvfrKIyR9LZrUFz8vjjOaDY6aMAUlfyoj31u5bLH4uWlCUkQyN2PqJMyePkXSElALWaFyNXTp0dvsEDVuu7kHiGP65D9g6uwFEuHSWEiiWv/SCA8fPhCySMJN/0BKnGgR8W3c7wy+fvy3dZt3SoRNmlsGtGslGL97+xY/lSlvNh0ACRwjhmoEiykgGtSuJilP2LZuvfqhTLkKZtu1eeN6SZEwd+FSi1OTqQ4KFSlmCI7j7DmsCJyzEXVNebQ2YW5PWtDQioMkztnCeXT69GmQwNE3mBo4Txa950RF4Dy5V51UN2MCx9vyrFkD88M4Ijy8RooUyZFX1LMqkbfTxgBvg2kvz0MqNW/UEDk7kbfpQY+bDA/HNB8+cOCAU0096NPD9sSPHx+lS5fGkSNH3J7I27i93PDOnDmDCxcuoEqVKg6tEcRozZo1QqqdrSEzDmISK1Ys6QcGMnFl3ztt0LqpIC2ISYUKFUBzoXv37knaDUeDmNy4969LSLibYFCfCUUESHhp0rlw6WrE+/77YDWhCWi5UkXEf840p6ezqq0InLOQdE053GOYP5b/VaxYUdbw1atXO/3ikpdXNLtncC3u3T/++KNrGuTEUhWBcyKYvlaUMYHjTXlAQICvNdFj26N3YrJBSgMX2K00lyTZoa08zfp4U+/KKJS87WdUSJoM0/eK5o3OTCNAE0WWz4uU8uXLS4S30IxCqW14Dx8+lA2PlzyOCk1B2Qa2Rc/7lr5nLg/cxYsXcfDgQWTLls2mhtDRdnjj86Z54M6ePYujR49KZFB788DZ0gx5Iy6qzu5FgIFWXr96hZy58wT78IljR0VLmy1HTpdVShE4l0Eb4oK5P3DN5uUS91ReMLkijQA1etu2bZMzAveyb4wC/IS4ES4sQO85UWngXNgpnlK0InCh1xN6J6YicIF9pplbFC5cGCVKlDAc2F1F4B48eCBkkaZoNL1g6HqKswjcH3/8IVqk4sWLg22ij0lo5oHjhrd9+3bRrnPDo4ZTr/AWnppEBk1iABRniDkCx3Jv3ryJDRs2CAn2d8sAUwJHfNgHnpDI2xljQJWhELAHAUXg7EHJ/c9orgLUvJK8aX66ziZw1Lhx76bGrVq1ahIV1ltE7zlREThv6eEQ1FMRuBCAF8JX9U5MfydwpuYW2bNnDxJFyhUEjsF9GCCFtvmlSpUKYuoTUgKn2eRfunRJTBSNbfJDi8AxChg3PCYUr169utM2PKZEoPlpjhw5Qqwhs0TgOD+4YdP/LkWKFJIQ3V/FHIHj4Wjo0KF2a+CohR01bLD4efFSgX5tCRImEl84LeGzHnxLFyuIFm3ao3TZ8npeD/YOo1pmz5kLyZKncEp5jhZiGuCkVbNGOH70MJKlSIkFi1c4Wpzu53t27STRI019/HQX6OQX0yX/QQIbfRE5svjdNf6lJRo3ayFfYY4/pjZgdE5GOaUvYvmKVdAhoJuhFoP795FgOacv/WmIjGqriorA2ULI/b8buwowyJLxOu1MAkdrA15E0i++TJkyVqMnux8F21/Ue05UBM42tl7/hCJwodeFeiemPxM4c+YWpj3oTAJHzRHN8k6ePIly5cpJVCwtVLf23ZAQOFs2+aFB4G7fvi0bHokx22wtXYCe2XPnzh2Jrpk5c+YQRai0RuBYL5IX+nrRP0672dVTX29+xxkEjuHuGfZ90vQ5hkMWg3o8evQQFatU0w1PmeKFJHKiswicXkLICxTOadN57WjDjFMMMIVCptRJcPPBM4fLZYAkEmVNuOY54hvWq1tnRIkS1WMJXPoUifHb8jUSsZKJxPNkS4/NO/dL1NICOTOjcrWaCOjeS3Dj+J03ewZ+adlG4JBALikSI2my5MiZO6/dbVQEztHR7NrnTV0FTC0lnEHgtByqNBn/6aefUKBAAYfnomtRsK90vedEReDsw9ern1IELvS6T+/E9AYCx6iAzrYx18L283BDkz5Lh/IpU6aIuZ8zzOeoeSOhYXAUmk6aE6bhoC+co+Hy2R4GB7Fmk08CN2vWLIlE6Q5hZEAmJi9ZsqRE1mREO1cIb1+ZZoDlk2Dp+Q6xYTSxzp07C/7mhIdh+t/xFpbPOJuMugIbZ5bJwz/XePqHkpBTHNHAMR1A/hyZcOHaTaupYiaMGYnVK5eB34sT5xuMmThNtCP0GWUkSUZ2pJSvVNUQhZEELnXadKKl+vfxY/xUtjwGDh0pz00cN1o0frzgiBAhguQR03ykmG+sW6d2uHfvrjzbuGkLiYTYvXM7fB07jmhu+gwYgqLFS8JSvagtu3jhnETVvHXzb6zeuF1C7mty7MhhdG7fSnLoMXpkk19aoUGTZmBI/PQZM6F5q7byqLG2S9PAUTNZvEAuMDk121e5ag3E+eZbieK4cOkqeY+RGbUcevv27Ebndi2RLUcunD51Ap269JByK1ergX17diFJ0uSYPGOOpALg38WXNFlywZi+uExjwAiZt2/dlBx5X38dW7R+1MCZagWnT5ko0T1ZHmXsyGFYunihzD+uQ+u27JJ1bPHCBZg5bZLgTzI4fMwEIVzUcjK6KPv44sXzssbO+XUpfkicWFJHtGhcHy9evhCixYiWxjn0NGyNCRz/rUi+HGjTIQBP/v0X8+fMwPa9hy1OAUY9HTKgD0ZPmCp5485dvW7XdNEI3Lx58zBixAiH12q7PqIesgsBrtuM4GzsKmD6Ivtr8uTJ4jOtZ29gedy3r1+/LhYkrs4nZ1fDdT6k95yoCJxOwL3pNUXgQq+39E5M1tjTg5gsWrTIJcmVGSbe1NzCtAcnTZok+QxDeqPOckkuaJv/vZnoadp3p02bJkmr9XyPDttVq1a1aKLIzW78+PHiC+cOIcGhf587IjkyKiIjVJIw6hX6IbZo0cIigfs8VyRwB7V+JBj+KHoJHHOIjRo+GPuPnrYI27LFv2H/3t0YM3GqHLZ4+F+1fAmWrFov5nBMcj19zgIJC858cO07dRXNHTVmJGcr1m0WUvJT0fxo2aaDpChgZMKvY8eWb5JMtWzaAEfPXJL+y5ExNXr0GWDIVfbo4UPE+vprkBDSJPOnMuXkPWv1IrEhWdh35DRix4kTrG1MVk7iRRJF4YUDTbwY7p4ETtMIUdtFc0WmBjAmS/+7cQP5cmQ0JBs3zaNmSuCYBHz91t3InTeffI8kh2H1x00OzNFKE1ZJx9Ctp/x9+OAB+Of+PYwYOxH1alVFipSp0b13P8Ga3yXhZJ1ME48zrcGpE8eFwLFOs6ZPFvJKnyC2kX8ePXwIo4cPxq9LV8mFx8H9+9CxTXMcOnleCFy3zu2w/+gZJEyUCP16dsO//z6Wvu/SsS3ixv0O7Tt3lTqSlDNVgamwbUxWTvPbC+fPoVThvJJXcOK4UYgYKRKMc+yZvlu/djUULFQU9Rs3RYFcWdB3wNAgydgtDVKNwDEnIi/j/O0ix5PWPM55RsW1Fr6f/cX8whyTeoWpSLh3J0qUSG8RHvGe3nOiInAe0X2urYQicK7F11rpeifm50Ppp9CrufUv8zDGBdgVwtthbgDWhLf+1G45Q7jR29LkheR7vPW2Zh7FQxsT7LpLeADXgrO445vOaB+TCRubnFmqN3Hk9/xRjMexIxo4UwL317VrqFurikQUzJ4zNyZNnw0eqkkKqPmiUGvFy4wDx86gYO6skr8tT77ARPSTJ4zF5YsXMH7KDCFcdeo1RI3aP8tvUyaOw4VzZzFx2izs3LYVI4cNwr+PH4l27fzZM7j77yv8ee0PVC3/U7Ak3Hzf1CTTWr1Itu7euW0gSKZjgnWZN2s6KlWtgfwFCxuiJ5LApcuQ0aCBs0TgmPDcONk4iY81DVzbFk1w8sLvhmqQ5MyYu1Dy3lEYhv/Z06fiN0ah71iiHxILSU4c72sholqS7ta/NEb8BAmDkUq+Z0zgSPyKlywdLKl47+4BWL7kN9FmavLgn/uSRJ3J2knOl6/dJD9tXL9WkrGv3rgNa1evRO9unVGlWk3kyVcABYsUNXupZewD92XkL9GyXUeUq1AJndq2RPgIETB05FizU5REPVOapDh39YaktKA28fDB/Zi9YLHNKa0ROOZEZH5Nd65xNivnZw/Ys8dw/+a+GhLhOcFRq5iQfM9V7+o9JyoC56oe8aByFYELvc7QOzFZY0/WwIUeourLCgGFgC0EHCFwlkwojTVKJEqFixRH3YaNg32aBG7g0FGGxNMkcFcuXRTiRMJVu24D1KxTV97TyB0Tev8QNyY2bNuDjJmzgGa9ib6NAeaio7mjNQJnrIGzVi9TzZQ5zKhF271jGxb/tgCp06TFyHGTJGl38pSp0KptB3mlc7tWiB3nm2DaLlMCt3TRQqxcvgSLV6yV99asXC5kav3WXaAJZfeA9th35D9NtKmZIQlc736DhRSZCgnc3sOnDAE9jAncyKGDcP/+PQwfPV5eo0nppYsXRANHfIoWL4U69RoEKZIELkL4COjVf1Cwb1kjonyYvn87t2/FmlXLxfR06eoNwcowbZv2wIK5szFv9nSLJpQcH4P79UKsrwM1sx8/fMDjx49w/vf/BUlSbq4vNQK3fPlyHD58WBE4W4uE+t1jENB7TlQEzmO60HUVUQTOddjaKlnvxGS5isDZQlf9rhBQCFg6zDoShbJJvdp4/eY1Jk6bbQhiQvKxZ9cO/LpkpZgq0vxtzcbtYjJHvymShPQZMqJvz65i1jdt9nzRJJcuVkBMKMtXqiImlNR1Qt5wAAAgAElEQVQMLluzUd4pU7ygmFAWLVEKSePHwYU//kacb74BScigfr2EwFHbmjtLOnTp0cdg3qiZUNasUh7lKlQ2EEJr9TL1DTPF6ferVwzRLLdv3Sz+Z7sPHhefsbNnTonWh98lsapZp14wbZcpgTt+9Aga1KmOY2cvi8ad5IkmhvYSOJpQUts0f/EK8VWjOer/blxHylSppdzkKVKhWy9qFe+ICWXjZi2lTiuXLcHUSeMkSAhNlmmqyWepOV3y26+YPmUCVm3YJhotRp7ln0cOHUST+rXlHWr1qLU+e/qUkGlrBI5kn4FIqGG5euUyShTKg79uP7SbwDFgCf0tq9Wsg45duks5/Lf5c2aiWYvW0u+8DChcrLihzIY/1xBNsGbSamnGKwKn1kJvRUDvOVEROG/tcQfqrQicA2A5+VG9E1MRONsdwfD3Eqr6iy/k4JgmTRp06dIF+/btw7Fjx7Bs2bIghbRtGxiUoEOHDuL/xcOMsfAwoR1wmEerXr16YHJr+sXMnTvXEOCEgT/of6eFROZz7dq1k6Loh8TcdZowqTJ9wBjamAnJGzZsKO/yMMuAE/TXcKcJSJ8+fTBo0CDx50uYMKFtkD3wCU/p9/r160MbUxwvTZs2lfHDQ3Tp0qUlkALNDOnfOGPGDEMUxICAAPHbcKU4ooFjPTgeSaIYVITkI0aMmIj9zTdo2yEAWbPnkKpOmzwBC+bOkv+n5qVO/UZo2aa9zD0GMTl5/Kj8VqFyNYkwSKEGLm36DDh25BAeP3qE0uUqYMCQEfLb+NEjRENFPzj6otGX7vrdx0IwGBwkoH1r3L17Rw75TZq1RL1GTbBl0wb07NJRzAy1ICaW6kUN3NOnTzB4+GizUAd0aIN9u3dKygSa5w4YMlK0iCRddWtWwb27d/BD4h/x1VcxkCRZcoMGTiuTBK5Arsy4fueRofyObVqIdop+Yjly5wETWK/bslM0cD26dMDewycNz5pqqWiWOmxQf6xfu8pglkgiTH/Be3fvShATaie/i/e9IYgJCRz7rlHdmrh4/hzifR8fyVOkFFJEAkcZN2o4Fi+cL2aqUaNExepN22XNIfEbP2aE+BzSXLPET2XQf/Bw8ZujKSiJO4WYMxgL2zFmxFAJcELMxF+va08h6qZiSQPH56jBI+k/sG8PokSNKm2tVqM28hUohJpVyuHSn7eCBLXYtGGdkHtrPposVyNwXKtpEs+LgAULFsi+oEns2LEl2rA9a9/ixYvFR4tzmms9STXXea7x5oTpYXjWYgRebR9x5RxXZfsOAnrPiYrA+c4YsNgSReBCr5P1TkzWWGngrPcbg52QHGnRp5gTjASJ/8aQwiRKDFBCoc8eg5Ts3LlTnPiZaJp5xIzFmMAxyTYP6HXr1sWKFSswbNgwCZJBKVSoENq3by/h960Jo3Ax0iM3dB6Kufnzz5EjR8rhh4d8ltG8eXO3DFB+k5gxqErevHlBMueN4on9Tof9okWLolWrVnKQzJYtG4YPHy79zzHHv3PcMQE5xx4TnjPhrKvEUQLnqnqochUC7kJAI3Djxo3DmDFj5ALOVOLEiSPREW0RuJkzZ2LUqFGSZzJlypRSzNWrVyVlSadOncw2ifM8Xbp0Eh1ZETh39bpvfEfvOVERON/of6utUAQu9DpZ78RUBM52n5ke5PlGt27dcO3aNbmVJtFq0yYwtxAd20mcSMIYdtgagaMGhSH9SfBI6uhsHS9ePBw4cEAO3SyXZKx8eevJiUnMeGvLW1wKNX902mZES36Dh3uGx6d2jt8ZOHCgHBDu3buHsWPHSm46huFnlC5qcJjjhsJbZbaFkiBBAjDFAet37tw5idZI0yuWX6tWLfTo0cMA5JYtW4S0UetHwkEcvFE8sd95K8+cdz179sSjR4+QJ08e8AbfXKRPHvIYRTV//sCgH64QReBcgaoq05MR0AjckCFDEDduXLmwo0UG/6MZb4kSJTBnzhwDgaMmnyRv27ZtYM6yRo0aGdbL+PHjY/bs2RIG31Exvgh09F31vH8ioPecqAicH4wXReBCr5P1TkxF4Gz3mbmDPG9MeYimxqxXr16G8PUkSxUrVkSzZs1sEjiaw9HE7fLly4ZK5MiRQ8qk+SQJHPPP0EyHeeN4YGBdjIUmTCRV+/fvN5jwkBDyoE9ixt/5Deazo3DT581x69atRWNDcshDPjWAdMqnNofk8/z58yhWrJiYATFH3uDBg+UbGzduFHLIgwr/Y/kkESyfdadUq1ZNtEQ09WN4Z/pIsSxvE0/sd2p7y5YtiwcPHgjx7927N7p2DQy1bixMnt6gQQO5zSe5d5UoAucqZFW5noqARuCofeOlFtfpVKlSyfpILRovwbT1nxo4riNcZ3lZRlP5JEmS4NKlS2JOSwLIeayZyTvSZkXgHEFLPUsE9J4TFYHzg/GjCFzodbLeiakInO0+M3eQp8aKxI3aKGqn1q9fD5rNkGgx0TPN2G7cuIGMGTOaNaGktouHa2sEjmZwvKH9P3vnATVFkbXhUkREWQO6iqxxjb85oIA5B8wZUQxrwjXnLOacFXPOri6imFkVE8GECQUFE2ZFTCuChP88pfVt0/TM9NT0dJh56xzOrtDTXf1U6PvWvXWLgsjiMNLhw4dPU2G8ZPwb2dBc4ToMBPZEsW+I8Emew8ovH31WgTkYnTrMNddc1ouGGAvW98orr7SilNVhCqE6GBsINvbYHXXUUfbgcO43ZswYK/AwWjBQODCVe7HH6KqrrrKGDV6iopU8tjur/ITr8r+0A8YjhiGC2RX6JGGzMF9jjT9Sx9erSMDVi6zum1cCQQFHlAMLJIxBFsQohJCzL4753Qm4++67z+5FpuBBZ15kngwLOELqWZwhFD+4sBfFQgIurz0kv/XytRMl4PLbponVTAIuMZRV38h3YPIg7YErj7tUKB1JRDCSTzzxROuVQhThVbvtttvsDQkxZGUVgeTOFuPjzEedf/vmm2/KhlCGa4WhgEcO0eUKnjrE2X777dfyd4TOsQrcpUsX+3cIukGDBpk777zTCi63b4I6I7LceWYIRlaSOegbA4M9HIQCUZyAo9541ngvkmdwPw7rJuySMFIMGYQtm/gpJEwg1C+4T7Dqzp3RD/LY7iwM0Mcw/CgkKkF8ExZLwevKvsybbrrJbLjhhnUnJwFXd8R6QM4IhAUcYg1vXDkBF9xDzT5V9r0R2sziH2M1GEIZXEgjegHxRyE6IhjJIAGXs45RgOr42okScAVo3FqrKAFXK0H/3/sOTAm4yswx5Pv169eyz4iPMYlHnnrqKZs0YtSoUVYszTnnnNZjFdxzxP/fddddWxKInHnmmebVV1+1CVAoeFC4F3skgiGMCB/EHhvVKSQ4YVN7MPMkz2WPHclL2rVr1/IihOvg+Tv99NPtHj0EFmGOeG3iCrhwCCWGBJ40PI1kQcOzw167kSNHWgYICAQc4hHjJGiQdO/e3V7v9glWJp6PK/LY7ux1Y18k4ZF4VxHw9ItddtnFel0333xzK97TClmVgMtHX1Ut0iMQDqHEa0boJHuXl1pqKUNiEha52PvrPHClBBxjlUUvIjr4LYUoC5I/hZNfBd+Q/dIsCvqGX6ZHS0/KEwFfO1ECLk+tWKe6SMDVCWyM2/oOTAm4ynAx5NmHFjxGgH1HzsPFHTCk8Y4RNhMsrKYiXPBA8dHlI01ooxNmXI+Ac8cI4PEiHbUzzjEWEF2EZ5KkBIHkColDeKbzkrm/x3Do1auX/TeEIOKJMEs8Ne6jj+cNDxyeNK6h4IEjBJTDjil47PCyUTBE2O/Bfrg33njD9OzZ09aL/Ry8F0K0a9euNlyTEFL+zZX+/ftbrxy/K1LJY7vDkAyUtB1hVoh19kZSEM3hzHfhVfuk+UvAJU1U98s7gbCAY88vSaH4JjDHsg+ahTwW6ioJON717rvvth48QtqZ5/nO7LvvvnbhL6oQHs2xMczvzMdk+3Xev7yzU/2yJeBrJ0rAZdtuqTxdAi4VzJEP8R2Y3EwhlNm1m54sAkUmIAFX5NZT3X0I6CBvH2r6TR4I+NqJEnB5aL0610ECrs6Ay9zed2BKwGXXZnqyCBSdgARc0VtQ9a+WgARctcR0fV4I+NqJEnB5acE61kMCro5wK9zad2BKwGXXZnqyCBSdgBNw+/zzCDNzmzZFfx3VXwQqEpg4YYK56epL7Z5lsv/ONttsFX+jC0QgDwR87UQJuDy0Xp3rIAFXZ8DywGUHWE8WARGYjoATcEIjAs1GQAKu2Vq8+O8rAVf8Nqz5DUhV/eGHH3IooE1iQAYmigRczWi9b+A7MHmg9sB5Y9cPRaCpCZDAhkx4JFYgsYqKCDQLAZKNDBw4UB64ZmnwBnhPXztRHrgGaPxJkyaZ0047zX6wOQ+KQvgA51+dfPLJZujQoTZTHYXzpDijSCUdAr4DUwIunfbRU0SgUQkg4tz3oFHfUe8lAlEEOAOOhWwVESgCAV870buH19s7gCAh1TfliCOOMCussEIR2iH1OvKRPuaYY2y686jSsWNHs+WWW0rApd4yfzzQd2BKwGXUYHqsCIiACIiACIiACKREwNdOlIBLqYHq9ZjrrrvObtgtVTj3iVBKwisp8sDVqyWi7+s7MCXg0m0nPU0EREAEREAEREAE0ibgaycWQsBxICKH2apMT+C3334zU6ZMKYtmpplmMoRZSsCl34N8B6YEXPptpSeKgAiIgAiIgAiIQJoEfO3EQgi4RRdd1Hz00Udp8izMs9q1a1dxk3rwGnng0m1a34EpAZduO+lpIiACIiACIiACIpA2AV87UQIu7ZbK4HmEUTovnQRcug3gOzAl4NJtJz1NBERABERABERABNIm4GsnFkLA7bfffgYvnMr0BO666y4zfPjwsmiCHkwJuHR7ke/AlIBLt530NBEQAREQAREQARFIm4CvnVgIAacslNN3p/9+9ZX59NlnzYJbbWWzULIXLqpwJkqPHj3MzTffbP9ZAi7doek7MCXg0m0nPU0EREAEREAEREAE0ibgaydKwKXdUgk8b8KPP5qXzz+fg57NOueea3744Qcr4giTdKGSJC7hz4UXXmjefvttHSOQAHefW/gOTAk4H9r6jQiIgAiIgAiIgAgUh4CvnSgBV5w2nqamU37/3bx46qlWwP1p7JsnnnjCvPfee1bYLbfccmaTTTaxh1kOHjxYAi6jdvYdmBJwGTWYHisCIiACIiACIiACKRHwtRMl4FJqoKQfExZw5e4vAZc0/fj38x2YEnDxGetKERABERABERABESgiAV87UQKuiK1tjJkyaZJ564YbzEoHHljxDSTgKiKq2wW+A1MCrm5NohuLgAiIgAiIgAiIQC4I+NqJEnC5aL76VkICrr58y93dd2BKwGXXZnqyCIiACIiACIiACKRBwNdOlIBLo3UyfoYEXHYN4Dsw8yLgTjvtNHPCCSeYNm3aRELkjEGS6Mw+++zT/Psrr7xiDjvsMPPmm2/avZgPPvhgdo2QwZMXWWQRQwZY/vz3v/81yy67rDnuuONM165dp6vNVVddZV577TVzyy23eNX0m2++MXvssYf58MMPbTtdffXVZu2117b32muvvcx//vMf89e//tX+N21x/vnntzzn3//+tzn99NPtvlnKI488YhZeeGFb54MPPtjW6/fffzfbbrutOffP/bZelYz4Ee+7zz77mOeff96stdZaSd02k/sUob1pz5deeqmFz4gRI2ySK/7+3nvvtf1i0qRJ9t/33ntvc+SRR2bCsl4PHThwoNl8883N0ksvbR9Bvz788MPNvvvuW69HGvrFww8/bFZYYYXYz7jtttvMnXfeaQYMGGB/89lnn5mFFlrIzg977rmn/bvzzjvPvPvuu+b222+f5r6ffPKJveaNN96wRy8NGzYs9nOLfGGp71Ct78Scuvjii5szzjjDnHzyyZG3K/rYT2Our9QORWD47bffmn/84x/m008/tXPH6quvbq699lr7ja91/vS1EyXgKvWsBvh3CbjsGtF3YFLjqc6qzq76hg/juHHjzBxzzDFNLSZPnmxatWpl/z1KwH3++efmiy++sAbE448/3nQCDuPpoYceajHcELBM/k8++aSd+IOlVgHHffkAnnrqqebVV1812223nfn4449t+2CIr7zyyubQQw+drhfRNhwx8uyzz5oOHTpY0UZ7tm3b1px00kmGNrz11lutUb/11lvb+u+4446J9UZE26yzzmr+9re/eYvXxCpT442K0N7BV/z666+tgU8/mXfeec2gQYOsocr//+mnn8yqq65qbrzxRrPuuuvWSCY/P0fAcSSREzX078UWW8yMHTvWzDbbbHWpKIz79etnVlxxxdj3HzNmjD3uh3mVTNJ33HGHXZRBeLpFnk033dSOXSfo3M2Zq0lkxm8ZwxJwsbFHXohoGzp0qF0cGz16dOQ1RR/7acz1lVqhCAxZ7GE8XnTRRTbb+xZbbGG/iwceeGDN86evnSgBV6lnNcC/S8Bl14i+AzMPAq5Xr142eykZTREDHTt2NPPNN5/9kLEaxQqwE3B/+ctfzPHHH29Y1Wc1ChFAQQAgZJrNAxf+IMECTybsOJMRzxPeSTxjeOcmTJjQYpzxgbj//vutcMKgvu666+wKfKkCe+7LtZTOnTubc845x2y44YZWwK200krWGxouu+++u1lvvfUiPRBbbbWVNRB33XVX+zO8M1dccYWZc845TevWrW0/OPHEE81BBx1kjf3XX3/dev9uuummFmMVw7NPnz52tbJdu3bmyiuvbBG0I0eOtPXDU7vMMsvYVU3eo6ilCO0dZEt7Ypj27ds3Ejntv9NOO1nPbpZl/Pjx1ov8zjvv1NzvwgJu+PDhVqCy0DTzzDPX/Jp8Z4899ljz888/W4/2mWeeaccdDPGmffXVV3bcYzBT+G/+HRHNe26zzTb2NxTENN61NdZYw47PLbfcsmV+ZTy1b9/eUP9S80L4XWt+uZzfILiQ+MEHH1ihTmQC8+r+++9v5ykK15111lnWK8oixmWXXWa/Y4yDH3/80dxwww0tixYsUrIw9tRTT9l58JJLLjEbbLDBdCSKPvbDcz3v+dxzz9nvdlqlCAyJSPjll1/s95hzlzfbbDN7fBdjM1yqnT997UQJuLR6aIbPkYDLDr7vwKTGefHAOQ8bhhSi48UXX2xZseaDiCGChwcxgpHP0RWuSMD9L3SKlXiMN1buMC5gg7ejS5cu9g/C7u677zYYX4RmwBYRdN9999nQxqiC9wAPFh8UV3bZZRcbKkZ7IeC4H+KI0EiMF+cNQHh169bNvPDCC7YefIgImeW5/O9bb71lxfjEiROt15BQLq6j0CcI08Kgefrpp836669vRSdeQDwAhOohIhHuGMc8g5VKDHEKhi6rmVyzww472I/hfvvtl91ArfHJUQZI3to7+Ip4cy6++GLbF8MFg3adddax7c+iTZaF/oPBxBE5tfY7xgH9famllrJ9mkUPvN9JhFB+//33diGCkOQ111zT1hVv2CqrrGKFGUKBsYrHj/Ex//zzG7xozAewZrGG8Udd8HIzFhiveIBoK0KZGdOIOgRnz549S3qEeHazCjg8qSxg3XXXXbadf/3115a5tVOnTnZuu/zyy80hhxxinnnmGds2LDIhsh944AFzwQUXmJdfftm232OPPWbnS7zTLD4NGTLE3jdcij72w3M9Y4R5nvk9rVIEhozn7bff3gp+Flx22203c8011yQyf/raiRJwafXQDJ8jAZcdfN+BSY3zJuAQA0sssYT1vLjCBxEhwIcwao+ABNz/BByrvKeccoqNmcegc/vUMBIwJgmP2nnnnW0YpAtZZRUYQYxwjiqVBBzGnjPCERT//Oc/zahRo2zoIvtyMBIxXHgO4SCEX7JajSDEo8pKLHWh3TEeMTDxWvCRp56EVRJe5Mpcc81lDX8MHsSo23vHvyNauRZv7oILLmjvveSSS1pxCgMMpKKWKAMkL+2NRyIYIoiY7t69uxXkwcUW2PN3iHL2OyKssy4fffSR9RIjbmrtd3wHwyGUiC1EImHGtZRHH33UGv/06WChX7AA48KmEXSIRhZRGFfLL798y+WEMLPowvzK2GFBhzmBBRlEBHMHApAxDRe8RaVKswo4+i9CDfHmCmIEkYLodQuORCvgcWO+Yq5jkYn9g0QrYKhTggtLzLN4ReFOFEK4jYPh8vxbkcZ+eK5fbbXVrLhl0SCtkuf5030vCWNm8YVFasYq30tEHF51V3znT187UQIurR6a4XMk4LKD7zswqXEeBVw4HI8PIt4U+hiGeDgMjg35CAeFUP4RQomA4YNw6aWX2pV3CuKFv8dgQ8CRaKQarwDhiQhAwhoprEBjgEeF+7Caf88991iDlTAPjBSMRgofKNoRr1+4kDTh7bfftoY/SVEwWBChhJUEBRyhXQg4PnKEWp599tnT3Yv+QEgS++7+7Ofmyy+/tGGYhJMWsZQKAcpTezuutDde23DbIAwIa0VAYOzmpeBFwctba7+LEjWIIwzWo48+uqbXxVtDWGqUgAsa9zwLzydCDvGAIRgVvklbsGjCGKMPMZ55fxZRiHhgMY0xRJgr8wlCnH9n/FGaVcCxfxDPJoZ0VAmGWhIOR/It9jNR+A17DwmBZYsA4aksQLHgRPnuu+9s8hgXjunu3whjP8iKd0So8O1OqxSBIYstLJoQLeO+lyyskHCIUsv86WsnSsCl1UMzfI4EXHbwfQfmn4btH6kBMyysEmO48zHD8MPwD+6nch9EVopZNSZhiTMiqHYze+CCyQsw4uDHfgo8XnijYMbqMJkpEV0uhBIDD2MMA4/9Lux1QTiXKhhz7NXo3bu33VOGF43VZAwPjJIFFljA/hQPF6INgxChjZBjLwjthgcOMYeoPOqoo6wRw+/x1LHqjLeQevPxIvwM4xKjl77Aai1eEv6dEEpCTAizZXWS/8XbhpFEQgW8tdSBP+xNcQVvH/dl/0URSzhZRR7bG670N8Qb4VF4c1xBQCP4aYdwYows24NEI3g88CD69jsWBvDKhEUNHhj6IyIpah9LNe+N14bFh3/96182oyr9nTBj7h8l4BhnLNTgAWTcOgOQvXO0D4XFFv4bbzbXImR5BkICr6pbAImqZ7MKOBdZwKKiW5hC4M4999x2Po0r4JiD8UAxN7pCGC+LG/SnYCn62A/P9YTDV5s5tZqxEnVtERgSZcR3mKzNfJf5XjJ+yS5d6/zpayfmQsAR/81+iGDB5Y0RQiHsoZo0vLV2pkb7vQRcdi3qOzCpcR48cKRPJu7fZQvEkAhmNMTIx3hhJZOsdYQHIVL4u4022sgaHcSL8wFlvweJUZqh8EEiqUfwGAGMYwQQK7942DCiWeElSQyb7RFwFAw2koFQmBsJ0WAOLFUQg3hMEFo8kzZw2QM33nhju1mfdiKxDCv57t8wDjF0CP/i3zEq2R/CXEzIJt5A/j9/+Hj179/fGpTUCZFI+5KCPpjEhD7g9tgROkaaeq7H+MZI5j0IRUJYYlC5wiIB98NgD38LitBfitDecCQpEcKdzKPBwp4r9jvSNq6QdS1rMYfRjOe61n5HeGNwDxzvyJjDA+cEVK39jKQwbvEDoUBCEuZKFnKc/eI8cIw1vDx4sJkH8KAhUmkfF1bJnlHmBOZS5l8KSU0QhiySRBXmW8IHeTfEOvML+7uiPOG1vm+efh/8DrFARd/FfmRhCgYIMfYdBq9jHmaBkmsozEnsY4QbQpk+E9wjyjzGYhh9MrigVvSxH57rESgcG5NmKQJDkg1hv+Bpo88wFtk/iQe91vnT107MVMCxYou4YLIDCKvQTKgUCbjkho8EXHIsq72T78DkOXkQcNW+r65vHgLNtsrfPC2b7zdVv8t3+6h2IiAC1RHwtRMzEXCEFxBqQyy3W/3gdRFyrJaQmYyVf3ngqusEpa6WgEuGo89dfAemBJwPbf0mTQLs98GjFg4pSrMOelbzEVC/a7421xuLQCMT8LUTMxFwQe9aVKPg6nZZgfh3hVDW1nUl4GrjV8uvfQemBFwt1Bvvt4RjRZ3Zxd8RfqLSWATU3o3VnnobEYhLQGM/LqnS1xWNoa+dmIqAI9OSi9lmjwMb9Pm7UoW9GmRWI0ZcAq72ziwBVztD3zv4DswiCDiSALDnik3QwXTxjhX7VtiPwH6MohXmH5JvhNOsJ/EeRCCw6RluLgNaEveNew/207BAFveML6Ik2F/HPrZmKbQ7iWHcgfTl3hs27PPMurA/h72oUYW2I6KFzfdZlnJ1DNeLzKr1mjvqOb7D78H4YX8VfaTUeGffJ+OSfaHBIx/KtRVsOLOMPVtZzCNZ9COitNhjC6PgHtpydWHvGvxJl1+JE3vkuDd748IZlaOewX4o+lKpced+g71LHYIRZz78mJf4LvHMUoV+RjZO3jvLEncO5Vvojm5Iur5851jgdJlEk75/UvfztRNTEXB0XLcpn47shFm5l6eDssmUIg9cbd1EAq42frX82ndg8sw874FjwmUxhmQIZJ0k42CU4cLBtmwML2LBqOKcMoyGpAof1U8//dQmd8mqDBgwwL5TMJtoubpgEJAmPXi2UlZ1T/O5GFuckxc+9ylcB5J/kIgm64JxylES7jiJYH1IdMO5fWSTzbKUq2OwXiz+kNGxnsdK1GN8h9kishDOlRY/EAJkcyWjbJxxhmjjPL9yC+FZtnO9ns2YpB+TXCTqmJTwcxFt1SxccN4bx+HwPSuX+dc9h+Qo9CMSPpUrZA1FqAQzNPsyYgwRpRa1aMo9eV8yb0bNA77P9P0d7cWcU05sk/Hy73//u+8jKv6OuYSMrhxrk9fiayemLuAQZaTBrVRYBXGTkwRcJVrl/10CrjZ+tfzad2DyzLwKOMYv+1cpLuMhZxKFC5M3mQiLnEE2yck/rjFXS3+L81sMIAx6MlDGKQg4PrL1NKbj1POGzcgAACAASURBVCOLa2h/MqiW81ZyDlAcYy+t+vPtxHsY9B7T3hwxgSDNQ2GBlrqU8nDDHePYZTStV515Dmd/1SPzKfYLizVxvC4siHEEC1k/Kwk4sspy/lTW3tR6tUm5+9JenDNJqnuy1pYr2Jrwr+R1C96D33DsA8n04oxpjsnBy0OW3nKe0yQFnKsvXr+oKBEEXKVjJtJsu0pzKGfEcixNPQtjkAy7cb3b9axL1L197cRUBByrGoRNUr7//nubNrfcihTuTlzYXEuRgKutO0nA1cavll/7DkyemTcBx0SMIRj0oJNanvOkolYgG0HA0Q68B564OOF0pfpKNcZcLf0tzm8l4OJQmvYawtsIxYkSHHkTcNQ87F3Km4CLqmOQeFoCLqnxHe5RpPonQoH3iFPiCjhC0jmfLO594zy7SNfEFXDYjng1qy3VCjjEB0J6nnnmKesRrIeA490QjwiTYJRI3gScawO+n3jawnNoGgKOOiDkOWORBbm8FV87MRUB52BN+PFHM9Oss5pDDjus7J4BOiUuYmWhTKabScAlw9HnLr4DM28Cjg8nISBuUcWxaAYB54y8SqEgpfqHCzf16T/1+I0EnB9VFhYR8uH9FHkUcH/OHy3epTwKuHAdsxJwbnzHCZet1HOiFrkq/YZ/ryTguC9nJY4YMSLO7Rr2mkoCjn/HQxknyisKkq+AQ8SxiFkqTK9eAs69AyGCzoucVwFHXRGaeJiDc2haAs7NN4RzIuTqsb/dd+D52ompCbgfP/rIPNazp9m2Xz/zW6tW9hgBt8ct+NKsdHKw5tVXXy0B59sbQr+TgEsIpMdtfAdmlIBj/xSJQ1hJIgTJeYT4aLEnghA9kooQWuEmJ8LfGGdMnHi1fUKFuD/7WHl2uDSLgHOTf6VwurARGvZYenShxH8iAeePlLFHyF3QUMurgHPCBNFJwpw8hVAGWyDKw52mB87VpVKoV6VeQ9/A6xY1T1b6bTkBx7zPIeHcu9lLOQEHf5J30Nd9i6+A43ks8HXp0iXy0fUWcIwhhNEss8xi7eY8hVCGgYTn0DQFnKsLnIgcqlbE0f/oX+ytxNGETeXuwd/xb/w3f+++EbQNHnn+lxDOqDBOXzsxNQH3zbBhZtRDD5mVDz7YtJ1nHmtwXnnllYbTzXkhNvUTz3vQQQfZDY+cBi8PnO80NO3vJOCS4ehzF9+BGRZwrPCxMdnF8zOR8MEgU1b4o4WAI9sbkwl7ANwEg4gj3KKazbw8j3tELbZQx2YScK79S4WCBPsH3BC9zHN5KxJwtbVIeD9FngWcE3EsjL7//vu52QMXboFwwpgsBJyrU7lw2VI9p9bkRKUEHAtwL774Yt2y9NU2EtL/dSkBx/cRexEjupZSi4CjDj169IgUBfUWcG6cE+ZNX8qzgHN1dXvSshBw1AG7CM9lNRkqWURBjLmCdsFzzxYJp1fcvXk/7k02Xbdflf7LfupwMhtfOzE1AWemTjVDzz/frLDvvlbA/WmgWsMQQweXZjAVa/CsOO2Bq2VKMiYo4Fil2X777b1viLgulf3I+6YN/EPfgRkWcMTzh8MX8aYh4tgXES7EmiO8wqmEmTiqSR/PpFQuY2IzCjhYR4WCuDaAOdzykFo+amhJwNU+4QT3U+RdwPG2hJThGWLVOa8l6AHLUsDBp1S4bBS7ajMdRt0jSsBhKCLessxYm7e+EiXgmGeZb5NIm1+LgIMV9lVU4p00BBzPZ14ivwR9p0OHDnlrvmnq4+bQ559/vu5JTEqBoA6IOLxplQrjnEX0YEEEIuCwz8J5PfDC4ekjCie4Z5XfkJk76P3ztRPTE3DGmKHnnTeNgCsHTAKuUneK/+9BAYfBj2HvU+hw3bp1MzvuuKPPz5vyN74DMyzgaLOwN4f2cO0ZniCi/p57Mqng6Y5TMPjw7pUrzSrgmPjxGqy88srT4SFpE/sF81ok4JJpGYw9wikZA3Ey1iXzVL+70CdJPoSxkueCWMEgmnfeeVPJQlnOsCs1vsO/YVW+1nOsogQcZ7wRoaTyPwJRAo5vlE/YahTXWgUcIhIvXLikJeB4Lh4fokTiLtRm2b/gzQJ0lpmqsas6d+5c0ROHl43xGLa1cD7xvQ9nm8U7RztE7cdEwAUTz/jaiakKuB9GjTJ/WWgh0yqG2pWAS25YJSXgqBECbqeddkqucg1+J9+BGRZweN+YJNzkgXhjgmDyIDQqPKmw7wXPdvCcIH7DpB73AFR+69LHlzoLrRkFHB9pzmAiRHWfffaZ7rwdNtE/+eSTdiW2mvCMtIaCBFztpPko9+vXz6YyZ7U77wIODwX9cvXVV6/95et0Bwypvn37mlVWWcWmced8vXofIxD1KoxvvAIYl/vuu68Vk+UK8y9jqhYjNErAvfvuuzZ7d949KXXqDpG3jRJw77zzjuFPEmcx1iLgWNQbOXKkPZ4lnLE4LQHHXHTbbbeZrbbaqq5nqyXR5m4OZXxtvfXWSdyyqnvQl4YPH26z4u+xxx4Vz6tFoNG+wWMpsKk4AoFjlbAHgvYZ5/CRk4DF96B9xt+Fj03wtRNTFXDV0JWAq4ZW+WuHDRtmP4YUBgsuXd+y7rrrWhGnEo+A78AMC7hwdi32aeBlI4wymKIewcDfI+7C+zIIfaX9q9m4i9HHIc54mqISoDSbgCOMgsPLmcx79uxpBXRUQTw/+uijplOnTnYSz1ORgKutNRBDzz33nFlnnXXMZpttZufWPAs45nuE0RprrFGTyKiNWvlfM8889dRTdp7Zdttt7ZhJ4xy4cK0Y308//bT96913373k+A7/juyQLOogkH0WbUrtgUOYwCVvWfPq2RfK3bvUHjgydJKkh/m21GJjnDrXIuDwcBPdEiVG0hBwLPAi+PFgIxDI5J7Xwhw6cOBAO4cSeljp7MOk34PvN/2FerBYFFf8EyHgtqbQz2hv8hBwPzzBLlcAW1Xgj62FUGV+o+8SUol9Fg7Z9LUTJeCS7hk5vV9wZSC4GlBtdasx/qu9dyNe7zswwwLOsaHt+BP1kWJlqNTf026+bYf3D08cK8zhiaeZBBweAjxrrKxxcHklTyZGGRu0+TgRupqXIgHn1xJuxZYFMUQGGecYU3neA0d4EsJi1VVXtZlrOeA7b4WQL+qIN3PjjTe2AiiLPXBufOPxYnzPOeecVaFiH8wjjzxiPYgYatWUclkoMQwfeOABO+/4ZBGuph55v7ZcFkoMaw5DX2211WLtaYp6V18B5+YGkvDRTuFSbwGHF3jIkCFm0003tZ43xlQePbfBOZTDz7t27WojGep9kHewPcjWjXjkf1mEjbulJHgPbK0om6rU35ez27ivr50oAZf3GUv1KzQB34FZSsBlBQPjhomW2O2gcdIsAg4vBivhiNjtttsu9qHebLBH/LoVuazaL/hcCbjqWwFvNgYSRjrGPXvfXMmjgMOQeOutt2xo2Q477GBX5fN2jAB1fPPNN20YE3XE8A4uVKXpgXMeQDypjO9qBZirNyFsLqteNQKw0jlwLKLdd999NrLCt27V9/r8/aKcgKO27PV03yk8I9UWXwFHu7NAErX/jTrUS8Dh+WFBidA+PEn037weIxA8DqN79+4tIYtpZqEkzJHvOMcB0VbhbJDV9pekrve1EyXgkmoB3UcEIgj4Dsy8CTjqQ6gmH0eX/pe/awYBhxeDPTFBD0E1nZ2PLB8NwrPSXGksVUcJuGpaz9h2KxdWlzcBR8rqQYMG2VB5wgDpc3k7yJs6vvTSSza8iDoypwRLmh4456XccMMNWzyA1fWQaa+mv7Bog4CLm0iikoDjCYRvYexyfx9xUss75eW3lQSc4/TQQw/Z7QJR3rBy7+Ir4NivuNdee5X0cNdDwBF6iOeaxVXGEEcIUPIo4OizzzzzjPWsh0OT0xJwzDVE0Cy77LI2EzvbUPJSfO1ECbi8tKDq0ZAEfAdmHgUcdSLsgI8jezIIC2xkARf0YjDh15IAgg8XWeUIdYnKXJlm55eAi087TlhdngQcHl/2wXDWI4aSC6PKk4CjjgMGDLD73KhjlJGdhoALeymDHsD4PST6ShZtaAcWvUgoVanEEXDcAy/GE088YffhzPPncUyV7t1I/x5HwAU50Q7VHJ3hI+Dw3hJOt//++5fcppC0gKNfMYbw+u22227THO2UNwFXaQ5NQ8DBhL3r66+/vg0zzVsosq+dKAHXSLOb3iV3BHwHZl4FHPVi9Zy9HoQf4Jlij9fOO+88HXs+noT91JKdLasGjfJiJFEXPuR8SOKkLU7ieVH3kICLRzZuWF1eBBxhXOwBYp8b4VTBc1XzIuCoIwKEOjJnBOsYbJV6C7hKHsB4PaTyVYw1QkTZF1duD3JcAffnd8Ge7Tp06NDYSVYq17QYV8QVcEFOH330Uexspj4CjjZmL1W5IzqSFHCE0zKG8KyTERwRFyx5EnDMOy4jc6mtB/UUcPQXwsjJNBnct5y33u5rJ0rA5a0lVZ+GIuA7MPMs4Kgb4gxjkQyVeKYIHwkXrrnsssvsJmGfzGxZdgT2FJCwJejFSKo+ZAsjxAomlRKhJPXM4H04t4qQFtJdxyl4TK666iobouObCCfOc/J0DXzYWxInbJY+jqcLr1dWBc/M66+/bpOVkBwgnPkUQ4o9ZT4b9pN6J+r42muv2fmCNOflsrNieF1yySV1mzvc+MbwrjbMrloeHP6LwY2hXSr0kXA49iwyj8bNyMc7ENLOnriwEV9tHYt0PTwJd2VsxinvvfeeDYFmsYA9hOWKG0fsc42TWbZPnz72DLrjjjuu7DcOAUcYfa2JRVh4YAytueaaZosttoiccxBwfF/ihu/GYehzTdw59P7777ee6nqEBbNnmeygeCnjjiufd631N752ogRcreT1exEoQ8B3YOZdwP1ZP2tAUFhdixJwGGGsuhetRHkxknwHPvp4J/nYZlEIgf3nP/8Z69EIuCuuuGK6g+Rj/bigFxFis+WWW06TWKPUq1x++eU1Hc2SBCKENeFB/InKRIuAu/HGG6c5FzKJ51ZzD+q4wQYbmPXWW69iqncE3MUXX1zzAdml6sdiRDkPYDXvFedavLkI6KhDfd3vWQBAOFRjaGKg3nPPPQavTDMVBBx9KW5h0Qr+cb5FjB+S6sQRcG6vI+nwyxUEHM9HqNdSqBuZWtdee+2Si2l8U26//Xa7SJdlYQ5FZFbaesARJ4jSehSihBhTWYvZSu/maydKwFUiq38XgRoI+A5MHjm1lvMeaqhztT9l1bJUTDkrhkUseMdqOU8ozjuzBwcvZVYlrseIbkgbN1uJyycvfbxcffPShnGZ0tfqyTWN8R0eL4z14CHAUePJp15x7ttoY5e5udqojmo4xb0/9+TaSpEJSc71lcZQXsY6fa5SXbmGb0u9TJ247Zj1+PC1EyXgsm45Pb+hCfgOzCIJuIZuQL2cCIiACIiACIiACNSJgK+dKAFXpwbRbUUAAr4DUwJO/UcEREAEREAEREAEGpuAr50oAdfY/UJvlzEB34EpAVe54Ti/ifOjzjjjDHPyySdX/oGuEAEREAEREAEREIEcEfC1EyXgctSIqkrjEfAdmBJwlfsCoo1U2gi50aNHV/6BrhCBJiZAYh4yFvKHrG8caEv2vK5du05HhayjJBa45ZZbvIiRNGWPPfawY5Nsk1dffbVNvEAh0yJnpP31r3+1/73JJpuY888/3/7/gw8+2B7w7cqIESPMhRdeaP+edOOnnXaa3XNEMggSJ5111lle9Svaj4redvB+7rnnzDHHHGMPJGfP00033WS6dOlStKbwrm8R2pCMjf/4xz/sOYPsQSUBybXXXmvnDArj7dZbb7X/v3v37k0z/rwbPeYPfe1ECbiYgHWZCPgQ8B2YjSTgMLhOOukk8+ijj1rDsXfv3qZHjx4W5yuvvGKOP/54m+GQDeEnnnii2XHHHc3HH39ss4Dtt99+Nv0y/0a2P7KPUfhvPoj8G1mmyHZZTVYyn7bUb0SgyATIvPjQQw+1nMuIIMJY45ymcKa4WgUc92V8nnrqqYZzshBbjGkST+y99972MPtDDz20LE4yN1JnfjfvvPMazuhy6fIxLtdaay1zwgkn2POdGr0Uve3Ilknqe45TIMsm7YeQm2OOORq96VrerwhtePjhh9uEZBdddJFNuEMWya233toceOCB9sxXMhfzzWYc056nn3666datW9O0Yb1e1NdOlICrV4voviKgPXC2DyDgTjnlFDvZc6hqp06d7BlbnMuD6OIsOc7HIcU251gNGTLETJgwwfz973+3q7QYfHja+JCwos9ZPvyG1cBBgwaZK6+80v7mrrvuUp8TAREoQSBsQHIZAgjv9c0332z22Wcf8+abb1rPGN45xqDzwGHQcV4TGeMQU9ddd51ZaKGFSrLmTCfuy7UUDq4/55xz7AIM45nFmcMOO6xsW+GVY9yTZjxcWPAhdTvh08wLjV6K3nYuxL1ZPKZR/bEIbXjkkUfahRLGN8cQbLbZZtZrypEqBx10kFl44YVbzg+95ppr7Pf3jjvuaPThV/f3k4CrO2I9QASqJ+A7MHlSUY4RqEQFAcfZNO4QYVbjt99+ezP33HNbTxwfNlfGjRtnQzRYvV9yySWtEelSNLNqj1hj5Z1zevi44KEbO3as3QuHOJxzzjkrVUf/LgJNSSDKgOQcR7zjrLQT9sjYQxwR2sYfhB3nVw0cONCGUjGWMdg4w/CRRx6J5Mh45JzB4DlUu+yyi9l8881t+CQCjvsh8jAIMepXXHHF6e619NJL27PgqJsrgwcPNgcccIDhXC28Afx7M5Sitx3zNW399ttv24U6wmnPO+88M+usszZD89l3LEIb8v3l2/zuu+8azv/kAGyEGoWFEv6bsUxhEZVFFkJjVWoj4GsnygNXG3f9WgTKEvAdmI0m4AiDciv2hDzttNNOZq655jJnn332NHteHEyuR8BhBLrz2BBwhHbx99wLT4E7CwijAIOAVUIVERCB6QlEGZB4t/COs8flsssua9mnhqjCg4YHjgOvCYN04W6EL7OogrcuqlQScITTuYN1EZAIMQQZnnVXXnjhBbvH5rPPPos8Y4vxjiignm5vXSO3edHbDuP/yy+/tHsfaWdEPFEX7G9slpLnNhw1apQV0+xVfe+998wVV1xhtzs40YZ3nv/PgivjUgIu2V7raydKwCXbDrqbCExDwHdgNpqAY98bfxBmhFC+/vrrdgWeUC1W9N3etjfeeMP+3eeff25DKDEg99xzT/Pyyy/bDwhGJZ4AEizgGXCFvRXsn+O+KiIgAvEEHCGUhCVjwF166aU2LJGCMOLv8cAh4Eg0su+++8bGyl41xup8881nf0MI5bnnnhu5TxVP2z333GP3xbmCpw4vHgs8pQqr/8wTGJuNXkqF3xWl7Q455BDTvn17G0ZPYT80C24I9WYpRWjD5Zdf3txwww0tyWUQdIRJ3nnnnTaREFE0JD6i8G9sXbj99tubpQnr9p6+dqIEXN2aRDcWAZ0DRx9we+D69+9vV/XIJEfiEQp74Y4++mgbBsnGdsJsWJVnlR6DDqNxwIABdu8Nhhp75hB4F1xwwTShVfz7AgssYDfJs79GRQREYFoCGJCMLReuSEIThBKJgB544AEbQsmCCSGUZKZEdLkQSkIV8Z7gNWecDh8+vOw4w8NCGDSLNiQ9IGyaMGo85njVGKsUDMCtttrKikUWdCg8H/HGYs5iiy3W8hIjR440SyyxhJ1Pfv75Zzv+WdzBO9DopehtR+grhj99aOaZZ7ZiAC8uIfHNUorQhttss40d1whtxjlebpKV0HaEShLhwmIq45itDEpikkzvlYBLhqPuIgKJEvAdmFSikfbA/fDDDzZpSdyCpw4BR0y+igiIQO0EMCBJ6R88RoAMsOx1I3EBiyWIJkKTl1tuObv/FAFHwdAmoRCFxRJE0xFHHFGyUojBnj172n2pPJPQ53XXXddev/HGGxsyTGIEtm3b1nrm3L/x79dff731yD377LPT3J+EJey9a926tc1CSxg2WS6boRS97WgjEuGwQEC707+IpKjmm1D0di5CG/Ld7dWrl11AZYytscYapk+fPlZ0U84888xpjhEo5yEvenulWX9fO1EeuDRbSc9qOgK+A7ORBBwfbIRYNR9rPiSrrLKK+f7775uuz+iFRUAEREAEREAEmoOAr50oAdcc/UNvmREB34HZSAIuI/R6rAiIgAiIgAiIgAjkmoCvnSgBl+tmVeWKTsB3YErAFb3lVX8RaGwChFNFndHG3wWPBmlsCsV8O7VdMdstWGu1YfHb0L2Br50oAdc4fUBvkkMCvgNTAi6HjakqiYAIiIAIiIAIiECCBHztRAm4BBtBtxKBMAHfgSkBp74kAiIgAiIgAiIgAo1NwNdOlIBr7H6ht8uYgO/AlIDLuOH0eBEQAREQAREQARGoMwFfO1ECrs4No9s3NwHfgSkB19z9Rm8vAiIgAiIgAiLQ+AR87UQJuMbvG3rDDAn4DkwJuAwbTY8WAREQAREQAREQgRQI+NqJEnApNI4e0bwEfAemBFzz9hm9uQiIgAiIgAiIQHMQ8LUTJeCao3/oLTMi4DswJeAyajA9VgREQAREQAREQARSIuBrJ0rApdRAekxzEvAdmBJwzdlf9NYiIAIiIAIiIALNQ8DXTpSAa54+ojfNgIDvwJSAy6Cx9EgREAEREAEREAERSJGAr50oAZdiI+lRzUfAd2BKwDVfX9Ebi4AIiIAIiIAINBcBXztRAq65+oneNmUCvgNTAi7lhtLjREAEREAEREAERCBlAr52ogRcyg2lxzUXAd+BKQHXXP1EbysCIiACIiACItB8BHztRAm45usreuMUCfgOTAm4FBtJjxIBERABERABERCBDAj42okScBk0lh7ZPAR8B6YEXPP0Eb2pCIiACIiACIhAcxLwtRMl4Jqzv+itUyLgOzDzLuBmnHFG88MPP5jZZ589NsmVV17ZXjtx4kQzYsQIs8IKK9j/Xnrppc0999wT+z5JXejzDjx7kUUWMePHjzeff/65mWmmmWx1nn32WbPhhhuaww47zFx66aX273jfF1980cw222xlqzx69Giz00472Wv4/Z577mnuu+8+c9FFF5mff/7ZtGvXznTs2NGcc845Zrnllkvq9XWfhAnceuuttv3+/ve/m6lTp9o/Z555ptl6660TflJxbvfaa6+ZCy+80Nx777220ow5+vAMM8xgfvvtN7PqqquaU045xfzf//1fai/10EMPmRNOOMHMMsss5vbbb7fzUNRcllV7PvLII+biiy+2c8qrr75q5wH4XXvttea6666bhtNnn31m5plnHvPee++lxi/ug0q1fatWrWzbb7zxxuayyy6zfSLPpVR7uP684oormmHDhrW8wi233GL22Wcf+x1gPohbKj2H78uAAQNabke7v/7662ahhRaK+4jUriv3Lj/++KM59thjzX/+8x8z88wzm7/85S/m9NNPN5tvvnlq9Sv1oDBT5nDm88UWW8zWt17F106UgKtXi+i+ImAMhkpux1gtDcRHd9y4cWaOOeao+jaffPKJWWmllezvsyy+77Dooouav/71r+b4448322+/vX2F3Xff3bz//vtmrbXWMpdccklVr3X++eebjz76yBpoFAyA8847z2BoIm4pfKi/++47s8kmm1R1b12cHIHJkycbjM9SBYP/4YcfNn379rWXDB061Gy11Vbmm2++Sa4SBb9TcNEE4+j666+3Yor+zcJIGgVDkUWS7t2728eVWsjJqj2Dxm85HsyjnTt3NnfeeafZaKON0kBX0zOCnH///Xez2mqr2UWpbt261XTfev+4XHvwTp06dbJz9yqrrGKrsvbaa9tFyt12280ceuihsatX6Tl8d6655pqWbwDfIERy3gVcEABjnm8kCzdOvL/55ptmiy22MIy3rPtxmCmiDbHJYi3zeb3mKF87MbfG5WmnnWaYoChHHHFEy2p97NGgC0UgBwR8ByZVn8psl2Hh43TSSSeZRx991Pz3v/81vXv3Nj169Ggxes4++2zTr18/8+2335pTTz3V7LXXXvbfPvjgAztmMVwnTJhg9t9/f3PQQQe1vMnHH39svVNpCLh6vAMf0mOOOcY89thjho8uK4p8xHfddVfrMXMeuKDBwsSP0cgK6ldffWVXaGGLB+C4444ziIMFFljA3HXXXWbTTTe1hu1mm22WYes316MR4CNHjrSG14ILLmhuuukm62VloaFXr1623fbYYw/rqXnrrbfs/37xxRdmiSWWsIZH+/bt7f8iuh988EEL74knnjCHH3649TY3QsETxNh2XiAYLL744i1zxLzzzmveeecd06ZNG/Ovf/3LME4GDhxo5wLnoYgSS4wbmF9wwQWJYSo17jGoWSDBUMOrjYfc1QlPAIsytBceL7zgSbZnFD/6z/Dhww32DsbiXHPNZQUAhjkeuDA/B4i+ueaaa1oRioFZ75JU27tFP/pO165dbVt06dKl3tWPvH8S7UHf6dOnj23Dq666yi7iMc+z8MbcgQcuqefccccd1jPLYgclCwFX67vQx/fee2+7YOmiV3gXhClj7rnnnqupLwwePNiOB77DLgIC734pe4QFN8Z869at7XeX/hgUxcxN6667rp0TiDY644wzaqpfqR/72okScHVpDt1UBP4g4Dsw8yLgCG8ivIEJF5GCIcaKHx8uPiZMjBi+rKbyUWbSZFUYIbLUUkuZX3/91X6oMYj5PSVtAZfEO2Bk3HzzzfYdMEwRrqywYthj5H366afWcIVBlIDjN9tss41ddRw7dqwNySDsaf7557d8EYF47hC9HTp08PZuatz5EcC7SfgMBe8ni4d82Gk3RDYCj4KhjbHx9ttvGwQLCxOTJk2yogYBh2DjN/R7Vm0RMnn3MMQlRh9dcsklrYjDmKGfv/vuu3ZFHUOVlfSFF17YetQwUaO+BwAAIABJREFU1PFKxBFwjAnGEQtFSRXmp1Ljfv3117fzlgtt5VoWVRB3GMVXXHGFDfFMuj1L8cPQR/Qj+Jk/MSQpzzzzTEkBx2ICYYj0rzRKUm1P+Cy8R40aZXbccUfLOKuSRHvwLizkrLHGGnY+Z5GTkDuEhBNwST2HMYVHnwUlFlKzEHC1vguLNLBxi1yu7bErWLj45ZdfvLvD999/b5ZZZhnz73//2y5uUPg7olaC9oj7lmPHcD2LOAjuG264wRxwwAHWPuHf+C0LVPw33wO8hHzn61F87UQJuHq0hu4pAn8S8B2YeRFwTFysjlO22247s8MOO1hj1hk9GLEUPBCsviNgEDmIN1d++ukna/j27NnT/lXaAi7pd3ACDoMLzxkC7u6777ahTOUEHCv6q6++umVAuA0rtnz4YcOHEYNYAi6bqePyyy+37YdRzB+MI/ZlIljwIruC2OYjzoIEBdFCGC37GMMeOFblCQlixRyh3ggFwYpRgwBijCMgML7o248//rh9RVa1r7zySivK4gg4Fi6efvrpxAVceNzTTsxBCDiENgsqFOYyQrr475NPPrmlmerRno4fz8dohBv7JLfddtuW+ZF+SP8q5YFDYGJsEtI166yzptatkmx7PIh8S1jcOPjgg1N7h/CDam0PF4bPYg9zOfMD44HFACfgeGYSz+HbwpzC9xePEB7kLEIoa3kXBNygQYPsAmiwJCHgWADi/kEvHgtMpewRtn+weMRCCWXKlCl2POFFRcAxhw0ZMsSKPwqL1IzVekTG+NqJEnCZTR16cDMQ8B2YsMlDCKVbjaI+GBkk28DzFA6FcquBhC6wgszm+lIlbQGX9Dsg4BBtvDNCDEMWQzUoxJxh6JIjuN+4xC18DPBgrrPOOvajHxR+fDzw6ORhU3czjFFWYAn/5WONFw4Bwko6RkZ4ryZthTDAG0t54403rCcBj0LY4Off8UYT0oOx2ggFbzueKwwfQqhhF37v4F6eOAKOMCU8d3g+kyrMT6XGfZQHjjbCM0DdCaWk1KM9g/zY//XCCy/YvoF4xKtGKSfgnn/+edvfMIIR0mmWpNv+6quvbglDT/M9gs+qtT3cdxBBteWWW9o5+7bbbrNhgkEBl9Rz8HyzkLreeuuZs846KxMBV8u7lAqhpC+wmFFLCCVbGthPHrwHgreUPdK/f3+7aFpKwNF+X3/9tU12RMG2Ye64//77E++uvnaiBFziTaEbisD/CPgOzLwIOAxZ/mAMsZLlsl6VEnCsCiJSMIjcnjiM27nnntvu76CkLeCSfoegGMPII8YeQ91XwIV/hzggcx8CwnkyWaEkzI/MbSrJEsBoJ9yO1WzCITGQCIuKEnC0FaGB7IPD+4z3AC8sYZVhg59FDMYCBsXyyy+fbKUzvBtGKoYRfXSXXXapWsC5fVCseONpwnvBvIKIS6owP5Ua91EeOBZQ2P+CJx1PYtSexqTaM8yPvkNIGQYoTPh3/jfsgSMklzn4xhtvtOFcWZSk2p4xg3BnsYp9VVkW3/agzsHvIJ5RwvXwrIYFHNcm9RxCNenDhGkT9ZJFEhPfdyFEmPBGl8SEpFAsgiF+GX+1fN+YV5ZddlkbFUBYN2OIbyZiN8oeoQ/SVi+99JL9zjKu2K+PfcK+fhZVggvRRMkQjfThhx+2hNsn1W997UQJuKRaQPcRgQgCvgMzLwIOw5aVKpKYYLzy0aUw8TJhumMEMGZJec3HhAmOECE8FUySeKowjFwYGRMknitizOtd3F6YJN8h7E1z7xDcyxZmVMkD5/bAuXsRvoeHjj0BbPZmtf3cc89NNd16vdsmL/dHtBGWhIBjoYGwR4xpjGqS7QT7KW2M0UTfx6AmxBLhxuIEK+/sBaOtKWTaO+SQQ+y+ikYq7DHhvcaMGWPnAd4bj7TLvkkoE33X7eE68sgjWxIvRB0jQAIkl201KU7lxn3YAxecyzDiCG1+8skn7Z60erQn/AixIxSXZ9P/eA4hpy6JCYI2zI+9TySyoM8FC/v1XGKLpPiVuk8Sbc87kywIDweLIc7jWe+6l3snn/aI+g66ZyDgmDuCWSh92z3qOfvuu68VPOxNz0LA1fIufOtIAhY8RgDbIolFCcKKjzrqKOstYw7AS8kCayl7xCUx4TgDQiNZPH3llVdsKCbhlCxSBQvecvbqcb8ki6+dKAGXZCvoXiIQIuA7MPMi4Ko96y1vHaBUivC81VP1KQaBsLe0GLVOtpZ4HVmMIYtqXkuex30R+JVq1yLXPYl3wruD1xiBX22phl0tz6m2Xj7XN9K7+Lx/0r/xtRMl4JJuCd1PBAIEfAdmHgRc2MtWxIZthHcoIvdGrXPYy9qo7xn1XoSVcpgwXkoM2EoH1GfJJo/jvkj8wm1X5LqX6ofVvhP7tNj7yf9Wc15ZWs9JY7w10rukwSvuM3ztRAm4uIR1nQh4EPAdmHkQcB6vq5+IgAiIgAiIgAiIgAjEJOBrJ0rAxQSsy0TAh4DvwCyCgGPvBumgOSagXbt25ssvv7T7OepZ2O9BvDrZAl1SlFqeR33Zi8LmZA7zpPBO7O9hDxObrlnNZx/a3/72t4ppuwk55T4wKZc6nj0gPJcU9ZXOa+fZ1IF9cLx/uPB7Nl2zYZuN281QHBMOQq6lcCYfe0nq3W9rqWNSv2XMkEG2bdu2Vd+S4y0mtJrJLDh3+6p/y6o9+2NrZcyeFM5PZE/tnHPOWbIetClzUb0KY5D5wu3/jfMcUpOTfIJU89Q/buE9eJ+kC32A88qi5pNqn0W7Ml+SVKPSXMbcyB+eTWG+Iqshe6V9Cnuv3VE21fye+Zk6V6pv8J4LLLBA2X4X9XzGDX/yXJLqC2N/m2he+26cmTiJtpxa8yvPM/l3M/eU380iiyzS8m32vSnfdPoYbeH6nrsXe/LpQ2keyRF+D187UQLOt0fodyIQg4DvwOTWWR8jUOn1SO5A6m0+wi5xQ6Xf1PrvGDMkJOE8pxVXXLHW29kEFSQMIYkF58JEfdjJosnmaM6/4SNeqpDohY/BU089Zc/oIUNfVOE6xFtco4VU7Qg0zqgKG1ywx0BGODZTIZ06fYEkO7UUhAVZKJMwZGupR1q/xUghwYI7tDzuc5986ikzdrY5TI81O8f9iTWOES6clZdEYdyQ/XLPPfcsm9Xz5ZdfNm3atLGLHvUqjLv55pvP/olT7r33XptUASZkkkSIxilkoqxG8MW5p7uGPk/ymFo4cWYiyapY7IpTmBt5LucIUvgdyYI4UNm3kASFRB7VjGFEI8Z8NYuAtB3jppozHfk+Zp2kJQ7XWvvCmF9+NcPG/piAbPtfbce8PNh8/dYwc/TRR9tvs29h3qCP8ofswccdd9w0t7r++uvt+awk1cmq+NqJEnBZtZie2xQEfAcmcPIu4MiAhYGGZyqtQnYpMkd17969ZgGHEYbgeuCBB8yBBx5ovWFfffXVdK/CWTBkzCon4PgthzlzTw6FJs0yBweHC0KL7IXVrPxyGOnbb79tzjjjjGmMFOc9ZOW72QqZIOECk1pKswk4WGGwk002rojgNwOfe878sNDiZqVZWplF5u9QETnjgKMvGBNJFfo77U7dyx3LkIaAc+9E5AHHH1QSD2ThxWvsCt50shRW+l09BRx1oZ1IoY7grbaQIZeFqGq8/mT4YzGKtO4UBBzzb63HbODVwxuPSI5TEHDUo5wnt9R9WARhwbJS2/H7ogg46urS6lfTF/iOjfjxF/P+j7/EwV7VNb98MMK88/RT9pgRXwGHvUAfpZ7s22WBln3MwYKA4/gBjjfIqvjaiRJwWbWYntsUBHwHJnDyLOAIAUS8pV2SFHAIH7xrpET+xz/+YVMPR5VKAg7DEkPVrUJzICnn3HDwuSsuzNEnnCZKwCEYWVGkHZqxSMDV3uosMrDyHKcMGjzYfNtxEWPGfmO2WaW855s+ieHKuEmy5FHA8X4IM1L7lxMPYQHH7wjbIiV5OQ9YvQWcM9wJK6vGU0TkAp7/aku9BJyrBx7FoFAuVb9aBBz3pM1o80reyyIJONcXEKdxwoMnT51qhn33g/n819+q7Qaxrq9VwLFYGjzHjUiADz74wB6HFBZwHDXAWMyq+NqJEnBZtZie2xQEfAdmngUc4gHvW617Wnw6QJICDiMEcUSIRTkvYjkBx+ozLIixdwVBSHinE3Bcg9eN8298SljAVRuC6fPMvP9GAi6ZFiKcr1u3bhW9CU7ATZk82XSZva2Zf565IytA33zhhRfsSnfSJa8CLo54iBJw/I5MnmQ0LOX1SEPAUQ/mqA4dOlTcT8ZCFPMh+219Sr0FHF4kPIqzzDJL2erVKuD+/D7bEFS3dzrqgUUTcHH7woTJk83Qb8eZcRPihc769JVaBFzUAgNRLLR7WMDdcMMNts+ss846PtVM5De+dqIEXCL4dRMRiCbgOzCjBBwr2xwgzMeW1VLCd9yHhAkLYYUh0L59+xaDjDAXd6gl8f4kAKmlZC0ekhRwhFZgiFRiUkrAYcywET5srAYFXBJhjkEB50Iwa2nDRvitBFxyrUiSgB49epQ1RFs8cHgfxn5rtlxlhekqwJ5E9msyD9Wj5F3AIR7wykQliSkl4OCExwgRF+X1SEvAufYqt5+M7w7zXalIhThtXm8BRx1oB8Jay4VHJiHg3LMIhy11pEYRBZxrx1LhwT9N/N0M/Wac+dUz8UycfsI1vgKO0Ngo738pAUcIJeN2vfXWq1g1vvnYYOz9xG7A1nKed2wjbAH+m77nFmWYt7DP+F+YRo1zXztRAq5ik+kCEfAn4DswwwIuuMfK1YaEGsSGs4/LZQAjLp8YfYyysLHPxMLHppJgKfW2Pvu3/MlF/zIpAQevDz/8MFb1Sgm4UhninIAjRI22ibvBv1RlnIBjD57vynesFy3QRRJwyTYWizuEEbtFofDdgwJu8sSJZr155zLt55i95TIWQ9h7Vs1+qGrfIO8Crpx4KCfgHAdCuMIJMtIWcNQlaj8Zcxgh2xiutZQ0BBz1ox8SoopXMaokJeDcs4jg4HzEcCmygIvqC9+Mn2Be+XacmTS19iyTlfpRtQIOccUWBQRcVCkn4LCLNthgg7JVcgu22CD8f2wtPL2EH7u9du4G/Bt7jFmcYZsGiwruN/TLcGIiXztRAs4Y89JLL5knnnjCsmfARyUyqNTZmuXfSZ+8//77N8vr1vyevgOTBwf3wBGCx8pPsBB/T/YtREIwKQaTBwIuKtMhhlq1SUcqTYw1Q6riBkkIOEIZiY2Pm0gkSsCVS1WOgOvSpYtNNZ6EQUv6cVZ4aWuVPwhIwCXfE1gxJkNlVGbFoIDjybOM+85sutLydgzRP2mPepciCLg/5+3pshXGEXCwJBMe4VyuZCHg3LPdfjLCwxFvSYTMpyXg3Dswb/ItDCccSVLAuTZ339bgs4ou4IJ94bPxE8xb3/+UaKbJcnNGNQKOscOCKp6uUqVcCCX7/iodzs7iBWIsWGhrvsvsBw0v1OJpQ+AhKMO2BklTgv3E106UgDPGpv0mlTgFJR5upHp/mIp0f1YVSGeuEo+A78D886PQssyFGAuH6rnJw2VZcjUq9ff8Ox47hEU1hQ83m3/jpr2v5t7VXlurgGMiZc9aNZkbwwKu0j0QcKzmVZOiuhwHEqRwLIHP+V3V8i3K9RJw9WkpFjDxxIX39YQF3LiPRptdOq9qw4WefvppQ6h2vUtRBBwcWGxDiLkzyuIIOH5HpMVmm23WskKflYCDNQZr586dy3o1qm3ztAUc/ZI2CJ/9lbSAgwMLgwgB5mpXGkHA8d3/dfxv5rcFFjE/TU7vnNFqBBziCRul3IJpOQ8c7bbxxhuX7c7umKDwYjmRUAi4sH3E4gHf7KioGY7NCCY98rUTJeAk4Kqag5kIaz17qaoHFvxi34EZFnDh8EVEGvsV+FiQDTI4cTExENPNpOLc/dyP37iwy2qxPvroozbjVrlz0Kq9p8/1tQo4nknqeEQRq2BxSpQHDq8DE3OUSEPAERtPHDypwmstnF1DCuT99tvPK+11rc/P4+8l4JJvFeYNvC2csxb0AvGksIB74193mb123snuNWJuYG8Ufb6epSgCjggezhZkRX/LLbe0SOIIOBaVOHqBpDKbbrqp/V0WAo691s8884w1SBHz1Iv/JrNu3DT9pfpBmgIOQfXQQw/ZIwo4eiKYnTJpAYeXheNtSIRB8irnXSm6gEMY0faMvaWWXtp8PfOspsMyy9VzmLfcuxoBx48ef/xxGwJeauG0lIC77rrrrPDeZJNNyr4X4wFbKyjUaGdsLWwE+psTd/w9i2FENRAhFSyIuvARLr52ogRcSMCtssoqZv3110+lgxbxIeyvCq9mFfE90qqz78AMCzj+G8HAh8IlMUFMOaMLA4rJlpVzPGxMEkw0hF66JCbEXVd7gK/jxL2YyPECho27tFjynCQEHPd54403bMIFPriVjJJSe+Cee+45y5kxESxuDxyrbFxDqvZK6abLMWQPHIYd4VV8nIIrvGmyz9OzJOCSaw36v1sA4kD7qPPhggLu5y8+Nz+/8Yrp1auXNVT5LXMDixpxD7b2qX0RBBzRO2zJwBjccMMNW+aWcgIOow/jj98i3oJzUtoCDi/DgAEDrCG82267tXgC+b5gICPiqjknLNzOaQk4wuk4u5P6brPNNtPVOUkBh4E+cOBAG3WB8A7O9UUWcAh3otNcX8B2GDxkiHntsy/NQqt39RnCVf2mWgHH/PDYY4/ZJCFRdk45AYdNW8kDR+XxSjMWWORgQQBbi+8/cyAL5oxj5kSej73F/2fxnf7INXjlsNvCR1342okScCEBR8gAHyYVEUiCgO/AjBJwSdSn1nu89tprdr8LCx1xDjKt9Xnh3ycl4LjvoEGDTL9+/cxOO+1UNjyxlIBjQmbvLIkHgiFnwSyUTNysxuOJq5TauhQrl8Skd+/e1ijhw8qqXzMXCbhkWh9jk/6EwYHBXkqABQXcO/0eMNtvspHh7KRgYVGEcEoMlHrMDXkWcMwFvP+IESPMDjvsYIVDsJQScIg3PHYkX9hll13MCitMm90zTQHH/iE8/RyIzZwYzqzIvz/44IO23eOcExbVQ9MQcITSMbcTispifNQCXRICjjZHFPBnu+22swt14X5fVAFHW/NtY0vRzjvvPE1yI9g9NnioWWSd9c2Mrf53OH0yM9L/7lKtgOOXtAmiE4EUTmBTTsDxniy4ZFV87UQJOAm4rPpsUzzXd2DmVcBRL8IP+dB37dq1bOrxejRwkgIObyKHbuMlO/jgg0tm4Ct3DhxeTwQaXgtnKITPgWNVDsOHa3z2xYXPgcMIwtuBKKyHoVyPdkv6nhJwtRNl0YHVYTzqGGnlDnJ2Am78d9+ar154xhxyyCGRXmX2l9LXWdSoxesc9XZ5FXDMAXjz2feGBzMqQiVKwGFskgmX/+3Zs6cNRw2XtAQcSZ0Q32uuuabZYostSs7rhNgSlojgL5XhsVzPrKeAI0qESAXC3OjPRCyUKrUKOPoiIpFvAwsfpRbUiijg6At41Pm+EwIcdc4d3qYHHnvCdFxrXdO67bQRKLXPTH/cwUfA/Wk32dBj9pMG9/uXEnDXXnutnQOzjLzztRMl4CTgkhpvuk8EAd+BmWcBR91YNb7xxhvNuuuu6yVKfDtLkgLOxfez0rjccstZwyuc3pd6lhNw/DthR9zDhZaGBRzX8DHp37+/NZKrzQIaFnDcj3ArPrKs+obDMXzZFul3EnC1tRZHibDK7oy0SkeLOAH33mMPm826di5rHBPmzcIIe0Ar3beat8ijgMN7ScghodSIN5e0JPxeYQHHu7AIw3zDGXylQtvrLeDwAOI1RFhhrHOUQaVFIeqOl4MwMjwX1ZR6CTjq4g6Qpx3Y01Su1CLgELGIXRYoeFb46Ifgc4sk4IJ9ARFPKG+5vsDiz/0PPWzarbCKmbX99EcoVNMvoq71FXDuXoQyM8e5vlBKwF1zzTXWqxznHLha36nU733tRAk4Cbh69Und94/EIbkdY7U2EMlAzjvvPJtco1pR4vvsJAUcdRgyZIhdUT722GPN888/bz/KpJ0OlkoCjms5VgADAoMmSsBxjdtHiEeOFN1xS5SAcyKajfN44potQ6UEXNzeM/11iCq8ReH9VuXuiID7aObZzGf/edwceeSRFT3vZP974IEHrHep1Nly1b5B3gQcDFm4wbNOyGE5D2ZQwLGYQ/9lj+yOO+5Y8hBo+NRTwDEfIag4IoDwTRax4haM/aFDh9rMf9UkaqqHgCu1b68eAo65mzYnTLh79+42q3O5UhQBR1/ge44Hnb5A4pc4hT1h/fr3N791WNDMtdD0HuQ49yh1Ta0Cjvuy5YM9iiyulvPAMRYRrFkVXzsxt8blaaedZicWyhFHHDFdbHiSoIPHCGgPXJJkdS/fgQm54DlweSSJmLrooousMch+gzTOKUtawLES+69//cvOMez5IBSKEJLgByyOgKN9OM5h+PDhdkXeZSOLardq9xGWEnDcmzBQQtZYZax3FsA89UEJOL/WYIECD0LUfqtKAu6pV14z3bqsbr2+cYoLL2b8JNE38yTgSA7FftS4Hkwn4BAbjGfmB+bMqPC0INt6CThEJIk38FzhRar2aBlXR+ZPFr7iJmpKWsC5vVpu317cxQIfDxxRJ9iKnTp1ikyMEjUmiiDg6AtsI0CM0Req/Y4j/gb85z/m40lTzfzLrRhnaoh1TRICjge5bwXfSr7taItgIYSSxYu11lorVr3qcZGvnSgBJw9cPfqj7vknAd+BWQQBxyrsVVddZY0ANu2ziZ8VySTDpsIdCSMOkcS+gxVXrP1jQWbPO+64w+y7774tK+hM+ITIkF2KP3zYWGlmn1ylYxSoG6GlfOSpY6nC3hf2zsGq3Mo9v8dIIcnBGWecERnS4sIzMSoxlCtl1WyEwYnHk7aDSS2FVWc8Rb7ZWWt5dtq/xYChv5Xab1WuPo88+qhdzT7u2GOrSsbDHIGBj3eaxBe1zA1481hNp/7lPAQvv/yyzQRZKpyxVu4IYPZZ4cEkhDxOkAUCjrBnBNxWW21l95rF+R1h0swpSQjg4Hszn9HnCd+s9d4kamIRibYmbLZcQQRh8BPxQEHkX3zxxTaCo9p5i/swV6+xxhpl9+1F1QcBx/EXcbOmMscS8hrOLlqpLyHg8FRGheZX+m1a/06/pA/QF+ae2z8U8pVXXzUDBg8x7Wafw8wyS5uaq//7zz+bMaM+MCeccEJFT2elhzFe+c7Tx04//fRpLucYAeYT+lFWxddOlICTgMuqzzbFc30HZhEEHHV88803rVFFqmbOK0OU1PvAbybhrbfeOnaYR7mOhlF4/fXX249XMLMaqYIx7N0BxaTORpDFCRV1G6hZYS9XEGV4//BgVirU7dBDDy1p9PEePJeN9c1SYHLYYYfV9Lqvv/66TT0dPJy1phvm+McYZyR38BGrGKKMhTiptqMQYGizB5QFmFoKHkTGfrlwP0Q5YW71alPqwH6xcMbIcu9FtlsEAHWPe/4k98MzggBOurDfl8yJ4SNQfJ9DaOH9999v06VXKiQ/IeyegoDr06ePPR6m2oIAZs9epb1aUfdFwDK/x+2PfHOYz8PZRSvVmRB9IjLyXJLsC4jBvn372j3fSRT6J33FN+tpsA60OW3xz3/+c5qq3XnnndaTHnV8ShLvEOcevnaiBJwEXJz+pWs8CfgOTB6X9xBKTyT6mQiIgAiIgAiIgAiIQA25EiTgJOA0gOpIQAKujnB1axEQAREQAREQAREoMAFfO1ECTgKuwN0+/1X3HZjywOW/bVVDERABERABERABEaiFgK+dKAEnAVdLv9NvKxDwHZhFFXDsFeB8mGpj1ombP+ecc+z+OZKGdOzY0WZ5i7PRP61OeOutt9ojB04++WSb9IRCghH2f7izZsjiddRRR9W9SmTSOvHEE21SiL322sum8o67H4x9NWwMJxkDCWg4XBz+9Ur6UHcYekDhCLBXluQV7NtkzxrZbO+55x57FhNjKLhHarXVVjOnnnqq2X777Qv3nllVmIQzF154obn33ntbqkDiJI45IfEPc1jRS/gd2aPJe5GchIzCfIs23HBDc+655zblWZlFb9806+9jt1x22WVm1113bUmEg33AN5g9ha6Q2fLKK6+s+Cq+dqIEnARcxc6lC/wJ+A7MIgs4knJUOh8nSJSN7yQEIKGES2dNRsuVVlrJH7zHLxGPJCgoVZyAI+OaK7fddpsVdQigNAsfHMd57733tqwqCTgSnbBpnyQCZLdzfNncjWDG6FERgTQIBAVc8HmcQ4XRQ9IU0pmfcsopNkkSGeRUaiOAuCEjIkJu9OjRtd0sZ78mYc36669vz/u64oorDEmnmOtuuukmuyCguS1nDZaz6vgIOBZtSU7ksmFjH3Auq48t4GsnSsBJwOVsKDVWdXwHZp4FHJPdSSedZFc6SYvdu3dvm8WRwr+dffbZdmL79ttv7co5HiIKQoHz1si+SJaq/fff3xx00EH2+AEyfJHBKu45Pkn1EupLHclEyOo/GeIQZaT2pz5k7Lv99tvNwgsvbKIEXNTfJVW3Uvfp1auXzZxJJj4EJ+IL4Ut9yZ7J37PyzhlTeOo4wJR24gwczjHiXTizbv755693VXX/JiOAJ41xTmpuCt54zugiFT59Ea86Hl+yB+JBwQPHeWTMC8OGDbO/oW+T0RDPCQdku+NJmgxly+tGMcWT5uZgPOdk+US0kNUWwzLMlMWpRRZZxI5/vAaXXHKJ2WCDDXKDtNZ35NgXvjMsAMw000y5eS9VpBgEnIDju3/88cebESNG2G/oNddcY/+XbKl8T1kc6NKli/ViY+cw1tq2bWttA+YvFnODC7xx397XTpSAk4CL28d0nQcB34GZdwHHyjjnqfDB5MwzJi9WzJkICY3CIOOsHcKfMOJYIe3cubO566677CpNObsMAAAgAElEQVTpr7/+aifCm2++2aZmxlDDmONcJc5jQRAiTOpdqG8wpIiJGIHE8QjUkzAkjM4nn3wyNwLOCWUXqorhAmv4EVJJWm3OrOvevbs1mm+44QZrBLuziGgbVqYxokmfzKHOGNkqIlArAcKJl1xySSviCKO+9NJL7cHVeHsxbFx6/0033dQ+Ck9wWGzw95yxRgp2jCfO3mrmUoopnkq87sxVLMoQFo1XnoOJw0xZoDrrrLPsMSOEdMGWuTgvpdZ3vOCCC6zX1sd4zgsD1SM7AtgBnFHIUT18JxFqbN/grFF37ApjhmgXjgOhIN4QbO44EWyHww8/vGU7Bdfwba4UGcN1vnaiBJwEXHajpgme7Dsw8y7gPvnkk5ZwR84T2mGHHWyoipsI3X4qDghldRixgdBDFLnCwcIIDA7mpSBC2J/1+OOPWwOEvTH1PpuF+uKZcmKRSZhwLVZ0KdSbw17Zl+fCJYNGQhYeuLCA46Pyf//3fy2H4x555JH2QFZW6OH7xRdfWK9GsOCBQ/DxnoR8IFA5XFhFBGolgFcdDxHGzNJLL23uu+8+c+aZZ5ptt922Zaxz9hKLCFEeOCfs+D1nS6oYG6kAUxZfmEPxtLFoBlvmSwrhW4gzzroKCzjmZ6Ic2Fs4duxYey8W3yodvJ0m+1reEQGHOCXyQ0UEqiWAHcBCMufZBveH4rHG08Zedzy7LJawdxxvd5SAkwfuT/IYHhiJFCatag7NrLbxaCQ2UFPwErACryICSRBoVAHHvhQ8bhQMMzxoHHQdjiVnNYtQKQ5qZdUdsRSnbL755nblnbFfzxKuL4KMkEk8A5SiCLjgHrhjjjnGhoASGoqXlHfAE1KqHHjggdZzd/nll9cTte7dJARYiOHAajb5k5johRdesAs8GEd77LGHpVBJwEV55ZoEX+RrBpliUL744ovTRQQE9xUG+RHKzlzNXOz2+OJZOO+886wwzEup5R1ZiGIhi719hLqpiEA1BLADjj32WOvFZRzx/STBFwvQLCoj7lhwZsHDRb5IwJUhLAFXTffTtXkl0KgCjn1v/EHI4VkjAYkLoQxmoXQCDg8XizBMkm5PHHu28BSxosVqsPMAEQZEaB/hi1tttVVdmzZKwCFoCDlkpZu9GRgHrHLnZQ8cQEgSw942mIezUB599NH2A0T7MI8SnuQEHHvhnn/+ebsaT4gI7BHLGNiHHHJIXVnr5s1DgD7FvjfGMCG67CXBc00oH8l0+Hf+t5QHTgJu+r4SZhqej0oJOELaWUS7++67W25KKCtZbJm381R835F3IPx+mWWWsQsHLonJLbfcYsPxlcQkT62cv7o4O4D+wjjhe4/HDfuFfeVEFLFwwtYRZ9+QvASPN1sWKLVE4/jaiQqhVAhl/kZTA9XId2CCYCobx3JYmOyYyPr372+TYyAS2BhPYYUXAeaOEWDiIxQSoUH2M8Ki8KyzqZ7JkcmSDcIkNEHEzTrrrDZ7GN48NhPXu4TrS5gkHrhwEhPqH5VxMqsslOzbYw8LvP72t79ZbyXx+5SwBw4BR9ICCqm2CXUljp/N17DHaCIEiXZVEYEkCPz73/+2/ZFQXcYYY5q9IIT3uSQmiAe3B46w36CYYNUb73veBEYSbHzvAVMWWTAoYRqee0gqhVgLMyUxE+N7iy22aHk07bHAAgvYPYlpZ/st9/6+7+jmNpfYhbmNBYItt9zSeoHLZRf2bQ/9rnEIBO2AG2+80Vx11VV2WwG2QJ8+few+OPaUH3fccS32DSHgjCsWBxB+LPoyx7kjhaBDCLmL7itHy9dOlICTgGucUZjDN/EdmHkXcD5nveWweaarUi2raEV4P9VRBNIgQBIdspxiUKskQ6AZmDbDOybTG3SXRiLgaydKwEnANdI4yN27+A7MPAu4sNcqd9BrqFBWHrUaqqyfikBuCJAwh8OTCY1mBVuha7U3TTMwbYZ3rL0n6A6NSsDXTpSAk4Br1DGRi/fyHZh5FnC5AKtKiIAIiIAIiIAIiEDBCfjaiRJwEnAF7/r5rr7vwJSAy3e7qnYiIAIiIAIiIAIiUCsBXztRAk4Crta+p9+XIeA7MCXg1K1EQAREQAREQAREoLEJ+NqJEnAScI09MjJ+O9+BKQGXccPp8SIgAiIgAiIgAiJQZwK+dqIEnARcnbtmc9/ed2BKwDV3v9Hbi4AIiIAIiIAIND4BXztRAk4CrvFHR4Zv6DswJeAybDQ9WgREQAREQAREQARSIOBrJ0rAScCl0D2b9xG+A1MCrnn7jN5cBERABERABESgOQj42okScBJwzTFCMnpL34EpAZdRg+mxIiACIiACIiACIpASAV87UQJOAi6lLtqcj/EdmBJwzdlf9NYiIAIiIAIiIALNQ8DXTpSAk4BrnlGSwZv6DkwJuAwaS48UAREQAREQAREQgRQJ+NqJEnAScCl20+Z7lO/AlIBrvr6iNxYBERABERABEWguAr52ogScBFxzjZSU39Z3YErApdxQepwIiIAIiIAIiIAIpEzA106UgJOAS7mrNtfjfAemBFxz9RO9rQiIgAiIgAiIQPMR8LUTJeAk4JpvtKT4xr4DUwIuxUbSo0RABERABERABEQgAwK+dqIEnARcBt21eR7pOzAl4Jqnj+hNRUAEREAEREAEmpOAr50oAScB15wjJqW39h2YEnApNZAeIwIiIAIiIAIiIAIZEfC1E5tOwL300kvmk08+MXPPPbfp0qWLmWOOOcxTTz1l7rnnHtt0nTt3Nr169cqoGfXYRiPgOzAl4BqtJ+h9REAEREAEREAERGBaAr52YtMIuEmTJpkjjzzSTJgwwUycONHMMMMMZrbZZjOHH364GT16tAScRlRdCPgOTAm4ujSHbioCIiACIiACIiACuSHgayc2hYCbOnWqOfTQQ80vv/wyXYO1adPGrLPOOmbAgAHywOWmOzdORXwHpgRc4/QBvYkIiIAIiIAIiIAIRBHwtRMLIeA6depkfvzxR++Wx/v26aefmsmTJ0feY5555jHfffedBJw3Yf2wFAHfgSkBpz4lAiIgAiIgAiIgAo1NwNdOLISAW3TRRc1HH33k3YKESv73v/8t+Xu8cIRWUrQHzhuzfhhBwHdgSsCpO4mACIiACIiACIhAYxPwtRObQsC1a9cuMnzSdYnWrVub33//XQKuscdIJm/nOzAl4DJpLj1UBERABERABERABFIj4Gsn5lbADR8+3Pz6668W4EwzzWQIg/Qt33zzjXniiSciRRzJTBZbbDEzatQoCThfwPpdSQK+A1MCTp1KBERABERABERABBqbgK+dmFsBF26uqZMnm3duucXM2KaNWXb33Y2ZYQZjpkwx7917r5n4yy9m+b33Nj9+9JEZ9fDDZtFNNzVzL7OMeee22wy/W27vvc2TAwaYvn37TiMEEW8LLbSQ6dq1q7n33nsl4Bp7jGTydr4DUwIuk+bSQ0VABERABERABEQgNQK+dmJhBNzohx82M7ZubSb89JOZrUMHs+C665ovBg0y373zjllogw3M6EceMTPNOqsVck8fdphZbs897b+1btfOtJ1nHnvNk08+ac98o/zwww9m9dVXNwcccIDNQKlz4FLrq031IN+BKQHXVN1ELysCIiACIiACItCEBHztxMIIuJcvuMAss/vuZuJPP5lPBw40K/XqZb58+WUzZuBAs8Daa5vhd9xhNu7Tx3rdvnn9dbNG797m8b33NjO0amU26tPHtOvY0XaLKVOmmLFjxxoyT+KBo+gg7yYcMSm9su/AlIBLqYH0GBEQAREQAREQARHIiICvnVgYATesTx8bGjnhxx/N2HffNcv07GkmT5xorGduppnM18OGmf/bdVfTfqmlzH8OPtjMs/zypmOXLqbNX/5iPnjoIdPpyCNLNo0EXEa9tgke6zswJeCaoHPoFUVABERABERABJqagK+dWBgB9/Onn5qh559v976t2bu3GfXII2bxrbc2L550kg2TXGGffez+N/a8IfKW3XNP88Y115hWbdpYYTffqqtKwDX1EMnm5X0HpgRcNu2lp4qACIiACIiACIhAWgR87cTCCDgLcsqUP3jOOOP/uIb/burUP/7tzyQn010f0SLywKXVTZvvOb4DUwKu+fqK3lgEREAEREAERKC5CPjaicUScHVqUwm4OoHVbdlnqTGmfiACIiACIiACIiACIjAdAV87UcalkphoONWRgO/AlAeujo2iW4uACIiACIiACIhADgj42okScBJwOei+jVsF34EpAde4fUJvJgIiIAIiIAIiIAIQ8LUTJeAk4DSC6kjAd2BKwNWxUXRrERABERABERABEcgBAV87UQJOAq7q7vvhhx+aTz75xP5u5plnNhMnTqz6HkX9waqrrmpmn3322NX3HZgScLER60IREAEREAEREAERKCQBXztRAk4CruoO37dvX9O/f397EPqSSy5pRo4cWfU9ivgD3vfEE080iy++eOzq+w5MCbjYiHWhCIiACIiACIiACBSSgK+dKAEnAVd1h3cCjh8utdRSTSPgeN+TTjpJAq7qHqMfiIAIiIAIiIAIiIAIhAlIwNXQJ3SMQHXwggJuxRVXNK1ataruBgW+eqeddjIdOnSI/Qa+A1MeuNiIdaEIiIAIiIAIiIAIFJKAr50oD5w8cFV3+KCA22CDDUzPnj2rvkez/MB3YErANUsP0XuKgAiIgAiIgAg0KwFfO1ECTgKu6jEjARcfme/AlICLz1hXioAIiIAIiIAIiEARCfjaiRJwEnBV93cJuPjIfAemBFx8xrpSBERABERABERABIpIwNdOlICTgKu6v0vAxUfmOzAl4OIz1pUiIAIiIAIiIAIiUEQCvnaiBJwEXNX9XQIuPjLfgSkBF5+xrhQBERABERABERCBIhLwtRMl4CTgqu7vEnDxkfkOTAm4+Ix1pQiIgAiIgAiIgAgUkYCvnSgBJwFXdX+XgIuPzHdgSsDFZ6wrRUAEREAEREAERKCIBHztRAk4Cbiq+7sEXHxkvgNTAi4+Y10pAiIgAiIgAiIgAkUk4GsnSsBJwFXd3yXg4iPzHZgScPEZ60oREAEREAEREAERKCIBXztRAk4Crur+LgEXH5nvwJSAi89YV4qACIiACIiACIhAEQn42okScBJwVfd3Cbj4yHwHpgRcfMa6UgREQAREQAREQASKSMDXTpSAk4Crur9LwMVH5jswJeDiM9aVIiACIiACIiACIlBEAr52ogScBFzV/V0CLj4y34EpARefsa4UAREQAREQAREQgSIS8LUTJeAk4Kru7xJw8ZH5DkwJuPiMdaUIiIAIiIAIiIAIFJGAr50oAScBV3V/l4CLj8x3YErAxWesK0VABERABERABESgiAR87UQJOAm4qvu7BFx8ZL4DUwIuPmNdKQIiIAIiIAIiIAJFJOBrJ0rAScBV3d8l4OIj8x2YEnDxGetKERABERABERABESgiAV87UQJOAq7q/i4BFx+Z78CUgIvPWFeKgAiIgAiIgAiIQBEJ+NqJEnAScFX3dwm4+Mh8B6YEXHzGulIEREAEREAEREAEikjA106UgJOAq7q/S8DFR+Y7MCXg4jPWlSIgAiIgAiIgAiJQRAK+dqIEnARc1f1dAi4+Mt+BKQEXn7GuFAEREAEREAEREIEiEvC1EyXgJOBK9vfffvvNfPzxx6Zdu3amY8eOZsYZZ7TXSsDFnyJ8B6YEXHzGulIEREAEREAEREAEikjA106UgJOAi+zvV1xxhXnnnXfMzDPPbCZNmmRmn312s/nmm5v1119fAq6KGcJ3YErAVQFZl4qACIiACIiACIhAAQn42okScBJw03X30047zXz66adm6tSp0/xbmzZtzM4772x++OEH079/f/tvG2ywgenZs2cBh0w6VfYdmBJw6bSPniICIiACIiACIiACWRHwtRMl4CTgpumzeN6GDRtWsh/PMsssZuWVVzaDBw+WgIsx2n0HpgRcDLi6RAREQAREQAREQAQKTMDXTpSACwm4ZZZZxrRu3brAXcG/6pMnTzYjRoywIZPlSvv27c33338vARcDte/AlICLAVeXiIAIiIAIiIAIiECBCfjaiRJwIQG3+OKLm1GjRhW4K9RWdTxsJC8pV9q2bWvGjx8vARcDte/AlICLAVeXiIAIiIAIiIAIiECBCfjaiRJwEnDTdHuSlkycOLHsUAiKPO2BKz9r+A5MCbgCz8aqugiIgAiIgAiIgAjEIOBrJ0rAGWM+++wzM3r0aIs5joCJ0R6FvITQyUcffdSMGzeubP0XXHBBM2bMGHngYrSy78CUgIsBV5eIgAiIgAiIgAiIQIEJ+NqJEnAFbvQkq/7buHHml88/N69/8YW5//77S+6Dm2GGGUy3bt2s0KPIA1e+FXwHpgRckr1b9xIBERABERABERCB/BHwtRMl4PLXlqnX6OcxY8zAo482i3brZpbdYw/zxBNPmAcffND8/vvv09SlVatW5vzzzzfPPfecjhGI2Uq+A1MCLiZgXSYCIiACIiACIiACBSXgaydKwBW0wZOs9qTx48249983Xw8bZpbbay976zfffNP069fPnvlGaOXyyy9v9txzT8NZcH379pWAi9kAvgNTAi4mYF0mAiIgAiIgAiIgAgUl4GsnSsAVtMGTrvZ377xjvnr11RYB5+4/ZcoUQ9gkf1yRgItP33dgSsDFZ6wrRUAEREAEREAERKCIBHztRAm4IrZ2HepcSsBFPUoCLn4D+A5MCbj4jHWlCIiACIiACIiACBSRgK+dKAFXxNbOuM4ScPEbwHdgSsDFZ6wrRUAEREAEREAERKCIBHztRAm4IrZ2xnWWgIvfAL4DUwIuPmNdKQIiIAIiIAIiIAJFJOBrJ0rAFbG1M66zBFz8BvAdmBJw8RnrShEQAREQAREQAREoIgFfO1ECroitnXGdJeDiN4DvwJSAi89YV4qACIiACIiACIhAEQn42okScEVs7YzrLAEXvwF8B6YEXHzGulIEREAEREAEREAEikjA106UgCtia2dcZwm4+A3gOzAl4OIz1pUiIAIiIAIiIAIiUEQCvnaiBFwRWzvjOkvAxW8A34EpARefsa4UAREQAREQAREQgSIS8LUTJeCK2NoZ11kCLn4D+A5MCbj4jHWlCIiACIiACIiACBSRgK+dKAFXxNbOuM4ScPEbwHdgSsDFZ6wrRUAEREAEREAERKCIBHztRAm4IrZ2xnWWgIvfAL4DUwIuPmNdKQIiIAIiIAIiIAJFJOBrJ0rAFbG1M66zBFz8BvAdmBJw8RnrShEQAREQAREQAREoIgFfO1ECroitnXGdJeDiN4DvwJSAi89YV4qACIiACIiACIhAEQn42okScEVs7YzrLAEXvwF8B6YEXHzGulIEREAEREAEREAEikjA106UgCtia2dcZwm4+A3gOzAl4OIz1pUiIAIiIAIiIAIiUEQCvnaiBFwRWzvjOkvAxW8A34EpARefsa4UAREQAREQAREQgSIS8LUTJeCK2NoZ11kCLn4D+A5MCbj4jHWlCIiACIiACIiACBSRgK+dKAFXxNbOuM4ScPEbwHdgSsDFZ6wrRUAEREAEREAERKCIBHztRAm4IrZ2xnWWgIvfAL4DUwIuPmNdKQIiIAIiIAIiIAJFJOBrJ0rAFbG1M66zBFz8BvAdmBJw8RnrShEQAREQAREQAREoIgFfO1ECroitnXGdJeDiN4DvwJSAi89YV4qACIiACIiACIhAEQn42okScEVs7YzrHBRwSy21lHn//fczrlHlx2+00UamR48elS9M+ArfgSkBl3BD6HYiIAIiIAIiIAIikDMCvnaiBFzOGrII1QkLuJEjR+a+2gi43XbbLfV6+g5MCbjUm0oPFAEREAEREAEREIFUCfjaiRJwqTZTYzxMAi5+O/oOTAm4+Ix1pQiIgAiIgAiIgAgUkYCvnSgBV8TWzrjOv/32mxk/frytxYwzzmimTJmScY0qP36WWWYxbdu2rXxhwlf4DkwJuIQbQrcTAREQAREQAREQgZwR8LUTJeBy1pCqTmMR8B2YEnCN1Q/0NiIgAiIgAiIgAiIQJuBrJ0rAqS+JQB0J+A5MCbg6NopuLQIiIAIiIAIiIAI5IOBrJ0rA5aDxVIXGJeA7MCXgGrdP6M1EQAREQAREQAREAAK+dqIEnPqPCNSRgO/AlICrY6Po1iIgAiIgAiIgAiKQAwK+dqIEXA4aT1VoXAK+A1MCrnH7hN5MBERABERABERABOSBUx8QgZwSkIDLacOoWiIgAiIgAiIgAiKQMQFfO1EeuIwbTo9vbAK+A1MeuMbuF3o7ERABERABERABEfC1EyXg1HdEoI4EfAemBFwdG0W3FgEREAEREAEREIEcEPC1EyXgctB4qkLjEvAdmBJwjdsn9GYiIAIiIAIiIAIiAAFfO1ECTv1HBOpIwHdgSsDVsVF0axEQAREQAREQARHIAQFfO1ECLgeNpyo0LgHfgSkB17h9Qm8mAiIgAiIgAiIgAvLAqQ+IQE4JSMDltGFULREQAREQAREQARHImICvnSgPXMYNp8c3NgHfgSkPXGP3C72dCIiACIiACIiACPjaiRJw6jsiUEcCvgNTAq6OjaJbi4AIiIAIiIAIiEAOCPjaiRJwOWg8VaFxCfgOTAm4xu0TejMREAEREAEREAERgICvnSgBp/4jAnUk4DswJeDq2Ci6tQiIgAiIgAiIgAjkgICvnSgBl4PGUxUal4DvwJSAa9w+oTcTAREQAREQAREQAXng1AdEIKcEJOBy2jCqlgiIgAiIgAiIgAhkTMDXTpQHLuOG0+Mbm4DvwJQHrrH7hd5OBERABERABERABHztRAk49R0RqCMB34EpAVfHRtGtRUAEREAEREAERCAHBHztRAm4HDSeqtC4BHwHpgRc4/YJvZkIiIAIiIAIiIAIQMDXTpSAU/8RgToS8B2YEnB1bBTdWgREQAREQAREQARyQMDXTpSAy0HjqQqNS8B3YErANW6f0JuJgAiIgAiIgAiIgDxw6gMikFMCEnA5bRhVSwREQAREQAREQAQyJuBrJ8oDl3HD6fGNTcB3YMoD19j9Qm8nAiIgAiIgAiIgAr52ogSc+o4I1JGA78CUgKtjo+jWIiACIiACIiACIpADAr52YioC7umnnzbLLbecmW+++XKAatoq3H///Wbbbbc1rVu3zl3dVKHiE/AdmBJwxW97vYEIiIAIiIAIiIAIlCPgayemIuCOOOIIs9tuu5lOnTrlqhWnTp1qunXrZhBx7dq1y1XdVJnGIOA7MCXgGqP99RYiIAIiIAIiIAIiUIqAr52YioA7/PDDze677y4Bp/7bdAR8B6YEXNN1Fb2wCIiACIiACIhAkxHwtRMl4OSBa7Khku7r+g5MCbh020lPEwEREAEREAEREIG0CfjaiakIuC222MIsscQSZoEFFkibS8Xn3XTTTeall14y7du3r3itLhCBagn4DkwJuGpJ63oREAEREAEREAERKBYBXzsxFQG3yy67mNVWW80svvjiuaLKHrizzjrLkGRlzjnnzFXdVJnGIOA7MCXgGqP99RYiIAIiIAIiIAIiUIqAr52YioDTHjh13GYl4DswJeCatcfovUVABERABERABJqFgK+dKAGnPXDNMkYyeU/fgSkBl0lz6aEiIAIiIAIiIAIikBoBXztRAk4CLrVO2owP8h2YEnDN2Fv0ziIgAiIgAiIgAs1EwNdOTEXAjRgxwnTs2NEMHTrUJgxZffXV7flrUeX333+357L16NFjmn+eNGmSOe+888zJJ5+caLu+9tprZqWVVjKtWrVK9L66mQhAwHdgSsCp/4iACIiACIiACIhAYxPwtRNTEXCg79evnzn//PPNUUcdZW688UYr0PbYY49pWmX48OGGrJBvv/22GTBgwDT/dsYZZ5jTTz/dTJ48ubFbUm/XUAR8B6YEXEN1A72MCIiACIiACIiACExHwNdOTEXAke1x0003td6ztdde27z11lvmgAMOMEOGDJnmRc4991yDiPv666+nEXB47a6//nrzzDPPmDFjxqj5RaAwBHwHpgRcYZpYFRUBERABERABERABLwK+dmJqAo4z4BBuc889txk/frw9E27s2LHTvCxC79VXXzUnnnhii4D74YcfzHbbbWf69u1rVlhhBQk4r+6hH2VFwHdgSsBl1WJ6rgiIgAiIgAiIgAikQ8DXTkxNwHXo0MF88MEHZvbZZzcTJkywe+LCAg5Ur7zySouAQ9DttddeZp999rGeu4UWWkgCLp3+pKckRMB3YErAJdQAuo0IiIAIiIAIiIAI5JSAr52YmoDjIO9bbrnFLL/88mbUqFFm1113tWItXMICrmvXrqZ169b2spdfftkmQHnyySfNrLPOmtOmULVE4H8EfAemBJx6kQiIgAiIgAiIgAg0NgFfOzEVAQf6Cy64wCYnOfvss03v3r3NiiuuaDjg+9prrzUbbrihWWKJJWwLBQXcn0as/Xu8cYsttpj58MMPyezX2K2pt2sYAr4DUwKuYbqAXkQEREAEREAEREAEIgn42oneSmgqiqqKwvEAJCkZPHiwDYc87rjjbOr+q666ymyyySZmySWXtHcjScljjz1mk5yEC+LvpJNOquKpulQEsiXgOzAl4LJtNz1dBERABERABERABOpNwNdOTE3AOQDoPnnQ6t0ddP+8EPAdmBJweWlB1UMEREAEREAEREAE6kPA105MXcDV5/V1VxHIJwHfgSkBl8/2VK1EQAREQAREQAREICkCvnaiBFxSLaD7iEAEAd+BKQGn7iQCIiACIiACIiACjU3A106UgGvsfqG3y5iA78CUgMu44fR4ERABERABERABEagzAV87UQKuzg2j2zc3Ad+BKQHX3P1Gby8CIiACIiACItD4BHztRAm4xu8besMMCfgOTAm4DBtNjxYBERABERABERCBFAj42k+k8iQAAAcmSURBVIkScCk0jh7RvAR8B6YEXPP2Gb25CIiACIiACIhAcxDwtRMl4Jqjf+gtMyLgOzAl4DJqMD1WBERABERABERABFIi4GsnSsCl1EB6THMS8B2YEnDN2V/01iIgAiIgAiIgAs1DwNdOlIBrnj6iN82AgO/AlIDLoLH0SBEQAREQAREQARFIkYCvnSgBl2Ij6VHNR8B3YErANV9f0RuLgAiIgAiIgAg0FwFfO1ECrrn6id42ZQK+A1MCLuWG+v/27MAEghgEAmD/XT/fwoKn0WngLhkVXOJ3BAgQIECAAIGPBdI9UYD7uFB+d0sgHUwB7lafuC0BAgQIECBwTyDdEwW4e73ixh8KpIMpwH1YJL8iQIAAAQIECDQIpHuiANdQLL+8I5AOpgB3p0fclAABAgQIELgpkO6JAtzNfnHrjwTSwRTgPiqQ3xAgQIAAAQIEmgTSPVGAayqY394QSAdTgLvRH25JgAABAgQI3BVI90QB7m7PuPkHAulgCnAfFMcvCBAgQIAAAQKNAumeKMA1Fs2v9wukgynA7e8NNyRAgAABAgRuC6R7ogB3u2/cvlggHUwBrrgwPk+AAAECBAgQaBZI90QBrrlwfr9bIB1MAW53X7gdAQIECBAgQCDdEwU4vUOgUCAdTAGusCg+TYAAAQIECBAYIJDuiQLcgOI5wl6BdDAFuL094WYECBAgQIAAgb9AuicKcPqHQKFAOpgCXGFRfJoAAQIECBAgMEAg3RMFuAHFc4S9AulgCnB7e8LNCBAgQIAAAQJe4PQAgaECAtzQwjgWAQIECBAgQKBZIN0TvcA1F87vdwukg+kFbndfuB0BAgQIECBAIN0TBTi9Q6BQIB1MAa6wKD5NgAABAgQIEBggkO6JAtyA4jnCXoF0MAW4vT3hZgQIECBAgACBv0C6Jwpw+odAoUA6mAJcYVF8mgABAgQIECAwQCDdEwW4AcVzhL0C6WAKcHt7ws0IECBAgAABAl7g9ACBoQIC3NDCOBYBAgQIECBAoFkg3RO9wDUXzu93C6SD6QVud1+4HQECBAgQIEAg3RMFOL1DoFAgHUwBrrAoPk2AAAECBAgQGCCQ7okC3IDiOcJegXQwBbi9PeFmBAgQIECAAIG/QLonCnD6h0ChQDqYAlxhUXyaAAECBAgQIDBAIN0TBbgBxXOEvQLpYApwe3vCzQgQIECAAAECXuD0AIGhAgLc0MI4FgECBAgQIECgWSDdE73ANRfO73cLpIPpBW53X7gdAQIECBAgQCDdEwU4vUOgUCAdTAGusCg+TYAAAQIECBAYIJDuiQLcgOI5wl6BdDAFuL094WYECBAgQIAAgb9AuicKcPqHQKFAOpgCXGFRfJoAAQIECBAgMEAg3RMFuAHFc4S9AulgCnB7e8LNCBAgQIAAAQJe4PQAgaECAtzQwjgWAQIECBAgQKBZIN0TvcA1F87vdwukg+kFbndfuB0BAgQIECBAIN0TBTi9Q6BQIB1MAa6wKD5NgAABAgQIEBggkO6JAtyA4jnCXoF0MAW4vT3hZgQIECBAgACBv0C6Jwpw+odAoUA6mAJcYVF8mgABAgQIECAwQCDdEwW4AcVzhL0C6WAKcHt7ws0IECBAgAABAl7g9ACBoQIC3NDCOBYBAgQIECBAoFkg3RO9wDUXzu93C6SD6QVud1+4HQECBAgQIEAg3RMFOL1DoFAgHUwBrrAoPk2AAAECBAgQGCCQ7okC3IDiOcJegXQwBbi9PeFmBAgQIECAAIG/QLonCnD6h0ChQDqYAlxhUXyaAAECBAgQIDBAIN0TBbgBxXOEvQLpYApwe3vCzQgQIECAAAECXuD0AIGhAgLc0MI4FgECBAgQIECgWSDdE73ANRfO73cLpIPpBW53X7gdAQIECBAgQCDdEwU4vUOgUCAdTAGusCg+TYAAAQIECBAYIJDuiQLcgOI5wl6BdDAFuL094WYECBAgQIAAgb9AuicKcPqHQKFAOpgCXGFRfJoAAQIECBAgMEAg3RMFuAHFc4S9AulgCnB7e8LNCBAgQIAAAQJe4PQAgaECAtzQwjgWAQIECBAgQKBZIN0T4xe45vv6PQECBAgQIECAAAECBM4JCHDnSu7CBAgQIECAAAECBAi8KiDAvVo55yZAgAABAgQIECBA4JyAAHeu5C5MgAABAgQIECBAgMCrAgLcq5VzbgIECBAgQIAAAQIEzgkIcOdK7sIECBAgQIAAAQIECLwqIMC9WjnnJkCAAAECBAgQIEDgnIAAd67kLkyAAAECBAgQIECAwKsCAtyrlXNuAgQIECBAgAABAgTOCQhw50ruwgQIECBAgAABAgQIvCogwL1aOecmQIAAAQIECBAgQOCcgAB3ruQuTIAAAQIECBAgQIDAqwIC3KuVc24CBAgQIECAAAECBM4JCHDnSu7CBAgQIECAAAECBAi8KvADFsRZSLck9QcAAAAASUVORK5CYII=\">"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 reading ../data/NCBI_genome_contigs/Gsulf/scaffolds/scaf_1.gbk\n",
      "1 reading ../data/NCBI_genome_contigs/DvulH/scaffolds/scaf_1.gbk\n",
      "2 reading ../data/NCBI_genome_contigs/DdesulfG20/scaffolds/scaf_1.gbk\n",
      "3 reading ../data/NCBI_genome_contigs/EcoliK12/scaffolds/scaf_1.gbk\n",
      "4 reading ../data/NCBI_genome_contigs/Bsub/scaffolds/scaf_1.gbk\n",
      "5 reading ../data/NCBI_genome_contigs/DaudaxMP104C/scaffolds/scaf_1.gbk\n"
     ]
    }
   ],
   "source": [
    "###############################################################################\n",
    "# KGB user input vars  (Preferably implement as separate upstream cell)\n",
    "###############################################################################        \n",
    "\n",
    "\"\"\"\n",
    "################################################################################################\n",
    "# Config #1: local notebook with NCBI annotated genbank format genomes and gene tree           #\n",
    "#            Domain Annotations: NO (KBase domain annotations require KBase genome annotation) #\n",
    "#            Gene Tree: YES                                                                    #\n",
    "#            Search Terms: YES                                                                 #\n",
    "################################################################################################\n",
    "\n",
    "%pylab notebook\n",
    "KBase_backend = False\n",
    "\n",
    "GenomeSet_names = [\"Gsulf\", \"DvulH\", \"DdesulfG20\", \"EcoliK12\", \"Bsub\", \"DaudaxMP104C\"]\n",
    "ContigSet_names = []\n",
    "PivotFeatures_IDs = [\"GSU2863\", \"DVU2928\", \"Dde_2997\", \"b3987\", \"BSU01070\", \"Daud0216\"]\n",
    "PrimaryAnchor_leafId = \"Gsulf rpoB\"\n",
    "PrimaryAnchor_locusTag = \"GSU2863\"\n",
    "\n",
    "genome_annotation_system = 'NCBI'\n",
    "genome_data_format = \"Genbank\"\n",
    "# if relative path to scaffolds is e.g.: ../data/NCBI_genome_contigs/Gsulf/scaffolds/scaf_1.gbk\n",
    "gbk_ext = \"gbk\"\n",
    "genome_data_base_path = \"./example_data/NCBI_genome_contigs\"\n",
    "genome_data_extra_subpath = \"/scaffolds\"\n",
    "\n",
    "domain_data_exists = False  # NCBI annotated genome\n",
    "domain_data_format = None\n",
    "domain_data_base_path = None\n",
    "domain_data_extra_subpath = None\n",
    "domain_family_desc_base_path = None\n",
    "\n",
    "tree_data_format = 'newick'\n",
    "# if relative path to tree is e.g.: ../data/trees/rpoB_tree-names.newick\n",
    "tree_data_base_path = './example_data/trees'\n",
    "tree_data_file = 'rpoB_tree-names.newick'\n",
    "\n",
    "Search_Terms = ['dna-directed polymerase',\n",
    "                '16S',\n",
    "                'DNA and methyltransferase',\n",
    "                '1.10.3.-',\n",
    "                'fucI',\n",
    "                'sulfate adenylyl transferase']\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "##########################################################################\n",
    "# Config #1: local notebook with KBase annotated genbank format plasmids #\n",
    "#            as separate contig gbk files in a directory for each sample #\n",
    "#            Domain Annotations: YES                                     #\n",
    "#            Gene Tree: NO                                               #\n",
    "##########################################################################\n",
    "\n",
    "%pylab notebook\n",
    "KBase_backend = False\n",
    "\n",
    "GenomeSet_names = [\"GW456\", \"GW460\"]\n",
    "ContigSet_names = []\n",
    "PivotFeatures_IDs = []\n",
    "PrimaryAnchor_leafId = \"GW456.scaffold_130\"\n",
    "PrimaryAnchor_locusTag = \"GW456_SEED_annot.CDS.1\"\n",
    "\n",
    "genome_annotation_system = 'KBase'\n",
    "genome_data_format = \"Genbank\"\n",
    "# if relative path to scaffolds is e.g.: ../data/plasmid_contigs/GW456_old/scaffolds/scaffold_130.gbk\n",
    "gbk_ext = \"gbk\"\n",
    "genome_data_base_path = \"../data/plasmid_contigs\"\n",
    "genome_data_extra_subpath = \"_old/scaffolds\"\n",
    "#genome_data_extra_subpath = \"_circular_2KbpMIN/scaffolds\"\n",
    "\n",
    "domain_data_exists = True\n",
    "domain_data_format = \"KBase_domains\"\n",
    "# if relative path to domain annotations is e.g.: ../data/plasmid_contigs/GW456_old/domains/GW456_Domain_annot.json\n",
    "domain_data_base_path = \"../data/plasmid_contigs\"\n",
    "domain_data_extra_subpath = \"_old/domains\"\n",
    "domain_family_desc_base_path = \"../data/domain_desc\"\n",
    "\n",
    "tree_data_format = 'newick'\n",
    "tree_data_base_path = None\n",
    "tree_data_file = None\n",
    "\n",
    "#Search_Terms = []\n",
    "Search_Terms = ['pil or conj or mobil or partition or t-dna or vir or plasmid',\n",
    "                'toxin',\n",
    "                'tra',\n",
    "                'rep'\n",
    "               ]\n",
    "\"\"\"\n",
    "\"\"\"\n",
    "# KBase genomes in KBase context\n",
    "#\n",
    "#KBase_backend = True\n",
    "#genome_data_format = \"KBase\"\n",
    "#GenomeSet_ref = '16750/58/1'\n",
    "#GenomeSet_refs = []\n",
    "#GenomeSet_refs = ['16750/7/1',\n",
    "#                   '16750/8/2',\n",
    "#                   '16750/9/1',\n",
    "#                   '16750/10/1',\n",
    "#                   '16750/11/1'\n",
    "#                   ]\n",
    "#GenomeSet_names = []\n",
    "#PivotFeatures_IDs = []\n",
    "#Search_Terms = []\n",
    "#domain_data_exists = True\n",
    "#domain_data_format = \"KBase_domains\"\n",
    "#domain_data_base_path = None\n",
    "#domain_family_desc_base_path = None\n",
    "#tree_data_format = 'newick'\n",
    "#tree_data_base_path = None\n",
    "#tree_data_file = None\n",
    "#PrimaryAnchor_leafId = None\n",
    "#PrimaryAnchor_locusTag = None\n",
    "\n",
    "\n",
    "# KBase genomes\n",
    "#\n",
    "#ref='ReferenceGenomeAnnotations/kb|g.3899'\n",
    "#ref='ReferenceGenomeAnnotations/kb|g.2624'\n",
    "#ref='ReferenceGenomeAnnotations/kb|g.2424'\n",
    "#GenomeSet_names = ['ReferenceGenomeAnnotations/kb|g.2424']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# KBase Genbank genomes\n",
    "#\n",
    "#GenomeSet_names = [\"Desulforudis_audaxviator\",\n",
    "#                   \"Pelotomaculum_thermopropionicum\",\n",
    "#                   \"Desulfotomaculum_reducens\",\n",
    "#                   \"Bacillus_subtilis\",\n",
    "#                   \"Desulfovibrio_desulfuricans\",\n",
    "#                   \"Desulfovibrio_vulgaris\",\n",
    "#                   \"Geobacter_sulfurreducens\",\n",
    "#                   \"Escherichia_coli\"]\n",
    "##                   \"Desulforudis_audaxviator\",\n",
    "##                   \"Bacillus_subtilis\",\n",
    "#                   #\"Moorella_thermoacetica\",\n",
    "#                   #\"Thermoanaerobacter_tengcongensis\",\n",
    "#\n",
    "#ContigSet_names = []\n",
    "#PivotFeatures_IDs = [\"kb|g.1713.peg.340\",\n",
    "#                         \"kb|g.1713.peg.368\",\n",
    "#                         \"kb|g.1106.peg.1477\",\n",
    "#                         \"kb|g.996.peg.3228\",\n",
    "#                         \"kb|g.423.peg.584\",\n",
    "#                         \"kb|g.423.peg.176\",\n",
    "#                         \"kb|g.357.peg.1850\",\n",
    "#                         \"kb|g.3562.peg.757\",\n",
    "#                         \"kb|g.483.peg.109\",\n",
    "#                         \"kb|g.26833.CDS.2562\"]\n",
    "#PrimaryAnchor_leafId = \"Daudaxviator-Sat\"\n",
    "#PrimaryAnchor_locusTag = \"kb|g.1713.peg.340\"\n",
    "#\n",
    "#genome_annotation_system = 'KBase'\n",
    "#genome_data_format = \"Genbank\"\n",
    "#genome_data_base_path = \"./data/KBase_genome_contigs\"\n",
    "#genome_data_extra_subpath = \"\"\n",
    "##genome_data_extra_subpath = \"/scaffolds\"\n",
    "#\n",
    "#domain_data_format = \"KBase_domains\"\n",
    "#domain_data_base_path = None\n",
    "#domain_data_base_path = \"./data/KBase_genome_contigs\"\n",
    "#domain_data_extra_subpath = \"\"\n",
    "##domain_data_extra_subpath = \"/domains\"\n",
    "#domain_family_desc_base_path = \"./data/domain_desc\"\n",
    "#\n",
    "#tree_data_format = 'newick'\n",
    "#tree_data_base_path = None\n",
    "#tree_data_file = None\n",
    "##tree_data_base_path = './data/trees'\n",
    "##tree_data_file = 'rpoB_tree-names.newick'\n",
    "#\n",
    "##Search_Terms = []\n",
    "#Search_Terms = ['dna-directed polymerase',\n",
    "#                '16S',\n",
    "#                'DNA and methyltransferase',\n",
    "#                '1.10.3.-',\n",
    "##                'polymerase or dna-directed',\n",
    "#                'fucI',\n",
    "#                'sulfate adenylyl transferase']\n",
    "\"\"\"\n",
    "\n",
    "###############################################################################\n",
    "# KGB\n",
    "###############################################################################\n",
    "\"\"\"\n",
    "## KGB Genome Browser (KGB)                                                  \n",
    "##                                                                              \n",
    "##  An IPython/Jupyter Notebook genome browser that enables comparative         \n",
    "##  browsing and searching of genome contig assemblies, with additional         \n",
    "##  support for                                                                 \n",
    "##                                                                              \n",
    "##  * domain structure visualization                                            \n",
    "##  * gene homology                                                             \n",
    "##  * phylogenetic trees                                                        \n",
    "##\n",
    "## source code available at http://github.com/dcchivian/KGB\n",
    "##                                                                              \n",
    "## Copyright 2015-2017 Dylan Chivian  \n",
    "##\n",
    "##  Initial Author: Dylan Chivian (DCChivian@lbl.gov)\n",
    "##  $Revision: 0.3 $\n",
    "##  $Date: 2017/05/29 00:00:00 $\n",
    "##  $Author: dylan $\n",
    "##\n",
    "\"\"\"\n",
    "###############################################################################\n",
    "# INIT\n",
    "###############################################################################\n",
    "\n",
    "# Init independent of KBase_backend\n",
    "#\n",
    "from __future__ import print_function\n",
    "from __future__ import division\n",
    "#import numpy as np # comes with pylab\n",
    "import sys  # for io and exit\n",
    "from os import walk  # for dir reading\n",
    "from os import path  # for file and dir existence check\n",
    "import re\n",
    "import json\n",
    "import csv\n",
    "#from math import pi  # get with pylab\n",
    "#from math import sqrt  # get with pylab\n",
    "from math import acos\n",
    "#from Bio import Phylo\n",
    "#import re\n",
    "#matplotlib.use('nbagg')\n",
    "import matplotlib.transforms as mtransforms\n",
    "import matplotlib.patheffects as path_effects\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Ellipse\n",
    "from matplotlib.patches import Arc\n",
    "from matplotlib.patches import FancyBboxPatch\n",
    "#from matplotlib.lines import Line2D\n",
    "\n",
    "# interact is cool but outside design.  Implement via event handling on plot elements.\n",
    "#from ipywidgets import interact, interactive, fixed\n",
    "#from ipywidgets import interact\n",
    "#import ipywidgets as widgets\n",
    "\n",
    "\n",
    "# Extra Init for KBase\n",
    "#\n",
    "if KBase_backend != None and KBase_backend:\n",
    "\n",
    "    # silence whining\n",
    "    import requests\n",
    "    requests.packages.urllib3.disable_warnings()\n",
    "\n",
    "    # Standard setup for accessing Data API\n",
    "    import doekbase.data_api\n",
    "    #from doekbase.data_api.core import ObjectAPI\n",
    "    from doekbase.data_api.annotation.genome_annotation.api import GenomeAnnotationAPI\n",
    "    from doekbase.data_api.sequence.assembly.api import AssemblyAPI\n",
    "    from doekbase.data_api.taxonomy.taxon.api import TaxonAPI\n",
    "    from biokbase.workspace.client import Workspace as workspaceService\n",
    "    import os\n",
    "\n",
    "    services = {\"workspace_service_url\": \"https://kbase.us/services/ws/\",\n",
    "                \"shock_service_url\": \"https://kbase.us/services/shock-api/\"}\n",
    "    session_token = os.environ[\"KB_AUTH_TOKEN\"]\n",
    "\n",
    "    try:\n",
    "        ws = workspaceService(services['workspace_service_url'], token=session_token)\n",
    "    except Exception as e:\n",
    "        raise ValueError('Unable to establish workspace connection: ' + str(e))\n",
    "        #to get the full stack trace: traceback.format_exc()\n",
    "\n",
    "    \n",
    "# Init SeqIO just for non-KBase use\n",
    "#\n",
    "if KBase_backend == None or not KBase_backend:\n",
    "    from Bio import SeqIO\n",
    "\n",
    "           \n",
    "# Caches\n",
    "#\n",
    "search_results = []\n",
    "search_done = []\n",
    "Contig_order = []\n",
    "Contig_order_lookup = {}\n",
    "Species_name_by_genome_ref = {}\n",
    "Global_KBase_Genomes = {}\n",
    "Global_KBase_Assemblies = {}\n",
    "Global_Genbank_Genomes = []\n",
    "#Global_Features = []  # not using yet\n",
    "domain_family_desc = {}\n",
    "Global_Domains = dict()\n",
    "\n",
    "\n",
    "# Build or append to GenomeSet_names\n",
    "#\n",
    "if KBase_backend:\n",
    "    PivotFeatures_IDs = []\n",
    "    GenomeSet_refs = []\n",
    "    GenomeSet_names = dict()\n",
    "\n",
    "    # Use FeatureSet to build GenomeSet\n",
    "    #\n",
    "    if FeatureSet_ref != None:\n",
    "        try:\n",
    "            featureSet_obj = ws.get_objects([{'ref':FeatureSet_ref}])\n",
    "            featureSet_data = featureSet_obj[0]['data']\n",
    "            featureSet_info = featureSet_obj[0]['info']\n",
    "\n",
    "        except Exception as e:\n",
    "            raise ValueError('Unable to fetch featureSet object from workspace: ' + str(e))\n",
    "            #to get the full stack trace: traceback.format_exc()\n",
    "\n",
    "        featureSet_id_list = []\n",
    "        if 'element_order' in featureSet_data:\n",
    "            featureSet_id_list = featureSet_data['element_order']\n",
    "        else:\n",
    "            featureSet_id_list = featureSet_data['elements'].keys()\n",
    "\n",
    "        for f_id in featureSet_id_list:\n",
    "            PivotFeatures_IDs.append(f_id)\n",
    "            if len(featureSet_data['elements'][f_id]) == 0:\n",
    "                raise ValueError(\"missing genome reference for feature \"+f_id+\" in FeatureSet \"+FeatureSet_ref)\n",
    "            for genome_ref in featureSet_data['elements'][f_id]:\n",
    "                if genome_ref not in GenomeSet_refs:\n",
    "                    GenomeSet_refs.append(genome_ref)\n",
    "                    #GenomeSet_names[genome_ref] = genome_ref  # FIX\n",
    "                    #print (genome_id+\" \"+genome_ref)\n",
    "\n",
    "    # or use GenomeSet\n",
    "    #\n",
    "    elif GenomeSet_ref != None:\n",
    "        try:\n",
    "            genomeSet_obj = ws.get_objects([{'ref':GenomeSet_ref}])\n",
    "            genomeSet_data = genomeSet_obj[0]['data']\n",
    "            genomeSet_info = genomeSet_obj[0]['info']\n",
    "\n",
    "        except Exception as e:\n",
    "            raise ValueError('Unable to fetch GenomeSet object from workspace: ' + str(e))\n",
    "            #to get the full stack trace: traceback.format_exc()\n",
    "\n",
    "        for genome_id in genomeSet_data['elements'].keys():\n",
    "            if 'ref' not in genomeSet_data['elements'][genome_id] or \\\n",
    "                    genomeSet_data['elements'][genome_id]['ref'] == None:\n",
    "                raise ValueError(\"missing reference for genome \"+genome_id+\" in GenomeSet \"+GenomeSet_ref)\n",
    "            genome_ref = genomeSet_data['elements'][genome_id]['ref']\n",
    "            if genome_ref not in GenomeSet_refs:\n",
    "                GenomeSet_refs.append(genome_ref)\n",
    "                #GenomeSet_names[genome_ref] = genome_id\n",
    "                #print (genome_id+\" \"+genome_ref)\n",
    "\n",
    "    # Need either FeatureSet_ref or GenomeSet_ref\n",
    "    #\n",
    "    else:\n",
    "        raise ValueError (\"KGB: FeatureSet_ref or GenomeSet_ref is required\")\n",
    "\n",
    "\n",
    "# Instantiate GenomeAnnotationAPI and Build GenomeSet_names\n",
    "#\n",
    "try:\n",
    "    are_there_genome_set_names = GenomeSet_names\n",
    "except:\n",
    "    GenomeSet_names = dict()\n",
    "if KBase_backend:\n",
    "    for genome_ref in GenomeSet_refs:\n",
    "        Global_KBase_Genomes[genome_ref] = ga = GenomeAnnotationAPI(services, token=session_token, ref=genome_ref)\n",
    "        Global_KBase_Assemblies[genome_ref] = ass = ga.get_assembly()\n",
    "        tax = ga.get_taxon()\n",
    "        #        if ws_genome_id.count('/') == 2:\n",
    "        #            [ws_id, genome_id, ver] = ws_genome_id.split('/')\n",
    "        #        elif ws_genome_id.count('/') == 1:\n",
    "        #            [ws_id, genome_id] = ws_genome_id.split('/')\n",
    "        #            ver = 'auto'\n",
    "        #        else:\n",
    "        #            print (\"badly formatted ws_genome_id\")\n",
    "        #            system.exit(-1)\n",
    "        #        Species_name_by_genome_id[genome_id] = tax.get_scientific_name()\n",
    "        Species_name_by_genome_ref[genome_ref] = tax.get_scientific_name()\n",
    "        #print (\"Getting Contig IDs for Genome: \"+ws_genome_id+\": \"+Species_name_by_genome_id[genome_id])  # DEBUG\n",
    "\n",
    "        try:\n",
    "            is_there_a_genome_set_name = GenomeSet_names[genome_ref]\n",
    "        except:\n",
    "            GenomeSet_names[genome_ref] = Species_name_by_genome_ref[genome_ref]\n",
    "\n",
    "\n",
    "# Build ContigSet_names from files or from KBase object\n",
    "#\n",
    "ContigSet_names = []\n",
    "genome_contig_id_delim = '/c:'\n",
    "Genome_ref_by_Contig_name = dict()\n",
    "if KBase_backend:\n",
    "    for genome_ref in GenomeSet_refs:\n",
    "        ass = Global_KBase_Assemblies[genome_ref]\n",
    "        for scaffold_id in ass.get_contig_ids():\n",
    "            contig_name = genome_ref+genome_contig_id_delim+scaffold_id\n",
    "            ContigSet_names.append(contig_name)\n",
    "            Genome_ref_by_Contig_name[contig_name] = genome_ref\n",
    "            \n",
    "elif genome_data_format == \"Genbank\":\n",
    "    for genome_id in GenomeSet_names:\n",
    "        genome_data_dir = genome_data_base_path+'/'+genome_id+genome_data_extra_subpath\n",
    "        if not path.exists(genome_data_dir):\n",
    "            print (\"DIRNOTFOUND: \"+genome_data_dir)\n",
    "            sys.exit(-1)\n",
    "        files = []\n",
    "        for (dirpath, dirnames, filenames) in walk(genome_data_dir):\n",
    "            files.extend(filenames)\n",
    "            break\n",
    "        for file in files:\n",
    "            if file.endswith(\".\"+gbk_ext):\n",
    "                scaffold_id = file[0:file.index(\".\"+gbk_ext)]\n",
    "                contig_name = genome_id+genome_contig_id_delim+scaffold_id\n",
    "                ContigSet_names.append(contig_name)\n",
    "                Genome_ref_by_Contig_name[contig_name] = genome_id\n",
    "                print(\"reading \"+contig_name+\" ...\", flush=True)\n",
    "else:\n",
    "    print (\"unknown genome_data_format: '\"+genome_data_format+\"'\")\n",
    "    sys.exit(0)\n",
    "\n",
    "\n",
    "# Init variables / objects\n",
    "#\n",
    "tool_title = \"KGB Genome Browser\"\n",
    "color_namespace_names_disp = ['Annot', 'EC']\n",
    "color_namespace_names = ['annot', 'ec']\n",
    "if domain_data_exists:\n",
    "    color_namespace_names_disp.extend(['COG', 'Pfam', 'Domains'])\n",
    "    color_namespace_names.extend(['cog', 'pfam', 'domains'])\n",
    "\n",
    "#mode_names_disp = ['Contigs', 'Genome', 'Homologs', 'Tree', 'Strains']\n",
    "#mode_names = ['contigs', 'genome', 'homologs', 'tree', strains']\n",
    "mode_names_disp = ['Contigs', 'Genome']\n",
    "mode_names = ['contigs', 'genome']    \n",
    "def_genomebrowser_mode = \"contigs\"\n",
    "if len(ContigSet_names) == 1:\n",
    "    def_genomebrowser_mode = \"genome\"\n",
    "if len(PivotFeatures_IDs) != 0:\n",
    "    mode_names_disp.append('Homologs')\n",
    "    mode_names.append('homologs')\n",
    "    def_genomebrowser_mode = \"homologs\"\n",
    "if tree_data_file != None and tree_data_file != '':\n",
    "    mode_names_disp.append('Tree')\n",
    "    mode_names.append('tree')\n",
    "    def_genomebrowser_mode = \"tree\"\n",
    "\n",
    "    \n",
    "# Configuration\n",
    "#\n",
    "num_genomes = len(ContigSet_names)\n",
    "def_genome_mode_n_rows = 7\n",
    "max_rows = 100\n",
    "#max_rows = 10  # DEBUG\n",
    "if max_rows < def_genome_mode_n_rows:\n",
    "    max_rows = def_genome_mode_n_rows\n",
    "if def_genome_mode_n_rows%2 == 0:\n",
    "    def_genome_mode_n_rows += 1\n",
    "    \n",
    "def_popup_zorder = 100\n",
    "def_genomebrowser_window_bp_width = 1000\n",
    "def_genomebrowser_zoom_tics = 7\n",
    "def_genomebrowser_zoom = 4  # zoom values are [0..zoom_tics] -> window width = is def_genomebrowser_window_bp_width * 2**i\n",
    "def_genomebrowser_xshift = 0\n",
    "genomebrowser_window_bp_width = def_genomebrowser_window_bp_width * 2**def_genomebrowser_zoom\n",
    "\n",
    "def_genomebrowser_color_namespace = \"annot\"\n",
    "#def_genomebrowser_color_namespace = \"domains\"\n",
    "#def_genomebrowser_color_namespace = \"cog\"\n",
    "genomebrowser_mode = def_genomebrowser_mode\n",
    "\n",
    "genomebrowser_color_namespace = def_genomebrowser_color_namespace\n",
    "\n",
    "KBASE_DOMAINHIT_GENE_ID_I        = 0\n",
    "#KBASE_DOMAINHIT_GENE_BEG_I       = 1  # not used\n",
    "#KBASE_DOMAINHIT_GENE_END_I       = 2  # not used\n",
    "#KBASE_DOMAINHIT_GENE_STRAND_I    = 3  # not used\n",
    "KBASE_DOMAINHIT_GENE_HITS_DICT_I = 4\n",
    "DOMHIT_BEG_I      = 0\n",
    "DOMHIT_END_I      = 1\n",
    "DOMHIT_EVALUE_I   = 2\n",
    "DOMHIT_BITSCORE_I = 3\n",
    "DOMHIT_ALNPERC_I  = 4 \n",
    "DOMHIT_DOMFAM_I   = 5\n",
    "#KB_LOC_CTG_I = 0\n",
    "#KB_LOC_BEG_I = 1\n",
    "#KB_LOC_STR_I = 2\n",
    "#KB_LOC_LEN_I = 3\n",
    "KB_LOC_CTG_I = 'contig_id'\n",
    "KB_LOC_BEG_I = 'start'\n",
    "KB_LOC_STR_I = 'strand'\n",
    "KB_LOC_LEN_I = 'length'\n",
    "\n",
    "\n",
    "# State\n",
    "#\n",
    "Global_State = { \\\n",
    "                \"ContigSet_names\":                   ContigSet_names, \\\n",
    "                \"PivotFeatures_IDs\":                 PivotFeatures_IDs, \\\n",
    "                \"Dataset_names_list\":                [\"none\"], \\\n",
    "                \"Dataset_species_list\":              [], \\\n",
    "                \"PrimaryContig_GCavg\":               0.0, \\\n",
    "                \"PrimaryContig_len\":                 0, \\\n",
    "                \"Contig_lens\":                       [], \\\n",
    "                \"PrimaryAnchor_leafId\":              PrimaryAnchor_leafId, \\\n",
    "                \"PrimaryAnchor_pivot_pos\":           0.0, \\\n",
    "                \"pivot_pos_list\":                    [], \\\n",
    "                \"genome_mode_n_rows\":                def_genome_mode_n_rows, \\\n",
    "                \"popup_zorder\":                      def_popup_zorder, \\\n",
    "                \"genomebrowser_pivot_pos\":           0.0, \\\n",
    "                \"def_genomebrowser_window_bp_width\": def_genomebrowser_window_bp_width, \\\n",
    "                \"genomebrowser_window_bp_width\":     def_genomebrowser_window_bp_width * 2**def_genomebrowser_zoom, \\\n",
    "                \"genomebrowser_zoom\":                def_genomebrowser_zoom, \\\n",
    "                \"genomebrowser_xshift\":              def_genomebrowser_xshift, \\\n",
    "                \"genomebrowser_mode\":                def_genomebrowser_mode, \\\n",
    "                \"genomebrowser_color_namespace\":     def_genomebrowser_color_namespace\n",
    "               }\n",
    "\n",
    "\n",
    "# Set up canvas and arrow config\n",
    "#\n",
    "total_rows = num_genomes\n",
    "if def_genome_mode_n_rows > total_rows:\n",
    "    total_rows = def_genome_mode_n_rows\n",
    "if total_rows > max_rows:\n",
    "    total_rows = max_rows\n",
    "#figure_width = 12.0\n",
    "figure_width = 11.0\n",
    "figure_height_scaling = 0.75\n",
    "top_nav_height = 1.5\n",
    "#top_margin = 1.0/total_rows\n",
    "top_margin = 0.0\n",
    "left_margin = 0.05\n",
    "right_margin = 0.05\n",
    "row_delta = 1.0/total_rows\n",
    "#row_delta = 1.0/(total_rows+1)\n",
    "bottom_margin=0.75*row_delta\n",
    "arrow_w = 0.25/total_rows\n",
    "head_w = arrow_w\n",
    "#head_l = 0.4*head_w\n",
    "head_l = 0.015\n",
    "base_arrow_label_fontsize = 10\n",
    "foi_arrow_label_fontsize = 13\n",
    "text_padding = 0.03/total_rows\n",
    "text_yshift = text_padding+0.5*arrow_w\n",
    "#arrow_label_scaling = 125.0  # about appropriate value for fontsize 12 and figure width 15 (or maybe 20)\n",
    "#arrow_label_scaling = 95.0  # appropriate value for fontsize 10 and figure width 12\n",
    "arrow_label_scaling = 100.0  # appropriate value for fontsize 10 and figure width 12\n",
    "text_disp_window_bp_limit = 80000\n",
    "\n",
    "\n",
    "# Colors\n",
    "#\n",
    "nav_bg_color = \"slategray\"\n",
    "popup_bg_color = \"aliceblue\"\n",
    "#popup_bg_color = \"cornsilk\"\n",
    "#popup_bg_color = \"lemonchiffon\"\n",
    "search_color_names = [\n",
    "    'magenta', \\\n",
    "    'orange', \\\n",
    "    'darkcyan', \\\n",
    "    'green', \\\n",
    "    'red', \\\n",
    "    'lightskyblue', \\\n",
    "    'crimson'\n",
    "]\n",
    "# feature colors names list (left out excessively light colors)\n",
    "color_names = [\n",
    "    #'aliceblue',\n",
    "    'aqua',\n",
    "    'aquamarine',\n",
    "    #'azure',\n",
    "    #'beige',\n",
    "    #'bisque',\n",
    "    #'blanchedalmond',\n",
    "    'blue',\n",
    "    'blueviolet',\n",
    "    'brown',\n",
    "    'burlywood',\n",
    "    'cadetblue',\n",
    "    'chartreuse',\n",
    "    'chocolate',\n",
    "    'coral',\n",
    "    'cornflowerblue',\n",
    "    #'cornsilk',\n",
    "    'crimson',\n",
    "    'cyan',\n",
    "    'darkblue',\n",
    "    'darkcyan',\n",
    "    'darkgoldenrod',\n",
    "    'darkgreen',\n",
    "    'darkkhaki',\n",
    "    'darkmagenta',\n",
    "    'darkolivegreen',\n",
    "    'darkorange',\n",
    "    'darkorchid',\n",
    "    'darkred',\n",
    "    'darksalmon',\n",
    "    'darkseagreen',\n",
    "    'darkslateblue',\n",
    "    #'darkslategray',\n",
    "    'darkturquoise',\n",
    "    'darkviolet',\n",
    "    'deeppink',\n",
    "    'deepskyblue',\n",
    "    'dodgerblue',\n",
    "    'firebrick',\n",
    "    'forestgreen',\n",
    "    'fuchsia',\n",
    "    #'gainsboro',\n",
    "    'gold',\n",
    "    'goldenrod',\n",
    "    'green',\n",
    "    'greenyellow',\n",
    "    #'honeydew',\n",
    "    'hotpink',\n",
    "    'indianred',\n",
    "    'indigo',\n",
    "    'khaki',\n",
    "    #'lavender',\n",
    "    #'lavenderblush',\n",
    "    'lawngreen',\n",
    "    #'lemonchiffon',\n",
    "    'lightblue',\n",
    "    #'lightcoral',\n",
    "    #'lightcyan'\n",
    "    #'lightgoldenrodyellow',\n",
    "    'lightgreen',\n",
    "    'lightpink',\n",
    "    'lightsalmon',\n",
    "    'lightseagreen',\n",
    "    'lightskyblue',\n",
    "    #'lightslategray',\n",
    "    #'lightsteelblue',\n",
    "    #'lightyellow',\n",
    "    'lime',\n",
    "    'limegreen',\n",
    "    'magenta',\n",
    "    'maroon',\n",
    "    'mediumaquamarine',\n",
    "    'mediumblue',\n",
    "    'mediumorchid',\n",
    "    'mediumpurple',\n",
    "    'mediumseagreen',\n",
    "    'mediumslateblue',\n",
    "    'mediumspringgreen',\n",
    "    'mediumturquoise',\n",
    "    'mediumvioletred',\n",
    "    'midnightblue',\n",
    "    #'mintcream',\n",
    "    #'mistyrose',\n",
    "    #'moccasin',\n",
    "    'navy',\n",
    "    #'oldlace',\n",
    "    'olive',\n",
    "    'olivedrab',\n",
    "    'orange',\n",
    "    'orangered',\n",
    "    'orchid',\n",
    "    #'palegoldenrod',\n",
    "    'palegreen',\n",
    "    'paleturquoise',\n",
    "    'palevioletred',\n",
    "    #'papayawhip',\n",
    "    'peachpuff',\n",
    "    #'peru',\n",
    "    'pink',\n",
    "    'plum',\n",
    "    'powderblue',\n",
    "    'purple',\n",
    "    'red',\n",
    "    'rosybrown',\n",
    "    'royalblue',\n",
    "    'saddlebrown',\n",
    "    'salmon',\n",
    "    'sandybrown',\n",
    "    'seagreen',\n",
    "    #'seashell',\n",
    "    'sienna',\n",
    "    'skyblue',\n",
    "    'slateblue',\n",
    "    'springgreen',\n",
    "    'steelblue',\n",
    "    #'tan',\n",
    "    'teal',\n",
    "    #'thistle',\n",
    "    'tomato',\n",
    "    'turquoise',\n",
    "    'violet',\n",
    "    #'wheat',\n",
    "    #'yellow',\n",
    "    #'yellowgreen'\n",
    "    ]\n",
    "\n",
    "###############################################################################\n",
    "# Subs\n",
    "###############################################################################\n",
    "\n",
    "# Utilities\n",
    "#\n",
    "def compute_GC (dna_seq):\n",
    "    seq_len = len(dna_seq)\n",
    "    GC_cnt = 0\n",
    "    AT_cnt = 0\n",
    "    other_cnt = 0\n",
    "    dna_seq.upper()\n",
    "    GC_cnt = dna_seq.count('G') + dna_seq.count('C')\n",
    "    AT_cnt = dna_seq.count('A') + dna_seq.count('T') + dna_seq.count('U')\n",
    "    other_cnt = seq_len - GC_cnt - AT_cnt\n",
    "    seq_len -= other_cnt\n",
    "    return GC_cnt / seq_len\n",
    "\n",
    "\n",
    "# KBase feature structure\n",
    "#\n",
    "def build_feature_rec_kbase (f, f_type='CDS', source_species='', contig_i=0, dna_seq=None):\n",
    "    feature_rec = {}\n",
    "    \n",
    "    id_delim = '.'\n",
    "        \n",
    "    # IDs and names\n",
    "    name = f['feature_id']\n",
    "    name_split = name.split(id_delim)\n",
    "    name = id_delim.join(name_split[2:])\n",
    "\n",
    "    locus_tag = f['feature_id']\n",
    "    aliases = []\n",
    "    for alias in f['feature_aliases'].keys():\n",
    "        if locus_tag == f['feature_id'] and 'IPR' not in alias:\n",
    "            locus_tag = alias  # fix this to match regexp \\D+_?\\d+ (but not IPR*), and stop assignment\n",
    "        aliases.append(alias)\n",
    "\n",
    "    feature_ID = f['feature_id']\n",
    "    #print (\"FEATURE_ID: '\"+feature_ID+\"'\")  # DEBUG\n",
    "    \n",
    "    # coords\n",
    "    strand = f['feature_locations'][0][KB_LOC_STR_I]\n",
    "    f_len = f['feature_locations'][0][KB_LOC_LEN_I]\n",
    "    if strand == '+':\n",
    "        beg = f['feature_locations'][0][KB_LOC_BEG_I]\n",
    "        end = beg + f_len - 1\n",
    "    else:\n",
    "        end = f['feature_locations'][0][KB_LOC_BEG_I]\n",
    "        beg = end - f_len + 1    \n",
    "    if end < beg:\n",
    "        print (\"WARNING: reversed gene: %s: %s (%d-%d)\"%(source_species, locus_tag, beg, end))\n",
    "        tmp_pos = end\n",
    "        end = beg\n",
    "        beg = tmp_pos\n",
    "\n",
    "    # annotation\n",
    "    annotation = f['feature_function']\n",
    "    EC_in_annotation = ''\n",
    "    in_paren = False\n",
    "    last_good_char = 0\n",
    "    close_paren_pos = 0\n",
    "    for i,c in enumerate(annotation[::-1]):\n",
    "        if in_paren == False and c != ' ' and c != ')':\n",
    "            break\n",
    "        if in_paren == False and c != ' ' and c == ')':\n",
    "            in_paren = True\n",
    "            close_paren_pos = i\n",
    "        if in_paren and c == '(':\n",
    "            last_good_char = i+1\n",
    "            break\n",
    "    if last_good_char > 0:\n",
    "        candidate_EC = annotation[-1-last_good_char+2:-1-close_paren_pos]\n",
    "        if candidate_EC[0:2] == 'EC':\n",
    "            #print (name + \" EC: '\" + candidate_EC +\"'\")  # DEBUG\n",
    "            EC_in_annotation = candidate_EC[3:]\n",
    "\n",
    "        annotation = annotation[0:-1-last_good_char]\n",
    "    annotation = annotation.strip()\n",
    "    annotation = annotation.lstrip()\n",
    "\n",
    "    # add domain hits to annotation\n",
    "    try:\n",
    "        domain_hits = Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[contig_i]]][ContigSet_names[contig_i]][feature_ID]\n",
    "        domfam_seen = {}\n",
    "        if annotation != \"\":\n",
    "            annotation += \"\\n\"\n",
    "        for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "            domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "            try:\n",
    "                seen = domfam_seen[domfam]\n",
    "                continue\n",
    "            except:\n",
    "                domfam_seen[domfam] = True\n",
    "                if domfam[0:3] == 'COG' or domfam[0:2] == 'PF' or domfam[0:4] == 'TIGR':\n",
    "                    if domfam[0:2] == 'PF':\n",
    "                        dom_namespace = 'PF'\n",
    "                        [domfam, ver] = domfam.split('.')\n",
    "                    elif domfam[0:3] == 'COG':\n",
    "                        dom_namespace = 'COG'\n",
    "                    elif domfam[0:4] == 'TIGR':\n",
    "                        dom_namespace = 'TIGR'\n",
    "                    line_break = \"\\n\"\n",
    "                    if annotation == \"\":\n",
    "                        line_break = \"\"\n",
    "                    try:\n",
    "                        domfam_desc = \": \"+domain_family_desc[dom_namespace][domfam]\n",
    "                    except:\n",
    "                        domfam_desc = \"\"\n",
    "                    annotation += line_break+domfam+domfam_desc\n",
    "    except:\n",
    "        domain_hits = []\n",
    "                \n",
    "    # pull out E.C.    \n",
    "    EC_number = ''\n",
    "    if EC_in_annotation != '':\n",
    "        EC_number = EC_in_annotation\n",
    "\n",
    "    # create feature_rec\n",
    "    feature_rec = {\"source_species\": source_species, \\\n",
    "                   \"name\": name, \\\n",
    "                   \"locus_tag\": locus_tag, \\\n",
    "                   \"ID\": feature_ID, # will probably need to make more unique \\\n",
    "                   \"aliases\": aliases, \\\n",
    "                   \"type\": f_type, \\\n",
    "                   \"beg_pos\": beg, \\\n",
    "                   \"end_pos\": end, \\\n",
    "                   \"strand\": strand, \\\n",
    "                   \"annot\": annotation, \\\n",
    "                   \"EC_number\": EC_number, \\\n",
    "                   \"dna_seq\": dna_seq\n",
    "                  }                         \n",
    "\n",
    "    return feature_rec\n",
    "\n",
    "\n",
    "# Genbank feature structure\n",
    "#\n",
    "def build_feature_rec_genbank (f, f_type='CDS', source_species='', contig_i=0, dna_seq=None):\n",
    "    feature_rec = {}\n",
    "    \n",
    "    id_delim = '.'\n",
    "\n",
    "    # IDs and names\n",
    "    if \"gene\" in f.qualifiers:\n",
    "        name = f.qualifiers['gene'][0]\n",
    "    elif \"locus_tag\" in f.qualifiers:\n",
    "        name = f.qualifiers['locus_tag'][0]\n",
    "    elif \"db_xref\" in f.qualifiers:\n",
    "        name = f.qualifiers['db_xref'][0]\n",
    "    else:\n",
    "        print (\"strangely formatted record.  SKIPPING\")\n",
    "        print (f)\n",
    "        return {}\n",
    "#    name_split = name.split(id_delim)\n",
    "#    name = id_delim.join(name_split[1:])\n",
    "    \n",
    "    if \"locus_tag\" in f.qualifiers:\n",
    "        locus_tag = f.qualifiers['locus_tag'][0]\n",
    "    elif \"gene\" in f.qualifiers:\n",
    "        locus_tag = f.qualifiers['gene'][0]\n",
    "    elif \"db_xref\" in f.qualifiers:\n",
    "        locus_tag = f.qualifiers['db_xref'][0]\n",
    "    else:\n",
    "        print (\"strangely formatted record.  SKIPPING\")\n",
    "        print (f)\n",
    "        return {}\n",
    "    locus_tag_split = locus_tag.split(id_delim)\n",
    "    locus_tag = id_delim.join(locus_tag_split[1:])\n",
    "    \n",
    "    # capture aliases\n",
    "    aliases = []\n",
    "    if \"db_xref\" in f.qualifiers:\n",
    "        for alias in f.qualifiers['db_xref']:\n",
    "            aliases.append(alias)\n",
    "        \n",
    "    # For now, feature ID will be locus_tag\n",
    "    feature_ID = locus_tag\n",
    "    \n",
    "    # coords\n",
    "    beg = f.location.start + 1  # BioPython is cute and shifts just the begin pos by -1 \n",
    "    end = f.location.end\n",
    "    if end < beg:\n",
    "        print (\"WARNING: reversed gene: %s: %s (%d-%d)\"%(source_species, locus_tag, beg, end))\n",
    "        tmp_pos = end\n",
    "        end = beg\n",
    "        beg = tmp_pos\n",
    "    if f.location.strand == -1:\n",
    "        strand = '-'\n",
    "    else :\n",
    "        strand = '+'\n",
    "\n",
    "    # annotation\n",
    "    annotation = ''\n",
    "    EC_in_annotation = ''\n",
    "    if f_type == \"CDS\" or f_type == \"rRNA\" or f_type == \"tRNA\":    \n",
    "        if genome_annotation_system == 'KBase':\n",
    "            annotation = f.qualifiers['function'][0]\n",
    "        else:\n",
    "            annotation = f.qualifiers['product'][0]\n",
    "        in_paren = False\n",
    "        last_good_char = 0\n",
    "        close_paren_pos = 0\n",
    "        for i,c in enumerate(annotation[::-1]):\n",
    "            if in_paren == False and c != ' ' and c != ')':\n",
    "                break\n",
    "            if in_paren == False and c != ' ' and c == ')':\n",
    "                in_paren = True\n",
    "                close_paren_pos = i\n",
    "            if in_paren and c == '(':\n",
    "                last_good_char = i+1\n",
    "                break\n",
    "        if last_good_char > 0:\n",
    "            candidate_EC = annotation[-1-last_good_char+2:-1-close_paren_pos]\n",
    "            if candidate_EC[0:2] == 'EC':\n",
    "                #print (name + \" EC: '\" + candidate_EC +\"'\")  # DEBUG\n",
    "                EC_in_annotation = candidate_EC[3:]\n",
    "\n",
    "            annotation = annotation[0:-1-last_good_char]\n",
    "        annotation = annotation.strip()\n",
    "        annotation = annotation.lstrip()\n",
    "\n",
    "    # add domain hits to annotation\n",
    "    try:\n",
    "        domain_hits = Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[contig_i]]][feature_ID]\n",
    "        domfam_seen = {}\n",
    "        if annotation != \"\":\n",
    "            annotation += \"\\n\"\n",
    "        for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "            domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "            try:\n",
    "                seen = domfam_seen[domfam]\n",
    "                continue\n",
    "            except:\n",
    "                domfam_seen[domfam] = True\n",
    "                if domfam[0:3] == 'COG' or domfam[0:2] == 'PF' or domfam[0:4] == 'TIGR':\n",
    "                    if domfam[0:2] == 'PF':\n",
    "                        dom_namespace = 'PF'\n",
    "                        [domfam, ver] = domfam.split('.')\n",
    "                    elif domfam[0:3] == 'COG':\n",
    "                        dom_namespace = 'COG'\n",
    "                    elif domfam[0:4] == 'TIGR':\n",
    "                        dom_namespace = 'TIGR'\n",
    "                    line_break = \"\\n\"\n",
    "                    if annotation == \"\":\n",
    "                        line_break = \"\"\n",
    "                    try:\n",
    "                        domfam_desc = \": \"+domain_family_desc[dom_namespace][domfam]\n",
    "                    except:\n",
    "                        domfam_desc = \"\"\n",
    "                    annotation += line_break+domfam+domfam_desc\n",
    "    except:\n",
    "        domain_hits = []\n",
    "                \n",
    "    # pull out E.C.    \n",
    "    EC_number = ''\n",
    "    if EC_in_annotation != '':\n",
    "        EC_number = EC_in_annotation\n",
    "    else:\n",
    "        if f_type == \"CDS\" and 'EC_number' in f.qualifiers:\n",
    "            EC_number = f.qualifiers['EC_number'][0]\n",
    "\n",
    "    # create feature_rec\n",
    "    feature_rec = {\"source_species\": source_species, \\\n",
    "                   \"name\": name, \\\n",
    "                   \"locus_tag\": locus_tag, \\\n",
    "                   \"ID\": feature_ID, # will probably need to make more unique \\\n",
    "                   \"aliases\": aliases, \\\n",
    "                   \"type\": f_type, \\\n",
    "                   \"beg_pos\": beg, \\\n",
    "                   \"end_pos\": end, \\\n",
    "                   \"strand\": strand, \\\n",
    "                   \"annot\": annotation, \\\n",
    "                   \"EC_number\": EC_number, \\\n",
    "                   \"dna_seq\": dna_seq\n",
    "                  }                         \n",
    "    \n",
    "    #if feature_rec['name'] == \"DVU2932\" or feature_rec['name'] == \"Daud0298\":\n",
    "    #    arrow_color = color_names[sum([ord(c) for c in feature_rec['annot']]) % len(color_names)]\n",
    "    #    print (feature_rec['source_species'] + \" \" + feature_rec['name']+\" \"+arrow_color)\n",
    "\n",
    "    return feature_rec\n",
    "\n",
    "\n",
    "# Load domain family descriptions\n",
    "#\n",
    "def read_domain_family_desc ():\n",
    "    #domain_family_desc = {}\n",
    "\n",
    "    # read Pfam descs\n",
    "    pfam_desc_file = domain_family_desc_base_path+'/Pfam-A.clans.tsv'\n",
    "    if path.isfile(pfam_desc_file):\n",
    "        domain_family_desc['PF'] = {}\n",
    "        with open(pfam_desc_file,'r') as f:\n",
    "            reader=csv.reader(f,delimiter='\\t')\n",
    "            for pfam,clan,clan_name,pfam_name,pfam_desc in reader:\n",
    "                domain_family_desc['PF'][pfam] = pfam_desc\n",
    "                \n",
    "    # read COG descs\n",
    "    COG_desc_file = domain_family_desc_base_path+'/COG_2014.tsv'\n",
    "    if path.isfile(COG_desc_file):\n",
    "        domain_family_desc['COG'] = {}\n",
    "        with open(COG_desc_file,'r') as f:\n",
    "            reader=csv.reader(f,delimiter='\\t')\n",
    "            for COG,COG_funcat,COG_desc in reader:\n",
    "                domain_family_desc['COG'][COG] = COG_desc\n",
    "                \n",
    "    # read TIGR descs\n",
    "    TIGR_desc_file = domain_family_desc_base_path+'/TIGRInfo.tsv'\n",
    "    if path.isfile(TIGR_desc_file):\n",
    "        domain_family_desc['TIGR'] = {}\n",
    "        with open(TIGR_desc_file,'r') as f:\n",
    "            reader=csv.reader(f,delimiter='\\t')\n",
    "            for id,tigrId,type,roleId,geneSymbol,ec,definition in reader:\n",
    "                domain_family_desc['TIGR'][tigrId] = definition\n",
    "                if ec != \"\":\n",
    "                    domain_family_desc['TIGR'][tigrId] += \" (EC \"+ec+\")\"\n",
    "                \n",
    "    #return domain_family_desc\n",
    "    \n",
    "\n",
    "# Check search terms\n",
    "#\n",
    "def search_term_match (feature_rec):    \n",
    "    term_match = []\n",
    "    for i,term in enumerate(Search_Terms):\n",
    "        term_uc = term.upper()  # FIX: should sanitize search terms once at beginning\n",
    "        term_match.append(False)\n",
    "                    \n",
    "        for k in ['name', 'locus_tag', 'ID', 'aliases', 'annot', 'EC_number']:\n",
    "            info_uc_list = []\n",
    "            \n",
    "            if k == 'aliases':\n",
    "                try:\n",
    "                    for f in feature_rec[k]:\n",
    "                        info_uc_list.append(f[k].upper())\n",
    "                except:\n",
    "                    continue                \n",
    "            else:\n",
    "                #print (\"CHECKING %s %s %s %s\"%(term,k,feature_rec['name'], feature_rec[k]))  # DEBUG\n",
    "                try:\n",
    "                    info_uc_list = [feature_rec[k].upper()]\n",
    "                except:\n",
    "                    continue\n",
    "                \n",
    "            search_words = term_uc.split(\" \")  # FIX: should sanitize search terms once at beginning\n",
    "            logical = 'AND'\n",
    "            if \" OR \" in term_uc:\n",
    "                logical = 'OR'\n",
    "\n",
    "            this_info_match = False\n",
    "\n",
    "            for info_uc in info_uc_list:\n",
    "                if logical == 'AND':\n",
    "                    this_info_match = True\n",
    "                    for word in search_words:\n",
    "                        #if word == 'AND' or word == 'OR':\n",
    "                        if word == 'AND' or word == 'OR' or word == '':  # FIX: should sanitize search terms once at beginning\n",
    "                            continue\n",
    "                        if word not in info_uc:\n",
    "                            this_info_match = False\n",
    "                            break\n",
    "                if logical == 'OR':\n",
    "                    this_info_match = False\n",
    "                    for word in search_words:\n",
    "                        #if word == 'AND' or word == 'OR':\n",
    "                        if word == 'AND' or word == 'OR' or word == '':  # FIX: should sanitize search terms once at beginning\n",
    "                            continue\n",
    "                        if word in info_uc:\n",
    "                            this_info_match = True\n",
    "                            break\n",
    "\n",
    "                if this_info_match:\n",
    "                    term_match[i] = True\n",
    "                    break\n",
    "\n",
    "            if this_info_match:\n",
    "                term_match[i] = True\n",
    "                break\n",
    "                    \n",
    "    return term_match\n",
    "    \n",
    "\n",
    "# sort rules\n",
    "#\n",
    "def sort_by_beg_pos_key (feature):\n",
    "    return feature['beg_pos']    \n",
    "\n",
    "def sort_by_bitscore_key (domhit):\n",
    "    return domhit[DOMHIT_BITSCORE_I]    \n",
    "\n",
    "def sort_by_domhit_len (domhit):\n",
    "    return domhit[DOMHIT_END_I]-domhit[DOMHIT_BEG_I]+1\n",
    "\n",
    "# Read KBase format domain hits\n",
    "#\n",
    "def getDomainHits (ContigSet_names, \\\n",
    "                   genomebrowser_mode=def_genomebrowser_mode,\n",
    "                   domain_data_format=domain_data_format):\n",
    "\n",
    "    #print (\"GETTING DOMAINHITS\")  # DEBUG\n",
    "\n",
    "    for i,contig_name in enumerate(ContigSet_names):\n",
    "        (genome_id,scaffold_id) = contig_name.split(genome_contig_id_delim)\n",
    "        #print(\"GETTING DOMAINHITS FOR \"+contig_name+\" \"+genome_id+\" c:\"+scaffold_id)  # DEBUG\n",
    "            \n",
    "        if Global_State['genomebrowser_mode'] == \"genome\" and i > 0:\n",
    "            break\n",
    "            \n",
    "        try:\n",
    "            Domain_Hits = Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[i]]]\n",
    "        except:\n",
    "            Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[i]]] = dict()\n",
    "\n",
    "            if KBase_backend:\n",
    "                genome_ref = Genome_ref_by_Contig_name[contig_name]\n",
    "                domain_data = None\n",
    "                found_domain_data = False\n",
    "\n",
    "                #print (\"CONTIG_NAME: '\"+contig_name+\"' GENOME_REF: '\"+genome_ref+\"'\")  # DEBUG\n",
    "\n",
    "                if genome_ref.count('/') == 2:\n",
    "                    (ws_id, ws_genome_id, ver) = genome_ref.split('/')\n",
    "                elif genome_ref.count('/') == 1:\n",
    "                    (ws_id, ws_genome_id) = genome_ref.split('/')\n",
    "                    ver = 'auto'\n",
    "\n",
    "                try:\n",
    "                    domain_annotation_obj_list = ws.list_objects({'ids':[ws_id],'type':\"KBaseGeneFamilies.DomainAnnotation\"})\n",
    "                except Exception as e:\n",
    "                    raise ValueError('Unable to list DomainAnnotation objects from workspace: ' + str(e))\n",
    "                    #to get the full stack trace: traceback.format_exc()\n",
    "                for info in domain_annotation_obj_list:\n",
    "                    # Object Info Contents\n",
    "                    # 0 - obj_id objid\n",
    "                    # 1 - obj_name name\n",
    "                    # 2 - type_string type\n",
    "                    # 3 - timestamp save_date\n",
    "                    # 4 - int version\n",
    "                    # 5 - username saved_by\n",
    "                    # 6 - ws_id wsid\n",
    "                    # 7 - ws_name workspace\n",
    "                    # 8 - string chsum\n",
    "                    # 9 - int size \n",
    "                    # 10 - usermeta meta\n",
    "                    # absolute ref = str(info[6]) + '/' + str(info[0]) + '/' + str(info[4])\n",
    "                    domain_annotation_ref = str(info[6])+'/'+str(info[0])+'/'+str(info[4])\n",
    "                    try:\n",
    "                        domain_data = ws.get_objects([{'ref':domain_annotation_ref}])[0]['data']  \n",
    "                    except Exception as e:\n",
    "                        raise ValueError('Unable to fetch DomainAnnotation object from workspace: ' + str(e))\n",
    "                    #to get the full stack trace: traceback.format_exc()\n",
    "\n",
    "                    # we found the correct DomainAnnotation object\n",
    "                    if domain_data['genome_ref'] == genome_ref:\n",
    "                        #print (\"FOUND DOM ANNOT \"+str(domain_annotation_ref)+\" FOR GENOME \"+str(genome_ref))  # DEBUG\n",
    "                        found_domain_data = True\n",
    "                        break\n",
    "\n",
    "                if not found_domain_data:\n",
    "                    continue\n",
    "\n",
    "                for scaffold_id_iter in domain_data['data'].keys():\n",
    "                    contig_id_iter = genome_ref+genome_contig_id_delim+scaffold_id_iter\n",
    "                    Global_Domains[genome_ref][contig_id_iter] = dict()\n",
    "\n",
    "                    for CDS_domain_list in domain_data['data'][scaffold_id_iter]:\n",
    "                        gene_ID   = CDS_domain_list[KBASE_DOMAINHIT_GENE_ID_I]\n",
    "                        #gene_name = re.sub ('^'+genome_object_name+'.', '', gene_ID) \n",
    "                        gene_name = gene_ID\n",
    "                        #(contig_name, gene_name) = (gene_ID[0:gene_ID.index(\".\")], gene_ID[gene_ID.index(\".\")+1:])\n",
    "                        #print (\"DOMAIN_HIT: \"+contig_name+\" \"+gene_name)  # DEBUG\n",
    "                        #print (\"DOMAIN_HIT for gene: \"+gene_name)  # DEBUG\n",
    "                        #gene_beg       = CDS_domain_list[KBASE_DOMAINHIT_GENE_BEG_I]\n",
    "                        #gene_end       = CDS_domain_list[KBASE_DOMAINHIT_GENE_END_I]\n",
    "                        #gene_strand    = CDS_domain_list[KBASE_DOMAINHIT_GENE_STRAND_I]\n",
    "                        gene_hits_dict = CDS_domain_list[KBASE_DOMAINHIT_GENE_HITS_DICT_I]\n",
    "                        gene_hits_list = []\n",
    "                        for domfam in gene_hits_dict.keys():\n",
    "                            # skip CD hits for now\n",
    "                            if domfam[0:2] != 'PF' and domfam[0:3] != 'COG' and domfam[0:4] != 'TIGR':\n",
    "                                continue\n",
    "                            #Global_Domains[i][gene_name] = gene_hits_dict\n",
    "                            for hit in gene_hits_dict[domfam]:\n",
    "                                list_format_hit = hit\n",
    "                                list_format_hit.append (domfam)\n",
    "                                #list_format_hit[DOMHIT_BEG_I]      = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_BEG_J]\n",
    "                                #list_format_hit[DOMHIT_END_I]      = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_END_J]\n",
    "                                #list_format_hit[DOMHIT_EVALUE_I]   = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_EVALUE_J]\n",
    "                                #list_format_hit[DOMHIT_BITSCORE_I] = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_BITSCORE_J]\n",
    "                                #list_format_hit[DOMHIT_ALNPERC_I]  = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_ALNPERC_J]\n",
    "                                gene_hits_list.append(list_format_hit)\n",
    "                                #print (\"   DOMAIN_HIT: \"+domfam)  # DEBUG\n",
    "                                #print (\"%s\\t%s\\t%s\\t%s\\t%d\\t%d\\t%f\"%(genome_id, scaffold_id, gene_name, domfam, hit_beg, hit_end, hit_evalue))\n",
    "                        Global_Domains[genome_ref][contig_id_iter][gene_name] = sorted (gene_hits_list, key=sort_by_bitscore_key, reverse=True)\n",
    "\n",
    "\n",
    "            # Or running outside KBase\n",
    "            #\n",
    "            elif domain_data_format == \"KBase_domains\" and domain_data_exists:\n",
    "                \n",
    "                domain_data_path = domain_data_base_path+'/'+genome_id+domain_data_extra_subpath+'/'+genome_id+\"_Domain_annot\"+'.json'\n",
    "                print (\"reading \"+domain_data_path+\" ...\")\n",
    "                with open(domain_data_path, 'r') as domain_file_handle:\n",
    "                    kbase_domains = json.load(domain_file_handle)\n",
    "\n",
    "                    for scaffold_id_iter in kbase_domains['data'].keys(): \n",
    "                        contig_id_iter = Genome_ref_by_Contig_name[ContigSet_names[i]]+genome_contig_id_delim+scaffold_id_iter\n",
    "                        Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[i]]][contig_id_iter] = dict()\n",
    "                        for CDS_domain_list in kbase_domains['data'][scaffold_id]:\n",
    "                            gene_ID        = CDS_domain_list[KBASE_DOMAINHIT_GENE_ID_I]\n",
    "                            (contig_name, gene_name) = (gene_ID[0:gene_ID.index(\".\")], gene_ID[gene_ID.index(\".\")+1:])\n",
    "                            #gene_beg       = CDS_domain_list[KBASE_DOMAINHIT_GENE_BEG_I]\n",
    "                            #gene_end       = CDS_domain_list[KBASE_DOMAINHIT_GENE_END_I]\n",
    "                            #gene_strand    = CDS_domain_list[KBASE_DOMAINHIT_GENE_STRAND_I]\n",
    "                            gene_hits_dict = CDS_domain_list[KBASE_DOMAINHIT_GENE_HITS_DICT_I]\n",
    "                            gene_hits_list = []\n",
    "                            for domfam in gene_hits_dict.keys():\n",
    "                                # skip CD hits for now\n",
    "                                if domfam[0:2] != 'PF' and domfam[0:3] != 'COG' and domfam[0:4] != 'TIGR':\n",
    "                                    continue\n",
    "                                #Global_Domains[i][gene_name] = gene_hits_dict\n",
    "                                for hit in gene_hits_dict[domfam]:\n",
    "                                    list_format_hit = hit\n",
    "                                    list_format_hit.append (domfam)\n",
    "                                    #list_format_hit[DOMHIT_BEG_I]      = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_BEG_J]\n",
    "                                    #list_format_hit[DOMHIT_END_I]      = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_END_J]\n",
    "                                    #list_format_hit[DOMHIT_EVALUE_I]   = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_EVALUE_J]\n",
    "                                    #list_format_hit[DOMHIT_BITSCORE_I] = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_BITSCORE_J]\n",
    "                                    #list_format_hit[DOMHIT_ALNPERC_I]  = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_ALNPERC_J]\n",
    "                                    gene_hits_list.append(list_format_hit)\n",
    "                                    #print (\"%s\\t%s\\t%s\\t%s\\t%d\\t%d\\t%f\"%(genome_id, scaffold_id, gene_name, domfam, hit_beg, hit_end, hit_evalue))\n",
    "                            Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[i]]][contig_id_iter][gene_name] = sorted (gene_hits_list, key=sort_by_bitscore_key, reverse=True)\n",
    "                        \n",
    "                            \n",
    "# Workhorse for feature retrieval (KBase flavor)\n",
    "#   does more than window slice, so perhaps should split up functionality\n",
    "#\n",
    "def getFeatureSlicesKBase (ContigSet_names, \\\n",
    "                      PivotFeatures_IDs, \\\n",
    "                      genomebrowser_mode=\"contigs\", \\\n",
    "                      genome_data_format=\"KBase\", \\\n",
    "                      window_size=10000, \\\n",
    "                      genomebrowser_xshift=0):\n",
    "\n",
    "    Feature_slices = []\n",
    "    \n",
    "    if genomebrowser_mode != \"genome\":\n",
    "\n",
    "        if KBase_backend:  \n",
    "            \n",
    "            for i,contig_name in enumerate(ContigSet_names):\n",
    "                if i >= max_rows:\n",
    "                    break\n",
    "\n",
    "                (genome_id,scaffold_id) = contig_name.split(genome_contig_id_delim)\n",
    "                #print (\"GENOME_ID: '\"+genome_id+\"' SCAFFOLD_ID: '\"+scaffold_id+\"' CONTIG_NAME: '\"+contig_name+\"'\")  # DEBUG\n",
    "\n",
    "                ga = Global_KBase_Genomes[genome_id]\n",
    "                ass = Global_KBase_Assemblies[genome_id]\n",
    "            \n",
    "                Feature_slice = []\n",
    "                Features_seen = set()\n",
    "                source = \"\"\n",
    "                pivot_pos = 0.0\n",
    "\n",
    "                try:\n",
    "                    this_search_done = search_done[i]\n",
    "                except:\n",
    "                    search_done.append([])                \n",
    "                    search_done[i] = False\n",
    "                    search_results.append([])\n",
    "                    search_results[i] = []\n",
    "                    for j,term in enumerate(Search_Terms):\n",
    "                        search_results[i].append([])\n",
    "\n",
    "                # Get genome length\n",
    "                #if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                contig_lens = ass.get_contig_lengths()\n",
    "                if i == 0:\n",
    "                    contig_GCs  = ass.get_contig_gc_content(contig_id_list=[scaffold_id])\n",
    "                    Global_State['PrimaryContig_len'] = contig_lens[scaffold_id]                    \n",
    "                    Global_State['PrimaryContig_GCavg'] = contig_GCs[scaffold_id]\n",
    "                    Global_State['Contig_lens'] = []\n",
    "                    Global_State['pivot_pos_list'] = []\n",
    "                    #print (\"%d\"%Global_State['PrimaryContig_len'])\n",
    "                Global_State['Contig_lens'].append(contig_lens[scaffold_id])\n",
    "                \n",
    "\n",
    "                # Find pivot feature and put in first position\n",
    "                #\n",
    "                contig_mode_xshift = 0\n",
    "                source = 'Is missingii'  # FIX THIS\n",
    "                        \n",
    "                try:\n",
    "                    this_pivotfeature_id = PivotFeatures_IDs[i]\n",
    "                except:\n",
    "                    PivotFeatures_IDs.append('')\n",
    "                    Global_State['PivotFeatures_IDs'].append('')\n",
    "                    \n",
    "                if Global_State['genomebrowser_mode'] == 'contigs' \\\n",
    "                    or PivotFeatures_IDs[i] == '':\n",
    "                        \n",
    "                    slice_beg = 1\n",
    "                    slice_end = 100000\n",
    "                    features = []\n",
    "                    feature_slice_ids_list = []\n",
    "\n",
    "# RESTORE\n",
    "#                    feature_slice_ids = ga.get_feature_ids(group_by='region', filters={ \"region_list\": [{'contig_id':scaffold_id, 'strand':'?', 'start':slice_beg, 'length':slice_end-slice_beg+1}]})\n",
    "                    feature_slice_ids = ga.get_feature_ids(group_by='region')\n",
    "\n",
    "\n",
    "                    # \"by_region\": dict<str contig_id, dict<str strand, dict<string range, list<string feature_id>>>>\n",
    "                    for ctg_id in feature_slice_ids['by_region'].keys():\n",
    "                        if ctg_id != scaffold_id:  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                            continue\n",
    "                        for strand in feature_slice_ids['by_region'][ctg_id].keys():\n",
    "                            for f_range in feature_slice_ids['by_region'][ctg_id][strand].keys():\n",
    "                                #print (\"%s\\t%s\\t%s\"%(ctg_id, strand, f_range))  # A\n",
    "                                [f_range_beg,f_range_end] = f_range.split('-')  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                                if int(f_range_beg) > slice_end or int(f_range_end) < slice_beg:\n",
    "                                    continue\n",
    "                                feature_slice_ids_list.extend(feature_slice_ids['by_region'][ctg_id][strand][f_range])\n",
    "                    features = ga.get_features(feature_id_list=feature_slice_ids_list)\n",
    "                    \n",
    "                    pivot_feature_rec = None\n",
    "                    lowest_beg = 10000000000\n",
    "\n",
    "                    for fid in features.keys():\n",
    "                        strand = features[fid]['feature_locations'][0][KB_LOC_STR_I]\n",
    "                        f_len = features[fid]['feature_locations'][0][KB_LOC_LEN_I]\n",
    "                        if strand == '+':\n",
    "                            beg = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                            end = beg + f_len - 1\n",
    "                        else:\n",
    "                            end = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                            beg = end - f_len + 1\n",
    "                        #print (\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\"%(contig_id, ctg_id, fid, beg, end, strand))\n",
    "\n",
    "                        if beg < lowest_beg:\n",
    "                            lowest_beg = beg\n",
    "                            f_type=features[fid]['feature_type']\n",
    "                            pivot_feature_rec = build_feature_rec_kbase(features[fid], f_type=f_type, source_species=source, contig_i=i)\n",
    "\n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                    if i == 0:\n",
    "                        Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "\n",
    "                    if Global_State['genomebrowser_mode'] == 'contigs':                    \n",
    "                        contig_mode_xshift = 0.5*window_size - 0.5*(pivot_feature_rec['end_pos']-pivot_feature_rec['beg_pos'])\n",
    "                            \n",
    "                else:  # genomebrowser_mode != 'contigs' and != 'genome', so use PivotFeatures_IDs\n",
    "                    print (\"FF \"+str(i))  # DEBUG\n",
    "                    fid = PivotFeatures_IDs[i]\n",
    "                    print (\"GG \"+str(fid))  # DEBUG\n",
    "                    features = ga.get_features(feature_id_list=[fid])\n",
    "                    print (\"HH \"+str(fid))  # DEBUG\n",
    "                    f = features[fid]\n",
    "                    print (\"II \"+str(fid))  # DEBUG\n",
    "                    f_type = f['feature_type']\n",
    "                    print (\"JJ \"+str(fid))  # DEBUG\n",
    "                    \n",
    "                    pivot_feature_rec = build_feature_rec_kbase(f, f_type=f_type, source_species=source, contig_i=i)\n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                    if i == 0:\n",
    "                        Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "\n",
    "                        \n",
    "                # Add in additional features within window.  Note: we want the duplicate pivot feature\n",
    "                #\n",
    "                slice_beg = pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size\n",
    "                slice_end = pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size\n",
    "                if slice_end < 1:\n",
    "                    Feature_slices.append([Feature_slice[0]])\n",
    "                    continue                 \n",
    "                if slice_beg < 1:\n",
    "                    slice_beg = 1\n",
    "                features = []\n",
    "                feature_slice_ids_list = []\n",
    "# RESTORE\n",
    "#                feature_slice_ids = ga.get_feature_ids(group_by='region', filters={ \"region_list\": [{'contig_id':scaffold_id, 'strand':'?', 'start':slice_beg, 'length':slice_end-slice_beg+1}]})\n",
    "                feature_slice_ids = ga.get_feature_ids(group_by='region')\n",
    "\n",
    "                #\"by_region\": dict<str contig_id, dict<str strand, dict<string range, list<string feature_id>>>>\n",
    "                for ctg_id in feature_slice_ids['by_region'].keys():\n",
    "                    if ctg_id != scaffold_id:  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                        continue\n",
    "                    for strand in feature_slice_ids['by_region'][ctg_id].keys():\n",
    "                        for f_range in feature_slice_ids['by_region'][ctg_id][strand].keys():                    \n",
    "                            #print (\"%s\\t%s\\t%s\"%(ctg_id, strand, f_range)) # B\n",
    "                            [f_range_beg,f_range_end] = f_range.split('-')  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                            if int(f_range_beg) > slice_end or int(f_range_end) < slice_beg:\n",
    "                                continue\n",
    "                            feature_slice_ids_list.extend(feature_slice_ids['by_region'][ctg_id][strand][f_range])\n",
    "                if len(feature_slice_ids_list) == 0:\n",
    "                    Feature_slices.append([Feature_slice[0]])\n",
    "                    continue  \n",
    "                features = ga.get_features(feature_id_list=feature_slice_ids_list)\n",
    "                \n",
    "                for fid in features.keys():\n",
    "                    f_type = features[fid]['feature_type']\n",
    "                    dna_seq = features[fid]['feature_dna_sequence']\n",
    "                    feature_rec = build_feature_rec_kbase(features[fid], f_type=f_type, source_species=source, contig_i=i, dna_seq=dna_seq)\n",
    "                    Feature_slice.append(feature_rec)                        \n",
    "\n",
    "                            \n",
    "                # Check features against Search Terms\n",
    "                #\n",
    "                slice_beg = 1\n",
    "                slice_end = 10000000000\n",
    "                features = []\n",
    "                feature_slice_ids_list = []\n",
    "# RESTORE\n",
    "#                feature_slice_ids = ga.get_feature_ids(group_by='region', filters={ \"region_list\": [{'contig_id':scaffold_id, 'strand':'?', 'start':slice_beg, 'length':slice_end-slice_beg+1}]})\n",
    "                feature_slice_ids = ga.get_feature_ids(group_by='region')\n",
    "\n",
    "                for ctg_id in feature_slice_ids['by_region'].keys():\n",
    "                    if ctg_id != scaffold_id:  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                        continue\n",
    "                    for strand in feature_slice_ids['by_region'][ctg_id].keys():\n",
    "                        for f_range in feature_slice_ids['by_region'][ctg_id][strand].keys():                    \n",
    "                            #print (\"%s\\t%s\\t%s\"%(ctg_id, strand, f_range)) # C\n",
    "                            [f_range_beg,f_range_end] = f_range.split('-')  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                            if int(f_range_beg) > slice_end or int(f_range_end) < slice_beg:\n",
    "                                continue\n",
    "                            feature_slice_ids_list.extend(feature_slice_ids['by_region'][ctg_id][strand][f_range])\n",
    "                if len(feature_slice_ids_list) == 0:\n",
    "                    Feature_slices.append([Feature_slice[0]])\n",
    "                    continue  \n",
    "                features = ga.get_features(feature_id_list=feature_slice_ids_list)    \n",
    "           \n",
    "                for fid in features.keys():\n",
    "                    strand = features[fid]['feature_locations'][0][KB_LOC_STR_I]\n",
    "                    f_len = features[fid]['feature_locations'][0][KB_LOC_LEN_I]\n",
    "                    if strand == '+':\n",
    "                        beg = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                        end = beg + f_len - 1\n",
    "                    else:\n",
    "                        end = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                        beg = end - f_len + 1                \n",
    "                    pos_key = \"%s,%d,%d\"%(strand,beg,end)\n",
    "                    if pos_key in Features_seen:\n",
    "                        continue\n",
    "                    else:\n",
    "                        Features_seen.add(pos_key)                    \n",
    "\n",
    "                    # check if a search hit\n",
    "                    if not search_done[i]:\n",
    "                        for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                            if match_flag:\n",
    "                                try:\n",
    "                                    term_hit_in_genome = search_results[i][j][0]\n",
    "                                except:\n",
    "                                    search_results[i][j] = []\n",
    "                                search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                             'end_pos': feature_rec['end_pos'],\n",
    "                                                             'name': feature_rec['name'],\n",
    "                                                             'annot': feature_rec['annot']\n",
    "                                                            })\n",
    "\n",
    "                # sort results by position, store contig slice, and mark search as done on this contig\n",
    "                #\n",
    "                Sorted_Feature_slice = [Feature_slice[0]]  # retain pivot\n",
    "                for f in sorted (Feature_slice[1:], key=sort_by_beg_pos_key):\n",
    "                    Sorted_Feature_slice.append (f)\n",
    "                    \n",
    "                Feature_slices.append(Sorted_Feature_slice)\n",
    "                search_done[i] = True   \n",
    "\n",
    "        else:\n",
    "            print (\"must use KBase_backend\")\n",
    "\n",
    "    # genomebrowser_mode == \"genome\"\n",
    "    else:  \n",
    "        if KBase_backend:  \n",
    "\n",
    "            for i in range (0,Global_State['genome_mode_n_rows']):\n",
    "                contig_name = ContigSet_names[0]\n",
    "                (genome_id,scaffold_id) = contig_name.split(genome_contig_id_delim)\n",
    "\n",
    "                ga = Global_KBase_Genomes[genome_id]\n",
    "                ass = Global_KBase_Assemblies[genome_id]\n",
    "            \n",
    "                Feature_slice = []\n",
    "                Features_seen = set()\n",
    "                source = \"\"\n",
    "                pivot_pos = 0.0\n",
    "                track_xshift = window_size * (i - (Global_State['genome_mode_n_rows']-1)/2)  # e.g. n_rows=7 -> -3,-2,-1,0,1,2,3                \n",
    "                \n",
    "                try:\n",
    "                    this_search_done = search_done[0]\n",
    "                except:\n",
    "                    search_done.append([])                \n",
    "                    search_done[0] = False\n",
    "                    search_results.append([])\n",
    "                    search_results[0] = []\n",
    "                    for j,term in enumerate(Search_Terms):\n",
    "                        search_results[0].append([])\n",
    "\n",
    "                # Get genome length\n",
    "                #if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                contig_lens = ass.get_contig_lengths()\n",
    "                if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                    contig_GCs  = ass.get_contig_gc_content(contig_id_list=[scaffold_id])\n",
    "                    Global_State['PrimaryContig_len'] = contig_lens[scaffold_id]                    \n",
    "                    Global_State['PrimaryContig_GCavg'] = contig_GCs[scaffold_id]\n",
    "                    #print (\"%d\"%Global_State['PrimaryContig_len'])\n",
    "                    Global_State['Contig_lens'].append(contig_lens[scaffold_id])\n",
    "                \n",
    "\n",
    "                # Find pivot feature and put in first position\n",
    "                #\n",
    "                if i == 0:\n",
    "                    try:\n",
    "                        this_pivotfeature_id = PivotFeatures_IDs[0]\n",
    "                    except:\n",
    "                        PivotFeatures_IDs.append('')\n",
    "                        Global_State['PivotFeatures_IDs'].append('')\n",
    "\n",
    "                if i > 0:\n",
    "                    pivot_feature_rec = Feature_slices[0][0]  # use homolog feature from first contig only\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                else:\n",
    "                    source = 'E. missingii'  # FIX THIS\n",
    "                    \n",
    "                    if PivotFeatures_IDs[0] == '':\n",
    "                        slice_beg = 1\n",
    "                        slice_end = 100000\n",
    "                        features = []\n",
    "                        feature_slice_ids_list = []\n",
    "# RESTORE\n",
    "#                        feature_slice_ids = ga.get_feature_ids(group_by='region', filters={ \"region_list\": [{'contig_id':scaffold_id, 'strand':'?', 'start':slice_beg, 'length':slice_end-slice_beg+1}]})\n",
    "                        feature_slice_ids = ga.get_feature_ids(group_by='region')\n",
    "\n",
    "                        #\"by_region\": dict<str contig_id, dict<str strand, dict<string range, list<string feature_id>>>>\n",
    "                        for ctg_id in feature_slice_ids['by_region'].keys():\n",
    "                            if ctg_id != scaffold_id:  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                                continue\n",
    "                            for strand in feature_slice_ids['by_region'][ctg_id].keys():\n",
    "                                for f_range in feature_slice_ids['by_region'][ctg_id][strand].keys():                    \n",
    "                                    #print (\"%s\\t%s\\t%s\"%(ctg_id, strand, f_range)) # D\n",
    "                                    [f_range_beg,f_range_end] = f_range.split('-')  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                                    if int(f_range_beg) > slice_end or int(f_range_end) < slice_beg:\n",
    "                                        continue\n",
    "                                    feature_slice_ids_list.extend(feature_slice_ids['by_region'][ctg_id][strand][f_range])\n",
    "                        features = ga.get_features(feature_id_list=feature_slice_ids_list)\n",
    "\n",
    "                        pivot_feature_rec = None\n",
    "                        lowest_beg = 10000000000\n",
    "\n",
    "                        for fid in features.keys():\n",
    "                            strand = features[fid]['feature_locations'][0][KB_LOC_STR_I]\n",
    "                            f_len = features[fid]['feature_locations'][0][KB_LOC_LEN_I]\n",
    "                            if strand == '+':\n",
    "                                beg = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                                end = beg + f_len - 1\n",
    "                            else:\n",
    "                                end = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                                beg = end - f_len + 1\n",
    "                            #print (\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\"%(contig_id, ctg_id, fid, beg, end, strand))\n",
    "\n",
    "                            if beg < lowest_beg:\n",
    "                                lowest_beg = beg\n",
    "                                f_type=features[fid]['feature_type']\n",
    "                                pivot_feature_rec = build_feature_rec_kbase(features[fid], f_type=f_type, source_species=source, contig_i=i)\n",
    "\n",
    "                        Feature_slice.append(pivot_feature_rec)\n",
    "                        pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                        Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                        if i == 0:\n",
    "                            Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "   \n",
    "                    else:  # PivotFeatures_IDs available\n",
    "                        fid = PivotFeatures_IDs[0]\n",
    "                        features = ga.get_features(feature_id_list=[fid])\n",
    "                        f = features[fid]\n",
    "                        f_type = f['feature_type']\n",
    "\n",
    "                        pivot_feature_rec = build_feature_rec_kbase(f, f_type=f_type, source_species=source, contig_i=0)\n",
    "                        Feature_slice.append(pivot_feature_rec)\n",
    "                        pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                        Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                        if i == 0:\n",
    "                            Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "\n",
    "                        \n",
    "                # Add in additional features within window.  Note: we want the duplicate pivot feature\n",
    "                #\n",
    "                slice_beg = int(round(pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size))\n",
    "                slice_end = int(round(pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size))\n",
    "                # DEBUG: temporary fix, but should make track_xshift correct for empty PivotFeatures_IDs\n",
    "                if slice_end < 1:\n",
    "                    Feature_slices.append([Feature_slice[0]])\n",
    "                    continue  \n",
    "                elif slice_beg < 1:\n",
    "                    slice_beg = 1\n",
    "                #print (\"%s\\t%s\\t%s\\t%s\"%(pivot_pos, genomebrowser_xshift, track_xshift, 0.5*window_size))\n",
    "                features = []\n",
    "                feature_slice_ids_list = []\n",
    "# RESTORE\n",
    "#                feature_slice_ids = ga.get_feature_ids(group_by='region', filters={ \"region_list\": [{'contig_id':scaffold_id, 'strand':'?', 'start':slice_beg, 'length':slice_end-slice_beg+1}]})\n",
    "                feature_slice_ids = ga.get_feature_ids(group_by='region')\n",
    "\n",
    "                #\"by_region\": dict<str contig_id, dict<str strand, dict<string range, list<string feature_id>>>>\n",
    "                for ctg_id in feature_slice_ids['by_region'].keys():\n",
    "                    if ctg_id != scaffold_id:  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                        continue\n",
    "                    for strand in feature_slice_ids['by_region'][ctg_id].keys():\n",
    "                        for f_range in feature_slice_ids['by_region'][ctg_id][strand].keys():                    \n",
    "                            #print (\"%s\\t%s\\t%s\"%(ctg_id, strand, f_range)) # E\n",
    "                            [f_range_beg,f_range_end] = f_range.split('-')  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                            if int(f_range_beg) > slice_end or int(f_range_end) < slice_beg:\n",
    "                                continue\n",
    "                            feature_slice_ids_list.extend(feature_slice_ids['by_region'][ctg_id][strand][f_range])\n",
    "                if len(feature_slice_ids_list) == 0:\n",
    "                    Feature_slices.append([Feature_slice[0]])\n",
    "                    continue                    \n",
    "                features = ga.get_features(feature_id_list=feature_slice_ids_list)\n",
    "                \n",
    "                for fid in features.keys():\n",
    "                    f_type = features[fid]['feature_type']\n",
    "                    dna_seq = features[fid]['feature_dna_sequence']\n",
    "                    feature_rec = build_feature_rec_kbase(features[fid], f_type=f_type, source_species=source, contig_i=i, dna_seq=dna_seq)\n",
    "                    Feature_slice.append(feature_rec)                        \n",
    "\n",
    "                            \n",
    "                # Check features against Search Terms\n",
    "                #\n",
    "                slice_beg = 1\n",
    "                slice_end = 100000000000\n",
    "                features = []\n",
    "                feature_slice_ids_list = []\n",
    "# RESTORE\n",
    "#                feature_slice_ids = ga.get_feature_ids(group_by='region', filters={ \"region_list\": [{'contig_id':scaffold_id, 'strand':'?', 'start':slice_beg, 'length':slice_end-slice_beg+1}]})\n",
    "                feature_slice_ids = ga.get_feature_ids(group_by='region')\n",
    "\n",
    "                for ctg_id in feature_slice_ids['by_region'].keys():\n",
    "                    if ctg_id != scaffold_id:  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                        continue\n",
    "                    for strand in feature_slice_ids['by_region'][ctg_id].keys():\n",
    "                        for f_range in feature_slice_ids['by_region'][ctg_id][strand].keys():                    \n",
    "                            #print (\"%s\\t%s\\t%s\"%(ctg_id, strand, f_range)) # F\n",
    "                            [f_range_beg,f_range_end] = f_range.split('-')  # SHOULDN'T BE NECESSARY IF get_feature_ids() WORKING\n",
    "                            if int(f_range_beg) > slice_end or int(f_range_end) < slice_beg:\n",
    "                                continue\n",
    "                            feature_slice_ids_list.extend(feature_slice_ids['by_region'][ctg_id][strand][f_range])\n",
    "                if len(feature_slice_ids_list) == 0:\n",
    "                    Feature_slices.append([Feature_slice[0]])\n",
    "                    continue\n",
    "                features = ga.get_features(feature_id_list=feature_slice_ids_list)    \n",
    "           \n",
    "                for fid in features.keys():\n",
    "                    strand = features[fid]['feature_locations'][0][KB_LOC_STR_I]\n",
    "                    f_len = features[fid]['feature_locations'][0][KB_LOC_LEN_I]\n",
    "                    if strand == '+':\n",
    "                        beg = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                        end = beg + f_len - 1\n",
    "                    else:\n",
    "                        end = features[fid]['feature_locations'][0][KB_LOC_BEG_I]\n",
    "                        beg = end - f_len + 1                \n",
    "                    pos_key = \"%s,%d,%d\"%(strand,beg,end)\n",
    "                    if pos_key in Features_seen:\n",
    "                        continue\n",
    "                    else:\n",
    "                        Features_seen.add(pos_key)                    \n",
    "\n",
    "                    # check if a search hit\n",
    "                    if not search_done[0]:\n",
    "                        for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                            if match_flag:\n",
    "                                try:\n",
    "                                    term_hit_in_genome = search_results[0][j][0]\n",
    "                                except:\n",
    "                                    search_results[0][j] = []\n",
    "                                search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                             'end_pos': feature_rec['end_pos'],\n",
    "                                                             'name': feature_rec['name'],\n",
    "                                                             'annot': feature_rec['annot']\n",
    "                                                            })\n",
    "\n",
    "                # sort results by position, store contig slice, and mark search as done on this contig\n",
    "                #\n",
    "                Sorted_Feature_slice = [Feature_slice[0]]  # retain pivot\n",
    "                for f in sorted (Feature_slice[1:], key=sort_by_beg_pos_key):\n",
    "                    Sorted_Feature_slice.append (f)\n",
    "                    \n",
    "                Feature_slices.append(Sorted_Feature_slice)\n",
    "                search_done[0] = True   \n",
    "\n",
    "        else:\n",
    "            print (\"must use KBase_backend\")\n",
    "            \n",
    "            \n",
    "    return Feature_slices\n",
    "\n",
    "\n",
    "# Workhorse for feature retrieval (Genbank flavor)\n",
    "#   does more than window slice, so perhaps should split up functionality\n",
    "#\n",
    "def getFeatureSlicesGenbank (ContigSet_names, \\\n",
    "                      PivotFeatures_IDs, \\\n",
    "                      genomebrowser_mode=\"contigs\", \\\n",
    "                      genome_data_format=\"Genbank\", \\\n",
    "                      window_size=10000, \\\n",
    "                      genomebrowser_xshift=0):\n",
    "\n",
    "    Feature_slices = []\n",
    "    \n",
    "    if genomebrowser_mode != \"genome\":\n",
    "    \n",
    "        if genome_data_format == \"Genbank\":\n",
    "            for i,contig_name in enumerate(ContigSet_names):\n",
    "                if i >= max_rows:\n",
    "                    break\n",
    "                    \n",
    "                try:\n",
    "                    t = Global_Genbank_Genomes[i]\n",
    "                except:\n",
    "                    (genome_id,scaffold_id) = contig_name.split(genome_contig_id_delim)\n",
    "                    #print (\"reading \" + contig_name + \" ...\")\n",
    "                    #genome_data_path = 'data/'+contig_name+'.'+gbk_ext\n",
    "                    genome_data_path = genome_data_base_path+'/'+genome_id+genome_data_extra_subpath+'/'+scaffold_id+'.'+gbk_ext\n",
    "                    print (\"%d \"%i+'reading '+genome_data_path)\n",
    "                    Global_Genbank_Genomes.append (SeqIO.read(genome_data_path, 'genbank'))\n",
    "                    #print (\"%d\"%len(Global_Genbank_Genomes))\n",
    "                    #contig_seq = Global_Genbank_Genomes[i].seq\n",
    "                    #print (contig_seq[-100:])\n",
    "                    #print (\"%d\"%len(contig_seq))\n",
    "                \n",
    "#                try:\n",
    "#                    feature_cache = Global_Features[i]\n",
    "#                except:\n",
    "#                    Global_Features[i] = {}\n",
    "                    \n",
    "                Feature_slice = []\n",
    "                Features_seen = set()\n",
    "                source = \"\"\n",
    "                pivot_pos = 0.0\n",
    "\n",
    "                try:\n",
    "                    this_search_done = search_done[i]\n",
    "                except:\n",
    "                    search_done.append([])                \n",
    "                    search_done[i] = False\n",
    "                    search_results.append([])\n",
    "                    search_results[i] = []\n",
    "                    for j,term in enumerate(Search_Terms):\n",
    "                        search_results[i].append([])\n",
    "\n",
    "                # Get genome length\n",
    "                #if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                if i == 0:\n",
    "                    Global_State['PrimaryContig_len'] = len(Global_Genbank_Genomes[i].seq)\n",
    "                    Global_State['PrimaryContig_GCavg'] = compute_GC (Global_Genbank_Genomes[i].seq)\n",
    "                    Global_State['Contig_lens'] = []\n",
    "                    Global_State['pivot_pos_list'] = []\n",
    "                    #print (\"%d\"%Global_State['PrimaryContig_len'])\n",
    "                Global_State['Contig_lens'].append(len(Global_Genbank_Genomes[i].seq))\n",
    "                \n",
    "                # Find pivot feature and put in first position\n",
    "                contig_mode_xshift = 0\n",
    "                for f in Global_Genbank_Genomes[i].features:                        \n",
    "                    if f.type == \"source\":\n",
    "                        source = f.qualifiers['organism'][0]\n",
    "                        break\n",
    "                        \n",
    "                try:\n",
    "                    this_pivotfeature_id = PivotFeatures_IDs[i]\n",
    "                except:\n",
    "                    PivotFeatures_IDs.append('')\n",
    "                    Global_State['PivotFeatures_IDs'].append('')\n",
    "                    \n",
    "                if Global_State['genomebrowser_mode'] == 'contigs' \\\n",
    "                    or PivotFeatures_IDs[i] == '':\n",
    "                        \n",
    "                    pivot_feature_rec = None\n",
    "                    lowest_beg = 10000000000\n",
    "                    for f in Global_Genbank_Genomes[i].features:                        \n",
    "                        #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers:                    \n",
    "#                        if (f.type == \"CDS\" and \"gene\" in f.qualifiers):\n",
    "                        if (genome_annotation_system == 'KBase' and f.type == \"CDS\" and \"gene\" in f.qualifiers) \\\n",
    "                            or (genome_annotation_system != 'KBase' and f.type == \"CDS\" and \"locus_tag\" in f.qualifiers) \\\n",
    "                            or f.type == 'rRNA' \\\n",
    "                            or f.type == 'tRNA' \\\n",
    "                            or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\") \\\n",
    "                            or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\") \\\n",
    "                            or (f.type == 'gene' and \"pseudo\" in f.qualifiers):\n",
    "\n",
    "                            if f.location.start == None or f.location.start <= 0:\n",
    "                                raise ValueError (\"Bad Feature.  gene:'\"+str(f.qualifiers.gene)+\"' locus_tag:'\"+str(f.qualifiers.locus_tag)+\"'\")\n",
    "\n",
    "                            if f.location.start < lowest_beg:\n",
    "                                lowest_beg = f.location.start\n",
    "                                pivot_feature_rec = build_feature_rec_genbank(f, f_type=f.type, source_species=source, contig_i=i)\n",
    "                                \n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                    if i == 0:\n",
    "                        Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "\n",
    "                    if Global_State['genomebrowser_mode'] == 'contigs':                    \n",
    "                        contig_mode_xshift = 0.5*window_size - 0.5*(pivot_feature_rec['end_pos']-pivot_feature_rec['beg_pos'])\n",
    "                            \n",
    "                else:             \n",
    "                    for f in Global_Genbank_Genomes[i].features:        \n",
    "                        \n",
    "                        #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers:                    \n",
    "                        if (genome_annotation_system == 'KBase' and f.type == \"CDS\" and \"gene\" in f.qualifiers) \\\n",
    "                            or (genome_annotation_system != 'KBase' and f.type == \"CDS\" and \"locus_tag\" in f.qualifiers):  # should it permit non-CDS anchor?             \n",
    "                            \n",
    "                            #elif f.qualifiers['locus_tag'][0] == PivotFeatures_IDs[i]:\n",
    "                            if (genome_annotation_system == 'KBase' and f.qualifiers['gene'][0] == PivotFeatures_IDs[i]) \\\n",
    "                                or (genome_annotation_system != 'KBase' and f.qualifiers['locus_tag'][0] == PivotFeatures_IDs[i]):  \n",
    "                                pivot_feature_rec = build_feature_rec_genbank(f, f_type='CDS', source_species=source, contig_i=i)\n",
    "                                Feature_slice.append(pivot_feature_rec)\n",
    "                                pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                                Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                                if i == 0:\n",
    "                                    Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "                                break\n",
    "                        \n",
    "                # Add in additional features within window.  Note: we want the duplicate pivot feature\n",
    "                for f in Global_Genbank_Genomes[i].features:\n",
    "                    \n",
    "                    if f.type == \"CDS\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        # fast enough for now to build recs for all features, but might want to only do for window\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='CDS', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "\n",
    "                        # add feature to slice if in viewable window\n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == \"rRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='rRNA', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                                    \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                    elif f.type == \"tRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)                        \n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='tRNA', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)                        \n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[i].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='CRISPR', source_species=source, contig_i=i, dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[i].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='CRISPR spacer', source_species=source, contig_i=i, dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"pseudo\" in f.qualifiers:\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)                        \n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='pseudogene', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                # sort results by position, store contig slice, and mark search as done on this contig\n",
    "                #\n",
    "                Sorted_Feature_slice = [Feature_slice[0]]  # retain pivot\n",
    "                for f in sorted (Feature_slice[1:], key=sort_by_beg_pos_key):\n",
    "                    Sorted_Feature_slice.append (f)\n",
    "                    \n",
    "                Feature_slices.append(Sorted_Feature_slice)\n",
    "                search_done[i] = True   \n",
    "\n",
    "        else:\n",
    "            print (\"unknown data format for Genomes\")\n",
    "\n",
    "    # genomebrowser_mode == \"genome\"\n",
    "    else:  \n",
    "\n",
    "        if genome_data_format == \"Genbank\":\n",
    "            #for i in range (0,total_rows):\n",
    "            for i in range (0,Global_State['genome_mode_n_rows']):\n",
    "                contig_name = ContigSet_names[0]\n",
    "                try:\n",
    "                    t=Global_Genbank_Genomes[0]\n",
    "                except:\n",
    "                    (genome_id,scaffold_id) = contig_name.split(genome_contig_id_delim)\n",
    "                    #print (\"reading \" + contig_name + \" ...\")\n",
    "                    #genome_data_path = 'data/'+contig_name+'.'+gbk_ext\n",
    "                    genome_data_path = genome_data_base_path+'/'+genome_id+genome_data_extra_subpath+'/'+scaffold_id+'.'+gbk_ext\n",
    "                    print (\"%d \"%i+'reading '+genome_data_path)\n",
    "                    Global_Genbank_Genomes.append (SeqIO.read(genome_data_path, 'genbank'))\n",
    "                    \n",
    "                Feature_slice = []\n",
    "                Features_seen = set()\n",
    "                source = \"\"\n",
    "                pivot_pos = 0.0\n",
    "                #track_xshift = window_size * (i - (total_rows-1)/2)  # e.g. total_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                track_xshift = window_size * (i - (Global_State['genome_mode_n_rows']-1)/2)  # e.g. n_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                #print (\"%d %f\"%(i,track_xshift))\n",
    "\n",
    "                try:\n",
    "                    this_search_done = search_done[0]\n",
    "                except:\n",
    "                    search_done.append([])                \n",
    "                    search_done[0] = False\n",
    "                    search_results.append([])\n",
    "                    search_results[0] = []\n",
    "                    for j,term in enumerate(Search_Terms):\n",
    "                        search_results[0].append([])                \n",
    "                \n",
    "                # Get genome length\n",
    "                if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                    Global_State['PrimaryContig_len'] = len(Global_Genbank_Genomes[i].seq)\n",
    "                    Global_State['PrimaryContig_GCavg'] = compute_GC (Global_Genbank_Genomes[i].seq)\n",
    "                    #print (\"%d\"%Global_State['PrimaryContig_len'])\n",
    "                    Global_State['Contig_lens'].append(len(Global_Genbank_Genomes[i].seq))\n",
    "\n",
    "                # Find pivot feature and put in first position\n",
    "                if i == 0:\n",
    "                    try:\n",
    "                        this_pivotfeature_id = PivotFeatures_IDs[0]\n",
    "                    except:\n",
    "                        PivotFeatures_IDs.append('')\n",
    "                        Global_State['PivotFeatures_IDs'].append('')              \n",
    "                \n",
    "                if i > 0:\n",
    "                    pivot_feature_rec = Feature_slices[0][0]  # use homolog feature from first contig only\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                else:\n",
    "                    for f in Global_Genbank_Genomes[0].features:  \n",
    "                        if f.type == \"source\":\n",
    "                            source = f.qualifiers['organism'][0]\n",
    "                            break\n",
    "                            \n",
    "                    if PivotFeatures_IDs[0] == '':      \n",
    "                        pivot_feature_rec = None\n",
    "                        lowest_beg = 10000000000\n",
    "                        for f in Global_Genbank_Genomes[0].features:                        \n",
    "                            #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers:                    \n",
    "    #                        if (f.type == \"CDS\" and \"gene\" in f.qualifiers):\n",
    "                            if (genome_annotation_system == 'KBase' and f.type == \"CDS\" and \"gene\" in f.qualifiers) \\\n",
    "                                or (genome_annotation_system != 'KBase' and f.type == \"CDS\" and \"locus_tag\" in f.qualifiers) \\\n",
    "                                or f.type == 'rRNA' \\\n",
    "                                or f.type == 'tRNA' \\\n",
    "                                or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\") \\\n",
    "                                or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\") \\\n",
    "                                or (f.type == 'gene' and \"pseudo\" in f.qualifiers):\n",
    "\n",
    "                                if f.location.start < lowest_beg:\n",
    "                                    lowest_beg = f.location.start\n",
    "                                    pivot_feature_rec = build_feature_rec_genbank(f, f_type=f.type, source_species=source, contig_i=0)\n",
    "\n",
    "                        Feature_slice.append(pivot_feature_rec)\n",
    "                        pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                        Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                        if i == 0:\n",
    "                            Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos  \n",
    "                            \n",
    "                    else:\n",
    "                        for f in Global_Genbank_Genomes[0].features:  # should it permit non-CDS anchor?    \n",
    "                            #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers and f.qualifiers['locus_tag'][0] == PivotFeatures_IDs[i]:\n",
    "                            if (genome_annotation_system == 'KBase' and f.type == \"CDS\" and \"gene\" in f.qualifiers) \\\n",
    "                                or (genome_annotation_system != 'KBase' and f.type == \"CDS\" and \"locus_tag\" in f.qualifiers):  \n",
    "                                #if f.qualifiers['gene'][0] == PivotFeatures_IDs[i]:\n",
    "                                if (genome_annotation_system == 'KBase' and f.qualifiers['gene'][0] == PivotFeatures_IDs[0]) \\\n",
    "                                    or (genome_annotation_system != 'KBase' and f.qualifiers['locus_tag'][0] == PivotFeatures_IDs[0]):\n",
    "                                    pivot_feature_rec = build_feature_rec_genbank(f, f_type='CDS', source_species=source, contig_i=0)\n",
    "                                    Feature_slice.append(pivot_feature_rec)\n",
    "                                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                                    if i == 0:\n",
    "                                        Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "                                    break\n",
    "                        \n",
    "                # Add in additional features within window.  Note: we want the duplicate pivot feature\n",
    "                for f in Global_Genbank_Genomes[0].features:\n",
    "                    if f.type == \"CDS\":  # we'll add other types later when we have icons for them\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        # fast enough for now to build recs for all features, but might want to only do for window\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='CDS', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                        \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            #print (\"%d %d %d\"%(i, feature_rec['beg_pos'], feature_rec['end_pos']))\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == \"rRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='rRNA', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                    elif f.type == \"tRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='tRNA', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)                    \n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[0].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='CRISPR', source_species=source, contig_i=0, dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[0].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='CRISPR spacer', source_species=source, contig_i=0, dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"pseudo\" in f.qualifiers:\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec_genbank(f, f_type='pseudogene', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                            \n",
    "                # sort results by position, store contig slice, and mark search as done on this contig\n",
    "                #\n",
    "                Sorted_Feature_slice = [Feature_slice[0]]  # retain pivot\n",
    "                for f in sorted (Feature_slice[1:], key=sort_by_beg_pos_key):\n",
    "                    Sorted_Feature_slice.append (f)\n",
    "                    \n",
    "                Feature_slices.append(Sorted_Feature_slice)\n",
    "                search_done[0] = True\n",
    "                \n",
    "        else:\n",
    "            print (\"unknown data format for Genomes\")\n",
    "            \n",
    "    return Feature_slices\n",
    "\n",
    "\n",
    "# Slider events\n",
    "#\n",
    "class DraggableRectangleSlider:\n",
    "    def __init__(self, rect, func='', move='x', min_pos=0.0, max_pos=1.0, step=0.1, tol=0.01):\n",
    "        self.rect = rect\n",
    "        self.press = None\n",
    "        self.func = func\n",
    "        self.move = move.lower()\n",
    "        self.min_pos = min_pos\n",
    "        self.max_pos = max_pos\n",
    "        self.step = step\n",
    "        self.tol = tol\n",
    "        self.n_steps = int(round((max_pos-min_pos)/step))\n",
    "        self.last_pos = 0\n",
    "        if self.move != 'x' and self.move != 'y':\n",
    "            print (\"bad DraggableRectSlider move: '\"+ self.move +\"' for func: '\"+ self.func +\"'\")\n",
    "\n",
    "    def connect(self):\n",
    "        'connect to all the events we need'\n",
    "        self.cidpress = self.rect.figure.canvas.mpl_connect('button_press_event', self.on_press)\n",
    "        self.cidrelease = self.rect.figure.canvas.mpl_connect('button_release_event', self.on_release)\n",
    "        self.cidmotion = self.rect.figure.canvas.mpl_connect('motion_notify_event', self.on_motion)\n",
    "\n",
    "    def on_press(self, event):\n",
    "        'on button press we will see if the mouse is over us and store some data'\n",
    "        if event.inaxes != self.rect.axes: return\n",
    "\n",
    "        contains, attrd = self.rect.contains(event)\n",
    "        if not contains: return\n",
    "        #print ('event contains', self.rect.xy)\n",
    "        x0, y0 = self.rect.xy\n",
    "        self.press = x0, y0, event.xdata, event.ydata\n",
    "        if self.move == 'x':\n",
    "            self.last_pos = x0\n",
    "        elif self.move == 'y':\n",
    "            self.last_pos = y0\n",
    "\n",
    "    def on_motion(self, event):\n",
    "        'on motion we will move the rect if the mouse is over us'\n",
    "        update_draggable = False\n",
    "        if self.press is None: return\n",
    "        if event.inaxes != self.rect.axes: return\n",
    "        x0, y0, xpress, ypress = self.press\n",
    "        #last_pos = self.last_pos\n",
    "        step = self.step\n",
    "        tol = self.tol\n",
    "        min_pos = self.min_pos\n",
    "        max_pos = self.max_pos\n",
    "        dx = event.xdata - xpress\n",
    "        dy = event.ydata - ypress\n",
    "        #print ('x0=%f, xpress=%f, event.xdata=%f, dx=%f, x0+dx=%f, last_pos, step'%(x0, xpress, event.xdata, dx, x0+dx, self.last_pos, self.step))\n",
    "        if self.move == 'x':\n",
    "            if x0+dx <= min_pos + tol:\n",
    "                if self.last_pos != min_pos:\n",
    "                    self.last_pos = min_pos\n",
    "                    nearest_step_i = 0\n",
    "                    update_draggable = True\n",
    "            elif x0+dx >= max_pos - tol:\n",
    "                if self.last_pos != max_pos:\n",
    "                    self.last_pos = max_pos\n",
    "                    nearest_step_i = int(round((max_pos-min_pos)/step))\n",
    "                    update_draggable = True\n",
    "            else:\n",
    "                nearest_step_i = int(round((x0+dx-min_pos)/step))\n",
    "                new_pos = min_pos + nearest_step_i*step\n",
    "                if new_pos != self.last_pos:\n",
    "                    self.last_pos = new_pos\n",
    "                    update_draggable = True\n",
    "                                     \n",
    "            if update_draggable:\n",
    "                self.rect.set_x(self.last_pos)\n",
    "                #self.rect.set_x(x0+dx)\n",
    "                #self.rect.set_y(y0)\n",
    "        elif self.move == 'y':\n",
    "            if y0+dy <= min_pos + tol:\n",
    "                if self.last_pos != min_pos:\n",
    "                    self.last_pos = min_pos\n",
    "                    nearest_step_i = 0\n",
    "                    update_draggable = True\n",
    "            elif y0+dy >= max_pos - tol:\n",
    "                if self.last_pos != max_pos:\n",
    "                    self.last_pos = max_pos\n",
    "                    nearest_step_i = int(round((max_pos-min_pos)/step))\n",
    "                    update_draggable = True\n",
    "            else:\n",
    "                nearest_step_i = int(round((y0+dy-min_pos)/step))\n",
    "                new_pos = min_pos + nearest_step_i*step\n",
    "                if new_pos != self.last_pos:\n",
    "                    self.last_pos = new_pos\n",
    "                    update_draggable = True\n",
    "                    \n",
    "            if update_draggable:\n",
    "                self.rect.set_y(self.last_pos)\n",
    "                #self.rect.set_x(x0)\n",
    "                #self.rect.set_y(y0+dy)\n",
    "\n",
    "        if update_draggable:\n",
    "            self.rect.figure.canvas.draw()\n",
    "            if self.func == 'zoom':\n",
    "                Global_State['genomebrowser_zoom'] = int(round((self.last_pos-min_pos)/step))\n",
    "                Global_State['genomebrowser_window_bp_width'] = Global_State['def_genomebrowser_window_bp_width']*(2**Global_State['genomebrowser_zoom'])\n",
    "            elif self.func == 'pan':\n",
    "                Global_State['genomebrowser_xshift'] = Global_State['PrimaryContig_len']*(self.last_pos-min_pos)/(max_pos-min_pos) - Global_State['PrimaryAnchor_pivot_pos']\n",
    "            elif self.func == 'mode':\n",
    "                Global_State['genomebrowser_mode'] = mode_names[nearest_step_i]\n",
    "            elif self.func == 'color':\n",
    "                Global_State['genomebrowser_color_namespace'] = color_namespace_names[nearest_step_i]\n",
    "                \n",
    "            update_genomebrowser_panel (ax_center)  # This should happen first so Features and Domains are updated\n",
    "            update_sidenav_panel (ax_left)\n",
    "            update_mode_panel (ax_top_left)\n",
    "\n",
    "    def on_release(self, event):\n",
    "        'on release we reset the press data'\n",
    "        self.press = None\n",
    "        self.rect.figure.canvas.draw()\n",
    "            \n",
    "    def disconnect(self):\n",
    "        'disconnect all the stored connection ids'\n",
    "        self.rect.figure.canvas.mpl_disconnect(self.cidpress)\n",
    "        self.rect.figure.canvas.mpl_disconnect(self.cidrelease)\n",
    "        self.rect.figure.canvas.mpl_disconnect(self.cidmotion)\n",
    "\n",
    "\n",
    "# handle contig nav events\n",
    "#\n",
    "def onpick_contig_nav (event):\n",
    "    artist = event.artist\n",
    "    mouseevent = event.mouseevent\n",
    "    ax = artist.axes\n",
    "    if artist.get_label() != \"circular_contig_nav\" and artist.get_label() != \"linear_contig_nav\":\n",
    "        return\n",
    "    artist_info = artist.gid.split(\",\")\n",
    "\n",
    "    repaint = False\n",
    "    if artist.get_label() == \"circular_contig_nav\":\n",
    "        genome_i = int(artist_info[0])\n",
    "        x0 = float(artist_info[1])\n",
    "        y0 = float(artist_info[2])\n",
    "        y_scale = float(artist_info[3])\n",
    "        dx = mouseevent.xdata - x0\n",
    "        dy = (mouseevent.ydata - y0) / y_scale\n",
    "        \n",
    "        cosine = dx / sqrt(dx*dx+dy*dy)\n",
    "        # Math.acos and Math.asin are in radians, and acos returns 0-pi\n",
    "        theta = acos(cosine)\n",
    "        if dy < 0:\n",
    "            theta = 2*pi - theta            \n",
    "        if theta < pi/2.0:\n",
    "            phi = pi/2.0 - theta\n",
    "        else:\n",
    "            phi = 2.0*pi - theta + pi/2.0\n",
    "        genome_frac_pos = phi / (2.0*pi)\n",
    "        \n",
    "        new_genome_pos = int(round(genome_frac_pos * Global_State['PrimaryContig_len']))\n",
    "\n",
    "        if (new_genome_pos <= 0):\n",
    "            new_genome_pos = 1\n",
    "        elif new_genome_pos > Global_State['PrimaryContig_len']:\n",
    "            new_genome_pos = Global_State['PrimaryContig_len']\n",
    "        \n",
    "        Global_State['genomebrowser_xshift'] = new_genome_pos - Global_State['PrimaryAnchor_pivot_pos']\n",
    "        repaint = True\n",
    "        \n",
    "    elif artist.get_label() == \"linear_contig_nav\":\n",
    "        genome_i = int(artist_info[0])\n",
    "        x0 = float(artist_info[1])\n",
    "        w = float(artist_info[2])\n",
    "        dx = mouseevent.xdata - x0\n",
    "        #dy = mouseevent.ydata - y0\n",
    "\n",
    "        new_genome_pos = int(round(Global_State['Contig_lens'][genome_i] * dx/w))\n",
    "        \n",
    "        if (new_genome_pos <= 0):\n",
    "            new_genome_pos = 1\n",
    "        elif new_genome_pos > Global_State['Contig_lens'][genome_i]:\n",
    "            new_genome_pos = Global_State['Contig_lens'][genome_i]\n",
    "            \n",
    "        Global_State['genomebrowser_xshift'] = new_genome_pos - Global_State['pivot_pos_list'][genome_i]        \n",
    "        repaint = True\n",
    "        \n",
    "    if repaint:\n",
    "        #update_control_panel(ax_top_center)  # for some reason this causes things to drag.  memory leak for event listeners?\n",
    "        # FIX THIS\n",
    "        #pos_knob_patch.set_x(self.last_pos)  # global ref (total kludge).  not working yet.        \n",
    "        update_genomebrowser_panel (ax_center)\n",
    "        update_sidenav_panel (ax_left)\n",
    "        \n",
    "        \n",
    "# handle pan events\n",
    "#\n",
    "def onpick_pan (event):\n",
    "    artist = event.artist\n",
    "    #mouseevent = event.mouseevent\n",
    "    #ax = artist.axes\n",
    "    if artist.get_label() != \"pan_button\":\n",
    "        return\n",
    "    control = artist.gid\n",
    "\n",
    "    repaint = False\n",
    "    if control == \"reset_pan\":\n",
    "        Global_State['genomebrowser_xshift'] = 0\n",
    "        repaint = True\n",
    "    elif control == \"left_pan\":\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            Global_State['genomebrowser_xshift'] -= 1.0*Global_State['genomebrowser_window_bp_width']\n",
    "        else:\n",
    "            Global_State['genomebrowser_xshift'] -= 0.4*Global_State['genomebrowser_window_bp_width']\n",
    "        repaint = True\n",
    "    elif control == \"right_pan\":\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            Global_State['genomebrowser_xshift'] += 1.0*Global_State['genomebrowser_window_bp_width']\n",
    "        else:\n",
    "            Global_State['genomebrowser_xshift'] += 0.4*Global_State['genomebrowser_window_bp_width']\n",
    "        repaint = True\n",
    "\n",
    "    if repaint:\n",
    "        #update_control_panel(ax_top_center)  # for some reason this causes things to drag.  memory leak for event listeners?\n",
    "        # FIX THIS\n",
    "        #pos_knob_patch.set_x(self.last_pos)  # global ref (total kludge).  not working yet.        \n",
    "        update_genomebrowser_panel (ax_center)\n",
    "        update_sidenav_panel (ax_left)\n",
    "        \n",
    "        \n",
    "# Feature events\n",
    "#\n",
    "\"\"\"\n",
    "# Something about the object solution here really makes ipython drag.\n",
    "#   Implemented solution with artist labels, feature_artist_label_to_popup_box and\n",
    "#   feature_artist_label_to_feature dicts and labels to link artists ugly but effective\n",
    "class InteractiveGenomeFeature:\n",
    "    def __init__(self, arrow, feature):\n",
    "        self.arrow = arrow\n",
    "        self.feature = feature\n",
    "        self.dialog_box = None\n",
    "        self.box = None\n",
    "        \n",
    "    def connect(self):\n",
    "        'connect to all the events we need'\n",
    "        self.cidpick = self.arrow.figure.canvas.mpl_connect('pick_event', self.on_pick)\n",
    "\n",
    "    def on_pick (self, event):\n",
    "        artist = event.artist\n",
    "        mouseevent = event.mouseevent\n",
    "        arrow = artist\n",
    "        ax = arrow.axes\n",
    "        #arrow.set_color(\"black\")\n",
    "        self.dialog_box = Rectangle((mouseevent.xdata, mouseevent.ydata), 0.1, 0.1, \\\n",
    "                                    facecolor=\"aliceblue\", edgecolor=\"gray\", alpha=1.0, zorder=4)\n",
    "        ax.add_patch(self.dialog_box)\n",
    "\n",
    "        #arrow.remove()\n",
    "        arrow.figure.canvas.draw()\n",
    "            \n",
    "    def disconnect(self):\n",
    "        'disconnect all the stored connection ids'\n",
    "        self.arrow.figure.canvas.mpl_disconnect(self.cidpick)\n",
    "\n",
    "\"\"\"\n",
    "# handle feature events\n",
    "#\n",
    "def onpick_feature (event):\n",
    "    artist = event.artist\n",
    "    mouseevent = event.mouseevent\n",
    "    ax = artist.axes\n",
    "    if artist.get_label() != \"feature\" and artist.get_label() != \"popup_destroy\":\n",
    "        return\n",
    "    feature_index = artist.gid\n",
    "    \n",
    "    # popup config\n",
    "    annot_word_limit = 5\n",
    "    popup_x_margin = 0.015\n",
    "    popup_y_margin = 0.020 * 7/(total_rows+1)\n",
    "    close_button_w = .02\n",
    "    close_button_h = .03 * 7/(total_rows+1)\n",
    "    close_button_marker_y_shift = 0.0037 * 7/(total_rows+1)\n",
    "    popup_text_w_scaling = 0.0085                     # appropriate value for fontsize 10\n",
    "    popup_text_h_scaling = 0.0300 * 7/(total_rows+1)  # appropriate value for fontsize 10\n",
    "    \n",
    "    #popup_zorder += 2   # why doesn't this work even when popup_zorder is a global?\n",
    "    Global_State['popup_zorder'] += 4\n",
    "    popup_zorder = Global_State['popup_zorder']\n",
    "    \n",
    "    if not feature_index in feature_artist_label_to_popup_box or \\\n",
    "        not feature_artist_label_to_popup_box[feature_index]:\n",
    "            \n",
    "        feature = feature_artist_label_to_feature[feature_index]\n",
    "\n",
    "        # build info strings\n",
    "        word_join = \" \"\n",
    "        row_join = \"\\n\"\n",
    "\n",
    "        # popup_info_1 (name)\n",
    "        popup_info_str1_row_split = [feature['name']]\n",
    "\n",
    "        # popup_info_2 (annot)\n",
    "        annot_processed = False\n",
    "        if not \"\\n\" in feature['annot']:\n",
    "            popup_info_str2_word_split = feature['annot'].split(\" \")\n",
    "            word_count = len(popup_info_str2_word_split)\n",
    "            if word_count <= annot_word_limit:\n",
    "                if feature['annot'] != '':\n",
    "                    popup_info_str2_row_split = [feature['annot']]\n",
    "                annot_processed = True\n",
    "        if not annot_processed:\n",
    "            popup_info_str2_row_split = []\n",
    "            for row in feature['annot'].split(\"\\n\"):\n",
    "                #if row == '':\n",
    "                #    continue\n",
    "                popup_info_str2_word_split = row.split(\" \")\n",
    "                word_count = len(popup_info_str2_word_split)\n",
    "                if word_count <= annot_word_limit:\n",
    "                    popup_info_str2_row_split.append(row)\n",
    "                else:\n",
    "                    imax = int(round(word_count/annot_word_limit))\n",
    "                    indent = ''\n",
    "                    for i in range(0,imax):\n",
    "                        row_slice = popup_info_str2_word_split[i*annot_word_limit:(i+1)*annot_word_limit]\n",
    "                        if i > 0:\n",
    "                            indent = '  '\n",
    "                        popup_info_str2_row_split.append(indent+word_join.join(row_slice))\n",
    "                    remainder = len(popup_info_str2_word_split)%annot_word_limit\n",
    "                    if remainder > 0:\n",
    "                        row_slice = popup_info_str2_word_split[imax*annot_word_limit:]\n",
    "                        indent = '  '\n",
    "                        popup_info_str2_row_split.append(indent+word_join.join(row_slice))\n",
    "            \n",
    "        # popup_info_3 (rest of info)\n",
    "        popup_info_str3_row_split = [\"locus_tag: %s\"%feature['locus_tag'], \\\n",
    "                                     \"id(s): %s\"%feature['ID'], \\\n",
    "                                     \"E.C.: %s\"%feature['EC_number'], \\\n",
    "                                     \"type: %s\"%feature['type'], \\\n",
    "                                     \"position: %d-%d (%s)\"%(feature['beg_pos'],feature['end_pos'],feature['strand']), \\\n",
    "                                     feature['source_species']\n",
    "                                    ]\n",
    "\n",
    "        # get dimensions and shift popup if hitting walls  \n",
    "        #\n",
    "        popup_info_row_cnt = len(popup_info_str1_row_split) + \\\n",
    "                             len(popup_info_str2_row_split) + \\\n",
    "                             1 + \\\n",
    "                             len(popup_info_str3_row_split)\n",
    "        popup_info_h = popup_text_h_scaling * popup_info_row_cnt\n",
    "        \n",
    "        longest_info_str = 0\n",
    "        for popup_info_str_row_split in [popup_info_str1_row_split, popup_info_str2_row_split, popup_info_str3_row_split]:\n",
    "        #for popup_info_str_row_split in [popup_info_str3_row_split]:\n",
    "            for popup_info_str in popup_info_str_row_split:\n",
    "                row_len = len(popup_info_str)\n",
    "                if row_len > longest_info_str:\n",
    "                    longest_info_str = row_len\n",
    "        popup_info_w = popup_text_w_scaling * longest_info_str\n",
    "\n",
    "        popup_box_w = popup_info_w + 2*popup_x_margin + 1.5*close_button_w\n",
    "        popup_box_h = popup_info_h + 2*popup_y_margin + 1.0*close_button_h\n",
    "        popup_x_shift = 0\n",
    "        popup_y_shift = 0\n",
    "        if mouseevent.xdata+popup_box_w > 1.0:\n",
    "            popup_x_shift = 1.0 - (mouseevent.xdata+popup_box_w) - 0.01  # extra .01 to not hit wall\n",
    "        if mouseevent.ydata+popup_box_h > 1.0:\n",
    "            popup_y_shift = 1.0 - (mouseevent.ydata+popup_box_h) - 0.01  # extra .01 to not hit wall\n",
    "\n",
    "            \n",
    "        # Add Box\n",
    "        #\n",
    "        popup_box = Rectangle((mouseevent.xdata+popup_x_shift, mouseevent.ydata+popup_y_shift), \\\n",
    "                              popup_box_w, \\\n",
    "                              popup_box_h, \\\n",
    "                              facecolor=popup_bg_color, edgecolor=\"gray\", \\\n",
    "                              alpha=1.0, zorder=popup_zorder)\n",
    "        popup_box.gid = feature_index  # if we want to make it and contents draggable later\n",
    "        popup_box.set_path_effects([path_effects.PathPatchEffect(offset=(1, -1), facecolor='black', alpha=0.75),\n",
    "                                    path_effects.PathPatchEffect(facecolor=popup_bg_color,\n",
    "                                                                 edgecolor=\"gray\")\n",
    "                                   ])\n",
    "        ax.add_patch(popup_box)\n",
    " \n",
    "        # Add Info\n",
    "        #\n",
    "        text_y_shift = popup_text_h_scaling*len(popup_info_str3_row_split) + \\\n",
    "                       popup_text_h_scaling*1.0 + \\\n",
    "                       popup_text_h_scaling*len(popup_info_str2_row_split) + \\\n",
    "                       popup_text_h_scaling*0.5\n",
    "        popup_info_1 = ax.text(mouseevent.xdata+popup_x_shift+popup_x_margin, \\\n",
    "                               mouseevent.ydata+popup_y_shift+popup_y_margin + text_y_shift, \\\n",
    "                                row_join.join(popup_info_str1_row_split), \\\n",
    "                                verticalalignment=\"bottom\", \\\n",
    "                                horizontalalignment=\"left\", \\\n",
    "                                transform=ax.transAxes, \\\n",
    "                                color=\"black\", \\\n",
    "                                fontsize=12, \\\n",
    "                                zorder=popup_zorder+1)\n",
    "        #popup_info_1.set_label(feature_index)\n",
    "        #ax.add_patch(popup_info_1)   # text cannot be added as patch to axes, or else things get wonky\n",
    "        \n",
    "        text_y_shift = popup_text_h_scaling*len(popup_info_str3_row_split) + \\\n",
    "                       popup_text_h_scaling*1.0\n",
    "        popup_info_2 = ax.text(mouseevent.xdata+popup_x_shift+popup_x_margin, \\\n",
    "                               mouseevent.ydata+popup_y_shift+popup_y_margin + text_y_shift, \\\n",
    "                                row_join.join(popup_info_str2_row_split), \\\n",
    "                                verticalalignment=\"bottom\", \\\n",
    "                                horizontalalignment=\"left\", \\\n",
    "                                transform=ax.transAxes, \\\n",
    "                                color=\"red\", \\\n",
    "                                fontsize=10, \\\n",
    "                                zorder=popup_zorder+1)\n",
    "        #popup_info_2.set_label(feature_index)\n",
    "        \n",
    "        text_y_shift = 0\n",
    "        popup_info_3 = ax.text(mouseevent.xdata+popup_x_shift+popup_x_margin, \\\n",
    "                               mouseevent.ydata+popup_y_shift+popup_y_margin + text_y_shift, \\\n",
    "                               row_join.join(popup_info_str3_row_split), \\\n",
    "                               verticalalignment=\"bottom\", \\\n",
    "                               horizontalalignment=\"left\", \\\n",
    "                               transform=ax.transAxes, \\\n",
    "                               color=\"black\", \\\n",
    "                               fontsize=10, \\\n",
    "                               zorder=popup_zorder+1)\n",
    "        #popup_info_3.set_label(feature_index)\n",
    "        \n",
    "        \n",
    "        # Add Close button\n",
    "        #\n",
    "        close_button = Rectangle((mouseevent.xdata+popup_x_shift+popup_box_w-popup_x_margin-close_button_w,\n",
    "                                  mouseevent.ydata+popup_y_shift+popup_box_h-popup_y_margin-close_button_h), \\\n",
    "                                  close_button_w, \\\n",
    "                                  close_button_h, \\\n",
    "                                  facecolor=\"white\", edgecolor=\"gray\", \\\n",
    "                                  picker=True, \\\n",
    "                                  alpha=1.0, zorder=popup_zorder+2)\n",
    "        close_button.gid = feature_index\n",
    "        close_button.set_label(\"popup_destroy\")\n",
    "        ax.add_patch(close_button)\n",
    "        #close_button.figure.canvas.mpl_connect ('pick_event', onpick_feature)  # already listening\n",
    "        close_button_x = ax.text(mouseevent.xdata+popup_x_shift+popup_box_w-popup_x_margin-0.5*close_button_w,\n",
    "                                 mouseevent.ydata+popup_y_shift+popup_box_h-popup_y_margin-0.5*close_button_h-close_button_marker_y_shift,\n",
    "                                 'x', \\\n",
    "                                 verticalalignment=\"center\", \\\n",
    "                                 horizontalalignment=\"center\", \\\n",
    "                                 transform=ax.transAxes, \\\n",
    "                                 color=\"gray\", \\\n",
    "                                 fontsize=10, \\\n",
    "                                 zorder=popup_zorder+3)\n",
    "        #close_button_x.set_label(feature_index)\n",
    "        \n",
    "        # Register artists for moving or later destruction\n",
    "        #\n",
    "        feature_artist_label_to_popup_box[feature_index] = []\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_box)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_info_1)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_info_2)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_info_3)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(close_button)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(close_button_x)\n",
    "\n",
    "    # Close popup (conflicts with close button)   \n",
    "    #\n",
    "    else: \n",
    "        popup_box = feature_artist_label_to_popup_box[feature_index][0]\n",
    "        popup_box.remove()\n",
    "        popup_info_1 = feature_artist_label_to_popup_box[feature_index][1]\n",
    "        popup_info_1.remove()\n",
    "        popup_info_2 = feature_artist_label_to_popup_box[feature_index][2]\n",
    "        popup_info_2.remove()\n",
    "        popup_info_3 = feature_artist_label_to_popup_box[feature_index][3]\n",
    "        popup_info_3.remove()\n",
    "        close_button = feature_artist_label_to_popup_box[feature_index][4]\n",
    "        close_button.remove()\n",
    "        close_button_x = feature_artist_label_to_popup_box[feature_index][5]\n",
    "        close_button_x.remove()\n",
    "        \n",
    "        feature_artist_label_to_popup_box[feature_index] = None\n",
    "        feature_index  = None\n",
    "    \n",
    "    artist.figure.canvas.draw()\n",
    "\n",
    "\n",
    "# Diplay coordinate transform\n",
    "#\n",
    "def disp_coord_transform (x, pivot_pos, genomebrowser_window_bp_width, genomebrowser_xshift, track_xshift, contig_mode_xshift):\n",
    "\n",
    "    track_disp_scale_factor = (1.0-right_margin-left_margin) / genomebrowser_window_bp_width\n",
    "    #transformed_x = track_disp_scale_factor * (x - (pivot_pos-0.5*genomebrowser_window_bp_width))\n",
    "    transformed_x = track_disp_scale_factor * (x - (pivot_pos+genomebrowser_xshift+track_xshift+contig_mode_xshift-0.5*genomebrowser_window_bp_width))\n",
    "\n",
    "    return left_margin + transformed_x\n",
    "\n",
    "\n",
    "# draw feature element\n",
    "#\n",
    "def draw_feature_element (ax, \\\n",
    "                     row_n, \\\n",
    "                     max_row_n, \\\n",
    "                     feature_i, \\\n",
    "                     feature_j, \\\n",
    "                     feature, \\\n",
    "                     pivot_pos, \\\n",
    "                     track_xshift, \\\n",
    "                     contig_mode_xshift, \\\n",
    "                     pivot_feature_flag, \\\n",
    "                     annot_repeat, \\\n",
    "                     seq_repeat, \\\n",
    "                     this_label_show_top, this_label_show_bot):\n",
    "    \n",
    "    left_cropped = False\n",
    "    right_cropped = False\n",
    "    feature_oi_flag = False\n",
    "    pivot_feature_flag = False\n",
    "    if feature['beg_pos'] <= pivot_pos and feature['end_pos'] >= pivot_pos:\n",
    "        pivot_feature_flag = True\n",
    "        \n",
    "    gene_id = feature['ID']\n",
    "    gene_name = feature['name']\n",
    "\n",
    "    # Determine if a search result feature\n",
    "    try:\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            contig_search_results = search_results[0]\n",
    "        else:\n",
    "            contig_search_results = search_results[feature_i]\n",
    "    except:\n",
    "        contig_search_results = []\n",
    "    for j,result_list in enumerate(contig_search_results):\n",
    "        for result in contig_search_results[j]:\n",
    "            if result['beg_pos'] == feature['beg_pos'] and result['end_pos'] == feature['end_pos']:\n",
    "                feature_oi_flag = True\n",
    "                highlight_color = search_color_names[j % len(search_color_names)]\n",
    "                break\n",
    "\n",
    "    # window coords\n",
    "    window_beg_pos = disp_coord_transform (feature['beg_pos'], \\\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "    window_end_pos = disp_coord_transform (feature['end_pos'],\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "    \n",
    "    # handle tiny number problems  (FIX: NOT QUITE WORKING)\n",
    "    if window_end_pos < window_beg_pos:\n",
    "        window_end_pos = window_beg_pos + .00000001\n",
    "    if window_beg_pos > 1.0-right_margin:\n",
    "        window_beg_pos = 1.0-right_margin\n",
    "    if window_end_pos < left_margin:\n",
    "        window_end_pos = left_margin\n",
    "\n",
    "    # handle features that exceed margins\n",
    "    if window_beg_pos < left_margin:\n",
    "        window_beg_pos = left_margin\n",
    "        left_cropped = True\n",
    "    if window_end_pos > 1.0-right_margin:\n",
    "        window_end_pos = 1.0-right_margin\n",
    "        right_cropped = True     \n",
    "        \n",
    "    # deal with short features\n",
    "    if feature['type'] != \"CRISPR\" and feature['type'] != \"CRISPR spacer\":\n",
    "        if window_end_pos - window_beg_pos < head_l:\n",
    "            feature_element_base_length = .00000001\n",
    "            arrow_params = {'head_width':head_w, 'head_length':window_end_pos-window_beg_pos}\n",
    "        else:\n",
    "            feature_element_base_length = window_end_pos - window_beg_pos - head_l\n",
    "            arrow_params = {'head_width':head_w, 'head_length':head_l}\n",
    "    elif feature['type'] == \"CRISPR\":\n",
    "        feature_element_base_length = .00000001\n",
    "        arrow_params = {'head_width':0.75*head_w, 'head_length':0.5*(window_end_pos-window_beg_pos)}\n",
    "    elif feature['type'] == \"CRISPR spacer\":\n",
    "        feature_element_base_length = window_end_pos - window_beg_pos\n",
    "\n",
    "        \n",
    "    # not going to change window_beg_pos again, so can set x_pos and y_pos now\n",
    "    x_pos = window_beg_pos   \n",
    "    # FIX\n",
    "    y_pos_base = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(total_rows-row_n)\n",
    "    #y_pos_base = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(max_row_n-row_n)\n",
    "    #y_pos_base = 1.0-0.02*row_n  # DEBUG\n",
    "\n",
    "    \n",
    "    # direction\n",
    "    if feature['strand'] == '-':\n",
    "        direction = 'rev'\n",
    "        feature_element_start_pos = x_pos+feature_element_base_length+head_l\n",
    "        #label_start_pos = x_pos+head_l\n",
    "        label_start_pos = x_pos\n",
    "        dir_adj = -1.0\n",
    "        drop_shadow_xshift = 1\n",
    "    else:\n",
    "        direction = 'fwd'\n",
    "        feature_element_start_pos = x_pos\n",
    "        label_start_pos = x_pos\n",
    "        dir_adj = 1.0\n",
    "        drop_shadow_xshift = 0\n",
    "\n",
    "        \n",
    "    # color\n",
    "    #\n",
    "    feature_element_color = \"lightgray\"\n",
    "    if feature['type'] == \"pseudogene\":\n",
    "        feature_element_color = \"gray\"\n",
    "        \n",
    "    elif feature['type'] == \"rRNA\" or feature['type'] == \"tRNA\":\n",
    "        feature_element_color = \"black\"\n",
    "        \n",
    "    elif feature['type'] == \"CRISPR\":\n",
    "        feature_element_color = \"black\"        \n",
    "        if feature['dna_seq'] in seq_repeat:\n",
    "            feature_element_color = color_names[sum([ord(c) for c in feature['dna_seq']]) % len(color_names)]           \n",
    "    elif feature['type'] == \"CRISPR spacer\":\n",
    "        feature_element_color = \"lightgray\"\n",
    "        if feature['dna_seq'] in seq_repeat:\n",
    "            feature_element_color = color_names[sum([ord(c) for c in feature['dna_seq']]) % len(color_names)]\n",
    "            \n",
    "    elif feature['type'] == \"CDS\":\n",
    "\n",
    "        if Global_State['genomebrowser_color_namespace'] == \"annot\":\n",
    "            if 'annot' in feature:\n",
    "                if feature['annot'] != '' and \\\n",
    "                    (feature['annot'].lower() == \"hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"conserved hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"orf, hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"orf, conserved hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"hypothetical protein; orphan\" \\\n",
    "                    or feature['annot'].lower() == \"hypothetical protein; genus orphan\"):\n",
    "                    \n",
    "                    feature_element_color = \"darkgray\"\n",
    "                elif feature['annot'] == '':\n",
    "                    feature_element_color = \"lightgray\"\n",
    "                else:\n",
    "                    if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                        or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                        if feature['annot'] in annot_repeat:\n",
    "                            feature_element_color = color_names[sum([ord(c) for c in feature['annot']]) % len(color_names)]\n",
    "                    else:\n",
    "                        feature_element_color = color_names[sum([ord(c) for c in feature['annot']]) % len(color_names)]\n",
    "\n",
    "        elif Global_State['genomebrowser_color_namespace'] == \"ec\":\n",
    "            if 'EC_number' in feature and feature['EC_number'] != '':\n",
    "                if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                    or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                    if feature['EC_number'] in annot_repeat:\n",
    "                        feature_element_color = color_names[sum([ord(c) for c in feature['EC_number']]) % len(color_names)]\n",
    "                else:\n",
    "                    feature_element_color = color_names[sum([ord(c) for c in feature['EC_number']]) % len(color_names)]\n",
    "\n",
    "        elif Global_State['genomebrowser_color_namespace'] == 'cog' \\\n",
    "            or Global_State['genomebrowser_color_namespace'] == 'pfam':\n",
    "                \n",
    "            if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "                contig_i = 0\n",
    "            else:\n",
    "                contig_i = feature_i\n",
    "            try:\n",
    "                domain_hits = Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[contig_i]]][ContigSet_names[contig_i]][feature['ID']]\n",
    "            except:\n",
    "                domain_hits = []\n",
    "                \n",
    "            top_domfam = None                \n",
    "            for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "                domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "                if (Global_State['genomebrowser_color_namespace'] == 'cog' and domfam[0:3] == 'COG') \\\n",
    "                    or (Global_State['genomebrowser_color_namespace'] == 'pfam' and domfam[0:2] == 'PF'):\n",
    "                        \n",
    "                    if (Global_State['genomebrowser_mode'] == \"tree\" or Global_State['genomebrowser_mode'] == \"homologs\") \\\n",
    "                          and not domfam in annot_repeat:\n",
    "                        continue                            \n",
    "                    top_domfam = domfam\n",
    "                    break\n",
    "                            \n",
    "            if top_domfam != None:\n",
    "                if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                    or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                    if top_domfam in annot_repeat:  # already checked, but safer to repeat in case logic above changes\n",
    "                        feature_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    "                else:\n",
    "                    feature_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    " \n",
    "        elif Global_State['genomebrowser_color_namespace'] == 'domains':\n",
    "            #feature_element_color = \"none\"\n",
    "            feature_element_color = \"lightgray\"\n",
    "            # individual domains will be drawn separately with rects below\n",
    "\n",
    "            \n",
    "    # edge color\n",
    "    feature_element_edge_lw = 1\n",
    "    feature_element_edge_color = \"none\"\n",
    "#    if (Global_State['genomebrowser_color_namespace'] == 'domains'):\n",
    "#        feature_element_edge_color = \"black\"\n",
    "    if pivot_feature_flag:\n",
    "        if Global_State['genomebrowser_mode'] == 'contigs':\n",
    "            feature_element_edge_color = \"black\"\n",
    "            feature_element_edge_lw = 1\n",
    "        else:\n",
    "            feature_element_edge_color = \"black\"\n",
    "            feature_element_edge_lw = 2\n",
    "    if feature_oi_flag:\n",
    "        feature_element_edge_color = \"black\"\n",
    "        feature_element_edge_lw = 2        \n",
    "                                \n",
    "                    \n",
    "    # Draw Rectangle (patch) or Arrow (from pylab) for features\n",
    "    #\n",
    "    if feature['type'] == \"CRISPR\":\n",
    "        new_arrow_l = ax.arrow (feature_element_start_pos+0.5*(window_end_pos-window_beg_pos), \\\n",
    "                          y_pos_base, \\\n",
    "                          -feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=0.75*arrow_w, \\\n",
    "                          color=feature_element_color, \\\n",
    "                          ec=\"none\",\n",
    "                          #lw=feature_element_edge_lw, \\\n",
    "                          alpha=0.75, \\\n",
    "                          zorder=1, \\\n",
    "                          picker=False, \\\n",
    "                          **arrow_params)\n",
    "        new_arrow_r = ax.arrow (feature_element_start_pos+0.5*(window_end_pos-window_beg_pos), \\\n",
    "                          y_pos_base, \\\n",
    "                          feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=0.75*arrow_w, \\\n",
    "                          color=feature_element_color, \\\n",
    "                          ec=\"none\", \\\n",
    "                          #lw=feature_element_edge_lw, \\\n",
    "                          alpha=0.75, \\\n",
    "                          zorder=1, \\\n",
    "                          picker=False, \\\n",
    "                          **arrow_params)\n",
    "        # FIX: listeners do work, but seems excessive so turning off for now\n",
    "        #feature_index = \"%d,%d\"%(feature_i,feature_j)\n",
    "        #feature_artist_label_to_feature[feature_index] = feature\n",
    "        #new_arrow_l.figure.canvas.mpl_connect ('pick_event', onpick_feature)\n",
    "        #new_arrow_l.gid = feature_index\n",
    "        #new_arrow_l.set_label(\"feature\")\n",
    "        #new_arrow_r.figure.canvas.mpl_connect ('pick_event', onpick_feature)  \n",
    "        #new_arrow_r.gid = feature_index  # not sure I can have 2 elements with the same gid?\n",
    "        new_arrow_r.set_label(\"feature\")\n",
    "        \n",
    "    elif feature['type'] == \"CRISPR spacer\":\n",
    "        rect_feature = Rectangle((x_pos, y_pos_base-0.325*arrow_w), \\\n",
    "                                 feature_element_base_length, 0.75*arrow_w, \\\n",
    "                                 facecolor=feature_element_color, ec=\"none\", alpha=0.75, zorder=2, \\\n",
    "                                 picker=False)\n",
    "        # FIX: listeners don't work as written below, but rect picker works elsewhere so should be doable\n",
    "        #feature_index = \"%d,%d\"%(feature_i,feature_j)\n",
    "        #feature_artist_label_to_feature[feature_index] = feature\n",
    "        #rect_feature.figure.canvas.mpl_connect ('pick_event', onpick_feature)\n",
    "        #rect_feature.gid = feature_index\n",
    "        #rect_feature.set_label(\"feature\")\n",
    "        ax.add_patch(rect_feature)\n",
    "        \n",
    "    else:\n",
    "        new_arrow = ax.arrow (feature_element_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=arrow_w, \\\n",
    "                          #color=feature_element_color, \\\n",
    "                          #ec=feature_element_edge_color, \\\n",
    "                          #lw=feature_element_edge_lw, \\\n",
    "                          #alpha=0.75, \\\n",
    "                          zorder=1, \\\n",
    "                          picker=True, \\\n",
    "                          **arrow_params)\n",
    "        \n",
    "        # Drop shadow baby!\n",
    "        new_arrow.set_path_effects([path_effects.PathPatchEffect(offset=(drop_shadow_xshift, -1),\n",
    "                                                                 facecolor='black', alpha=0.4),\n",
    "                                    path_effects.PathPatchEffect(facecolor=feature_element_color,\n",
    "                                                                 edgecolor=feature_element_edge_color,\n",
    "                                                                 lw=feature_element_edge_lw)])\n",
    "                                                         \n",
    "\n",
    "        # add popup onclick\n",
    "        #\n",
    "        # the object approach causes iPython to hang.  Use artist label to index into hashes instead\n",
    "        #ax.add_patch (new_arrow)\n",
    "        #active_arrow = InteractiveGenomeFeature (new_arrow, feature)\n",
    "        #active_arrow.connect()\n",
    "        #\n",
    "        feature_index = \"%d,%d\"%(feature_i,feature_j)\n",
    "        feature_artist_label_to_feature[feature_index] = feature    \n",
    "        new_arrow.figure.canvas.mpl_connect ('pick_event', onpick_feature)\n",
    "        new_arrow.gid = feature_index\n",
    "        new_arrow.set_label(\"feature\")\n",
    "    \n",
    "\n",
    "    # Add Domains\n",
    "    #\n",
    "    if Global_State['genomebrowser_color_namespace'] == 'domains':\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            contig_i = 0\n",
    "        else:\n",
    "            contig_i = feature_i\n",
    "        try:\n",
    "            domain_hits = Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[contig_i]]][ContigSet_names[contig_i]][feature['ID']]\n",
    "        except:\n",
    "            domain_hits = []\n",
    "\n",
    "        uncovered_frac_thresh = 0.5\n",
    "        domhits_disp = []\n",
    "        # needs to extend beyond by 1 because protein numbering starting from 1\n",
    "        covered_mask = []\n",
    "        for i in range(0,int(round(feature['end_pos']-feature['beg_pos']+1)/3)+2):\n",
    "            covered_mask.append(False)\n",
    "\n",
    "        for dom_namespace in ['TIGR', 'PF', 'COG']:\n",
    "#        for dom_namespace in ['TIGR']:\n",
    "#        for dom_namespace in ['PF']:\n",
    "#        for dom_namespace in ['COG']:\n",
    "            for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "                domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "                if domfam[0:len(dom_namespace)] != dom_namespace:\n",
    "                    continue\n",
    "                if (Global_State['genomebrowser_mode'] == \"tree\" or Global_State['genomebrowser_mode'] == \"homologs\") \\\n",
    "                      and not domfam in annot_repeat:\n",
    "                    continue\n",
    "                    \n",
    "                occluded = False\n",
    "                uncovered_pos_cnt = 0\n",
    "                for i in range(domhit[DOMHIT_BEG_I],domhit[DOMHIT_END_I]+1):\n",
    "                    if not covered_mask[i]:\n",
    "                        uncovered_pos_cnt += 1\n",
    "                    \n",
    "                if uncovered_pos_cnt/(domhit[DOMHIT_END_I]-domhit[DOMHIT_BEG_I]+1) < uncovered_frac_thresh:\n",
    "                    occluded = True\n",
    "                            \n",
    "                if not occluded:\n",
    "                    domhits_disp.append(domhit)\n",
    "                    for i in range(domhit[DOMHIT_BEG_I],domhit[DOMHIT_END_I]+1):\n",
    "                        covered_mask[i] = True\n",
    "\n",
    "        dom_zorder = 3                    \n",
    "        for domhit in sorted(domhits_disp, key=sort_by_domhit_len, reverse=True):\n",
    "            top_domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "            \n",
    "            if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                  or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                if top_domfam in annot_repeat:  # already checked, but safer to repeat in case logic above changes\n",
    "                    dom_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    "            else:\n",
    "                dom_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    "     \n",
    "            # get coords\n",
    "            if direction == 'fwd':\n",
    "                dom_beg_pos = feature['beg_pos']+3*(domhit[DOMHIT_BEG_I]-1)\n",
    "                dom_end_pos = feature['beg_pos']+3*(domhit[DOMHIT_END_I]-1)\n",
    "            else:\n",
    "                dom_beg_pos = feature['end_pos']-3*(domhit[DOMHIT_END_I]-1)\n",
    "                dom_end_pos = feature['end_pos']-3*(domhit[DOMHIT_BEG_I]-1)\n",
    "                    \n",
    "            dom_window_beg_pos = disp_coord_transform (dom_beg_pos, \\\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "            dom_window_end_pos = disp_coord_transform (dom_end_pos, \\\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "            \n",
    "            if dom_window_end_pos <= left_margin or dom_window_beg_pos >= 1.0-right_margin:\n",
    "                continue\n",
    "            if dom_window_beg_pos < left_margin:\n",
    "                dom_window_beg_pos = left_margin\n",
    "            if dom_window_end_pos > 1.0-right_margin:\n",
    "                dom_window_end_pos = 1.0-right_margin\n",
    "\n",
    "            # rect representation  # DEBUG\n",
    "            \"\"\"\n",
    "            ##y_shift_tweak = -0.0005  # rectangles and arrows don't necessarily wind up on same pixel for close calls\n",
    "            y_shift_tweak = 0.0  # rectangles and arrows don't necessarily wind up on same pixel for close calls\n",
    "            dom_x_pos = dom_window_beg_pos\n",
    "            dom_y_pos = y_pos_base - 0.5*arrow_w + y_shift_tweak\n",
    "            dom_w = dom_window_end_pos - dom_window_beg_pos + .00000001\n",
    "            dom_h = arrow_w\n",
    "                    \n",
    "            rect_feature = Rectangle((dom_x_pos, dom_y_pos), dom_w, dom_h, \\\n",
    "                                     facecolor=dom_element_color, ec=\"black\", lw=1, alpha=1.0, zorder=dom_zorder, \\\n",
    "                                     picker=False)  # Probably will want picker\n",
    "            ax.add_patch(rect_feature)\n",
    "            \"\"\"\n",
    "            \n",
    "            # arrow representation of domains\n",
    "            #\n",
    "            dom_y_scale = 0.9\n",
    "            #dom_y_scale = 1.0  # DEBUG\n",
    "            dom_arrow_w = dom_y_scale * arrow_w\n",
    "            dom_element_edge_color = 'none'\n",
    "            dom_element_edge_lw = 0\n",
    "            add_tip_crop = False\n",
    "                \n",
    "            # direction and position            \n",
    "            if direction == 'fwd':\n",
    "                \n",
    "                # domain completely in arrow head\n",
    "                if window_end_pos - dom_window_beg_pos <= head_l:\n",
    "                    dom_element_base_length = .00000001\n",
    "                    dom_head_l = window_end_pos - dom_window_beg_pos\n",
    "                    if window_end_pos - window_beg_pos <= head_l:\n",
    "                        this_head_l = window_end_pos - window_beg_pos\n",
    "                    else:\n",
    "                        this_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * dom_head_l * (head_w/this_head_l)\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_end_pos < window_end_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_end_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = window_end_pos - dom_window_end_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/this_head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                # domain completely *not* in arrow head\n",
    "                elif dom_window_end_pos <= window_end_pos - head_l:\n",
    "                    dom_element_base_length = dom_window_end_pos - dom_window_beg_pos\n",
    "                    dom_head_l = 0\n",
    "                    dom_head_w = 0\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "\n",
    "                # domain in both regions\n",
    "                else:  # dom_window_beg_pos <= window_end_pos - head_l and dom_window_end_pos > window_end_pos - head_l\n",
    "                    dom_element_base_length = window_end_pos - dom_window_beg_pos - head_l\n",
    "                    if dom_element_base_length < 0:\n",
    "                        dom_element_base_length = .00000001\n",
    "                    dom_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * head_w\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_end_pos < window_end_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_end_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = window_end_pos - dom_window_end_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                dom_element_start_pos = dom_window_beg_pos\n",
    "                dir_adj = 1.0\n",
    "                \n",
    "            else:  # direction == \"rev\"  \n",
    "                \n",
    "                # domain completely in arrow head\n",
    "                if dom_window_end_pos - window_beg_pos <= head_l:\n",
    "                    dom_element_base_length = .00000001\n",
    "                    dom_head_l = dom_window_end_pos - window_beg_pos\n",
    "                    if window_end_pos - window_beg_pos <= head_l:\n",
    "                        this_head_l = window_end_pos - window_beg_pos\n",
    "                    else:\n",
    "                        this_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * dom_head_l * (head_w/this_head_l)\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_beg_pos > window_beg_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_beg_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = dom_window_beg_pos - window_beg_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/this_head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                # domain completely *not* in arrow head\n",
    "                elif dom_window_beg_pos >= window_beg_pos + head_l:\n",
    "                    dom_element_base_length = dom_window_end_pos - dom_window_beg_pos\n",
    "                    dom_head_l = head_l\n",
    "                    dom_head_w = 0\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':0}\n",
    "\n",
    "                # domain in both regions\n",
    "                else:  # dom_window_beg_pos <= window_end_pos - head_l and dom_window_end_pos > window_end_pos - head_l\n",
    "                    dom_element_base_length = dom_window_end_pos - window_beg_pos - head_l\n",
    "                    if dom_element_base_length < 0:\n",
    "                        dom_element_base_length = .00000001\n",
    "                    dom_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * head_w\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_beg_pos > window_beg_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_beg_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = dom_window_beg_pos - window_beg_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                dom_element_start_pos = window_beg_pos+dom_element_base_length+dom_head_l\n",
    "                dir_adj = -1.0 \n",
    "\n",
    "                \n",
    "            # draw domain element \n",
    "            #\n",
    "            dom_arrow = ax.arrow (dom_element_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*dom_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=dom_arrow_w, \\\n",
    "                          color=dom_element_color, \\\n",
    "                          ec=dom_element_edge_color, \\\n",
    "                          #lw=dom_element_edge_lw, \\\n",
    "                          alpha=1.0, \\\n",
    "                          zorder=dom_zorder, \\\n",
    "                          picker=False, \\\n",
    "                          **dom_arrow_params)\n",
    "            if add_tip_crop:\n",
    "                tip_arrow = ax.arrow (tip_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*tip_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=dom_arrow_w, \\\n",
    "                          color=feature_element_color, \\\n",
    "                          ec='none', \\\n",
    "                          #lw=dom_element_edge_lw, \\\n",
    "                          alpha=1.0, \\\n",
    "                          zorder=dom_zorder+1, \\\n",
    "                          picker=False, \\\n",
    "                          **tip_arrow_params)\n",
    "            \n",
    "            dom_zorder += 2\n",
    "            \n",
    "        # restore border for highlighted features           \n",
    "        if pivot_feature_flag or feature_oi_flag:\n",
    "            border_arrow = ax.arrow (feature_element_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=arrow_w, \\\n",
    "                          color='none', \\\n",
    "                          ec=feature_element_edge_color, \\\n",
    "                          lw=feature_element_edge_lw, \\\n",
    "                          alpha=1.0, \\\n",
    "                          zorder=dom_zorder+1, \\\n",
    "                          picker=False, \\\n",
    "                          **arrow_params)\n",
    "                \n",
    "    # Feature label\n",
    "    #\n",
    "    if Global_State['genomebrowser_window_bp_width'] <= text_disp_window_bp_limit or pivot_feature_flag:\n",
    "        \n",
    "        label_color = \"black\"\n",
    "        arrow_label_fontsize = base_arrow_label_fontsize        \n",
    "        \n",
    "        # indicate pivot feature\n",
    "        if pivot_feature_flag and Global_State['genomebrowser_mode'] != \"contigs\":\n",
    "            #label_color = 'red'\n",
    "            label_color = 'black'            \n",
    "            arrow_label_fontsize = foi_arrow_label_fontsize\n",
    "            if this_label_show_top == False and this_label_show_bot == False:\n",
    "                this_label_show_top = True\n",
    "\n",
    "        if feature_oi_flag:\n",
    "            label_color = highlight_color            \n",
    "            arrow_label_fontsize = foi_arrow_label_fontsize\n",
    "            if this_label_show_top == False and this_label_show_bot == False:\n",
    "                this_label_show_top = True\n",
    "\n",
    "        # label position\n",
    "        label_y_pos = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(total_rows-row_n)\n",
    "\n",
    "        if Global_State['genomebrowser_window_bp_width'] > text_disp_window_bp_limit or this_label_show_top:\n",
    "            vert_align = \"bottom\"\n",
    "            label_y_pos = y_pos_base + 0.9*text_yshift\n",
    "            if direction == \"rev\":\n",
    "                label_start_pos += .0015\n",
    "            #else:\n",
    "            #    label_start_pos += .0005\n",
    "        else:\n",
    "            vert_align = \"top\"\n",
    "            label_y_pos = y_pos_base - 1.25*text_yshift\n",
    "            if direction == \"rev\":\n",
    "                label_start_pos += .0050\n",
    "            #else:\n",
    "            #    label_start_pos += .0005\n",
    "\n",
    "        # draw label\n",
    "        if pivot_feature_flag or feature_oi_flag or this_label_show_top or this_label_show_bot:\n",
    "            feature_label = ax.text(label_start_pos, \\\n",
    "                                    label_y_pos, \\\n",
    "                                    gene_name, \\\n",
    "                                    #row_n, \\  # DEBUG\n",
    "                                    verticalalignment=vert_align, \\\n",
    "                                    horizontalalignment=\"left\", \\\n",
    "                                    transform=ax.transAxes, \\\n",
    "                                    color=label_color, \\\n",
    "                                    fontsize=arrow_label_fontsize, \\\n",
    "                                    zorder=1)\n",
    "\n",
    "        #print (feature['source_species']+\" \"+feature['name']+\" %f %f %f\"%(arrow_start_pos, arrow_base_length, y_pos_base))\n",
    "        \n",
    "    # add ellipsis to indicate cropped feature\n",
    "    #\n",
    "    if left_cropped:\n",
    "        ellipsis_spacing = 0.2 * left_margin\n",
    "        x_center_base = left_margin - 2*ellipsis_spacing\n",
    "        y_center_base = y_pos_base\n",
    "        diameter = 0.20 * arrow_w\n",
    "        ellipse_to_circle_scaling = 0.75*figure_width / (figure_height_scaling*(total_rows+1))        \n",
    "        x_diameter = 1.0 * diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * diameter\n",
    "        for i in range(0,3):\n",
    "            ellipse_center = (x_center_base-i*ellipsis_spacing, y_center_base)\n",
    "            ellipsis_dot_patch = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                         facecolor=feature_element_color, edgecolor=feature_element_color, \\\n",
    "                                         lw=1, alpha=1.0, zorder=1)\n",
    "        \n",
    "            ax.add_patch(ellipsis_dot_patch)\n",
    "            \n",
    "    if right_cropped:\n",
    "        ellipsis_spacing = 0.2 * left_margin\n",
    "        x_center_base = 1.0-right_margin + 2*ellipsis_spacing\n",
    "        y_center_base = y_pos_base\n",
    "        diameter = 0.20 * arrow_w\n",
    "        ellipse_to_circle_scaling = 0.75*figure_width / (figure_height_scaling*(total_rows+1))        \n",
    "        x_diameter = 1.0 * diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * diameter\n",
    "        for i in range(0,3):\n",
    "            ellipse_center = (x_center_base+i*ellipsis_spacing, y_center_base)\n",
    "            ellipsis_dot_patch = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                         facecolor=feature_element_color, edgecolor=feature_element_color, \\\n",
    "                                         lw=1, alpha=1.0, zorder=1)\n",
    "        \n",
    "            ax.add_patch(ellipsis_dot_patch)\n",
    "\n",
    "\n",
    "# Paint methods\n",
    "#\n",
    "def draw_mode_panel (ax, contig_name, genomebrowser_mode, data_set_name):\n",
    "    \n",
    "    # Config\n",
    "    field_name_fontsize=12\n",
    "    field_val_fontsize=11\n",
    "    field_color=\"white\"\n",
    "    value_color=\"lightsalmon\"\n",
    "    #shadow_color=\"magenta\"\n",
    "    shadow_color=\"blue\"\n",
    "    #shadow_color=\"gray\"\n",
    "    base_zorder=5\n",
    "\n",
    "    # Draw background with rounded corners\n",
    "    #\n",
    "    bb = mtransforms.Bbox([[0.06, 0.10], [0.94, 0.89]])\n",
    "    bb_shadow = mtransforms.Bbox([[0.07, 0.08], [0.95, 0.87]])\n",
    "\n",
    "    p_fancy = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.04\",\n",
    "                             mutation_aspect=0.25*figure_width/top_nav_height,\n",
    "                             fc=nav_bg_color,\n",
    "                             ec='white',\n",
    "                             lw=2,\n",
    "                             zorder=base_zorder)\n",
    "    ax.add_patch(p_fancy)\n",
    "    \"\"\"\n",
    "    p_fancy_back = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.06\",\n",
    "                             mutation_aspect=0.25*figure_width/top_nav_height,\n",
    "                             fc='white',\n",
    "                             ec='black',\n",
    "                             lw=1,\n",
    "                             zorder=base_zorder-1)\n",
    "    ax.add_patch(p_fancy_back)\n",
    "    \"\"\"\n",
    "    p_fancy_shadow = FancyBboxPatch((bb_shadow.xmin, bb_shadow.ymin),\n",
    "                             abs(bb_shadow.width), abs(bb_shadow.height),\n",
    "                             boxstyle=\"round,pad=0.04\",\n",
    "                             mutation_aspect=0.25*figure_width/top_nav_height,\n",
    "                             fc=shadow_color,\n",
    "                             ec=shadow_color,\n",
    "                             lw=1,\n",
    "                             alpha=0.5,\n",
    "                             zorder=base_zorder-2)\n",
    "    ax.add_patch(p_fancy_shadow)\n",
    "    \n",
    "    # Tool title\n",
    "    ax.text(0.05, \\\n",
    "            0.9, \\\n",
    "            tool_title, \\\n",
    "            verticalalignment=\"top\", \\\n",
    "            horizontalalignment=\"left\", \\\n",
    "            #transform=ax_top_left.transAxes, \\\n",
    "            color=\"white\", \\\n",
    "            style=\"normal\", # style=\"oblique\", \\\n",
    "            fontsize=14,\n",
    "            zorder=base_zorder+1)\n",
    "    \n",
    "    # Genome\n",
    "    name_disp = contig_name\n",
    "    if KBase_backend:\n",
    "        #[ws_id, genome_contig_id] = contig_name.split('/')\n",
    "        #[genome_id, contig_id] = genome_contig_id.split(genome_contig_id_delim)\n",
    "        [genome_ref, scaffold_id] = contig_name.split(genome_contig_id_delim)\n",
    "        name_disp = Species_name_by_genome_ref[genome_ref]\n",
    "    ax.text(0.30, 0.5, \"Genome\", verticalalignment=\"bottom\", horizontalalignment=\"right\", color=field_color, fontsize=field_name_fontsize, zorder=base_zorder+1)\n",
    "    ax.text(0.35, 0.5, name_disp, verticalalignment=\"bottom\", horizontalalignment=\"left\", color=value_color, fontsize=field_val_fontsize, zorder=base_zorder+1)\n",
    "        \n",
    "    # Mode\n",
    "    ax.text(0.30, 0.3, \"Mode\", verticalalignment=\"bottom\", horizontalalignment=\"right\", color=field_color, fontsize=field_name_fontsize, zorder=base_zorder+1)\n",
    "    ax.text(0.35, 0.3, genomebrowser_mode, verticalalignment=\"bottom\", horizontalalignment=\"left\", color=value_color, fontsize=field_val_fontsize, zorder=base_zorder+1)\n",
    "\n",
    "    # Dataset\n",
    "    #ax.text(0.30, 0.1, \"Data\", verticalalignment=\"bottom\", horizontalalignment=\"right\", color=field_color, fontsize=field_name_fontsize, zorder=base_zorder+1)\n",
    "    #ax.text(0.35, 0.1, data_set_name, verticalalignment=\"bottom\", horizontalalignment=\"left\", color=value_color, fontsize=field_val_fontsize, zorder=base_zorder+1)\n",
    "\n",
    "    \n",
    "def draw_control_panel (ax, genomebrowser_zoom, genomebrowser_xshift, genomebrowser_color_namespace):\n",
    "\n",
    "    # Config\n",
    "    text_color = \"white\"\n",
    "    base_zorder = 5\n",
    "    #shadow_color = 'magenta'\n",
    "    shadow_color = 'blue'\n",
    "    #shadow_color = 'gray'\n",
    "    \n",
    "    # Draw background with rounded corners\n",
    "    #\n",
    "    bb = mtransforms.Bbox([[0.02, 0.11], [0.98, 0.88]])\n",
    "    bb_shadow = mtransforms.Bbox([[0.0225, 0.09], [0.9825, 0.86]])\n",
    "\n",
    "    p_fancy = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.015\",\n",
    "                             mutation_aspect=0.75*figure_width/top_nav_height,\n",
    "                             fc=nav_bg_color,\n",
    "                             ec='white',\n",
    "                             lw=2,\n",
    "                             zorder=base_zorder)\n",
    "    ax.add_patch(p_fancy)\n",
    "    \"\"\"\n",
    "    p_fancy_back = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.020\",\n",
    "                             mutation_aspect=0.75*figure_width/top_nav_height,\n",
    "                             fc='white',\n",
    "                             ec='black',\n",
    "                             lw=1,\n",
    "                             zorder=base_zorder-1)\n",
    "    ax.add_patch(p_fancy_back)\n",
    "    \"\"\"\n",
    "    p_fancy_shadow = FancyBboxPatch((bb_shadow.xmin, bb_shadow.ymin),\n",
    "                             abs(bb_shadow.width), abs(bb_shadow.height),\n",
    "                             boxstyle=\"round,pad=0.015\",\n",
    "                             mutation_aspect=0.75*figure_width/top_nav_height,\n",
    "                             fc=shadow_color,\n",
    "                             ec=shadow_color,\n",
    "                             lw=1,\n",
    "                             alpha=0.5,\n",
    "                             zorder=base_zorder-2)\n",
    "    ax.add_patch(p_fancy_shadow)\n",
    "\n",
    "    \n",
    "    # Color slider\n",
    "    #\n",
    "    slider_name = \"Color\"\n",
    "    slider_name_h_align = \"right\"\n",
    "    tic_label_v_align=\"bottom\"\n",
    "    #slider_x = 0.64\n",
    "    slider_x = 0.13\n",
    "    slider_y = 0.55\n",
    "    #track_w = 0.25\n",
    "    #track_w = 0.45\n",
    "    track_w = 0.40\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    slider_margin = 0.02\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    \n",
    "    num_tics = len(color_namespace_names)\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    knob_initial_pos = 0\n",
    "    for i in range(0,num_tics):\n",
    "        knob_initial_pos = i\n",
    "        if Global_State['genomebrowser_color_namespace'] == color_namespace_names[i]:\n",
    "            break\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        tic_labels.append(color_namespace_names_disp[i])\n",
    "    \n",
    "    ax.text(slider_x-1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y+slider_h+1.00*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    #color_knob_patch = Rectangle((slider_x+0.5*track_w-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=4)\n",
    "    color_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5)\n",
    "    ax.add_patch (color_knob_patch)\n",
    "    color_knob = DraggableRectangleSlider(color_knob_patch, func='color', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    color_knob.connect()\n",
    "    \n",
    "    \n",
    "    # Mode slider  \n",
    "    #\n",
    "    slider_name = \"Mode\"\n",
    "    slider_name_h_align = \"left\"\n",
    "    tic_label_v_align=\"bottom\"\n",
    "    #slider_x = 0.13\n",
    "    slider_x = 0.64\n",
    "    slider_y = 0.55\n",
    "    #track_w = 0.45\n",
    "    track_w = 0.25\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    slider_margin = 0.02\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    num_tics = len(mode_names)\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    knob_initial_pos = 0\n",
    "    for i in range(0,num_tics):\n",
    "        knob_initial_pos = i\n",
    "        if Global_State['genomebrowser_mode'] == mode_names[i]:\n",
    "            break\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        tic_labels.append(mode_names_disp[i])\n",
    "\n",
    "    ax.text(slider_x+track_w+1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y+slider_h+1.00*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    mode_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5, transform=ax.transAxes)\n",
    "    ax.add_patch (mode_knob_patch)\n",
    "    mode_knob = DraggableRectangleSlider(mode_knob_patch, func='mode', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    mode_knob.connect()\n",
    "    \n",
    "    \n",
    "    # Position slider  \n",
    "    #\n",
    "    slider_name = \"Position\"\n",
    "    slider_name_h_align = \"right\"\n",
    "    tic_label_v_align=\"top\"\n",
    "    pan_button_w = 0.025\n",
    "    pan_button_h = 0.2\n",
    "    pan_button_margin = 0.01\n",
    "    slider_margin = 0.02\n",
    "    slider_x = 0.13 + 3*(pan_button_w+pan_button_margin)\n",
    "    slider_y = 0.25\n",
    "    #track_w = 0.45 - 3*(pan_button_w+pan_button_margin)\n",
    "    track_w = 0.40 - 3*(pan_button_w+pan_button_margin)\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    num_tics = 11\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    label_bp_disp = 1000000\n",
    "    genome_step = int(round(Global_State['PrimaryContig_len'] / (num_tics-1)))\n",
    "    knob_initial_pos = (num_tics-1)*(1.0-(Global_State['PrimaryContig_len']-(Global_State['PrimaryAnchor_pivot_pos']+Global_State['genomebrowser_xshift']))/Global_State['PrimaryContig_len'])\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        if i == num_tics-1:\n",
    "            tic_val = \"%5.1f\"%(Global_State['PrimaryContig_len']/label_bp_disp)\n",
    "        else:\n",
    "        #    tic_val = i*genome_step/label_bp_disp      \n",
    "        #tic_labels.append(\"%4.1f\"%tic_val)\n",
    "             tic_val = \"%5.1f\"%(i*genome_step/label_bp_disp)        \n",
    "        tic_labels.append(tic_val.strip())\n",
    "        \n",
    "    ax.text(slider_x-3*(pan_button_w+pan_button_margin)-1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y-1.25*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    pos_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5, transform=ax.transAxes)\n",
    "    ax.add_patch (pos_knob_patch)\n",
    "    pos_knob = DraggableRectangleSlider(pos_knob_patch, func='pan', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    pos_knob.connect()\n",
    "    \n",
    "    # pan buttons\n",
    "    pan_button_left = Rectangle((slider_x-3*(pan_button_w+pan_button_margin)-slider_margin, slider_y), pan_button_w, pan_button_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+3, transform=ax.transAxes, picker=True)\n",
    "    ax.text(slider_x-3*(pan_button_w+pan_button_margin)-slider_margin+0.5*pan_button_w, slider_y+0.5*pan_button_h, \"<\", verticalalignment=\"center\", horizontalalignment=\"center\", color=\"white\", fontsize=12, zorder=base_zorder+4)\n",
    "    pan_button_left.set_label(\"pan_button\")\n",
    "    pan_button_left.gid = \"left_pan\"\n",
    "    ax.add_patch(pan_button_left)\n",
    "\n",
    "    pan_button_reset = Rectangle((slider_x-2*(pan_button_w+pan_button_margin)-slider_margin, slider_y), pan_button_w, pan_button_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+3, transform=ax.transAxes, picker=True)\n",
    "    ax.text(slider_x-2*(pan_button_w+pan_button_margin)-slider_margin+0.5*pan_button_w, slider_y+0.5*pan_button_h, \"o\", verticalalignment=\"center\", horizontalalignment=\"center\", color=\"white\", fontsize=14, zorder=base_zorder+4)\n",
    "    pan_button_reset.set_label(\"pan_button\")\n",
    "    pan_button_reset.gid = \"reset_pan\"\n",
    "    ax.add_patch(pan_button_reset)\n",
    "    \n",
    "    pan_button_right = Rectangle((slider_x-1*(pan_button_w+pan_button_margin)-slider_margin, slider_y), pan_button_w, pan_button_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+3, transform=ax.transAxes, picker=True)\n",
    "    ax.text(slider_x-1*(pan_button_w+pan_button_margin)-slider_margin+0.5*pan_button_w, slider_y+0.5*pan_button_h, \">\", verticalalignment=\"center\", horizontalalignment=\"center\", color=\"white\", fontsize=12, zorder=base_zorder+4)\n",
    "    pan_button_right.set_label(\"pan_button\")\n",
    "    pan_button_right.gid = \"right_pan\"\n",
    "    ax.add_patch(pan_button_right)\n",
    "    \n",
    "    pan_button_reset.figure.canvas.mpl_connect ('pick_event', onpick_pan)\n",
    "\n",
    "    \n",
    "    # Zoom slider\n",
    "    #\n",
    "    slider_name = \"Zoom\"\n",
    "    slider_name_h_align = \"left\"\n",
    "    tic_label_v_align=\"top\"\n",
    "    slider_x = 0.64\n",
    "    slider_y = 0.25\n",
    "    track_w = 0.25\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    slider_margin = 0.02\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    num_tics = def_genomebrowser_zoom_tics\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    knob_initial_pos = Global_State['genomebrowser_zoom']\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        tic_val = int(round(0.001*Global_State['def_genomebrowser_window_bp_width']*(2**i)))\n",
    "        tic_labels.append(\"%d\"%tic_val)\n",
    "    \n",
    "    ax.text(slider_x+track_w+1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y-1.25*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    zoom_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5)\n",
    "    ax.add_patch (zoom_knob_patch)\n",
    "    zoom_knob = DraggableRectangleSlider(zoom_knob_patch, func='zoom', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    zoom_knob.connect()\n",
    "\n",
    "    \n",
    "def draw_sidenav_panel (ax, genomebrowser_mode):\n",
    "    \n",
    "    # Tree representation\n",
    "    #\n",
    "    if genomebrowser_mode == \"tree\":\n",
    "\n",
    "        \"\"\"\n",
    "        # BioPython Phylo version screws up height scaling\n",
    "        treeObj = Phylo.read(tree_data_base_path+'/'+tree_data_file, tree_data_format)\n",
    "        #print (treeObj)\n",
    "        treeObj.ladderize()   # Flip branches so deeper clades are displayed at top\n",
    "        # TODO: reorder genome tracks following species tree?  Must put primary contig on top.\n",
    "        Phylo.draw(treeObj, axes=ax)\n",
    "        \"\"\"\n",
    "        \n",
    "        # Draw Tree with ETE3\n",
    "        import ete3\n",
    "        newick_string = ''\n",
    "        tree_path = path.join(tree_data_base_path,tree_data_file)\n",
    "        tree_img_path = path.join('.','tree.png')\n",
    "        #tree_img_path = path.join('.','tree.pdf')  # can't put pdf into plot\n",
    "        #with open (tree_path, 'r', 0) as tree_file_handle:   # can't have unbuffered text I/O\n",
    "        with open (tree_path, 'r') as tree_file_handle:\n",
    "            for tree_line in tree_file_handle:\n",
    "                newick_string += tree_line\n",
    "\n",
    "        # ETE3 customization\n",
    "        treeObj = ete3.Tree(newick_string)\n",
    "        #treeObj.ladderize()  # read row order from leaves?\n",
    "        ts = ete3.TreeStyle()\n",
    "        #ts.show_leaf_name = True\n",
    "        ts.show_leaf_name = False\n",
    "        ts.show_branch_length = False\n",
    "        ts.show_branch_support = True\n",
    "        #ts.scale = 50 # 50 pixels per branch length unit\n",
    "        ts.branch_vertical_margin = 70 # pixels between adjacent branches\n",
    "        #ts.title.add_face(ete3.TextFace(params['output_name']+\": \"+params['desc'], fsize=10), column=0)\n",
    "        node_style = ete3.NodeStyle()\n",
    "        node_style[\"fgcolor\"] = \"#606060\"  # for node balls\n",
    "        node_style[\"size\"] = 10  # for node balls (gets reset based on support)\n",
    "        node_style[\"vt_line_color\"] = \"#606060\"\n",
    "        node_style[\"hz_line_color\"] = \"#606060\"\n",
    "        node_style[\"vt_line_width\"] = 4\n",
    "        node_style[\"hz_line_width\"] = 4\n",
    "        node_style[\"vt_line_type\"] = 0 # 0 solid, 1 dashed, 2 dotted\n",
    "        node_style[\"hz_line_type\"] = 0\n",
    "\n",
    "        leaf_style = ete3.NodeStyle()\n",
    "        leaf_style[\"fgcolor\"] = \"#ffffff\"  # for node balls\n",
    "        leaf_style[\"size\"] = 0  # for node balls (we're not using it to add space)\n",
    "        leaf_style[\"vt_line_color\"] = \"#606060\"  # unecessary\n",
    "        leaf_style[\"hz_line_color\"] = \"#606060\"\n",
    "        leaf_style[\"vt_line_width\"] = 4\n",
    "        leaf_style[\"hz_line_width\"] = 4\n",
    "        leaf_style[\"vt_line_type\"] = 0 # 0 solid, 1 dashed, 2 dotted\n",
    "        leaf_style[\"hz_line_type\"] = 0\n",
    "        for n in treeObj.traverse():\n",
    "            if n.is_leaf():\n",
    "                style = leaf_style\n",
    "                #n.name = genome_sci_name_by_id[genome_id]\n",
    "                #n.name = None\n",
    "                #n.add_face(ete3.TextFace(leaf_name_disp, fsize=10), column=0, position=\"branch-right\")\n",
    "            else:\n",
    "                style = ete3.NodeStyle()\n",
    "                for k in node_style.keys():\n",
    "                    style[k] = node_style[k]\n",
    "\n",
    "                if n.support > 0.95:\n",
    "                    style[\"size\"] = 12\n",
    "                elif n.support > 0.90:\n",
    "                    style[\"size\"] = 10\n",
    "                elif n.support > 0.80:\n",
    "                    style[\"size\"] = 8\n",
    "                else:\n",
    "                    style[\"size\"] = 4\n",
    "\n",
    "            n.set_style(style)\n",
    "\n",
    "        # write tree to image file\n",
    "        dpi = 600\n",
    "        img_units = \"in\"\n",
    "        img_pix_width = 200\n",
    "        img_in_width = round(float(img_pix_width)/float(dpi), 1)\n",
    "        #img_in_height = img_in_width * total_rows / 5.0\n",
    "        #treeObj.render(tree_img_path, w=img_in_width, h=img_in_height, units=img_units, dpi=dpi, tree_style=ts)\n",
    "        treeObj.render(tree_img_path, w=img_in_width, units=img_units, dpi=dpi, tree_style=ts)\n",
    "\n",
    "        # load and display tree\n",
    "        from matplotlib.offsetbox import (OffsetImage, AnnotationBbox)\n",
    "        xy = [0.5, 0.51]\n",
    "        arr_img = plt.imread(tree_img_path, format='png')\n",
    "        #ax.imshow(arr_img)\n",
    "        \n",
    "        imagebox = OffsetImage(arr_img, zoom=1.0)\n",
    "        imagebox.image.axes = ax\n",
    "\n",
    "        ab = AnnotationBbox(imagebox, xy, frameon=False)\n",
    "        ax.add_artist(ab)\n",
    "        \n",
    "        \n",
    "    # Line representation\n",
    "    #\n",
    "    elif genomebrowser_mode == \"contigs\" \\\n",
    "        or genomebrowser_mode == \"homologs\":\n",
    "            \n",
    "        contig_label_fontsize = 12\n",
    "        contig_label_x_margin = 0.0075\n",
    "        contig_label_y_margin = 0.06 / (figure_height_scaling*(total_rows+1))\n",
    "        contig_color = \"steelblue\"\n",
    "        contig_disp_margin = 0.10\n",
    "        contig_h           = 0.17 / (figure_height_scaling*(total_rows+1))\n",
    "        clickable_y_margin = 0.75 * contig_h\n",
    "            \n",
    "\n",
    "        longest_contig_len = 0\n",
    "        for i,contig_len in enumerate(Global_State['Contig_lens']):\n",
    "            if i >= max_rows:\n",
    "                break\n",
    "            if longest_contig_len < contig_len:\n",
    "                longest_contig_len = contig_len\n",
    "                    \n",
    "        for i,contig_len in enumerate(Global_State['Contig_lens']):\n",
    "            if i >= max_rows:\n",
    "                break\n",
    "            pivot_pos = Global_State['pivot_pos_list'][i]\n",
    "\n",
    "            # Contig coords\n",
    "            #\n",
    "            x0 = contig_disp_margin\n",
    "            y0 = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(total_rows-(i+1)) - 0.5*contig_h\n",
    "            w = (1.0 - 2*contig_disp_margin) * contig_len / longest_contig_len\n",
    "            h = contig_h\n",
    "                \n",
    "            # Add clickable base\n",
    "            #\n",
    "            x0_click = 0.5*contig_disp_margin\n",
    "            y0_click = y0 - 0.75*clickable_y_margin\n",
    "            w_click = (1.0 - 2*contig_disp_margin) * contig_len / longest_contig_len + contig_disp_margin\n",
    "            h_click = contig_h + 3.0*clickable_y_margin\n",
    "            clickable_contig_rect = Rectangle((x0_click,y0_click), w_click, h_click, \\\n",
    "                                                  facecolor=\"none\", edgecolor=\"none\", \\\n",
    "                                                  alpha=0.0, zorder=0, \\\n",
    "                                                  picker=True)         \n",
    "            clickable_contig_rect.set_label(\"linear_contig_nav\")\n",
    "            clickable_contig_rect.gid = \"%d,%f,%f\"%(i,x0,w)\n",
    "            ax.add_patch(clickable_contig_rect)\n",
    "            clickable_contig_rect.figure.canvas.mpl_connect ('pick_event', onpick_contig_nav)\n",
    "        \n",
    "            # Show contig\n",
    "            #\n",
    "            contig_rect = Rectangle((x0,y0), w, h, \\\n",
    "                                    facecolor=contig_color, edgecolor=contig_color, \\\n",
    "                                        alpha=1.0, zorder=1)                 \n",
    "            # drop shadow\n",
    "            contig_rect.set_path_effects([path_effects.PathPatchEffect(offset=(2, -2),\n",
    "                                                                       facecolor='magenta', edgecolor='none', alpha=0.5),\n",
    "                                          path_effects.PathPatchEffect(facecolor=contig_color, edgecolor=contig_color)]) \n",
    "            ax.add_patch (contig_rect)\n",
    "\n",
    "            # Add label\n",
    "            #\n",
    "            #print (\"CONTIGSET_NAME: '\"+ContigSet_names[i]+\"'\\n\")  # DEBUG\n",
    "            disp_name = ContigSet_names[i]\n",
    "            if KBase_backend:\n",
    "                #full_name_split = ContigSet_names[i].split('/')\n",
    "                #[genome_id, scaffold_id] = full_name_split[1].split(genome_contig_id_delim)\n",
    "                [genome_id, scaffold_id] = ContigSet_names[i].split(genome_contig_id_delim)\n",
    "                scaffold_id_disp = scaffold_id.split('.')[-1]\n",
    "                disp_name = Species_name_by_genome_ref[genome_id]+' c:'+scaffold_id_disp\n",
    "            ax.text(x0+contig_label_x_margin, y0-contig_label_y_margin,\n",
    "                    disp_name, verticalalignment=\"top\", horizontalalignment=\"left\",\n",
    "                    color=\"black\", fontsize=contig_label_fontsize, zorder=1)\n",
    "\n",
    "            # Mark pivot\n",
    "            #\n",
    "            if Global_State['genomebrowser_mode'] == 'homologs':\n",
    "                mark_h = 1.25 * contig_h\n",
    "                mark_w = 0.01\n",
    "                mark_margin = 0.02 * mark_h\n",
    "                disp_mark_x0 = x0 + w * (1-(contig_len-pivot_pos)/contig_len) - 0.5*mark_w\n",
    "                disp_mark_y0 = y0 + contig_h + mark_margin\n",
    "                pivot_feature_rect = Rectangle((disp_mark_x0,disp_mark_y0), mark_w, mark_h,\n",
    "                                               facecolor=\"black\", edgecolor=\"none\", alpha=1.0, zorder=2) \n",
    "                ax.add_patch (pivot_feature_rect)\n",
    "\n",
    "            # Mark search results\n",
    "            #\n",
    "            try:\n",
    "                contig_search_results = search_results[i]\n",
    "            except:\n",
    "                contig_search_results = []\n",
    "            for j,result_list in enumerate(contig_search_results):\n",
    "                mark_color = search_color_names[j % len(search_color_names)]\n",
    "                for result in contig_search_results[j]:\n",
    "                    mid_pos = 0.5 * (result['beg_pos'] + result['end_pos'])\n",
    "                    mark_h = 1.25 * contig_h\n",
    "                    mark_w = 0.01\n",
    "                    mark_margin = 0.02 * mark_h\n",
    "                    disp_mark_x0 = x0 + w * (1-(contig_len-mid_pos)/contig_len) - 0.5*mark_w\n",
    "                    disp_mark_y0 = y0 + contig_h + mark_margin\n",
    "                    search_feature_rect = Rectangle((disp_mark_x0,disp_mark_y0), mark_w, mark_h,\n",
    "                                                    facecolor=mark_color, edgecolor=\"none\", alpha=1.0, zorder=3) \n",
    "                    ax.add_patch (search_feature_rect)\n",
    "                    \n",
    "            # Indicate window viewed by track panel\n",
    "            #\n",
    "            contig_mode_xshift = 0.0\n",
    "            if Global_State['genomebrowser_mode'] == 'contigs':\n",
    "                    #contig_mode_xshift = 0.5*Global_State['genomebrowser_window_bp_width'] - 0.5*(Feature_slices[i][0]['end_pos']-Feature_slices[i][0]['beg_pos'])\n",
    "                contig_mode_xshift = 0.5*Global_State['genomebrowser_window_bp_width']\n",
    "                    \n",
    "            #if (pivot_pos+Global_State['genomebrowser_xshift']) >= -.0000001 \\\n",
    "            #    and (pivot_pos+Global_State['genomebrowser_xshift']) <= (contig_len + .0000001):\n",
    "            track_window_h = 1.6 * contig_h\n",
    "            track_window_w = w * Global_State['genomebrowser_window_bp_width'] / contig_len\n",
    "            track_window_x0 = x0 - 0.5*track_window_w + w * ((pivot_pos+Global_State['genomebrowser_xshift']+contig_mode_xshift) / contig_len)\n",
    "            track_window_y0 = y0 + 0.5*h - 0.5*track_window_h\n",
    "            if track_window_x0 > x0+w or track_window_x0 + track_window_w < x0:\n",
    "                continue                            \n",
    "            track_window_rect = Rectangle((track_window_x0,track_window_y0), track_window_w, track_window_h, \\\n",
    "                                              facecolor=\"red\", edgecolor=\"red\", alpha=0.4, zorder=4) \n",
    "            ax.add_patch (track_window_rect)\n",
    "\n",
    "                \n",
    "    # Circle representation\n",
    "    #\n",
    "    elif genomebrowser_mode == \"genome\":\n",
    "\n",
    "        # set up circle genome representation\n",
    "        contig_label_fontsize = 12\n",
    "        contig_label_y_margin = 0.5 / (figure_height_scaling*(total_rows+1))\n",
    "        contig_color = \"steelblue\"\n",
    "        diameter = 0.75\n",
    "        clickable_diameter = 1.25 * diameter\n",
    "        genome_lw = 10\n",
    "        marker_lw = 15\n",
    "        track_window_lw = 20\n",
    "        marker_len = 0.1\n",
    "        GC_window_bp = 5000\n",
    "        GC_bar_high_lw = 20\n",
    "        GC_bar_high_perc = 0.20\n",
    "        GC_base_diameter = 0.55\n",
    "        GC_lw_to_coord_scale = 0.01\n",
    "        ellipse_center_x = 0.50\n",
    "        #ellipse_center_y = 0.70\n",
    "        ellipse_center_y = (total_rows+1 - 2.5) / (total_rows+1)\n",
    "        ellipse_center = (ellipse_center_x, ellipse_center_y)\n",
    "        ellipse_to_circle_scaling = 0.25*figure_width / (figure_height_scaling*(total_rows+1))        \n",
    "\n",
    "        \n",
    "        # Add clickable base\n",
    "        #\n",
    "        x_diameter = 1.0 * clickable_diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * clickable_diameter\n",
    "        #Ellipse(xy, width, height, angle=0.0, **kwargs)\n",
    "        clickable_contig_image = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                          facecolor=\"none\", edgecolor=\"none\", alpha=0.0, zorder=0,\n",
    "                                          picker=True)\n",
    "        clickable_contig_image.set_label(\"circular_contig_nav\")\n",
    "        clickable_contig_image.gid = \"%d,%f,%f,%f\"%(0,ellipse_center_x,ellipse_center_y,ellipse_to_circle_scaling)\n",
    "        ax.add_patch(clickable_contig_image)\n",
    "        clickable_contig_image.figure.canvas.mpl_connect ('pick_event', onpick_contig_nav)\n",
    "        \n",
    "        # Draw circle genome representation\n",
    "        #\n",
    "        x_diameter = 1.0 * diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * diameter\n",
    "        #Ellipse(xy, width, height, angle=0.0, **kwargs)\n",
    "        contig_image = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                facecolor=\"white\", edgecolor=contig_color, lw=genome_lw, alpha=1.0, zorder=1,\n",
    "                               picker=False)\n",
    "        ax.add_patch(contig_image)\n",
    "\n",
    "        # Add label\n",
    "        #\n",
    "        disp_name = ContigSet_names[0]\n",
    "        scaffold_id = None\n",
    "        if KBase_backend:\n",
    "            #full_name_split = ContigSet_names[0].split('/')\n",
    "            #[genome_id, scaffold_id] = full_name_split[1].split(genome_contig_id_delim)\n",
    "            [genome_id, scaffold_id] = ContigSet_names[i].split(genome_contig_id_delim)\n",
    "            scaffold_id_disp = scaffold_id.split('.')[-1]\n",
    "            disp_name = Species_name_by_genome_ref[genome_id] + ' c:'+scaffold_id_disp\n",
    "        ax.text(ellipse_center_x, ellipse_center_y - 0.5*y_diameter - contig_label_y_margin,\n",
    "                disp_name, verticalalignment=\"top\", horizontalalignment=\"center\",\n",
    "                color=\"black\", fontsize=contig_label_fontsize)\n",
    "        if scaffold_id != None:\n",
    "            ax.text(ellipse_center_x, ellipse_center_y - 0.5*y_diameter - 1.5*contig_label_y_margin,\n",
    "                    scaffold_id, verticalalignment=\"top\", horizontalalignment=\"center\",\n",
    "                    color=\"black\", fontsize=contig_label_fontsize)            \n",
    "        \n",
    "        # Mark contig edge\n",
    "        #\n",
    "        mark_width = 1.0\n",
    "        arc_end = 90 + 0.5*mark_width\n",
    "        arc_beg = 90 - 0.5*mark_width\n",
    "        contig_edge_arc = Arc (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=\"white\", lw=genome_lw, alpha=1.0, zorder=2)\n",
    "        ax.add_patch (contig_edge_arc)        \n",
    "\n",
    "        # Mark pivot\n",
    "        #\n",
    "        mark_width = 1.0\n",
    "        mark_pad = 0.125\n",
    "        mark_x_diameter = x_diameter + 1.0 * mark_pad\n",
    "        mark_y_diameter = y_diameter + ellipse_to_circle_scaling * mark_pad\n",
    "        # pivot feature\n",
    "        arc_end = 90 + 0.5*mark_width - 360 * (Global_State['PrimaryAnchor_pivot_pos'] / Global_State['PrimaryContig_len'])\n",
    "        arc_beg = 90 - 0.5*mark_width - 360 * (Global_State['PrimaryAnchor_pivot_pos'] / Global_State['PrimaryContig_len'])\n",
    "        pivot_feature_arc = Arc (ellipse_center, mark_x_diameter, mark_y_diameter, \\\n",
    "                                    theta1=arc_beg, theta2=arc_end, \\\n",
    "                                    edgecolor=\"black\", lw=marker_lw, alpha=1.0, zorder=3)\n",
    "        ax.add_patch (pivot_feature_arc)\n",
    "        \n",
    "        # Mark search features\n",
    "        #\n",
    "        try:\n",
    "            contig_search_results = search_results[0]\n",
    "        except:\n",
    "            contig_search_results = []\n",
    "        for j,result_list in enumerate(contig_search_results):\n",
    "            mark_color = search_color_names[j % len(search_color_names)]\n",
    "            for result in contig_search_results[j]:\n",
    "                mid_pos = 0.5 * (result['beg_pos'] + result['end_pos'])\n",
    "                arc_end = 90 + 0.5*mark_width - 360 * (mid_pos / Global_State['PrimaryContig_len'])\n",
    "                arc_beg = 90 - 0.5*mark_width - 360 * (mid_pos / Global_State['PrimaryContig_len'])\n",
    "                search_result_arc = Arc (ellipse_center, mark_x_diameter, mark_y_diameter, \\\n",
    "                                    theta1=arc_beg, theta2=arc_end, \\\n",
    "                                    edgecolor=mark_color, lw=marker_lw, alpha=1.0, zorder=4)\n",
    "                ax.add_patch (search_result_arc)\n",
    "        \n",
    "        # Indicate window viewed by track panel\n",
    "        #\n",
    "        track_extra = 0.5*(total_rows-1)*Global_State['genomebrowser_window_bp_width']\n",
    "        arc_end = 90 - 360 * ((Global_State['PrimaryAnchor_pivot_pos']+Global_State['genomebrowser_xshift']-track_extra-0.5*Global_State['genomebrowser_window_bp_width']) / Global_State['PrimaryContig_len'])\n",
    "        arc_beg = 90 - 360 * ((Global_State['PrimaryAnchor_pivot_pos']+Global_State['genomebrowser_xshift']+track_extra+0.5*Global_State['genomebrowser_window_bp_width']) / Global_State['PrimaryContig_len']) \n",
    "        if arc_end > 90:\n",
    "            arc_end = 90 - 0.5\n",
    "        if arc_beg < -270:\n",
    "            arc_beg = -270 + 0.5\n",
    "                \n",
    "        track_window_arc = Arc (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=\"red\", lw=track_window_lw, alpha=0.40, zorder=5)  # facecolor does nothing (no fill for Arc)\n",
    "        ax.add_patch (track_window_arc)\n",
    "        \n",
    "        # Add GC\n",
    "        #       \n",
    "        GC_bar_scale = GC_bar_high_lw / GC_bar_high_perc\n",
    "        GC_steps = int(round(Global_State['PrimaryContig_len'] / GC_window_bp))\n",
    "        for GC_i in range(1,GC_steps-2):\n",
    "            arc_end = 90 - 360 * (GC_i*GC_window_bp / Global_State['PrimaryContig_len'])\n",
    "            arc_beg = 90 - 360 * ((GC_i+1)*GC_window_bp / Global_State['PrimaryContig_len'])\n",
    "            GC_in_window = compute_GC (Global_Genbank_Genomes[0].seq[(GC_i-1)*GC_window_bp:(GC_i+2)*GC_window_bp])\n",
    "            #GC_in_window = 0.5  # DEBUG\n",
    "            GC_delta = GC_in_window - Global_State['PrimaryContig_GCavg']\n",
    "            if (GC_delta > 0):\n",
    "                GC_color = \"blue\"\n",
    "                GC_bar_lw = int(round(GC_bar_scale * GC_delta))\n",
    "                GC_bar_diameter = GC_base_diameter + 0.5*GC_bar_lw*GC_lw_to_coord_scale\n",
    "                GC_x_diameter = 1.0 * GC_bar_diameter\n",
    "                GC_y_diameter = ellipse_to_circle_scaling * GC_bar_diameter\n",
    "                GC_arc = Arc (ellipse_center, GC_x_diameter, GC_y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=GC_color, lw=GC_bar_lw, alpha=1.0, zorder=1)  # facecolor does nothing (no fill for Arc)\n",
    "                ax.add_patch (GC_arc)\n",
    "            else:\n",
    "                GC_color = \"red\"\n",
    "                GC_bar_lw = int(round(GC_bar_scale * -GC_delta))\n",
    "                GC_bar_diameter = GC_base_diameter - 0.5*GC_bar_lw*GC_lw_to_coord_scale\n",
    "                GC_x_diameter = 1.0 * GC_bar_diameter\n",
    "                GC_y_diameter = ellipse_to_circle_scaling * GC_bar_diameter\n",
    "                GC_arc = Arc (ellipse_center, GC_x_diameter, GC_y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=GC_color, lw=GC_bar_lw, alpha=1.0, zorder=1)  # facecolor does nothing (no fill for Arc)\n",
    "                ax.add_patch (GC_arc)\n",
    "\n",
    "                \n",
    "def draw_genomebrowser_panel (ax, \\\n",
    "                               ContigSet_names, \\\n",
    "                               PivotFeatures_IDs, \\\n",
    "                               genomebrowser_window_bp_width, \\\n",
    "                               genomebrowser_xshift, \\\n",
    "                               genomebrowser_mode):                  \n",
    "    \n",
    "    # Read Domains\n",
    "    #\n",
    "    if Global_State['genomebrowser_color_namespace'] == 'domains' \\\n",
    "        or Global_State['genomebrowser_color_namespace'] == 'cog' \\\n",
    "        or Global_State['genomebrowser_color_namespace'] == 'pfam':\n",
    "\n",
    "        # update Global_Domains, but really should be method of Domains object\n",
    "        getDomainHits (ContigSet_names, \\\n",
    "                       genomebrowser_mode=genomebrowser_mode, \\\n",
    "                       domain_data_format=domain_data_format)\n",
    "        \n",
    "    \n",
    "    # Read Features\n",
    "    #\n",
    "    if KBase_backend:\n",
    "        Feature_slices = getFeatureSlicesKBase (ContigSet_names, \\\n",
    "                                                    PivotFeatures_IDs, \\\n",
    "                                                    genomebrowser_mode=genomebrowser_mode, \\\n",
    "                                                    genome_data_format=genome_data_format, \\\n",
    "                                                    window_size=genomebrowser_window_bp_width, \\\n",
    "                                                    genomebrowser_xshift=genomebrowser_xshift)\n",
    "    elif genome_data_format == 'Genbank':\n",
    "        Feature_slices = getFeatureSlicesGenbank (ContigSet_names, \\\n",
    "                                                    PivotFeatures_IDs, \\\n",
    "                                                    genomebrowser_mode=genomebrowser_mode, \\\n",
    "                                                    genome_data_format=genome_data_format, \\\n",
    "                                                    window_size=genomebrowser_window_bp_width, \\\n",
    "                                                    genomebrowser_xshift=genomebrowser_xshift)\n",
    "    \n",
    "\n",
    "    # Add features elements to figure\n",
    "    #\n",
    "    if genomebrowser_mode == \"contigs\" \\\n",
    "        or genomebrowser_mode == \"genome\" \\\n",
    "        or genomebrowser_mode == \"homologs\" \\\n",
    "        or genomebrowser_mode == \"tree\":\n",
    "\n",
    "        # Determine repeats\n",
    "        #\n",
    "        annot_repeat = set()\n",
    "        annot_seen   = set()\n",
    "        seq_repeat   = set()\n",
    "        seq_seen     = set()\n",
    "        for i in range(0,len(Feature_slices)):\n",
    "            if i >= max_rows:\n",
    "                break\n",
    "                \n",
    "            for j in range(1,len(Feature_slices[i])):  # skip [0]. it's the pivot and if in window will be repeated\n",
    "                if Feature_slices[i][j]['type'] == \"CDS\":\n",
    "                    if Global_State['genomebrowser_color_namespace'] == 'annot' \\\n",
    "                        and 'annot' in Feature_slices[i][j] \\\n",
    "                        and Feature_slices[i][j]['annot'] != '':\n",
    "                        if Feature_slices[i][j]['annot'] in annot_seen:\n",
    "                            annot_repeat.add(Feature_slices[i][j]['annot'])\n",
    "                        else:\n",
    "                            annot_seen.add(Feature_slices[i][j]['annot'])\n",
    "                            \n",
    "                    elif Global_State['genomebrowser_color_namespace'] == 'ec' \\\n",
    "                        and 'EC_number' in Feature_slices[i][j] \\\n",
    "                        and Feature_slices[i][j]['EC_number'] != '':\n",
    "                        if Feature_slices[i][j]['EC_number'] in annot_seen:\n",
    "                            annot_repeat.add(Feature_slices[i][j]['EC_number'])\n",
    "                        else:\n",
    "                            annot_seen.add(Feature_slices[i][j]['EC_number'])\n",
    "                            \n",
    "                    elif Global_State['genomebrowser_color_namespace'] == 'cog' \\\n",
    "                        or Global_State['genomebrowser_color_namespace'] == 'pfam' \\\n",
    "                        or Global_State['genomebrowser_color_namespace'] == 'domains':\n",
    "                            \n",
    "                        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "                            contig_i = 0\n",
    "                        else:\n",
    "                            contig_i = i\n",
    "                        try:\n",
    "                            domain_hits = Global_Domains[Genome_ref_by_Contig_name[ContigSet_names[contig_i]]][ContigSet_names[contig_i]][Feature_slices[i][j]['ID']]\n",
    "                        except:\n",
    "                            domain_hits = []\n",
    "                        for domhit in domain_hits:\n",
    "                            domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "                            if domfam in annot_seen:\n",
    "                                annot_repeat.add(domfam)\n",
    "                            else:\n",
    "                                annot_seen.add(domfam)\n",
    "                            \n",
    "                elif (Feature_slices[i][j]['type'] == \"CRISPR\" or Feature_slices[i][j]['type'] == \"CRISPR spacer\") \\\n",
    "                    and 'dna_seq' in Feature_slices[i][j]:\n",
    "                    if Feature_slices[i][j]['dna_seq'] in seq_seen:\n",
    "                        seq_repeat.add(Feature_slices[i][j]['dna_seq'])\n",
    "                    else:\n",
    "                        seq_seen.add(Feature_slices[i][j]['dna_seq'])\n",
    "                        \n",
    "\n",
    "        # Draw feature elements\n",
    "        #\n",
    "        for i in range(0,len(Feature_slices)):\n",
    "            if i >= max_rows:\n",
    "                break\n",
    "                \n",
    "            pivot_pos = 0.5 * (Feature_slices[i][0]['beg_pos']+Feature_slices[i][0]['end_pos'])\n",
    "            track_xshift = 0\n",
    "            if genomebrowser_mode == \"genome\":\n",
    "                #track_xshift = genomebrowser_window_bp_width * (i - (total_rows-1)/2)  # e.g. total_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                track_xshift = genomebrowser_window_bp_width * (i - (Global_State['genome_mode_n_rows']-1)/2)  # e.g. n_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                #print (\"%d %f\"%(i,track_xshift))\n",
    "            contig_mode_xshift = 0\n",
    "            if genomebrowser_mode == \"contigs\":\n",
    "                contig_mode_xshift = 0.5*genomebrowser_window_bp_width - 0.5*(Feature_slices[i][0]['end_pos']-Feature_slices[i][0]['beg_pos'])\n",
    "            prev_top_beg_pos = 0\n",
    "            prev_bot_beg_pos = 0\n",
    "            prev_top_name = ''\n",
    "            prev_bot_name = ''\n",
    "            \n",
    "            for j in range(1,len(Feature_slices[i])):  # skip [0]. it's the pivot and if in window will be repeated\n",
    "                pivot_feature_flag = False\n",
    "                if Feature_slices[i][j]['beg_pos'] == Feature_slices[i][0]['beg_pos'] and \\\n",
    "                   Feature_slices[i][j]['end_pos'] == Feature_slices[i][0]['end_pos']:\n",
    "                    pivot_feature_flag = True\n",
    "                \n",
    "                # show label on top or bottom, if it fits\n",
    "                this_label_show_top = True\n",
    "                this_label_show_bot = False\n",
    "                this_beg_pos = disp_coord_transform (Feature_slices[i][j]['beg_pos'], \\\n",
    "                                                     pivot_pos, \\\n",
    "                                                     genomebrowser_window_bp_width, \\\n",
    "                                                     genomebrowser_xshift, \\\n",
    "                                                     track_xshift, \\\n",
    "                                                     contig_mode_xshift)\n",
    "                #if j == 1:\n",
    "                if this_beg_pos < left_margin:\n",
    "                    this_beg_pos = left_margin\n",
    "                elif len(prev_top_name) > arrow_label_scaling*(this_beg_pos-prev_top_beg_pos):\n",
    "                    this_label_show_top = False\n",
    "                if this_label_show_top == False and len(prev_bot_name) <= arrow_label_scaling*(this_beg_pos-prev_bot_beg_pos):\n",
    "                    this_label_show_bot = True\n",
    "\n",
    "                # FIX\n",
    "                max_row_n = total_rows  \n",
    "                if num_genomes < total_rows:\n",
    "                    max_row_n = num_genomes\n",
    "                    \n",
    "                # finally draw arrow!  (or other shape)\n",
    "                draw_feature_element (ax, \\\n",
    "                                 i+1, \\\n",
    "                                 max_row_n, \\\n",
    "                                 i, \\\n",
    "                                 j, \\\n",
    "                                 Feature_slices[i][j], \\\n",
    "                                 pivot_pos, \\\n",
    "                                 track_xshift, \\\n",
    "                                 contig_mode_xshift, \\\n",
    "                                 pivot_feature_flag, \\\n",
    "                                 annot_repeat, \\\n",
    "                                 seq_repeat, \\\n",
    "                                 this_label_show_top, this_label_show_bot)\n",
    "\n",
    "                if this_label_show_top:\n",
    "                    prev_top_beg_pos = this_beg_pos\n",
    "                    prev_top_name = Feature_slices[i][j]['name']\n",
    "                if this_label_show_bot:\n",
    "                    prev_bot_beg_pos = this_beg_pos\n",
    "                    prev_bot_name = Feature_slices[i][j]['name']\n",
    "\n",
    "                    \n",
    "        \n",
    "# Panel updates\n",
    "#\n",
    "def update_mode_panel (ax):\n",
    "    ax.cla()\n",
    "    draw_mode_panel (ax, \\\n",
    "                     Global_State['ContigSet_names'][0], \\\n",
    "                     Global_State['genomebrowser_mode'], \\\n",
    "                     Global_State['Dataset_names_list'][0])    \n",
    "    pyplot.show()\n",
    "\n",
    "def update_control_panel (ax):\n",
    "    ax.cla()\n",
    "    draw_control_panel (ax, \\\n",
    "                        Global_State['genomebrowser_zoom'], \\\n",
    "                        Global_State['genomebrowser_xshift'], \\\n",
    "                        Global_State['genomebrowser_color_namespace'])    \n",
    "    pyplot.show()\n",
    "\n",
    "def update_sidenav_panel (ax):\n",
    "    ax.cla()\n",
    "    draw_sidenav_panel (ax, \\\n",
    "                        Global_State['genomebrowser_mode'])\n",
    "    pyplot.show()\n",
    "    \n",
    "def update_genomebrowser_panel (ax):\n",
    "\n",
    "    # clear panel\n",
    "    ax.cla()\n",
    "    \n",
    "    # Update contig order\n",
    "    #   might be used elsewhere later, likely when track ordering changes\n",
    "    #\n",
    "    Contig_order = []\n",
    "    Contig_order_lookup = {}\n",
    "    for i,contig_name in enumerate(ContigSet_names):\n",
    "        (genome_id,scaffold_id) = contig_name.split(genome_contig_id_delim)        \n",
    "        try:\n",
    "            col = Contig_order_lookup['genome_id']\n",
    "        except:\n",
    "            Contig_order_lookup['genome_id'] = {}\n",
    "        Contig_order_lookup['genome_id']['scaffold_id'] = i\n",
    "        Contig_order.append(contig_name)\n",
    "        \n",
    "    # Reset feature and popup_box storage\n",
    "    #\n",
    "    feature_artist_label_to_feature = {}\n",
    "    feature_artist_label_to_popup_box = {}\n",
    "    Global_State['popup_zorder'] = def_popup_zorder\n",
    "    \n",
    "    # Add Genome tracks\n",
    "    #\n",
    "    draw_genomebrowser_panel (ax, \\\n",
    "                              ContigSet_names = Global_State['ContigSet_names'], \\\n",
    "                              PivotFeatures_IDs = Global_State['PivotFeatures_IDs'], \\\n",
    "                              genomebrowser_window_bp_width = Global_State['genomebrowser_window_bp_width'], \\\n",
    "                              genomebrowser_xshift = Global_State['genomebrowser_xshift'], \\\n",
    "                              genomebrowser_mode = Global_State['genomebrowser_mode'])\n",
    "    pyplot.show()\n",
    "    \n",
    "\n",
    "###############################################################################                \n",
    "# Main\n",
    "###############################################################################\n",
    "    \n",
    "# Instantiate fig_KGB\n",
    "#   \n",
    "fig_KGB = pyplot.figure(1, (figure_width, top_nav_height + figure_height_scaling*(total_rows+1)))\n",
    "#fig_KGB.suptitle(tool_title, fontsize=20)\n",
    "\n",
    "# want to maintain constant top nav height, so adjust proportion according to number of rows\n",
    "# e.g. with 6 tracks, top nav should be 2 units, so top rowspan=2 when bottom rowspan = 6 and grid is 8\n",
    "top_nav_rows = int(round(1000*top_nav_height/(top_nav_height+figure_height_scaling*(total_rows+1))))\n",
    "\n",
    "# subplot2grid(shape, loc, rowspan=1, colspan=1)\n",
    "ax_top_left   = pyplot.subplot2grid((1000,4),            (0,0), rowspan=top_nav_rows,        colspan=1)\n",
    "ax_top_center = pyplot.subplot2grid((1000,4),            (0,1), rowspan=top_nav_rows,        colspan=3)\n",
    "ax_left       = pyplot.subplot2grid((1000,4), (top_nav_rows,0), rowspan=(1000-top_nav_rows), colspan=1)\n",
    "ax_center     = pyplot.subplot2grid((1000,4), (top_nav_rows,1), rowspan=(1000-top_nav_rows), colspan=3)\n",
    "\n",
    "\n",
    "# Let's turn off visibility of all tick labels and boxes here\n",
    "for ax in fig_KGB.axes:\n",
    "    ax.xaxis.set_visible(False)  # remove axis labels and ticks\n",
    "    ax.yaxis.set_visible(False)\n",
    "    for t in ax.get_xticklabels()+ax.get_yticklabels():  # remove tick labels\n",
    "        t.set_visible(False)\n",
    "#for ax in [ax_top_left, ax_top_center]:\n",
    "#    ax.set_axis_bgcolor(nav_bg_color)\n",
    "for ax in [ax_top_left, ax_top_center, ax_left, ax_center]:\n",
    "    ax.spines['top'].set_visible(False) # Get rid of top axis line\n",
    "    ax.spines['bottom'].set_visible(False) #  bottom axis line\n",
    "    ax.spines['left'].set_visible(False) #  Get rid of bottom axis line\n",
    "    ax.spines['right'].set_visible(False) #  Get rid of bottom axis line\n",
    "\n",
    "fig_KGB.tight_layout()  # left justify and space subplots reasonably.  Must follow axis creation\n",
    "\n",
    "\n",
    "# Load domain family descriptions\n",
    "#\n",
    "#domain_family_desc = read_domain_family_desc ()  # these need to be global (kludge)\n",
    "if domain_family_desc_base_path != None:\n",
    "    read_domain_family_desc ()  # these need to be global (kludge)\n",
    "\n",
    "\n",
    "# Draw genome browser panel (must happen first to load Global_State)\n",
    "#\n",
    "feature_artist_label_to_feature = {}     # these need to be global (kludge)\n",
    "feature_artist_label_to_popup_box = {}   # these need to be global (kludge)\n",
    "pos_knob_patch = None                    # these need to be global (kludge)\n",
    "update_genomebrowser_panel (ax_center)\n",
    "    \n",
    "    \n",
    "# Draw sidenav (tree, circles, etc.)\n",
    "#\n",
    "update_sidenav_panel (ax_left)\n",
    "    \n",
    "\n",
    "# Draw mode panel\n",
    "#\n",
    "update_mode_panel (ax_top_left)\n",
    "\n",
    "\n",
    "# Draw control panel (currently, must occur after primary genome read in update_genomebrowser_panel)\n",
    "#\n",
    "update_control_panel (ax_top_center)\n",
    "\n",
    "###############################################################################\n",
    "# End KGB\n",
    "###############################################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
