{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Killer Genome Browser v0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: MacOSX\n",
      "Populating the interactive namespace from numpy and matplotlib\n",
      "reading ./data/plasmid_contigs/GW456_old/domains/GW456_Domain_annot.json ...\n",
      "0 reading ./data/plasmid_contigs/GW456_old/scaffolds/scaffold_130.gbk\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "## Killer Genome Browser (KGB)                                                  \n",
    "##                                                                              \n",
    "##  An IPython/Jupyter Notebook genome browser that enables comparative         \n",
    "##  browsing and searching of genome contig assemblies, with additional         \n",
    "##  support for                                                                 \n",
    "##                                                                              \n",
    "##  * domain structure visualization                                            \n",
    "##  * gene homology                                                             \n",
    "##  * phylogenetic trees                                                        \n",
    "##                                                                              \n",
    "## Copyright 2015,2016 Dylan Chivian  \n",
    "##\n",
    "##  Initial Author: Dylan Chivian (DCChivian@lbl.gov)\n",
    "##  $Revision: 0.1 $\n",
    "##  $Date: 2015/01/01 00:00:00 $\n",
    "##  $Author: dylan $\n",
    "##\n",
    "\"\"\"\n",
    "\n",
    "# Init system\n",
    "#\n",
    "#import numpy as np # comes with pylab\n",
    "from __future__ import print_function\n",
    "from __future__ import division\n",
    "#import sys  # for io and exit\n",
    "from os import walk  # for dir reading\n",
    "from os import path  # for file existence check\n",
    "import json\n",
    "import csv\n",
    "#from math import pi  # get with pylab\n",
    "#from math import sqrt  # get with pylab\n",
    "from math import acos\n",
    "from Bio import Phylo\n",
    "from Bio import SeqIO\n",
    "#import re\n",
    "#%pylab notebook\n",
    "%pylab\n",
    "#matplotlib.use('nbagg')\n",
    "import matplotlib.transforms as mtransforms\n",
    "import matplotlib.patheffects as path_effects\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Ellipse\n",
    "from matplotlib.patches import Arc\n",
    "from matplotlib.patches import FancyBboxPatch\n",
    "#from matplotlib.lines import Line2D\n",
    "\n",
    "#from ipywidgets import interact, interactive, fixed\n",
    "#from ipywidgets import interact\n",
    "#import ipywidgets as widgets\n",
    "\n",
    "\n",
    "# Input variables\n",
    "#\n",
    "\n",
    "#Search_Terms = []\n",
    "Search_Terms = ['pil or conj or mobil or partition or t-dna or vir or plasmid',\n",
    "                'toxin'\n",
    "                #'tra',\n",
    "                #'rep'\n",
    "               ]\n",
    "#Search_Terms = ['dna-directed polymerase',\n",
    "#                '16S',\n",
    "#                'DNA and methyltransferase',\n",
    "#                '1.10.3.-',\n",
    "#                'polymerase or dna-directed',\n",
    "#                'fucI',\n",
    "#                'sulfate adenylyl transferase']\n",
    "\n",
    "GenomeSet_names = [\"GW456\", \"GW460\"]\n",
    "ContigSet_names = []\n",
    "OrthologSet_locusTags = []\n",
    "\n",
    "#ContigSet_names = [\"GW456.scaffold_130\", \"GW456.scaffold_437\", \"GW456.scaffold_206\", \"GW456.scaffold_1786\", \"GW460.scaffold_49\", \"GW460.scaffold_575\"]\n",
    "#OrthologSet_locusTags = [\"GW456_SEED_annot.CDS.1\", \"GW456_SEED_annot.CDS.233\", \"GW456_SEED_annot.CDS.128\", \"GW456_SEED_annot.CDS.112\", \"GW460_SEED_annot.CDS.186\", \"GW460_SEED_annot.CDS.299\"]\n",
    "\n",
    "#ContigSet_names = [\"Gsulf\", \"DvulH\", \"DdesulfG20\", \"EcoliK12\", \"Bsub\", \"DaudaxMP104C\"]\n",
    "#OrthologSet_locusTags = [\"GSU2863\", \"DVU2928\", \"Dde_2997\", \"b3987\", \"BSU01070\", \"Daud0216\"]\n",
    "#ContigSet_names = [\"Gsulf\", \"DvulH\", \"DdesulfG20\", \"EcoliK12\", \"Bsub\", \"DaudaxMP104C\", \"Gsulf\", \"DvulH\", \"DdesulfG20\", \"EcoliK12\", \"Bsub\", \"DaudaxMP104C\"]\n",
    "#OrthologSet_locusTags = [\"GSU2863\", \"DVU2928\", \"Dde_2997\", \"b3987\", \"BSU01070\", \"Daud0216\", \"GSU2863\", \"DVU2928\", \"Dde_2997\", \"b3987\", \"BSU01070\", \"Daud0216\"]\n",
    "#ContigSet_names = [\"Gsulf\", \"DvulH\"]\n",
    "#ContigSet_names = [\"Gsulf\", \"DvulH\"]\n",
    "#OrthologSet_locusTags = [\"GSU2863\", \"DVU2928\"]\n",
    "\n",
    "PrimaryAnchor_leafId = \"GW456.scaffold_130\"\n",
    "PrimaryAnchor_locusTag = \"GW456_SEED_annot.CDS.1\"\n",
    "\n",
    "genome_data_base_path = \"./data/plasmid_contigs\"\n",
    "genome_data_extra_subpath = \"_old\"\n",
    "#genome_data_extra_subpath = \"_circular_2KbpMIN\"\n",
    "domain_data_base_path = \"./data/plasmid_contigs\"\n",
    "domain_family_desc_base_path = \"./data/domain_desc\"\n",
    "\n",
    "tree_data_base_path = './data'\n",
    "tree_data_file = 'rpoB_tree-names.newick'\n",
    "tree_data_format = 'newick'\n",
    "\n",
    "genome_data_format = \"Genbank\"\n",
    "domain_data_format = \"KBase_domains\"\n",
    "\n",
    "# build ContigSet_names from files\n",
    "if genome_data_format == \"Genbank\":\n",
    "    for genome_id in GenomeSet_names:\n",
    "        genome_data_dir = genome_data_base_path+'/'+genome_id+genome_data_extra_subpath+'/scaffolds'\n",
    "        files = []\n",
    "        for (dirpath, dirnames, filenames) in walk(genome_data_dir):\n",
    "            files.extend(filenames)\n",
    "            break\n",
    "        for file in files:\n",
    "            if \".gbk\" in file:\n",
    "                scaffold_id = file[0:file.index(\".gbk\")]\n",
    "                contig_id = genome_id+'.'+scaffold_id\n",
    "                ContigSet_names.append(contig_id)\n",
    "\n",
    "\n",
    "# Init variables / objects\n",
    "#\n",
    "tool_title = \"Flexible Genome Browser\"\n",
    "\n",
    "#color_namespace_names_disp = ['Annot', 'EC', 'COG', 'Pfam', 'Domains', 'Local']\n",
    "#color_namespace_names = ['annot', 'ec', 'cog', 'pfam', 'domains', 'local']\n",
    "color_namespace_names_disp = ['Annot']\n",
    "color_namespace_names = ['annot']\n",
    "if domain_data_base_path != None:\n",
    "    color_namespace_names_disp.extend(['COG', 'Pfam', 'Domains'])\n",
    "    color_namespace_names.extend(['cog', 'pfam', 'domains'])\n",
    "    \n",
    "#mode_names_disp = ['Contigs', 'Genome', 'Homologs', 'Tree', 'Strains']\n",
    "#mode_names = ['contigs', 'genome', 'homologs', 'tree', strains']\n",
    "mode_names_disp = ['Contigs', 'Genome']\n",
    "mode_names = ['contigs', 'genome']    \n",
    "if len(OrthologSet_locusTags) != 0:\n",
    "    mode_names_disp.append('Homologs')\n",
    "    mode_names.append('homologs')\n",
    "if tree_data_file != None:\n",
    "    mode_names_disp.append('Tree')\n",
    "    mode_names.append('tree')\n",
    "\n",
    "\n",
    "# Configuration\n",
    "#\n",
    "num_genomes = len(ContigSet_names)\n",
    "#def_genome_mode_n_rows = 5\n",
    "def_genome_mode_n_rows = 7\n",
    "max_rows = 100\n",
    "#max_rows = 10  # DEBUG\n",
    "if max_rows < def_genome_mode_n_rows:\n",
    "    max_rows = def_genome_mode_n_rows\n",
    "def_popup_zorder = 100\n",
    "def_genomebrowser_window_bp_width = 1000\n",
    "def_genomebrowser_zoom_tics = 7\n",
    "def_genomebrowser_zoom = 4  # zoom values are [0..zoom_tics] -> window width = is def_genomebrowser_window_bp_width * 2**i\n",
    "def_genomebrowser_xshift = 0\n",
    "#def_genomebrowser_mode = \"tree\"\n",
    "#def_genomebrowser_mode = \"homologs\"\n",
    "def_genomebrowser_mode = \"genome\"\n",
    "#def_genomebrowser_mode = \"contigs\"\n",
    "#def_genomebrowser_color_namespace = \"annot\"\n",
    "def_genomebrowser_color_namespace = \"domains\"\n",
    "genomebrowser_mode = def_genomebrowser_mode\n",
    "genomebrowser_window_bp_width = def_genomebrowser_window_bp_width * 2**def_genomebrowser_zoom\n",
    "genomebrowser_color_namespace = def_genomebrowser_color_namespace\n",
    "if def_genome_mode_n_rows%2 == 0:\n",
    "    def_genome_mode_n_rows += 1\n",
    "KBASE_DOMAINHIT_GENE_ID_I        = 0\n",
    "#KBASE_DOMAINHIT_GENE_BEG_I       = 1\n",
    "#KBASE_DOMAINHIT_GENE_END_I       = 2\n",
    "#KBASE_DOMAINHIT_GENE_STRAND_I    = 3\n",
    "KBASE_DOMAINHIT_GENE_HITS_DICT_I = 4\n",
    "#KBASE_DOMAINHIT_GENE_HITS_DICT_BEG_J      = 0\n",
    "#KBASE_DOMAINHIT_GENE_HITS_DICT_END_J      = 1\n",
    "#KBASE_DOMAINHIT_GENE_HITS_DICT_EVALUE_J   = 2\n",
    "#KBASE_DOMAINHIT_GENE_HITS_DICT_BITSCORE_J = 3\n",
    "#KBASE_DOMAINHIT_GENE_HITS_DICT_ALNPERC_J  = 4\n",
    "DOMHIT_BEG_I                              = 0\n",
    "DOMHIT_END_I                              = 1\n",
    "DOMHIT_EVALUE_I                           = 2\n",
    "DOMHIT_BITSCORE_I                         = 3\n",
    "DOMHIT_ALNPERC_I                          = 4 \n",
    "DOMHIT_DOMFAM_I                           = 5\n",
    "\n",
    "\n",
    "\n",
    "# Caches and State\n",
    "#\n",
    "search_results = []\n",
    "search_done = []\n",
    "Contig_order = []\n",
    "Contig_order_lookup = {}\n",
    "Global_Genbank_Genomes = []\n",
    "#Global_Features = []\n",
    "domain_family_desc = {}\n",
    "Global_Domains = []\n",
    "Global_State = { \\\n",
    "                \"ContigSet_names\":                   ContigSet_names, \\\n",
    "                \"OrthologSet_locusTags\":             OrthologSet_locusTags, \\\n",
    "                \"Dataset_names_list\":                [\"none\"], \\\n",
    "                \"Dataset_species_list\":              [], \\\n",
    "                \"PrimaryContig_GCavg\":               0.0, \\\n",
    "                \"PrimaryContig_len\":                 0, \\\n",
    "                \"Contig_lens\":                       [], \\\n",
    "                \"PrimaryAnchor_leafId\":              PrimaryAnchor_leafId, \\\n",
    "                \"PrimaryAnchor_pivot_pos\":           0.0, \\\n",
    "                \"pivot_pos_list\":                    [], \\\n",
    "                \"genome_mode_n_rows\":                def_genome_mode_n_rows, \\\n",
    "                \"popup_zorder\":                      def_popup_zorder, \\\n",
    "                \"genomebrowser_pivot_pos\":           0.0, \\\n",
    "                \"def_genomebrowser_window_bp_width\": def_genomebrowser_window_bp_width, \\\n",
    "                \"genomebrowser_window_bp_width\":     def_genomebrowser_window_bp_width * 2**def_genomebrowser_zoom, \\\n",
    "                \"genomebrowser_zoom\":                def_genomebrowser_zoom, \\\n",
    "                \"genomebrowser_xshift\":              def_genomebrowser_xshift, \\\n",
    "                \"genomebrowser_mode\":                def_genomebrowser_mode, \\\n",
    "                \"genomebrowser_color_namespace\":     def_genomebrowser_color_namespace\n",
    "               }\n",
    "\n",
    "\n",
    "# Set up canvas and arrow config\n",
    "#\n",
    "total_rows = num_genomes\n",
    "if def_genome_mode_n_rows > total_rows:\n",
    "    total_rows = def_genome_mode_n_rows\n",
    "if total_rows > max_rows:\n",
    "    total_rows = max_rows\n",
    "figure_width = 12.0\n",
    "figure_height_scaling = 0.75\n",
    "top_nav_height = 1.5\n",
    "#top_margin = 1.0/total_rows\n",
    "top_margin = 0.0\n",
    "left_margin = 0.05\n",
    "right_margin = 0.05\n",
    "row_delta = 1.0/total_rows\n",
    "#row_delta = 1.0/(total_rows+1)\n",
    "bottom_margin=0.75*row_delta\n",
    "arrow_w = 0.25/total_rows\n",
    "head_w = arrow_w\n",
    "#head_l = 0.4*head_w\n",
    "head_l = 0.015\n",
    "base_arrow_label_fontsize = 10\n",
    "foi_arrow_label_fontsize = 13\n",
    "text_padding = 0.03/total_rows\n",
    "text_yshift = text_padding+0.5*arrow_w\n",
    "#arrow_label_scaling = 125.0  # about appropriate value for fontsize 12 and figure width 15 (or maybe 20)\n",
    "#arrow_label_scaling = 95.0  # appropriate value for fontsize 10 and figure width 12\n",
    "arrow_label_scaling = 100.0  # appropriate value for fontsize 10 and figure width 12\n",
    "text_disp_window_bp_limit = 80000\n",
    "\n",
    "\n",
    "# Colors\n",
    "#\n",
    "nav_bg_color = \"slategray\"\n",
    "popup_bg_color = \"aliceblue\"\n",
    "#popup_bg_color = \"cornsilk\"\n",
    "#popup_bg_color = \"lemonchiffon\"\n",
    "search_color_names = [\n",
    "    'magenta', \\\n",
    "    'orange', \\\n",
    "    'darkcyan', \\\n",
    "    'green', \\\n",
    "    'red', \\\n",
    "    'lightskyblue', \\\n",
    "    'crimson'\n",
    "]\n",
    "# feature colors names list (left out excessively light colors)\n",
    "color_names = [\n",
    "    #'aliceblue',\n",
    "    'aqua',\n",
    "    'aquamarine',\n",
    "    #'azure',\n",
    "    #'beige',\n",
    "    #'bisque',\n",
    "    #'blanchedalmond',\n",
    "    'blue',\n",
    "    'blueviolet',\n",
    "    'brown',\n",
    "    'burlywood',\n",
    "    'cadetblue',\n",
    "    'chartreuse',\n",
    "    'chocolate',\n",
    "    'coral',\n",
    "    'cornflowerblue',\n",
    "    #'cornsilk',\n",
    "    'crimson',\n",
    "    'cyan',\n",
    "    'darkblue',\n",
    "    'darkcyan',\n",
    "    'darkgoldenrod',\n",
    "    'darkgreen',\n",
    "    'darkkhaki',\n",
    "    'darkmagenta',\n",
    "    'darkolivegreen',\n",
    "    'darkorange',\n",
    "    'darkorchid',\n",
    "    'darkred',\n",
    "    'darksalmon',\n",
    "    'darkseagreen',\n",
    "    'darkslateblue',\n",
    "    #'darkslategray',\n",
    "    'darkturquoise',\n",
    "    'darkviolet',\n",
    "    'deeppink',\n",
    "    'deepskyblue',\n",
    "    'dodgerblue',\n",
    "    'firebrick',\n",
    "    'forestgreen',\n",
    "    'fuchsia',\n",
    "    #'gainsboro',\n",
    "    'gold',\n",
    "    'goldenrod',\n",
    "    'green',\n",
    "    'greenyellow',\n",
    "    #'honeydew',\n",
    "    'hotpink',\n",
    "    'indianred',\n",
    "    'indigo',\n",
    "    'khaki',\n",
    "    #'lavender',\n",
    "    #'lavenderblush',\n",
    "    'lawngreen',\n",
    "    #'lemonchiffon',\n",
    "    'lightblue',\n",
    "    #'lightcoral',\n",
    "    #'lightcyan'\n",
    "    #'lightgoldenrodyellow',\n",
    "    'lightgreen',\n",
    "    'lightpink',\n",
    "    'lightsalmon',\n",
    "    'lightseagreen',\n",
    "    'lightskyblue',\n",
    "    #'lightslategray',\n",
    "    #'lightsteelblue',\n",
    "    #'lightyellow',\n",
    "    'lime',\n",
    "    'limegreen',\n",
    "    'magenta',\n",
    "    'maroon',\n",
    "    'mediumaquamarine',\n",
    "    'mediumblue',\n",
    "    'mediumorchid',\n",
    "    'mediumpurple',\n",
    "    'mediumseagreen',\n",
    "    'mediumslateblue',\n",
    "    'mediumspringgreen',\n",
    "    'mediumturquoise',\n",
    "    'mediumvioletred',\n",
    "    'midnightblue',\n",
    "    #'mintcream',\n",
    "    #'mistyrose',\n",
    "    #'moccasin',\n",
    "    'navy',\n",
    "    #'oldlace',\n",
    "    'olive',\n",
    "    'olivedrab',\n",
    "    'orange',\n",
    "    'orangered',\n",
    "    'orchid',\n",
    "    #'palegoldenrod',\n",
    "    'palegreen',\n",
    "    'paleturquoise',\n",
    "    'palevioletred',\n",
    "    #'papayawhip',\n",
    "    'peachpuff',\n",
    "    #'peru',\n",
    "    'pink',\n",
    "    'plum',\n",
    "    'powderblue',\n",
    "    'purple',\n",
    "    'red',\n",
    "    'rosybrown',\n",
    "    'royalblue',\n",
    "    'saddlebrown',\n",
    "    'salmon',\n",
    "    'sandybrown',\n",
    "    'seagreen',\n",
    "    #'seashell',\n",
    "    'sienna',\n",
    "    'skyblue',\n",
    "    'slateblue',\n",
    "    'springgreen',\n",
    "    'steelblue',\n",
    "    #'tan',\n",
    "    'teal',\n",
    "    #'thistle',\n",
    "    'tomato',\n",
    "    'turquoise',\n",
    "    'violet',\n",
    "    #'wheat',\n",
    "    #'yellow',\n",
    "    #'yellowgreen'\n",
    "    ]\n",
    "\n",
    "                \n",
    "##############################################################################################################\n",
    "# Subs\n",
    "##############################################################################################################\n",
    "\n",
    "\n",
    "# Utilities\n",
    "#\n",
    "def compute_GC (dna_seq):\n",
    "    seq_len = len(dna_seq)\n",
    "    GC_cnt = 0\n",
    "    AT_cnt = 0\n",
    "    other_cnt = 0\n",
    "    dna_seq.upper()\n",
    "    GC_cnt = dna_seq.count('G') + dna_seq.count('C')\n",
    "    AT_cnt = dna_seq.count('A') + dna_seq.count('T') + dna_seq.count('U')\n",
    "    other_cnt = seq_len - GC_cnt - AT_cnt\n",
    "    seq_len -= other_cnt\n",
    "    return GC_cnt / seq_len\n",
    "\n",
    "\n",
    "# Features\n",
    "#\n",
    "def build_feature_rec (f, f_type='CDS', source_species='', contig_i=0, fwd_dna_seq=None):\n",
    "    feature_rec = {}\n",
    "    \n",
    "    id_delim = '.'\n",
    "\n",
    "    # IDs and names\n",
    "    if \"gene\" in f.qualifiers:\n",
    "        name = f.qualifiers['gene'][0]\n",
    "    elif \"locus_tag\" in f.qualifiers:\n",
    "        name = f.qualifiers['locus_tag'][0]\n",
    "    elif \"db_xref\" in f.qualifiers:\n",
    "        name = f.qualifiers['db_xref'][0]\n",
    "    else:\n",
    "        print (\"strangely formatted record.  SKIPPING\")\n",
    "        print (f)\n",
    "        return {}\n",
    "    name_split = name.split(id_delim)\n",
    "    name = id_delim.join(name_split[1:])\n",
    "    \n",
    "    if \"locus_tag\" in f.qualifiers:\n",
    "        locus_tag = f.qualifiers['locus_tag'][0]\n",
    "    elif \"gene\" in f.qualifiers:\n",
    "        locus_tag = f.qualifiers['gene'][0]\n",
    "    elif \"db_xref\" in f.qualifiers:\n",
    "        locus_tag = f.qualifiers['db_xref'][0]\n",
    "    else:\n",
    "        print (\"strangely formatted record.  SKIPPING\")\n",
    "        print (f)\n",
    "        return {}\n",
    "    locus_tag_split = locus_tag.split(id_delim)\n",
    "    locus_tag = id_delim.join(locus_tag_split[1:])\n",
    "    \n",
    "    # For now, feature ID will be locus_tag\n",
    "    feature_ID = locus_tag\n",
    "    \n",
    "    # coords\n",
    "    beg = f.location.start + 1  # BioPython is cute and shifts just the begin pos by -1 \n",
    "    end = f.location.end\n",
    "    if end < beg:\n",
    "        print (\"WARNING: reversed gene: %s: %s (%d-%d)\"%(source_species, locus_tag, beg, end))\n",
    "        tmp_pos = end\n",
    "        end = beg\n",
    "        beg = tmp_pos\n",
    "    if f.location.strand == -1:\n",
    "        strand = '-'\n",
    "    else :\n",
    "        strand = '+'\n",
    "\n",
    "    # annotation\n",
    "    annotation = ''\n",
    "    EC_in_annotation = ''\n",
    "    if f_type == \"CDS\" or f_type == \"rRNA\" or f_type == \"tRNA\":    \n",
    "        #annotation = f.qualifiers['product'][0]\n",
    "        annotation = f.qualifiers['function'][0]\n",
    "        in_paren = False\n",
    "        last_good_char = 0\n",
    "        close_paren_pos = 0\n",
    "        for i,c in enumerate(annotation[::-1]):\n",
    "            if in_paren == False and c != ' ' and c != ')':\n",
    "                break\n",
    "            if in_paren == False and c != ' ' and c == ')':\n",
    "                in_paren = True\n",
    "                close_paren_pos = i\n",
    "            if in_paren and c == '(':\n",
    "                last_good_char = i+1\n",
    "                break\n",
    "        if last_good_char > 0:\n",
    "            candidate_EC = annotation[-1-last_good_char+2:-1-close_paren_pos]\n",
    "            if candidate_EC[0:2] == 'EC':\n",
    "                #print (name + \" EC: '\" + candidate_EC +\"'\")  # DEBUG\n",
    "                EC_in_annotation = candidate_EC[3:]\n",
    "\n",
    "            annotation = annotation[0:-1-last_good_char]\n",
    "        annotation = annotation.strip()\n",
    "        annotation = annotation.lstrip()\n",
    "\n",
    "    # add domain hits to annotation\n",
    "    try:\n",
    "        domain_hits = Global_Domains[contig_i][feature_ID]\n",
    "        domfam_seen = {}\n",
    "        if annotation != \"\":\n",
    "            annotation += \"\\n\"\n",
    "        for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "            domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "            try:\n",
    "                seen = domfam_seen[domfam]\n",
    "                continue\n",
    "            except:\n",
    "                domfam_seen[domfam] = True\n",
    "                if domfam[0:3] == 'COG' or domfam[0:2] == 'PF' or domfam[0:4] == 'TIGR':\n",
    "                    if domfam[0:2] == 'PF':\n",
    "                        dom_namespace = 'PF'\n",
    "                        [domfam, ver] = domfam.split('.')\n",
    "                    elif domfam[0:3] == 'COG':\n",
    "                        dom_namespace = 'COG'\n",
    "                    elif domfam[0:4] == 'TIGR':\n",
    "                        dom_namespace = 'TIGR'\n",
    "                    line_break = \"\\n\"\n",
    "                    if annotation == \"\":\n",
    "                        line_break = \"\"\n",
    "                    try:\n",
    "                        domfam_desc = \": \"+domain_family_desc[dom_namespace][domfam]\n",
    "                    except:\n",
    "                        domfam_desc = \"\"\n",
    "                    annotation += line_break+domfam+domfam_desc\n",
    "    except:\n",
    "        domain_hits = []\n",
    "                \n",
    "    # pull out E.C.    \n",
    "    EC_number = ''\n",
    "    if EC_in_annotation != '':\n",
    "        EC_number = EC_in_annotation\n",
    "    else:\n",
    "        if f_type == \"CDS\" and 'EC_number' in f.qualifiers:\n",
    "            EC_number = f.qualifiers['EC_number'][0]\n",
    "\n",
    "    # create feature_rec\n",
    "    feature_rec = {\"source_species\": source_species, \\\n",
    "                   \"name\": name, \\\n",
    "                   \"locus_tag\": locus_tag, \\\n",
    "                   \"ID\": feature_ID, # will probably need to make more unique \\\n",
    "                   \"type\": f_type, \\\n",
    "                   \"beg_pos\": beg, \\\n",
    "                   \"end_pos\": end, \\\n",
    "                   \"strand\": strand, \\\n",
    "                   \"annot\": annotation, \\\n",
    "                   \"EC_number\": EC_number, \\\n",
    "                   \"fwd_dna_seq\": fwd_dna_seq\n",
    "                  }                         \n",
    "    \n",
    "    #if feature_rec['name'] == \"DVU2932\" or feature_rec['name'] == \"Daud0298\":\n",
    "    #    arrow_color = color_names[sum([ord(c) for c in feature_rec['annot']]) % len(color_names)]\n",
    "    #    print (feature_rec['source_species'] + \" \" + feature_rec['name']+\" \"+arrow_color)\n",
    "\n",
    "    return feature_rec\n",
    "\n",
    "\n",
    "# Load domain family descriptions\n",
    "#\n",
    "def read_domain_family_desc ():\n",
    "    #domain_family_desc = {}\n",
    "\n",
    "    # read Pfam descs\n",
    "    pfam_desc_file = domain_family_desc_base_path+'/Pfam-A.clans.tsv'\n",
    "    if path.isfile(pfam_desc_file):\n",
    "        domain_family_desc['PF'] = {}\n",
    "        with open(pfam_desc_file,'r') as f:\n",
    "            reader=csv.reader(f,delimiter='\\t')\n",
    "            for pfam,clan,clan_name,pfam_name,pfam_desc in reader:\n",
    "                domain_family_desc['PF'][pfam] = pfam_desc\n",
    "                \n",
    "    # read COG descs\n",
    "    COG_desc_file = domain_family_desc_base_path+'/COG_2014.tsv'\n",
    "    if path.isfile(COG_desc_file):\n",
    "        domain_family_desc['COG'] = {}\n",
    "        with open(COG_desc_file,'r') as f:\n",
    "            reader=csv.reader(f,delimiter='\\t')\n",
    "            for COG,COG_funcat,COG_desc in reader:\n",
    "                domain_family_desc['COG'][COG] = COG_desc\n",
    "                \n",
    "    # read TIGR descs\n",
    "    TIGR_desc_file = domain_family_desc_base_path+'/TIGRInfo.tsv'\n",
    "    if path.isfile(TIGR_desc_file):\n",
    "        domain_family_desc['TIGR'] = {}\n",
    "        with open(TIGR_desc_file,'r') as f:\n",
    "            reader=csv.reader(f,delimiter='\\t')\n",
    "            for id,tigrId,type,roleId,geneSymbol,ec,definition in reader:\n",
    "                domain_family_desc['TIGR'][tigrId] = definition\n",
    "                if ec != \"\":\n",
    "                    domain_family_desc['TIGR'][tigrId] += \" (EC \"+ec+\")\"\n",
    "                \n",
    "    #return domain_family_desc\n",
    "    \n",
    "\n",
    "# Check search terms\n",
    "#\n",
    "def search_term_match (feature_rec):    \n",
    "    term_match = []\n",
    "    for i,term in enumerate(Search_Terms):\n",
    "        term_uc = term.upper()  # FIX: should sanitize search terms once at beginning\n",
    "        term_match.append(False)\n",
    "        for k in ['name', 'locus_tag', 'ID', 'annot', 'EC_number']:\n",
    "\n",
    "            #print (\"CHECKING %s %s %s %s\"%(term,k,feature_rec['name'], feature_rec[k]))  # DEBUG\n",
    "            try:\n",
    "                info_uc = feature_rec[k].upper()\n",
    "            except:\n",
    "                continue\n",
    "                \n",
    "            search_words = term_uc.split(\" \")  # FIX: should sanitize search terms once at beginning\n",
    "            logical = 'AND'\n",
    "            if \" OR \" in term_uc:\n",
    "                logical = 'OR'\n",
    "                \n",
    "            this_info_match = False\n",
    "            if logical == 'AND':\n",
    "                this_info_match = True\n",
    "                for word in search_words:\n",
    "                    #if word == 'AND' or word == 'OR':\n",
    "                    if word == 'AND' or word == 'OR' or word == '':  # FIX: should sanitize search terms once at beginning\n",
    "                        continue\n",
    "                    if word not in info_uc:\n",
    "                        this_info_match = False\n",
    "                        break\n",
    "            if logical == 'OR':\n",
    "                this_info_match = False\n",
    "                for word in search_words:\n",
    "                    #if word == 'AND' or word == 'OR':\n",
    "                    if word == 'AND' or word == 'OR' or word == '':  # FIX: should sanitize search terms once at beginning\n",
    "                        continue\n",
    "                    if word in info_uc:\n",
    "                        this_info_match = True\n",
    "                        break\n",
    "                        \n",
    "            if this_info_match:\n",
    "                term_match[i] = True\n",
    "                break\n",
    "                \n",
    "    return term_match\n",
    "    \n",
    "\n",
    "# sort rules\n",
    "#\n",
    "def sort_by_beg_pos_key (feature):\n",
    "    return feature['beg_pos']    \n",
    "\n",
    "def sort_by_bitscore_key (domhit):\n",
    "    return domhit[DOMHIT_BITSCORE_I]    \n",
    "\n",
    "def sort_by_domhit_len (domhit):\n",
    "    return domhit[DOMHIT_END_I]-domhit[DOMHIT_BEG_I]+1\n",
    "\n",
    "# Read KBase format domain hits\n",
    "#\n",
    "def getDomainHits (ContigSet_names, \\\n",
    "                   genomebrowser_mode=def_genomebrowser_mode,\n",
    "                   domain_data_format=domain_data_format):\n",
    "\n",
    "    for i,genome_name in enumerate(ContigSet_names):\n",
    "        (genome_id,scaffold_id) = genome_name.split(\".\")\n",
    "            \n",
    "        if Global_State['genomebrowser_mode'] == \"genome\" and i > 0:\n",
    "            break\n",
    "            \n",
    "        try:\n",
    "            Domain_Hits = Global_Domains[i]\n",
    "        except:\n",
    "            Global_Domains.append({})\n",
    "\n",
    "            if domain_data_format == \"KBase_domains\":\n",
    "                \n",
    "                domain_data_path = domain_data_base_path+'/'+genome_id+genome_data_extra_subpath+'/domains'+'/'+genome_id+\"_Domain_annot\"+'.json'\n",
    "                print (\"reading \"+domain_data_path+\" ...\")\n",
    "                with open(domain_data_path, 'r') as domain_file_handle:\n",
    "                    kbase_domains = json.load(domain_file_handle)\n",
    "\n",
    "                    for CDS_domain_list in kbase_domains['data'][scaffold_id]:\n",
    "                        gene_ID        = CDS_domain_list[KBASE_DOMAINHIT_GENE_ID_I]\n",
    "                        (genome_name, gene_name) = (gene_ID[0:gene_ID.index(\".\")], gene_ID[gene_ID.index(\".\")+1:])\n",
    "                        #gene_beg       = CDS_domain_list[KBASE_DOMAINHIT_GENE_BEG_I]\n",
    "                        #gene_end       = CDS_domain_list[KBASE_DOMAINHIT_GENE_END_I]\n",
    "                        #gene_strand    = CDS_domain_list[KBASE_DOMAINHIT_GENE_STRAND_I]\n",
    "                        gene_hits_dict = CDS_domain_list[KBASE_DOMAINHIT_GENE_HITS_DICT_I]\n",
    "                        gene_hits_list = []\n",
    "                        for domfam in gene_hits_dict.keys():\n",
    "                            # skip CD hits for now\n",
    "                            if domfam[0:2] != 'PF' and domfam[0:3] != 'COG' and domfam[0:4] != 'TIGR':\n",
    "                                continue\n",
    "                            #Global_Domains[i][gene_name] = gene_hits_dict\n",
    "                            for hit in gene_hits_dict[domfam]:\n",
    "                                list_format_hit = hit\n",
    "                                list_format_hit.append (domfam)\n",
    "                                #list_format_hit[DOMHIT_BEG_I]      = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_BEG_J]\n",
    "                                #list_format_hit[DOMHIT_END_I]      = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_END_J]\n",
    "                                #list_format_hit[DOMHIT_EVALUE_I]   = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_EVALUE_J]\n",
    "                                #list_format_hit[DOMHIT_BITSCORE_I] = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_BITSCORE_J]\n",
    "                                #list_format_hit[DOMHIT_ALNPERC_I]  = hit[KBASE_DOMAINHIT_GENE_HITS_DICT_ALNPERC_J]\n",
    "                                gene_hits_list.append(list_format_hit)\n",
    "                                #print (\"%s\\t%s\\t%s\\t%s\\t%d\\t%d\\t%f\"%(genome_id, scaffold_id, gene_name, domfam, hit_beg, hit_end, hit_evalue))\n",
    "\n",
    "                        Global_Domains[i][gene_name] = sorted (gene_hits_list, key=sort_by_bitscore_key, reverse=True)\n",
    "                        #for hit in Global_Domains[i][gene_name]:\n",
    "                            #print (\"%s\\t%s\\t%s\\t%s\\t%d\\t%d\\t%16.14f\\t%f\"%(genome_id, \\\n",
    "                            #                                     scaffold_id, \\\n",
    "                            #                                     gene_name, \\\n",
    "                            #                                     hit[DOMHIT_DOMFAM_I], \\\n",
    "                            #                                     hit[DOMHIT_BEG_I], \\\n",
    "                            #                                     hit[DOMHIT_END_I], \\\n",
    "                            #                                     hit[DOMHIT_EVALUE_I], \\\n",
    "                            #                                     hit[DOMHIT_BITSCORE_I] \\\n",
    "                            #                                     ))\n",
    "\n",
    "                        \n",
    "                            \n",
    "# Workhorse for feature retrieval\n",
    "#   does more than window slice, so perhaps should split up functionality\n",
    "#\n",
    "def getFeatureSlices (ContigSet_names, \\\n",
    "                      OrthologSet_locusTags, \\\n",
    "                      genomebrowser_mode=\"tree\", \\\n",
    "                      genome_data_format=\"Genbank\", \\\n",
    "                      window_size=10000, \\\n",
    "                      genomebrowser_xshift=0):\n",
    "    Feature_slices = []\n",
    "    \n",
    "    if genomebrowser_mode != \"genome\":\n",
    "        if genome_data_format == \"Genbank\":\n",
    "            for i,genome_name in enumerate(ContigSet_names):\n",
    "                if i >= max_rows:\n",
    "                    break\n",
    "                    \n",
    "                try:\n",
    "                    t = Global_Genbank_Genomes[i]\n",
    "                except:\n",
    "                    (genome_id,scaffold_id) = genome_name.split(\".\")\n",
    "                    #print (\"reading \" + genome_name + \" ...\")\n",
    "                    #genome_data_path = 'data/'+genome_name+'.gbk'\n",
    "                    genome_data_path = genome_data_base_path+'/'+genome_id+genome_data_extra_subpath+'/scaffolds'+'/'+scaffold_id+'.gbk'\n",
    "                    print (\"%d \"%i+'reading '+genome_data_path)\n",
    "                    Global_Genbank_Genomes.append (SeqIO.read(genome_data_path, 'genbank'))\n",
    "                    #print (\"%d\"%len(Global_Genbank_Genomes))\n",
    "                    #contig_seq = Global_Genbank_Genomes[i].seq\n",
    "                    #print (contig_seq[-100:])\n",
    "                    #print (\"%d\"%len(contig_seq))\n",
    "                \n",
    "#                try:\n",
    "#                    feature_cache = Global_Features[i]\n",
    "#                except:\n",
    "#                    Global_Features[i] = {}\n",
    "                    \n",
    "                Feature_slice = []\n",
    "                Features_seen = set()\n",
    "                source = \"\"\n",
    "                pivot_pos = 0.0\n",
    "\n",
    "                try:\n",
    "                    this_search_done = search_done[i]\n",
    "                except:\n",
    "                    search_done.append([])                \n",
    "                    search_done[i] = False\n",
    "                    search_results.append([])\n",
    "                    search_results[i] = []\n",
    "                    for j,term in enumerate(Search_Terms):\n",
    "                        search_results[i].append([])\n",
    "\n",
    "                # Get genome length\n",
    "                #if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                if i == 0:\n",
    "                    Global_State['PrimaryContig_len'] = len(Global_Genbank_Genomes[i].seq)\n",
    "                    Global_State['PrimaryContig_GCavg'] = compute_GC (Global_Genbank_Genomes[i].seq)\n",
    "                    Global_State['Contig_lens'] = []\n",
    "                    Global_State['pivot_pos_list'] = []\n",
    "                    #print (\"%d\"%Global_State['PrimaryContig_len'])\n",
    "                Global_State['Contig_lens'].append(len(Global_Genbank_Genomes[i].seq))\n",
    "                \n",
    "                # Find ortholog feature and put in first position\n",
    "                contig_mode_xshift = 0\n",
    "                for f in Global_Genbank_Genomes[i].features:                        \n",
    "                    if f.type == \"source\":\n",
    "                        source = f.qualifiers['organism'][0]\n",
    "                        break\n",
    "                        \n",
    "                try:\n",
    "                    this_ortholog_locustag = OrthologSet_locusTags[i]\n",
    "                except:\n",
    "                    OrthologSet_locusTags.append('')\n",
    "                    Global_State['OrthologSet_locusTags'].append('')\n",
    "                    \n",
    "                if Global_State['genomebrowser_mode'] == 'contigs' \\\n",
    "                    or OrthologSet_locusTags[i] == '':\n",
    "                        \n",
    "                    for f in Global_Genbank_Genomes[i].features:                        \n",
    "                        pivot_feature_rec = None\n",
    "                        lowest_beg = 10000000000\n",
    "                        #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers:                    \n",
    "#                        if (f.type == \"CDS\" and \"gene\" in f.qualifiers):\n",
    "                        if (f.type == \"CDS\" and \"gene\" in f.qualifiers) \\\n",
    "                            or f.type == 'rRNA' \\\n",
    "                            or f.type == 'tRNA' \\\n",
    "                            or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\") \\\n",
    "                            or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\") \\\n",
    "                            or (f.type == 'gene' and \"pseudo\" in f.qualifiers):\n",
    "\n",
    "                            if f.location.start < lowest_beg:\n",
    "                                lowest_beg = f.location.start\n",
    "                                pivot_feature_rec = build_feature_rec(f, f_type=f.type, source_species=source, contig_i=i)\n",
    "                                \n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                    if i == 0:\n",
    "                        Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "\n",
    "                    if Global_State['genomebrowser_mode'] == 'contigs':                    \n",
    "                        contig_mode_xshift = 0.5*window_size - 0.5*(pivot_feature_rec['end_pos']-pivot_feature_rec['beg_pos'])\n",
    "                            \n",
    "                else:             \n",
    "                    for f in Global_Genbank_Genomes[i].features:        \n",
    "                        \n",
    "                        #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers:                    \n",
    "                        if f.type == \"CDS\" and \"gene\" in f.qualifiers:  # should it permit non-CDS anchor?             \n",
    "                            \n",
    "                            #elif f.qualifiers['locus_tag'][0] == OrthologSet_locusTags[i]:\n",
    "                            if f.qualifiers['gene'][0] == OrthologSet_locusTags[i]:  \n",
    "                                pivot_feature_rec = build_feature_rec(f, f_type='CDS', source_species=source, contig_i=i)\n",
    "                                Feature_slice.append(pivot_feature_rec)\n",
    "                                pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                                Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                                if i == 0:\n",
    "                                    Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "                                break\n",
    "                        \n",
    "                # Add in additional features within window.  Note: we want the duplicate ortholog feature\n",
    "                for f in Global_Genbank_Genomes[i].features:\n",
    "                    \n",
    "                    if f.type == \"CDS\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        # fast enough for now to build recs for all features, but might want to only do for window\n",
    "                        feature_rec = build_feature_rec(f, f_type='CDS', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "\n",
    "                        # add feature to slice if in viewable window\n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == \"rRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec(f, f_type='rRNA', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                                    \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                    elif f.type == \"tRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)                        \n",
    "                        feature_rec = build_feature_rec(f, f_type='tRNA', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)                        \n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[i].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec(f, f_type='CRISPR', source_species=source, contig_i=i, fwd_dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[i].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec(f, f_type='CRISPR spacer', source_species=source, contig_i=i, fwd_dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"pseudo\" in f.qualifiers:\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)                        \n",
    "                        feature_rec = build_feature_rec(f, f_type='pseudogene', source_species=source, contig_i=i)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[i]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[i][j][0]\n",
    "                                    except:\n",
    "                                        search_results[i][j] = []\n",
    "                                    search_results[i][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })\n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + contig_mode_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + contig_mode_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                # sort results by position, store contig slice, and mark search as done on this contig\n",
    "                #\n",
    "                Sorted_Feature_slice = [Feature_slice[0]]  # retain pivot\n",
    "                for f in sorted (Feature_slice[1:], key=sort_by_beg_pos_key):\n",
    "                    Sorted_Feature_slice.append (f)\n",
    "                    \n",
    "                Feature_slices.append(Sorted_Feature_slice)\n",
    "                search_done[i] = True   \n",
    "\n",
    "        else:\n",
    "            print (\"unknown data format for Genomes\")\n",
    "\n",
    "    # genomebrowser_mode == \"genome\"\n",
    "    else:  \n",
    "        if genome_data_format == \"Genbank\":\n",
    "            #for i in range (0,total_rows):\n",
    "            for i in range (0,Global_State['genome_mode_n_rows']):\n",
    "                genome_name = ContigSet_names[0]\n",
    "                try:\n",
    "                    t=Global_Genbank_Genomes[0]\n",
    "                except:\n",
    "                    (genome_id,scaffold_id) = genome_name.split(\".\")\n",
    "                    #print (\"reading \" + genome_name + \" ...\")\n",
    "                    #genome_data_path = 'data/'+genome_name+'.gbk'\n",
    "                    genome_data_path = genome_data_base_path+'/'+genome_id+genome_data_extra_subpath+'/scaffolds/'+scaffold_id+'.gbk'\n",
    "                    print (\"%d \"%i+'reading '+genome_data_path)\n",
    "                    Global_Genbank_Genomes.append (SeqIO.read(genome_data_path, 'genbank'))\n",
    "                    \n",
    "                Feature_slice = []\n",
    "                Features_seen = set()\n",
    "                source = \"\"\n",
    "                pivot_pos = 0.0\n",
    "                #track_xshift = window_size * (i - (total_rows-1)/2)  # e.g. total_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                track_xshift = window_size * (i - (Global_State['genome_mode_n_rows']-1)/2)  # e.g. n_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                #print (\"%d %f\"%(i,track_xshift))\n",
    "\n",
    "                try:\n",
    "                    this_search_done = search_done[0]\n",
    "                except:\n",
    "                    search_done.append([])                \n",
    "                    search_done[0] = False\n",
    "                    search_results.append([])\n",
    "                    search_results[0] = []\n",
    "                    for j,term in enumerate(Search_Terms):\n",
    "                        search_results[0].append([])                \n",
    "                \n",
    "                # Get genome length\n",
    "                if Global_State['PrimaryContig_len'] == 0 and i == 0:\n",
    "                    Global_State['PrimaryContig_len'] = len(Global_Genbank_Genomes[i].seq)\n",
    "                    Global_State['PrimaryContig_GCavg'] = compute_GC (Global_Genbank_Genomes[i].seq)\n",
    "                    #print (\"%d\"%Global_State['PrimaryContig_len'])\n",
    "                    Global_State['Contig_lens'].append(len(Global_Genbank_Genomes[i].seq))\n",
    "\n",
    "                # Find ortholog feature and put in first position\n",
    "                if i == 0:\n",
    "                    try:\n",
    "                        this_ortholog_locustag = OrthologSet_locusTags[0]\n",
    "                    except:\n",
    "                        OrthologSet_locusTags.append('')\n",
    "                        Global_State['OrthologSet_locusTags'].append('')              \n",
    "                \n",
    "                if i > 0:\n",
    "                    pivot_feature_rec = Feature_slices[0][0]  # use homolog feature from first contig only\n",
    "                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                    Feature_slice.append(pivot_feature_rec)\n",
    "                else:\n",
    "                    for f in Global_Genbank_Genomes[0].features:  \n",
    "                        if f.type == \"source\":\n",
    "                            source = f.qualifiers['organism'][0]\n",
    "                            break\n",
    "                            \n",
    "                    if OrthologSet_locusTags[0] == '':      \n",
    "                        for f in Global_Genbank_Genomes[0].features:                        \n",
    "                            pivot_feature_rec = None\n",
    "                            lowest_beg = 10000000000\n",
    "                            #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers:                    \n",
    "    #                        if (f.type == \"CDS\" and \"gene\" in f.qualifiers):\n",
    "                            if (f.type == \"CDS\" and \"gene\" in f.qualifiers) \\\n",
    "                                or f.type == 'rRNA' \\\n",
    "                                or f.type == 'tRNA' \\\n",
    "                                or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\") \\\n",
    "                                or (f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\") \\\n",
    "                                or (f.type == 'gene' and \"pseudo\" in f.qualifiers):\n",
    "\n",
    "                                if f.location.start < lowest_beg:\n",
    "                                    lowest_beg = f.location.start\n",
    "                                    pivot_feature_rec = build_feature_rec(f, f_type=f.type, source_species=source, contig_i=0)\n",
    "\n",
    "                        Feature_slice.append(pivot_feature_rec)\n",
    "                        pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                        Global_State['pivot_pos_list'].append(pivot_pos)\n",
    "                        if i == 0:\n",
    "                            Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos  \n",
    "                            \n",
    "                    else:\n",
    "                        for f in Global_Genbank_Genomes[0].features:  # should it permit non-CDS anchor?    \n",
    "                            #if f.type == \"CDS\" and \"locus_tag\" in f.qualifiers and f.qualifiers['locus_tag'][0] == OrthologSet_locusTags[i]:\n",
    "                            if f.type == \"CDS\" and \"gene\" in f.qualifiers:  \n",
    "                                if f.qualifiers['gene'][0] == OrthologSet_locusTags[i]:\n",
    "                                    pivot_feature_rec = build_feature_rec(f, f_type='CDS', source_species=source, contig_i=0)\n",
    "                                    Feature_slice.append(pivot_feature_rec)\n",
    "                                    pivot_pos = 0.5 * (pivot_feature_rec['beg_pos']+pivot_feature_rec['end_pos'])\n",
    "                                    if i == 0:\n",
    "                                        Global_State['PrimaryAnchor_pivot_pos'] = pivot_pos\n",
    "                                    break\n",
    "                        \n",
    "                # Add in additional features within window.  Note: we want the duplicate ortholog feature\n",
    "                for f in Global_Genbank_Genomes[0].features:\n",
    "                    if f.type == \"CDS\":  # we'll add other types later when we have icons for them\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        # fast enough for now to build recs for all features, but might want to only do for window\n",
    "                        feature_rec = build_feature_rec(f, f_type='CDS', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                        \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            #print (\"%d %d %d\"%(i, feature_rec['beg_pos'], feature_rec['end_pos']))\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == \"rRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec(f, f_type='rRNA', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                    elif f.type == \"tRNA\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec(f, f_type='tRNA', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)                    \n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[0].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec(f, f_type='CRISPR', source_species=source, contig_i=0, fwd_dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"comment\" in f.qualifiers and f.qualifiers['comment'][0] == \"CRISPR spacer\":\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        fwd_dna_seq = Global_Genbank_Genomes[0].seq[f.location.start:f.location.end-1]  # BioPython shifts start by -1 but not end?\n",
    "                        feature_rec = build_feature_rec(f, f_type='CRISPR spacer', source_species=source, contig_i=0, fwd_dna_seq=fwd_dna_seq)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "                            \n",
    "                    elif f.type == 'gene' and \"pseudo\" in f.qualifiers:\n",
    "                        pos_key = \"%d,%d,%d\"%(f.location.strand,f.location.start+1,f.location.end)\n",
    "                        if pos_key in Features_seen:\n",
    "                            continue\n",
    "                        else:\n",
    "                            Features_seen.add(pos_key)\n",
    "                        feature_rec = build_feature_rec(f, f_type='pseudogene', source_species=source, contig_i=0)\n",
    "\n",
    "                        # check if a search hit\n",
    "                        if not search_done[0]:\n",
    "                            for j,match_flag in enumerate(search_term_match(feature_rec)):\n",
    "                                if match_flag:\n",
    "                                    try:\n",
    "                                        term_hit_in_genome = search_results[0][j][0]\n",
    "                                    except:\n",
    "                                        search_results[0][j] = []\n",
    "                                    search_results[0][j].append({'beg_pos': feature_rec['beg_pos'],\n",
    "                                                                 'end_pos': feature_rec['end_pos'],\n",
    "                                                                 'name': feature_rec['name'],\n",
    "                                                                 'annot': feature_rec['annot']\n",
    "                                                                })                          \n",
    "                        \n",
    "                        if feature_rec['end_pos'] >= pivot_pos + genomebrowser_xshift + track_xshift - 0.5*window_size and \\\n",
    "                            feature_rec['beg_pos'] <= pivot_pos + genomebrowser_xshift + track_xshift + 0.5*window_size:\n",
    "\n",
    "                            Feature_slice.append(feature_rec)\n",
    "\n",
    "                            \n",
    "                # sort results by position, store contig slice, and mark search as done on this contig\n",
    "                #\n",
    "                Sorted_Feature_slice = [Feature_slice[0]]  # retain pivot\n",
    "                for f in sorted (Feature_slice[1:], key=sort_by_beg_pos_key):\n",
    "                    Sorted_Feature_slice.append (f)\n",
    "                    \n",
    "                Feature_slices.append(Sorted_Feature_slice)\n",
    "                search_done[0] = True\n",
    "                \n",
    "        else:\n",
    "            print (\"unknown data format for Genomes\")\n",
    "            \n",
    "\n",
    "    return Feature_slices\n",
    "\n",
    "\n",
    "# Slider events\n",
    "#\n",
    "class DraggableRectangleSlider:\n",
    "    def __init__(self, rect, func='', move='x', min_pos=0.0, max_pos=1.0, step=0.1, tol=0.01):\n",
    "        self.rect = rect\n",
    "        self.press = None\n",
    "        self.func = func\n",
    "        self.move = move.lower()\n",
    "        self.min_pos = min_pos\n",
    "        self.max_pos = max_pos\n",
    "        self.step = step\n",
    "        self.tol = tol\n",
    "        self.n_steps = int(round((max_pos-min_pos)/step))\n",
    "        self.last_pos = 0\n",
    "        if self.move != 'x' and self.move != 'y':\n",
    "            print (\"bad DraggableRectSlider move: '\"+ self.move +\"' for func: '\"+ self.func +\"'\")\n",
    "\n",
    "    def connect(self):\n",
    "        'connect to all the events we need'\n",
    "        self.cidpress = self.rect.figure.canvas.mpl_connect('button_press_event', self.on_press)\n",
    "        self.cidrelease = self.rect.figure.canvas.mpl_connect('button_release_event', self.on_release)\n",
    "        self.cidmotion = self.rect.figure.canvas.mpl_connect('motion_notify_event', self.on_motion)\n",
    "\n",
    "    def on_press(self, event):\n",
    "        'on button press we will see if the mouse is over us and store some data'\n",
    "        if event.inaxes != self.rect.axes: return\n",
    "\n",
    "        contains, attrd = self.rect.contains(event)\n",
    "        if not contains: return\n",
    "        #print ('event contains', self.rect.xy)\n",
    "        x0, y0 = self.rect.xy\n",
    "        self.press = x0, y0, event.xdata, event.ydata\n",
    "        if self.move == 'x':\n",
    "            self.last_pos = x0\n",
    "        elif self.move == 'y':\n",
    "            self.last_pos = y0\n",
    "\n",
    "    def on_motion(self, event):\n",
    "        'on motion we will move the rect if the mouse is over us'\n",
    "        update_draggable = False\n",
    "        if self.press is None: return\n",
    "        if event.inaxes != self.rect.axes: return\n",
    "        x0, y0, xpress, ypress = self.press\n",
    "        #last_pos = self.last_pos\n",
    "        step = self.step\n",
    "        tol = self.tol\n",
    "        min_pos = self.min_pos\n",
    "        max_pos = self.max_pos\n",
    "        dx = event.xdata - xpress\n",
    "        dy = event.ydata - ypress\n",
    "        #print ('x0=%f, xpress=%f, event.xdata=%f, dx=%f, x0+dx=%f, last_pos, step'%(x0, xpress, event.xdata, dx, x0+dx, self.last_pos, self.step))\n",
    "        if self.move == 'x':\n",
    "            if x0+dx <= min_pos + tol:\n",
    "                if self.last_pos != min_pos:\n",
    "                    self.last_pos = min_pos\n",
    "                    nearest_step_i = 0\n",
    "                    update_draggable = True\n",
    "            elif x0+dx >= max_pos - tol:\n",
    "                if self.last_pos != max_pos:\n",
    "                    self.last_pos = max_pos\n",
    "                    nearest_step_i = int(round((max_pos-min_pos)/step))\n",
    "                    update_draggable = True\n",
    "            else:\n",
    "                nearest_step_i = int(round((x0+dx-min_pos)/step))\n",
    "                new_pos = min_pos + nearest_step_i*step\n",
    "                if new_pos != self.last_pos:\n",
    "                    self.last_pos = new_pos\n",
    "                    update_draggable = True\n",
    "                                     \n",
    "            if update_draggable:\n",
    "                self.rect.set_x(self.last_pos)\n",
    "                #self.rect.set_x(x0+dx)\n",
    "                #self.rect.set_y(y0)\n",
    "        elif self.move == 'y':\n",
    "            if y0+dy <= min_pos + tol:\n",
    "                if self.last_pos != min_pos:\n",
    "                    self.last_pos = min_pos\n",
    "                    nearest_step_i = 0\n",
    "                    update_draggable = True\n",
    "            elif y0+dy >= max_pos - tol:\n",
    "                if self.last_pos != max_pos:\n",
    "                    self.last_pos = max_pos\n",
    "                    nearest_step_i = int(round((max_pos-min_pos)/step))\n",
    "                    update_draggable = True\n",
    "            else:\n",
    "                nearest_step_i = int(round((y0+dy-min_pos)/step))\n",
    "                new_pos = min_pos + nearest_step_i*step\n",
    "                if new_pos != self.last_pos:\n",
    "                    self.last_pos = new_pos\n",
    "                    update_draggable = True\n",
    "                    \n",
    "            if update_draggable:\n",
    "                self.rect.set_y(self.last_pos)\n",
    "                #self.rect.set_x(x0)\n",
    "                #self.rect.set_y(y0+dy)\n",
    "\n",
    "        if update_draggable:\n",
    "            self.rect.figure.canvas.draw()\n",
    "            if self.func == 'zoom':\n",
    "                Global_State['genomebrowser_zoom'] = int(round((self.last_pos-min_pos)/step))\n",
    "                Global_State['genomebrowser_window_bp_width'] = Global_State['def_genomebrowser_window_bp_width']*(2**Global_State['genomebrowser_zoom'])\n",
    "            elif self.func == 'pan':\n",
    "                Global_State['genomebrowser_xshift'] = Global_State['PrimaryContig_len']*(self.last_pos-min_pos)/(max_pos-min_pos) - Global_State['PrimaryAnchor_pivot_pos']\n",
    "            elif self.func == 'mode':\n",
    "                Global_State['genomebrowser_mode'] = mode_names[nearest_step_i]\n",
    "            elif self.func == 'color':\n",
    "                Global_State['genomebrowser_color_namespace'] = color_namespace_names[nearest_step_i]\n",
    "                \n",
    "            update_genomebrowser_panel (ax_center)  # This should happen first so Features and Domains are updated\n",
    "            update_sidenav_panel (ax_left)\n",
    "            update_mode_panel (ax_top_left)\n",
    "\n",
    "    def on_release(self, event):\n",
    "        'on release we reset the press data'\n",
    "        self.press = None\n",
    "        self.rect.figure.canvas.draw()\n",
    "            \n",
    "    def disconnect(self):\n",
    "        'disconnect all the stored connection ids'\n",
    "        self.rect.figure.canvas.mpl_disconnect(self.cidpress)\n",
    "        self.rect.figure.canvas.mpl_disconnect(self.cidrelease)\n",
    "        self.rect.figure.canvas.mpl_disconnect(self.cidmotion)\n",
    "\n",
    "\n",
    "# handle contig nav events\n",
    "#\n",
    "def onpick_contig_nav (event):\n",
    "    artist = event.artist\n",
    "    mouseevent = event.mouseevent\n",
    "    ax = artist.axes\n",
    "    if artist.get_label() != \"circular_contig_nav\" and artist.get_label() != \"linear_contig_nav\":\n",
    "        return\n",
    "    artist_info = artist.gid.split(\",\")\n",
    "\n",
    "    repaint = False\n",
    "    if artist.get_label() == \"circular_contig_nav\":\n",
    "        genome_i = int(artist_info[0])\n",
    "        x0 = float(artist_info[1])\n",
    "        y0 = float(artist_info[2])\n",
    "        y_scale = float(artist_info[3])\n",
    "        dx = mouseevent.xdata - x0\n",
    "        dy = (mouseevent.ydata - y0) / y_scale\n",
    "        \n",
    "        cosine = dx / sqrt(dx*dx+dy*dy)\n",
    "        # Math.acos and Math.asin are in radians, and acos returns 0-pi\n",
    "        theta = acos(cosine)\n",
    "        if dy < 0:\n",
    "            theta = 2*pi - theta            \n",
    "        if theta < pi/2.0:\n",
    "            phi = pi/2.0 - theta\n",
    "        else:\n",
    "            phi = 2.0*pi - theta + pi/2.0\n",
    "        genome_frac_pos = phi / (2.0*pi)\n",
    "        \n",
    "        new_genome_pos = int(round(genome_frac_pos * Global_State['PrimaryContig_len']))\n",
    "\n",
    "        if (new_genome_pos <= 0):\n",
    "            new_genome_pos = 1\n",
    "        elif new_genome_pos > Global_State['PrimaryContig_len']:\n",
    "            new_genome_pos = Global_State['PrimaryContig_len']\n",
    "        \n",
    "        Global_State['genomebrowser_xshift'] = new_genome_pos - Global_State['PrimaryAnchor_pivot_pos']\n",
    "        repaint = True\n",
    "        \n",
    "    elif artist.get_label() == \"linear_contig_nav\":\n",
    "        genome_i = int(artist_info[0])\n",
    "        x0 = float(artist_info[1])\n",
    "        w = float(artist_info[2])\n",
    "        dx = mouseevent.xdata - x0\n",
    "        #dy = mouseevent.ydata - y0\n",
    "\n",
    "        new_genome_pos = int(round(Global_State['Contig_lens'][genome_i] * dx/w))\n",
    "        \n",
    "        if (new_genome_pos <= 0):\n",
    "            new_genome_pos = 1\n",
    "        elif new_genome_pos > Global_State['Contig_lens'][genome_i]:\n",
    "            new_genome_pos = Global_State['Contig_lens'][genome_i]\n",
    "            \n",
    "        Global_State['genomebrowser_xshift'] = new_genome_pos - Global_State['pivot_pos_list'][genome_i]        \n",
    "        repaint = True\n",
    "        \n",
    "    if repaint:\n",
    "        #update_control_panel(ax_top_center)  # for some reason this causes things to drag.  memory leak for event listeners?\n",
    "        # FIX THIS\n",
    "        #pos_knob_patch.set_x(self.last_pos)  # global ref (total kludge).  not working yet.        \n",
    "        update_genomebrowser_panel (ax_center)\n",
    "        update_sidenav_panel (ax_left)\n",
    "        \n",
    "        \n",
    "# handle pan events\n",
    "#\n",
    "def onpick_pan (event):\n",
    "    artist = event.artist\n",
    "    #mouseevent = event.mouseevent\n",
    "    #ax = artist.axes\n",
    "    if artist.get_label() != \"pan_button\":\n",
    "        return\n",
    "    control = artist.gid\n",
    "\n",
    "    repaint = False\n",
    "    if control == \"reset_pan\":\n",
    "        Global_State['genomebrowser_xshift'] = 0\n",
    "        repaint = True\n",
    "    elif control == \"left_pan\":\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            Global_State['genomebrowser_xshift'] -= 1.0*Global_State['genomebrowser_window_bp_width']\n",
    "        else:\n",
    "            Global_State['genomebrowser_xshift'] -= 0.4*Global_State['genomebrowser_window_bp_width']\n",
    "        repaint = True\n",
    "    elif control == \"right_pan\":\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            Global_State['genomebrowser_xshift'] += 1.0*Global_State['genomebrowser_window_bp_width']\n",
    "        else:\n",
    "            Global_State['genomebrowser_xshift'] += 0.4*Global_State['genomebrowser_window_bp_width']\n",
    "        repaint = True\n",
    "\n",
    "    if repaint:\n",
    "        #update_control_panel(ax_top_center)  # for some reason this causes things to drag.  memory leak for event listeners?\n",
    "        # FIX THIS\n",
    "        #pos_knob_patch.set_x(self.last_pos)  # global ref (total kludge).  not working yet.        \n",
    "        update_genomebrowser_panel (ax_center)\n",
    "        update_sidenav_panel (ax_left)\n",
    "        \n",
    "        \n",
    "# Feature events\n",
    "#\n",
    "\"\"\"\n",
    "# Something about the object solution here really makes ipython drag.\n",
    "#   Implemented solution with artist labels, feature_artist_label_to_popup_box and\n",
    "#   feature_artist_label_to_feature dicts and labels to link artists ugly but effective\n",
    "class InteractiveGenomeFeature:\n",
    "    def __init__(self, arrow, feature):\n",
    "        self.arrow = arrow\n",
    "        self.feature = feature\n",
    "        self.dialog_box = None\n",
    "        self.box = None\n",
    "        \n",
    "    def connect(self):\n",
    "        'connect to all the events we need'\n",
    "        self.cidpick = self.arrow.figure.canvas.mpl_connect('pick_event', self.on_pick)\n",
    "\n",
    "    def on_pick (self, event):\n",
    "        artist = event.artist\n",
    "        mouseevent = event.mouseevent\n",
    "        arrow = artist\n",
    "        ax = arrow.axes\n",
    "        #arrow.set_color(\"black\")\n",
    "        self.dialog_box = Rectangle((mouseevent.xdata, mouseevent.ydata), 0.1, 0.1, \\\n",
    "                                    facecolor=\"aliceblue\", edgecolor=\"gray\", alpha=1.0, zorder=4)\n",
    "        ax.add_patch(self.dialog_box)\n",
    "\n",
    "        #arrow.remove()\n",
    "        arrow.figure.canvas.draw()\n",
    "            \n",
    "    def disconnect(self):\n",
    "        'disconnect all the stored connection ids'\n",
    "        self.arrow.figure.canvas.mpl_disconnect(self.cidpick)\n",
    "\n",
    "\"\"\"\n",
    "# handle feature events\n",
    "#\n",
    "def onpick_feature (event):\n",
    "    artist = event.artist\n",
    "    mouseevent = event.mouseevent\n",
    "    ax = artist.axes\n",
    "    if artist.get_label() != \"feature\" and artist.get_label() != \"popup_destroy\":\n",
    "        return\n",
    "    feature_index = artist.gid\n",
    "    \n",
    "    # popup config\n",
    "    annot_word_limit = 5\n",
    "    popup_x_margin = 0.015\n",
    "    popup_y_margin = 0.020 * 7/(total_rows+1)\n",
    "    close_button_w = .02\n",
    "    close_button_h = .03 * 7/(total_rows+1)\n",
    "    close_button_marker_y_shift = 0.0037 * 7/(total_rows+1)\n",
    "    popup_text_w_scaling = 0.0085                     # appropriate value for fontsize 10\n",
    "    popup_text_h_scaling = 0.0300 * 7/(total_rows+1)  # appropriate value for fontsize 10\n",
    "    \n",
    "    #popup_zorder += 2   # why doesn't this work even when popup_zorder is a global?\n",
    "    Global_State['popup_zorder'] += 4\n",
    "    popup_zorder = Global_State['popup_zorder']\n",
    "    \n",
    "    if not feature_index in feature_artist_label_to_popup_box or \\\n",
    "        not feature_artist_label_to_popup_box[feature_index]:\n",
    "            \n",
    "        feature = feature_artist_label_to_feature[feature_index]\n",
    "\n",
    "        # build info strings\n",
    "        word_join = \" \"\n",
    "        row_join = \"\\n\"\n",
    "\n",
    "        # popup_info_1 (name)\n",
    "        popup_info_str1_row_split = [feature['name']]\n",
    "\n",
    "        # popup_info_2 (annot)\n",
    "        annot_processed = False\n",
    "        if not \"\\n\" in feature['annot']:\n",
    "            popup_info_str2_word_split = feature['annot'].split(\" \")\n",
    "            word_count = len(popup_info_str2_word_split)\n",
    "            if word_count <= annot_word_limit:\n",
    "                if feature['annot'] != '':\n",
    "                    popup_info_str2_row_split = [feature['annot']]\n",
    "                annot_processed = True\n",
    "        if not annot_processed:\n",
    "            popup_info_str2_row_split = []\n",
    "            for row in feature['annot'].split(\"\\n\"):\n",
    "                #if row == '':\n",
    "                #    continue\n",
    "                popup_info_str2_word_split = row.split(\" \")\n",
    "                word_count = len(popup_info_str2_word_split)\n",
    "                if word_count <= annot_word_limit:\n",
    "                    popup_info_str2_row_split.append(row)\n",
    "                else:\n",
    "                    imax = int(round(word_count/annot_word_limit))\n",
    "                    indent = ''\n",
    "                    for i in range(0,imax):\n",
    "                        row_slice = popup_info_str2_word_split[i*annot_word_limit:(i+1)*annot_word_limit]\n",
    "                        if i > 0:\n",
    "                            indent = '  '\n",
    "                        popup_info_str2_row_split.append(indent+word_join.join(row_slice))\n",
    "                    remainder = len(popup_info_str2_word_split)%annot_word_limit\n",
    "                    if remainder > 0:\n",
    "                        row_slice = popup_info_str2_word_split[imax*annot_word_limit:]\n",
    "                        indent = '  '\n",
    "                        popup_info_str2_row_split.append(indent+word_join.join(row_slice))\n",
    "            \n",
    "        # popup_info_3 (rest of info)\n",
    "        popup_info_str3_row_split = [\"locus_tag: %s\"%feature['locus_tag'], \\\n",
    "                                     \"id(s): %s\"%feature['ID'], \\\n",
    "                                     \"E.C.: %s\"%feature['EC_number'], \\\n",
    "                                     \"type: %s\"%feature['type'], \\\n",
    "                                     \"position: %d-%d (%s)\"%(feature['beg_pos'],feature['end_pos'],feature['strand']), \\\n",
    "                                     feature['source_species']\n",
    "                                    ]\n",
    "\n",
    "        # get dimensions and shift popup if hitting walls  \n",
    "        #\n",
    "        popup_info_row_cnt = len(popup_info_str1_row_split) + \\\n",
    "                             len(popup_info_str2_row_split) + \\\n",
    "                             1 + \\\n",
    "                             len(popup_info_str3_row_split)\n",
    "        popup_info_h = popup_text_h_scaling * popup_info_row_cnt\n",
    "        \n",
    "        longest_info_str = 0\n",
    "        for popup_info_str_row_split in [popup_info_str1_row_split, popup_info_str2_row_split, popup_info_str3_row_split]:\n",
    "        #for popup_info_str_row_split in [popup_info_str3_row_split]:\n",
    "            for popup_info_str in popup_info_str_row_split:\n",
    "                row_len = len(popup_info_str)\n",
    "                if row_len > longest_info_str:\n",
    "                    longest_info_str = row_len\n",
    "        popup_info_w = popup_text_w_scaling * longest_info_str\n",
    "\n",
    "        popup_box_w = popup_info_w + 2*popup_x_margin + 1.5*close_button_w\n",
    "        popup_box_h = popup_info_h + 2*popup_y_margin + 1.0*close_button_h\n",
    "        popup_x_shift = 0\n",
    "        popup_y_shift = 0\n",
    "        if mouseevent.xdata+popup_box_w > 1.0:\n",
    "            popup_x_shift = 1.0 - (mouseevent.xdata+popup_box_w) - 0.01  # extra .01 to not hit wall\n",
    "        if mouseevent.ydata+popup_box_h > 1.0:\n",
    "            popup_y_shift = 1.0 - (mouseevent.ydata+popup_box_h) - 0.01  # extra .01 to not hit wall\n",
    "\n",
    "            \n",
    "        # Add Box\n",
    "        #\n",
    "        popup_box = Rectangle((mouseevent.xdata+popup_x_shift, mouseevent.ydata+popup_y_shift), \\\n",
    "                              popup_box_w, \\\n",
    "                              popup_box_h, \\\n",
    "                              facecolor=popup_bg_color, edgecolor=\"gray\", \\\n",
    "                              alpha=1.0, zorder=popup_zorder)\n",
    "        popup_box.gid = feature_index  # if we want to make it and contents draggable later\n",
    "        popup_box.set_path_effects([path_effects.PathPatchEffect(offset=(1, -1), facecolor='black', alpha=0.75),\n",
    "                                    path_effects.PathPatchEffect(facecolor=popup_bg_color,\n",
    "                                                                 edgecolor=\"gray\")\n",
    "                                   ])\n",
    "        ax.add_patch(popup_box)\n",
    " \n",
    "        # Add Info\n",
    "        #\n",
    "        text_y_shift = popup_text_h_scaling*len(popup_info_str3_row_split) + \\\n",
    "                       popup_text_h_scaling*1.0 + \\\n",
    "                       popup_text_h_scaling*len(popup_info_str2_row_split) + \\\n",
    "                       popup_text_h_scaling*0.5\n",
    "        popup_info_1 = ax.text(mouseevent.xdata+popup_x_shift+popup_x_margin, \\\n",
    "                               mouseevent.ydata+popup_y_shift+popup_y_margin + text_y_shift, \\\n",
    "                                row_join.join(popup_info_str1_row_split), \\\n",
    "                                verticalalignment=\"bottom\", \\\n",
    "                                horizontalalignment=\"left\", \\\n",
    "                                transform=ax.transAxes, \\\n",
    "                                color=\"black\", \\\n",
    "                                fontsize=12, \\\n",
    "                                zorder=popup_zorder+1)\n",
    "        #popup_info_1.set_label(feature_index)\n",
    "        #ax.add_patch(popup_info_1)   # text cannot be added as patch to axes, or else things get wonky\n",
    "        \n",
    "        text_y_shift = popup_text_h_scaling*len(popup_info_str3_row_split) + \\\n",
    "                       popup_text_h_scaling*1.0\n",
    "        popup_info_2 = ax.text(mouseevent.xdata+popup_x_shift+popup_x_margin, \\\n",
    "                               mouseevent.ydata+popup_y_shift+popup_y_margin + text_y_shift, \\\n",
    "                                row_join.join(popup_info_str2_row_split), \\\n",
    "                                verticalalignment=\"bottom\", \\\n",
    "                                horizontalalignment=\"left\", \\\n",
    "                                transform=ax.transAxes, \\\n",
    "                                color=\"red\", \\\n",
    "                                fontsize=10, \\\n",
    "                                zorder=popup_zorder+1)\n",
    "        #popup_info_2.set_label(feature_index)\n",
    "        \n",
    "        text_y_shift = 0\n",
    "        popup_info_3 = ax.text(mouseevent.xdata+popup_x_shift+popup_x_margin, \\\n",
    "                               mouseevent.ydata+popup_y_shift+popup_y_margin + text_y_shift, \\\n",
    "                               row_join.join(popup_info_str3_row_split), \\\n",
    "                               verticalalignment=\"bottom\", \\\n",
    "                               horizontalalignment=\"left\", \\\n",
    "                               transform=ax.transAxes, \\\n",
    "                               color=\"black\", \\\n",
    "                               fontsize=10, \\\n",
    "                               zorder=popup_zorder+1)\n",
    "        #popup_info_3.set_label(feature_index)\n",
    "        \n",
    "        \n",
    "        # Add Close button\n",
    "        #\n",
    "        close_button = Rectangle((mouseevent.xdata+popup_x_shift+popup_box_w-popup_x_margin-close_button_w,\n",
    "                                  mouseevent.ydata+popup_y_shift+popup_box_h-popup_y_margin-close_button_h), \\\n",
    "                                  close_button_w, \\\n",
    "                                  close_button_h, \\\n",
    "                                  facecolor=\"white\", edgecolor=\"gray\", \\\n",
    "                                  picker=True, \\\n",
    "                                  alpha=1.0, zorder=popup_zorder+2)\n",
    "        close_button.gid = feature_index\n",
    "        close_button.set_label(\"popup_destroy\")\n",
    "        ax.add_patch(close_button)\n",
    "        #close_button.figure.canvas.mpl_connect ('pick_event', onpick_feature)  # already listening\n",
    "        close_button_x = ax.text(mouseevent.xdata+popup_x_shift+popup_box_w-popup_x_margin-0.5*close_button_w,\n",
    "                                 mouseevent.ydata+popup_y_shift+popup_box_h-popup_y_margin-0.5*close_button_h-close_button_marker_y_shift,\n",
    "                                 'x', \\\n",
    "                                 verticalalignment=\"center\", \\\n",
    "                                 horizontalalignment=\"center\", \\\n",
    "                                 transform=ax.transAxes, \\\n",
    "                                 color=\"gray\", \\\n",
    "                                 fontsize=10, \\\n",
    "                                 zorder=popup_zorder+3)\n",
    "        #close_button_x.set_label(feature_index)\n",
    "        \n",
    "        # Register artists for moving or later destruction\n",
    "        #\n",
    "        feature_artist_label_to_popup_box[feature_index] = []\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_box)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_info_1)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_info_2)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(popup_info_3)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(close_button)\n",
    "        feature_artist_label_to_popup_box[feature_index].append(close_button_x)\n",
    "\n",
    "    # Close popup (conflicts with close button)   \n",
    "    #\n",
    "    else: \n",
    "        popup_box = feature_artist_label_to_popup_box[feature_index][0]\n",
    "        popup_box.remove()\n",
    "        popup_info_1 = feature_artist_label_to_popup_box[feature_index][1]\n",
    "        popup_info_1.remove()\n",
    "        popup_info_2 = feature_artist_label_to_popup_box[feature_index][2]\n",
    "        popup_info_2.remove()\n",
    "        popup_info_3 = feature_artist_label_to_popup_box[feature_index][3]\n",
    "        popup_info_3.remove()\n",
    "        close_button = feature_artist_label_to_popup_box[feature_index][4]\n",
    "        close_button.remove()\n",
    "        close_button_x = feature_artist_label_to_popup_box[feature_index][5]\n",
    "        close_button_x.remove()\n",
    "        \n",
    "        feature_artist_label_to_popup_box[feature_index] = None\n",
    "        feature_index  = None\n",
    "    \n",
    "    artist.figure.canvas.draw()\n",
    "\n",
    "\n",
    "# Diplay coordinate transform\n",
    "#\n",
    "def disp_coord_transform (x, pivot_pos, genomebrowser_window_bp_width, genomebrowser_xshift, track_xshift, contig_mode_xshift):\n",
    "\n",
    "    track_disp_scale_factor = (1.0-right_margin-left_margin) / genomebrowser_window_bp_width\n",
    "    #transformed_x = track_disp_scale_factor * (x - (pivot_pos-0.5*genomebrowser_window_bp_width))\n",
    "    transformed_x = track_disp_scale_factor * (x - (pivot_pos+genomebrowser_xshift+track_xshift+contig_mode_xshift-0.5*genomebrowser_window_bp_width))\n",
    "\n",
    "    return left_margin + transformed_x\n",
    "\n",
    "\n",
    "# draw feature element\n",
    "#\n",
    "def draw_feature_element (ax, \\\n",
    "                     row_n, \\\n",
    "                     max_row_n, \\\n",
    "                     feature_i, \\\n",
    "                     feature_j, \\\n",
    "                     feature, \\\n",
    "                     pivot_pos, \\\n",
    "                     track_xshift, \\\n",
    "                     contig_mode_xshift, \\\n",
    "                     pivot_feature_flag, \\\n",
    "                     annot_repeat, \\\n",
    "                     seq_repeat, \\\n",
    "                     this_label_show_top, this_label_show_bot):\n",
    "    \n",
    "    left_cropped = False\n",
    "    right_cropped = False\n",
    "    feature_oi_flag = False\n",
    "    pivot_feature_flag = False\n",
    "    if feature['beg_pos'] <= pivot_pos and feature['end_pos'] >= pivot_pos:\n",
    "        pivot_feature_flag = True\n",
    "        \n",
    "    gene_id = feature['ID']\n",
    "    gene_name = feature['name']\n",
    "\n",
    "    # Determine if a search result feature\n",
    "    try:\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            contig_search_results = search_results[0]\n",
    "        else:\n",
    "            contig_search_results = search_results[feature_i]\n",
    "    except:\n",
    "        contig_search_results = []\n",
    "    for j,result_list in enumerate(contig_search_results):\n",
    "        for result in contig_search_results[j]:\n",
    "            if result['beg_pos'] == feature['beg_pos'] and result['end_pos'] == feature['end_pos']:\n",
    "                feature_oi_flag = True\n",
    "                highlight_color = search_color_names[j % len(search_color_names)]\n",
    "                break\n",
    "\n",
    "    # window coords\n",
    "    window_beg_pos = disp_coord_transform (feature['beg_pos'], \\\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "    window_end_pos = disp_coord_transform (feature['end_pos'],\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "    \n",
    "    # handle tiny number problems  (FIX: NOT QUITE WORKING)\n",
    "    if window_end_pos < window_beg_pos:\n",
    "        window_end_pos = window_beg_pos + .00000001\n",
    "    if window_beg_pos > 1.0-right_margin:\n",
    "        window_beg_pos = 1.0-right_margin\n",
    "    if window_end_pos < left_margin:\n",
    "        window_end_pos = left_margin\n",
    "\n",
    "    # handle features that exceed margins\n",
    "    if window_beg_pos < left_margin:\n",
    "        window_beg_pos = left_margin\n",
    "        left_cropped = True\n",
    "    if window_end_pos > 1.0-right_margin:\n",
    "        window_end_pos = 1.0-right_margin\n",
    "        right_cropped = True     \n",
    "        \n",
    "    # deal with short features\n",
    "    if feature['type'] != \"CRISPR\" and feature['type'] != \"CRISPR spacer\":\n",
    "        if window_end_pos - window_beg_pos < head_l:\n",
    "            feature_element_base_length = .00000001\n",
    "            arrow_params = {'head_width':head_w, 'head_length':window_end_pos-window_beg_pos}\n",
    "        else:\n",
    "            feature_element_base_length = window_end_pos - window_beg_pos - head_l\n",
    "            arrow_params = {'head_width':head_w, 'head_length':head_l}\n",
    "    elif feature['type'] == \"CRISPR\":\n",
    "        feature_element_base_length = .00000001\n",
    "        arrow_params = {'head_width':0.75*head_w, 'head_length':0.5*(window_end_pos-window_beg_pos)}\n",
    "    elif feature['type'] == \"CRISPR spacer\":\n",
    "        feature_element_base_length = window_end_pos - window_beg_pos\n",
    "\n",
    "        \n",
    "    # not going to change window_beg_pos again, so can set x_pos and y_pos now\n",
    "    x_pos = window_beg_pos   \n",
    "    # FIX\n",
    "    y_pos_base = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(total_rows-row_n)\n",
    "    #y_pos_base = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(max_row_n-row_n)\n",
    "    #y_pos_base = 1.0-0.02*row_n  # DEBUG\n",
    "\n",
    "    \n",
    "    # direction\n",
    "    if feature['strand'] == '-':\n",
    "        direction = 'rev'\n",
    "        feature_element_start_pos = x_pos+feature_element_base_length+head_l\n",
    "        #label_start_pos = x_pos+head_l\n",
    "        label_start_pos = x_pos\n",
    "        dir_adj = -1.0\n",
    "        drop_shadow_xshift = 1\n",
    "    else:\n",
    "        direction = 'fwd'\n",
    "        feature_element_start_pos = x_pos\n",
    "        label_start_pos = x_pos\n",
    "        dir_adj = 1.0\n",
    "        drop_shadow_xshift = 0\n",
    "\n",
    "        \n",
    "    # color\n",
    "    #\n",
    "    feature_element_color = \"lightgray\"\n",
    "    if feature['type'] == \"pseudogene\":\n",
    "        feature_element_color = \"gray\"\n",
    "        \n",
    "    elif feature['type'] == \"rRNA\" or feature['type'] == \"tRNA\":\n",
    "        feature_element_color = \"black\"\n",
    "        \n",
    "    elif feature['type'] == \"CRISPR\":\n",
    "        feature_element_color = \"black\"        \n",
    "        if feature['fwd_dna_seq'] in seq_repeat:\n",
    "            feature_element_color = color_names[sum([ord(c) for c in feature['fwd_dna_seq']]) % len(color_names)]           \n",
    "    elif feature['type'] == \"CRISPR spacer\":\n",
    "        feature_element_color = \"lightgray\"\n",
    "        if feature['fwd_dna_seq'] in seq_repeat:\n",
    "            feature_element_color = color_names[sum([ord(c) for c in feature['fwd_dna_seq']]) % len(color_names)]\n",
    "            \n",
    "    elif feature['type'] == \"CDS\":\n",
    "\n",
    "        if Global_State['genomebrowser_color_namespace'] == \"annot\":\n",
    "            if 'annot' in feature:\n",
    "                if feature['annot'] != '' and \\\n",
    "                    (feature['annot'].lower() == \"hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"conserved hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"orf, hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"orf, conserved hypothetical protein\" \\\n",
    "                    or feature['annot'].lower() == \"hypothetical protein; orphan\" \\\n",
    "                    or feature['annot'].lower() == \"hypothetical protein; genus orphan\"):\n",
    "                    \n",
    "                    feature_element_color = \"darkgray\"\n",
    "                elif feature['annot'] == '':\n",
    "                    feature_element_color = \"lightgray\"\n",
    "                else:\n",
    "                    if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                        or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                        if feature['annot'] in annot_repeat:\n",
    "                            feature_element_color = color_names[sum([ord(c) for c in feature['annot']]) % len(color_names)]\n",
    "                    else:\n",
    "                        feature_element_color = color_names[sum([ord(c) for c in feature['annot']]) % len(color_names)]\n",
    "\n",
    "        elif Global_State['genomebrowser_color_namespace'] == \"ec\":\n",
    "            if 'EC_number' in feature and feature['EC_number'] != '':\n",
    "                if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                    or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                    if feature['EC_number'] in annot_repeat:\n",
    "                        feature_element_color = color_names[sum([ord(c) for c in feature['EC_number']]) % len(color_names)]\n",
    "                else:\n",
    "                    feature_element_color = color_names[sum([ord(c) for c in feature['EC_number']]) % len(color_names)]\n",
    "\n",
    "        elif Global_State['genomebrowser_color_namespace'] == 'cog' \\\n",
    "            or Global_State['genomebrowser_color_namespace'] == 'pfam':\n",
    "                \n",
    "            if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "                contig_i = 0\n",
    "            else:\n",
    "                contig_i = feature_i\n",
    "            try:\n",
    "                domain_hits = Global_Domains[contig_i][feature['ID']]\n",
    "            except:\n",
    "                domain_hits = []\n",
    "                \n",
    "            top_domfam = None                \n",
    "            for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "                domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "                if (Global_State['genomebrowser_color_namespace'] == 'cog' and domfam[0:3] == 'COG') \\\n",
    "                    or (Global_State['genomebrowser_color_namespace'] == 'pfam' and domfam[0:2] == 'PF'):\n",
    "                        \n",
    "                    if (Global_State['genomebrowser_mode'] == \"tree\" or Global_State['genomebrowser_mode'] == \"homologs\") \\\n",
    "                          and not domfam in annot_repeat:\n",
    "                        continue                            \n",
    "                    top_domfam = domfam\n",
    "                    break\n",
    "                            \n",
    "            if top_domfam != None:\n",
    "                if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                    or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                    if top_domfam in annot_repeat:  # already checked, but safer to repeat in case logic above changes\n",
    "                        feature_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    "                else:\n",
    "                    feature_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    " \n",
    "        elif Global_State['genomebrowser_color_namespace'] == 'domains':\n",
    "            #feature_element_color = \"none\"\n",
    "            feature_element_color = \"lightgray\"\n",
    "            # individual domains will be drawn separately with rects below\n",
    "\n",
    "            \n",
    "    # edge color\n",
    "    feature_element_edge_lw = 1\n",
    "    feature_element_edge_color = \"none\"\n",
    "#    if (Global_State['genomebrowser_color_namespace'] == 'domains'):\n",
    "#        feature_element_edge_color = \"black\"\n",
    "    if pivot_feature_flag:\n",
    "        if Global_State['genomebrowser_mode'] == 'contigs':\n",
    "            feature_element_edge_color = \"black\"\n",
    "            feature_element_edge_lw = 1\n",
    "        else:\n",
    "            feature_element_edge_color = \"black\"\n",
    "            feature_element_edge_lw = 2\n",
    "    if feature_oi_flag:\n",
    "        feature_element_edge_color = \"black\"\n",
    "        feature_element_edge_lw = 2        \n",
    "                                \n",
    "                    \n",
    "    # Draw Rectangle (patch) or Arrow (from pylab) for features\n",
    "    #\n",
    "    if feature['type'] == \"CRISPR\":\n",
    "        new_arrow_l = ax.arrow (feature_element_start_pos+0.5*(window_end_pos-window_beg_pos), \\\n",
    "                          y_pos_base, \\\n",
    "                          -feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=0.75*arrow_w, \\\n",
    "                          color=feature_element_color, \\\n",
    "                          ec=\"none\",\n",
    "                          #lw=feature_element_edge_lw, \\\n",
    "                          alpha=0.75, \\\n",
    "                          zorder=1, \\\n",
    "                          picker=False, \\\n",
    "                          **arrow_params)\n",
    "        new_arrow_r = ax.arrow (feature_element_start_pos+0.5*(window_end_pos-window_beg_pos), \\\n",
    "                          y_pos_base, \\\n",
    "                          feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=0.75*arrow_w, \\\n",
    "                          color=feature_element_color, \\\n",
    "                          ec=\"none\", \\\n",
    "                          #lw=feature_element_edge_lw, \\\n",
    "                          alpha=0.75, \\\n",
    "                          zorder=1, \\\n",
    "                          picker=False, \\\n",
    "                          **arrow_params)\n",
    "        # FIX: listeners do work, but seems excessive so turning off for now\n",
    "        #feature_index = \"%d,%d\"%(feature_i,feature_j)\n",
    "        #feature_artist_label_to_feature[feature_index] = feature\n",
    "        #new_arrow_l.figure.canvas.mpl_connect ('pick_event', onpick_feature)\n",
    "        #new_arrow_l.gid = feature_index\n",
    "        #new_arrow_l.set_label(\"feature\")\n",
    "        #new_arrow_r.figure.canvas.mpl_connect ('pick_event', onpick_feature)  \n",
    "        #new_arrow_r.gid = feature_index  # not sure I can have 2 elements with the same gid?\n",
    "        new_arrow_r.set_label(\"feature\")\n",
    "        \n",
    "    elif feature['type'] == \"CRISPR spacer\":\n",
    "        rect_feature = Rectangle((x_pos, y_pos_base-0.325*arrow_w), \\\n",
    "                                 feature_element_base_length, 0.75*arrow_w, \\\n",
    "                                 facecolor=feature_element_color, ec=\"none\", alpha=0.75, zorder=2, \\\n",
    "                                 picker=False)\n",
    "        # FIX: listeners don't work as written below, but rect picker works elsewhere so should be doable\n",
    "        #feature_index = \"%d,%d\"%(feature_i,feature_j)\n",
    "        #feature_artist_label_to_feature[feature_index] = feature\n",
    "        #rect_feature.figure.canvas.mpl_connect ('pick_event', onpick_feature)\n",
    "        #rect_feature.gid = feature_index\n",
    "        #rect_feature.set_label(\"feature\")\n",
    "        ax.add_patch(rect_feature)\n",
    "        \n",
    "    else:\n",
    "        new_arrow = ax.arrow (feature_element_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=arrow_w, \\\n",
    "                          #color=feature_element_color, \\\n",
    "                          #ec=feature_element_edge_color, \\\n",
    "                          #lw=feature_element_edge_lw, \\\n",
    "                          #alpha=0.75, \\\n",
    "                          zorder=1, \\\n",
    "                          picker=True, \\\n",
    "                          **arrow_params)\n",
    "        \n",
    "        # Drop shadow baby!\n",
    "        new_arrow.set_path_effects([path_effects.PathPatchEffect(offset=(drop_shadow_xshift, -1),\n",
    "                                                                 facecolor='black', alpha=0.4),\n",
    "                                    path_effects.PathPatchEffect(facecolor=feature_element_color,\n",
    "                                                                 edgecolor=feature_element_edge_color,\n",
    "                                                                 lw=feature_element_edge_lw)])\n",
    "                                                         \n",
    "\n",
    "        # add popup onclick\n",
    "        #\n",
    "        # the object approach causes iPython to hang.  Use artist label to index into hashes instead\n",
    "        #ax.add_patch (new_arrow)\n",
    "        #active_arrow = InteractiveGenomeFeature (new_arrow, feature)\n",
    "        #active_arrow.connect()\n",
    "        #\n",
    "        feature_index = \"%d,%d\"%(feature_i,feature_j)\n",
    "        feature_artist_label_to_feature[feature_index] = feature    \n",
    "        new_arrow.figure.canvas.mpl_connect ('pick_event', onpick_feature)\n",
    "        new_arrow.gid = feature_index\n",
    "        new_arrow.set_label(\"feature\")\n",
    "    \n",
    "\n",
    "    # Add Domains\n",
    "    #\n",
    "    if Global_State['genomebrowser_color_namespace'] == 'domains':\n",
    "        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "            contig_i = 0\n",
    "        else:\n",
    "            contig_i = feature_i\n",
    "        try:\n",
    "            domain_hits = Global_Domains[contig_i][feature['ID']]\n",
    "        except:\n",
    "            domain_hits = []\n",
    "\n",
    "        uncovered_frac_thresh = 0.5\n",
    "        domhits_disp = []\n",
    "        # needs to extend beyond by 1 because protein numbering starting from 1\n",
    "        covered_mask = []\n",
    "        for i in range(0,int(round(feature['end_pos']-feature['beg_pos']+1)/3)+2):\n",
    "            covered_mask.append(False)\n",
    "\n",
    "        for dom_namespace in ['TIGR', 'PF', 'COG']:\n",
    "#        for dom_namespace in ['TIGR']:\n",
    "#        for dom_namespace in ['PF']:\n",
    "#        for dom_namespace in ['COG']:\n",
    "            for domhit in domain_hits:  # already reverse sorted by bitscore\n",
    "                domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "                if domfam[0:len(dom_namespace)] != dom_namespace:\n",
    "                    continue\n",
    "                if (Global_State['genomebrowser_mode'] == \"tree\" or Global_State['genomebrowser_mode'] == \"homologs\") \\\n",
    "                      and not domfam in annot_repeat:\n",
    "                    continue\n",
    "                    \n",
    "                occluded = False\n",
    "                uncovered_pos_cnt = 0\n",
    "                for i in range(domhit[DOMHIT_BEG_I],domhit[DOMHIT_END_I]+1):\n",
    "                    if not covered_mask[i]:\n",
    "                        uncovered_pos_cnt += 1\n",
    "                    \n",
    "                if uncovered_pos_cnt/(domhit[DOMHIT_END_I]-domhit[DOMHIT_BEG_I]+1) < uncovered_frac_thresh:\n",
    "                    occluded = True\n",
    "                            \n",
    "                if not occluded:\n",
    "                    domhits_disp.append(domhit)\n",
    "                    for i in range(domhit[DOMHIT_BEG_I],domhit[DOMHIT_END_I]+1):\n",
    "                        covered_mask[i] = True\n",
    "\n",
    "        dom_zorder = 3                    \n",
    "        for domhit in sorted(domhits_disp, key=sort_by_domhit_len, reverse=True):\n",
    "            top_domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "            \n",
    "            if Global_State['genomebrowser_mode'] == \"tree\" \\\n",
    "                  or Global_State['genomebrowser_mode'] == \"homologs\":\n",
    "                if top_domfam in annot_repeat:  # already checked, but safer to repeat in case logic above changes\n",
    "                    dom_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    "            else:\n",
    "                dom_element_color = color_names[sum([ord(c) for c in top_domfam]) % len(color_names)]\n",
    "     \n",
    "            # get coords\n",
    "            if direction == 'fwd':\n",
    "                dom_beg_pos = feature['beg_pos']+3*(domhit[DOMHIT_BEG_I]-1)\n",
    "                dom_end_pos = feature['beg_pos']+3*(domhit[DOMHIT_END_I]-1)\n",
    "            else:\n",
    "                dom_beg_pos = feature['end_pos']-3*(domhit[DOMHIT_END_I]-1)\n",
    "                dom_end_pos = feature['end_pos']-3*(domhit[DOMHIT_BEG_I]-1)\n",
    "                    \n",
    "            dom_window_beg_pos = disp_coord_transform (dom_beg_pos, \\\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "            dom_window_end_pos = disp_coord_transform (dom_end_pos, \\\n",
    "                                           pivot_pos, \\\n",
    "                                           Global_State['genomebrowser_window_bp_width'], \\\n",
    "                                           Global_State['genomebrowser_xshift'], \\\n",
    "                                           track_xshift, \\\n",
    "                                           contig_mode_xshift)\n",
    "            \n",
    "            if dom_window_end_pos <= left_margin or dom_window_beg_pos >= 1.0-right_margin:\n",
    "                continue\n",
    "            if dom_window_beg_pos < left_margin:\n",
    "                dom_window_beg_pos = left_margin\n",
    "            if dom_window_end_pos > 1.0-right_margin:\n",
    "                dom_window_end_pos = 1.0-right_margin\n",
    "\n",
    "            # rect representation  # DEBUG\n",
    "            \"\"\"\n",
    "            ##y_shift_tweak = -0.0005  # rectangles and arrows don't necessarily wind up on same pixel for close calls\n",
    "            y_shift_tweak = 0.0  # rectangles and arrows don't necessarily wind up on same pixel for close calls\n",
    "            dom_x_pos = dom_window_beg_pos\n",
    "            dom_y_pos = y_pos_base - 0.5*arrow_w + y_shift_tweak\n",
    "            dom_w = dom_window_end_pos - dom_window_beg_pos + .00000001\n",
    "            dom_h = arrow_w\n",
    "                    \n",
    "            rect_feature = Rectangle((dom_x_pos, dom_y_pos), dom_w, dom_h, \\\n",
    "                                     facecolor=dom_element_color, ec=\"black\", lw=1, alpha=1.0, zorder=dom_zorder, \\\n",
    "                                     picker=False)  # Probably will want picker\n",
    "            ax.add_patch(rect_feature)\n",
    "            \"\"\"\n",
    "            \n",
    "            # arrow representation of domains\n",
    "            #\n",
    "            dom_y_scale = 0.9\n",
    "            #dom_y_scale = 1.0  # DEBUG\n",
    "            dom_arrow_w = dom_y_scale * arrow_w\n",
    "            dom_element_edge_color = 'none'\n",
    "            dom_element_edge_lw = 0\n",
    "            add_tip_crop = False\n",
    "                \n",
    "            # direction and position            \n",
    "            if direction == 'fwd':\n",
    "                \n",
    "                # domain completely in arrow head\n",
    "                if window_end_pos - dom_window_beg_pos <= head_l:\n",
    "                    dom_element_base_length = .00000001\n",
    "                    dom_head_l = window_end_pos - dom_window_beg_pos\n",
    "                    if window_end_pos - window_beg_pos <= head_l:\n",
    "                        this_head_l = window_end_pos - window_beg_pos\n",
    "                    else:\n",
    "                        this_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * dom_head_l * (head_w/this_head_l)\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_end_pos < window_end_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_end_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = window_end_pos - dom_window_end_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/this_head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                # domain completely *not* in arrow head\n",
    "                elif dom_window_end_pos <= window_end_pos - head_l:\n",
    "                    dom_element_base_length = dom_window_end_pos - dom_window_beg_pos\n",
    "                    dom_head_l = 0\n",
    "                    dom_head_w = 0\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "\n",
    "                # domain in both regions\n",
    "                else:  # dom_window_beg_pos <= window_end_pos - head_l and dom_window_end_pos > window_end_pos - head_l\n",
    "                    dom_element_base_length = window_end_pos - dom_window_beg_pos - head_l\n",
    "                    if dom_element_base_length < 0:\n",
    "                        dom_element_base_length = .00000001\n",
    "                    dom_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * head_w\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_end_pos < window_end_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_end_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = window_end_pos - dom_window_end_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                dom_element_start_pos = dom_window_beg_pos\n",
    "                dir_adj = 1.0\n",
    "                \n",
    "            else:  # direction == \"rev\"  \n",
    "                \n",
    "                # domain completely in arrow head\n",
    "                if dom_window_end_pos - window_beg_pos <= head_l:\n",
    "                    dom_element_base_length = .00000001\n",
    "                    dom_head_l = dom_window_end_pos - window_beg_pos\n",
    "                    if window_end_pos - window_beg_pos <= head_l:\n",
    "                        this_head_l = window_end_pos - window_beg_pos\n",
    "                    else:\n",
    "                        this_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * dom_head_l * (head_w/this_head_l)\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_beg_pos > window_beg_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_beg_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = dom_window_beg_pos - window_beg_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/this_head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                # domain completely *not* in arrow head\n",
    "                elif dom_window_beg_pos >= window_beg_pos + head_l:\n",
    "                    dom_element_base_length = dom_window_end_pos - dom_window_beg_pos\n",
    "                    dom_head_l = head_l\n",
    "                    dom_head_w = 0\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':0}\n",
    "\n",
    "                # domain in both regions\n",
    "                else:  # dom_window_beg_pos <= window_end_pos - head_l and dom_window_end_pos > window_end_pos - head_l\n",
    "                    dom_element_base_length = dom_window_end_pos - window_beg_pos - head_l\n",
    "                    if dom_element_base_length < 0:\n",
    "                        dom_element_base_length = .00000001\n",
    "                    dom_head_l = head_l\n",
    "                    dom_head_w = dom_y_scale * head_w\n",
    "                    dom_arrow_params = {'head_width':dom_head_w, 'head_length':dom_head_l}\n",
    "                    if dom_window_beg_pos > window_beg_pos:\n",
    "                        add_tip_crop = True\n",
    "                        tip_start_pos = dom_window_beg_pos\n",
    "                        tip_base_length = .00000001\n",
    "                        tip_head_l = dom_window_beg_pos - window_beg_pos\n",
    "                        tip_head_w = tip_head_l * (head_w/head_l)\n",
    "                        tip_arrow_params = {'head_width':tip_head_w, 'head_length':tip_head_l}\n",
    "                        \n",
    "                dom_element_start_pos = window_beg_pos+dom_element_base_length+dom_head_l\n",
    "                dir_adj = -1.0 \n",
    "\n",
    "                \n",
    "            # draw domain element \n",
    "            #\n",
    "            dom_arrow = ax.arrow (dom_element_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*dom_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=dom_arrow_w, \\\n",
    "                          color=dom_element_color, \\\n",
    "                          ec=dom_element_edge_color, \\\n",
    "                          #lw=dom_element_edge_lw, \\\n",
    "                          alpha=1.0, \\\n",
    "                          zorder=dom_zorder, \\\n",
    "                          picker=False, \\\n",
    "                          **dom_arrow_params)\n",
    "            if add_tip_crop:\n",
    "                tip_arrow = ax.arrow (tip_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*tip_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=dom_arrow_w, \\\n",
    "                          color=feature_element_color, \\\n",
    "                          ec='none', \\\n",
    "                          #lw=dom_element_edge_lw, \\\n",
    "                          alpha=1.0, \\\n",
    "                          zorder=dom_zorder+1, \\\n",
    "                          picker=False, \\\n",
    "                          **tip_arrow_params)\n",
    "            \n",
    "            dom_zorder += 2\n",
    "            \n",
    "        # restore border for highlighted features           \n",
    "        if pivot_feature_flag or feature_oi_flag:\n",
    "            border_arrow = ax.arrow (feature_element_start_pos, \\\n",
    "                          y_pos_base, \\\n",
    "                          dir_adj*feature_element_base_length, \\\n",
    "                          0.0, \\\n",
    "                          width=arrow_w, \\\n",
    "                          color='none', \\\n",
    "                          ec=feature_element_edge_color, \\\n",
    "                          lw=feature_element_edge_lw, \\\n",
    "                          alpha=1.0, \\\n",
    "                          zorder=dom_zorder+1, \\\n",
    "                          picker=False, \\\n",
    "                          **arrow_params)\n",
    "                \n",
    "    # Feature label\n",
    "    #\n",
    "    if Global_State['genomebrowser_window_bp_width'] <= text_disp_window_bp_limit or pivot_feature_flag:\n",
    "        \n",
    "        label_color = \"black\"\n",
    "        arrow_label_fontsize = base_arrow_label_fontsize        \n",
    "        \n",
    "        # indicate pivot feature\n",
    "        if pivot_feature_flag and Global_State['genomebrowser_mode'] != \"contigs\":\n",
    "            #label_color = 'red'\n",
    "            label_color = 'black'            \n",
    "            arrow_label_fontsize = foi_arrow_label_fontsize\n",
    "            if this_label_show_top == False and this_label_show_bot == False:\n",
    "                this_label_show_top = True\n",
    "\n",
    "        if feature_oi_flag:\n",
    "            label_color = highlight_color            \n",
    "            arrow_label_fontsize = foi_arrow_label_fontsize\n",
    "            if this_label_show_top == False and this_label_show_bot == False:\n",
    "                this_label_show_top = True\n",
    "\n",
    "        # label position\n",
    "        label_y_pos = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(total_rows-row_n)\n",
    "\n",
    "        if Global_State['genomebrowser_window_bp_width'] > text_disp_window_bp_limit or this_label_show_top:\n",
    "            vert_align = \"bottom\"\n",
    "            label_y_pos = y_pos_base + 0.9*text_yshift\n",
    "            if direction == \"rev\":\n",
    "                label_start_pos += .0015\n",
    "            #else:\n",
    "            #    label_start_pos += .0005\n",
    "        else:\n",
    "            vert_align = \"top\"\n",
    "            label_y_pos = y_pos_base - 1.25*text_yshift\n",
    "            if direction == \"rev\":\n",
    "                label_start_pos += .0050\n",
    "            #else:\n",
    "            #    label_start_pos += .0005\n",
    "\n",
    "        # draw label\n",
    "        if pivot_feature_flag or feature_oi_flag or this_label_show_top or this_label_show_bot:\n",
    "            feature_label = ax.text(label_start_pos, \\\n",
    "                                    label_y_pos, \\\n",
    "                                    gene_name, \\\n",
    "                                    #row_n, \\  # DEBUG\n",
    "                                    verticalalignment=vert_align, \\\n",
    "                                    horizontalalignment=\"left\", \\\n",
    "                                    transform=ax.transAxes, \\\n",
    "                                    color=label_color, \\\n",
    "                                    fontsize=arrow_label_fontsize, \\\n",
    "                                    zorder=1)\n",
    "\n",
    "        #print (feature['source_species']+\" \"+feature['name']+\" %f %f %f\"%(arrow_start_pos, arrow_base_length, y_pos_base))\n",
    "        \n",
    "    # add ellipsis to indicate cropped feature\n",
    "    #\n",
    "    if left_cropped:\n",
    "        ellipsis_spacing = 0.2 * left_margin\n",
    "        x_center_base = left_margin - 2*ellipsis_spacing\n",
    "        y_center_base = y_pos_base\n",
    "        diameter = 0.20 * arrow_w\n",
    "        ellipse_to_circle_scaling = 0.75*figure_width / (figure_height_scaling*(total_rows+1))        \n",
    "        x_diameter = 1.0 * diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * diameter\n",
    "        for i in range(0,3):\n",
    "            ellipse_center = (x_center_base-i*ellipsis_spacing, y_center_base)\n",
    "            ellipsis_dot_patch = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                         facecolor=feature_element_color, edgecolor=feature_element_color, \\\n",
    "                                         lw=1, alpha=1.0, zorder=1)\n",
    "        \n",
    "            ax.add_patch(ellipsis_dot_patch)\n",
    "            \n",
    "    if right_cropped:\n",
    "        ellipsis_spacing = 0.2 * left_margin\n",
    "        x_center_base = 1.0-right_margin + 2*ellipsis_spacing\n",
    "        y_center_base = y_pos_base\n",
    "        diameter = 0.20 * arrow_w\n",
    "        ellipse_to_circle_scaling = 0.75*figure_width / (figure_height_scaling*(total_rows+1))        \n",
    "        x_diameter = 1.0 * diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * diameter\n",
    "        for i in range(0,3):\n",
    "            ellipse_center = (x_center_base+i*ellipsis_spacing, y_center_base)\n",
    "            ellipsis_dot_patch = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                         facecolor=feature_element_color, edgecolor=feature_element_color, \\\n",
    "                                         lw=1, alpha=1.0, zorder=1)\n",
    "        \n",
    "            ax.add_patch(ellipsis_dot_patch)\n",
    "\n",
    "\n",
    "# Paint methods\n",
    "#\n",
    "def draw_mode_panel (ax, genome_name, genomebrowser_mode, data_set_name):\n",
    "    \n",
    "    # Config\n",
    "    field_name_fontsize=12\n",
    "    field_val_fontsize=11\n",
    "    field_color=\"white\"\n",
    "    value_color=\"lightsalmon\"\n",
    "    #shadow_color=\"magenta\"\n",
    "    shadow_color=\"blue\"\n",
    "    #shadow_color=\"gray\"\n",
    "    base_zorder=5\n",
    "\n",
    "    # Draw background with rounded corners\n",
    "    #\n",
    "    bb = mtransforms.Bbox([[0.06, 0.10], [0.94, 0.89]])\n",
    "    bb_shadow = mtransforms.Bbox([[0.07, 0.08], [0.95, 0.87]])\n",
    "\n",
    "    p_fancy = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.04\",\n",
    "                             mutation_aspect=0.25*figure_width/top_nav_height,\n",
    "                             fc=nav_bg_color,\n",
    "                             ec='white',\n",
    "                             lw=2,\n",
    "                             zorder=base_zorder)\n",
    "    ax.add_patch(p_fancy)\n",
    "    \"\"\"\n",
    "    p_fancy_back = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.06\",\n",
    "                             mutation_aspect=0.25*figure_width/top_nav_height,\n",
    "                             fc='white',\n",
    "                             ec='black',\n",
    "                             lw=1,\n",
    "                             zorder=base_zorder-1)\n",
    "    ax.add_patch(p_fancy_back)\n",
    "    \"\"\"\n",
    "    p_fancy_shadow = FancyBboxPatch((bb_shadow.xmin, bb_shadow.ymin),\n",
    "                             abs(bb_shadow.width), abs(bb_shadow.height),\n",
    "                             boxstyle=\"round,pad=0.04\",\n",
    "                             mutation_aspect=0.25*figure_width/top_nav_height,\n",
    "                             fc=shadow_color,\n",
    "                             ec=shadow_color,\n",
    "                             lw=1,\n",
    "                             alpha=0.5,\n",
    "                             zorder=base_zorder-2)\n",
    "    ax.add_patch(p_fancy_shadow)\n",
    "    \n",
    "    # Tool title\n",
    "    ax.text(0.05, \\\n",
    "            0.9, \\\n",
    "            tool_title, \\\n",
    "            verticalalignment=\"top\", \\\n",
    "            horizontalalignment=\"left\", \\\n",
    "            #transform=ax_top_left.transAxes, \\\n",
    "            color=\"white\", \\\n",
    "            style=\"normal\", # style=\"oblique\", \\\n",
    "            fontsize=14,\n",
    "            zorder=base_zorder+1)\n",
    "    \n",
    "    # Genome\n",
    "    ax.text(0.30, 0.5, \"Genome\", verticalalignment=\"bottom\", horizontalalignment=\"right\", color=field_color, fontsize=field_name_fontsize, zorder=base_zorder+1)\n",
    "    ax.text(0.35, 0.5, genome_name, verticalalignment=\"bottom\", horizontalalignment=\"left\", color=value_color, fontsize=field_val_fontsize, zorder=base_zorder+1)\n",
    "        \n",
    "    # Mode\n",
    "    ax.text(0.30, 0.3, \"Mode\", verticalalignment=\"bottom\", horizontalalignment=\"right\", color=field_color, fontsize=field_name_fontsize, zorder=base_zorder+1)\n",
    "    ax.text(0.35, 0.3, genomebrowser_mode, verticalalignment=\"bottom\", horizontalalignment=\"left\", color=value_color, fontsize=field_val_fontsize, zorder=base_zorder+1)\n",
    "\n",
    "    # Dataset\n",
    "    #ax.text(0.30, 0.1, \"Data\", verticalalignment=\"bottom\", horizontalalignment=\"right\", color=field_color, fontsize=field_name_fontsize, zorder=base_zorder+1)\n",
    "    #ax.text(0.35, 0.1, data_set_name, verticalalignment=\"bottom\", horizontalalignment=\"left\", color=value_color, fontsize=field_val_fontsize, zorder=base_zorder+1)\n",
    "\n",
    "    \n",
    "def draw_control_panel (ax, genomebrowser_zoom, genomebrowser_xshift, genomebrowser_color_namespace):\n",
    "\n",
    "    # Config\n",
    "    text_color = \"white\"\n",
    "    base_zorder = 5\n",
    "    #shadow_color = 'magenta'\n",
    "    shadow_color = 'blue'\n",
    "    #shadow_color = 'gray'\n",
    "    \n",
    "    # Draw background with rounded corners\n",
    "    #\n",
    "    bb = mtransforms.Bbox([[0.02, 0.11], [0.98, 0.88]])\n",
    "    bb_shadow = mtransforms.Bbox([[0.0225, 0.09], [0.9825, 0.86]])\n",
    "\n",
    "    p_fancy = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.015\",\n",
    "                             mutation_aspect=0.75*figure_width/top_nav_height,\n",
    "                             fc=nav_bg_color,\n",
    "                             ec='white',\n",
    "                             lw=2,\n",
    "                             zorder=base_zorder)\n",
    "    ax.add_patch(p_fancy)\n",
    "    \"\"\"\n",
    "    p_fancy_back = FancyBboxPatch((bb.xmin, bb.ymin),\n",
    "                             abs(bb.width), abs(bb.height),\n",
    "                             boxstyle=\"round,pad=0.020\",\n",
    "                             mutation_aspect=0.75*figure_width/top_nav_height,\n",
    "                             fc='white',\n",
    "                             ec='black',\n",
    "                             lw=1,\n",
    "                             zorder=base_zorder-1)\n",
    "    ax.add_patch(p_fancy_back)\n",
    "    \"\"\"\n",
    "    p_fancy_shadow = FancyBboxPatch((bb_shadow.xmin, bb_shadow.ymin),\n",
    "                             abs(bb_shadow.width), abs(bb_shadow.height),\n",
    "                             boxstyle=\"round,pad=0.015\",\n",
    "                             mutation_aspect=0.75*figure_width/top_nav_height,\n",
    "                             fc=shadow_color,\n",
    "                             ec=shadow_color,\n",
    "                             lw=1,\n",
    "                             alpha=0.5,\n",
    "                             zorder=base_zorder-2)\n",
    "    ax.add_patch(p_fancy_shadow)\n",
    "\n",
    "    \n",
    "    # Color slider\n",
    "    #\n",
    "    slider_name = \"Color\"\n",
    "    slider_name_h_align = \"right\"\n",
    "    tic_label_v_align=\"bottom\"\n",
    "    #slider_x = 0.64\n",
    "    slider_x = 0.13\n",
    "    slider_y = 0.55\n",
    "    #track_w = 0.25\n",
    "    #track_w = 0.45\n",
    "    track_w = 0.40\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    slider_margin = 0.02\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    \n",
    "    num_tics = len(color_namespace_names)\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    knob_initial_pos = 0\n",
    "    for i in range(0,num_tics):\n",
    "        knob_initial_pos = i\n",
    "        if Global_State['genomebrowser_color_namespace'] == color_namespace_names[i]:\n",
    "            break\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        tic_labels.append(color_namespace_names_disp[i])\n",
    "    \n",
    "    ax.text(slider_x-1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y+slider_h+1.00*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    #color_knob_patch = Rectangle((slider_x+0.5*track_w-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=4)\n",
    "    color_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5)\n",
    "    ax.add_patch (color_knob_patch)\n",
    "    color_knob = DraggableRectangleSlider(color_knob_patch, func='color', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    color_knob.connect()\n",
    "    \n",
    "    \n",
    "    # Mode slider  \n",
    "    #\n",
    "    slider_name = \"Mode\"\n",
    "    slider_name_h_align = \"left\"\n",
    "    tic_label_v_align=\"bottom\"\n",
    "    #slider_x = 0.13\n",
    "    slider_x = 0.64\n",
    "    slider_y = 0.55\n",
    "    #track_w = 0.45\n",
    "    track_w = 0.25\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    slider_margin = 0.02\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    num_tics = len(mode_names)\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    knob_initial_pos = 0\n",
    "    for i in range(0,num_tics):\n",
    "        knob_initial_pos = i\n",
    "        if Global_State['genomebrowser_mode'] == mode_names[i]:\n",
    "            break\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        tic_labels.append(mode_names_disp[i])\n",
    "\n",
    "    ax.text(slider_x+track_w+1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y+slider_h+1.00*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    mode_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5, transform=ax.transAxes)\n",
    "    ax.add_patch (mode_knob_patch)\n",
    "    mode_knob = DraggableRectangleSlider(mode_knob_patch, func='mode', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    mode_knob.connect()\n",
    "    \n",
    "    \n",
    "    # Position slider  \n",
    "    #\n",
    "    slider_name = \"Position\"\n",
    "    slider_name_h_align = \"right\"\n",
    "    tic_label_v_align=\"top\"\n",
    "    pan_button_w = 0.025\n",
    "    pan_button_h = 0.2\n",
    "    pan_button_margin = 0.01\n",
    "    slider_margin = 0.02\n",
    "    slider_x = 0.13 + 3*(pan_button_w+pan_button_margin)\n",
    "    slider_y = 0.25\n",
    "    #track_w = 0.45 - 3*(pan_button_w+pan_button_margin)\n",
    "    track_w = 0.40 - 3*(pan_button_w+pan_button_margin)\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    num_tics = 11\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    label_bp_disp = 1000000\n",
    "    genome_step = int(round(Global_State['PrimaryContig_len'] / (num_tics-1)))\n",
    "    knob_initial_pos = (num_tics-1)*(1.0-(Global_State['PrimaryContig_len']-(Global_State['PrimaryAnchor_pivot_pos']+Global_State['genomebrowser_xshift']))/Global_State['PrimaryContig_len'])\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        if i == num_tics-1:\n",
    "            tic_val = \"%5.1f\"%(Global_State['PrimaryContig_len']/label_bp_disp)\n",
    "        else:\n",
    "        #    tic_val = i*genome_step/label_bp_disp      \n",
    "        #tic_labels.append(\"%4.1f\"%tic_val)\n",
    "             tic_val = \"%5.1f\"%(i*genome_step/label_bp_disp)        \n",
    "        tic_labels.append(tic_val.strip())\n",
    "        \n",
    "    ax.text(slider_x-3*(pan_button_w+pan_button_margin)-1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y-1.25*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    pos_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5, transform=ax.transAxes)\n",
    "    ax.add_patch (pos_knob_patch)\n",
    "    pos_knob = DraggableRectangleSlider(pos_knob_patch, func='pan', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    pos_knob.connect()\n",
    "    \n",
    "    # pan buttons\n",
    "    pan_button_left = Rectangle((slider_x-3*(pan_button_w+pan_button_margin)-slider_margin, slider_y), pan_button_w, pan_button_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+3, transform=ax.transAxes, picker=True)\n",
    "    ax.text(slider_x-3*(pan_button_w+pan_button_margin)-slider_margin+0.5*pan_button_w, slider_y+0.5*pan_button_h, \"<\", verticalalignment=\"center\", horizontalalignment=\"center\", color=\"white\", fontsize=12, zorder=base_zorder+4)\n",
    "    pan_button_left.set_label(\"pan_button\")\n",
    "    pan_button_left.gid = \"left_pan\"\n",
    "    ax.add_patch(pan_button_left)\n",
    "\n",
    "    pan_button_reset = Rectangle((slider_x-2*(pan_button_w+pan_button_margin)-slider_margin, slider_y), pan_button_w, pan_button_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+3, transform=ax.transAxes, picker=True)\n",
    "    ax.text(slider_x-2*(pan_button_w+pan_button_margin)-slider_margin+0.5*pan_button_w, slider_y+0.5*pan_button_h, \"o\", verticalalignment=\"center\", horizontalalignment=\"center\", color=\"white\", fontsize=14, zorder=base_zorder+4)\n",
    "    pan_button_reset.set_label(\"pan_button\")\n",
    "    pan_button_reset.gid = \"reset_pan\"\n",
    "    ax.add_patch(pan_button_reset)\n",
    "    \n",
    "    pan_button_right = Rectangle((slider_x-1*(pan_button_w+pan_button_margin)-slider_margin, slider_y), pan_button_w, pan_button_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+3, transform=ax.transAxes, picker=True)\n",
    "    ax.text(slider_x-1*(pan_button_w+pan_button_margin)-slider_margin+0.5*pan_button_w, slider_y+0.5*pan_button_h, \">\", verticalalignment=\"center\", horizontalalignment=\"center\", color=\"white\", fontsize=12, zorder=base_zorder+4)\n",
    "    pan_button_right.set_label(\"pan_button\")\n",
    "    pan_button_right.gid = \"right_pan\"\n",
    "    ax.add_patch(pan_button_right)\n",
    "    \n",
    "    pan_button_reset.figure.canvas.mpl_connect ('pick_event', onpick_pan)\n",
    "\n",
    "    \n",
    "    # Zoom slider\n",
    "    #\n",
    "    slider_name = \"Zoom\"\n",
    "    slider_name_h_align = \"left\"\n",
    "    tic_label_v_align=\"top\"\n",
    "    slider_x = 0.64\n",
    "    slider_y = 0.25\n",
    "    track_w = 0.25\n",
    "    slider_h = 0.2\n",
    "    track_h  = 0.05\n",
    "    knob_w = 0.025\n",
    "    knob_h = 0.25\n",
    "    slider_margin = 0.02\n",
    "    tic_margin = 0.05\n",
    "    tic_w = 0.004\n",
    "    num_tics = def_genomebrowser_zoom_tics\n",
    "    step = track_w / (num_tics-1)\n",
    "    tol = 0.1 * step\n",
    "    knob_initial_pos = Global_State['genomebrowser_zoom']\n",
    "    tic_labels = []\n",
    "    for i in range (0,num_tics):\n",
    "        tic_val = int(round(0.001*Global_State['def_genomebrowser_window_bp_width']*(2**i)))\n",
    "        tic_labels.append(\"%d\"%tic_val)\n",
    "    \n",
    "    ax.text(slider_x+track_w+1.5*slider_margin, slider_y+0.5*slider_h, slider_name, verticalalignment=\"center\", horizontalalignment=slider_name_h_align, color=text_color, fontsize=12, zorder=base_zorder+1)\n",
    "    ax.add_patch(Rectangle((slider_x-slider_margin, slider_y), track_w+2*slider_margin, slider_h, facecolor=\"gray\", alpha=0.5, zorder=base_zorder+2))\n",
    "    ax.add_patch(Rectangle((slider_x, slider_y+0.5*slider_h-0.5*track_h), track_w, track_h, facecolor=\"black\", alpha=1.0, zorder=base_zorder+4))\n",
    "    for i in range (0,num_tics):\n",
    "        tic_shift = 0\n",
    "        if i == 0:\n",
    "            tic_shift = 0.5*tic_w\n",
    "        elif i == num_tics-1:\n",
    "            tic_shift = -0.5*tic_w\n",
    "        ax.add_patch(Rectangle((slider_x+i*step-0.5*tic_w+tic_shift, slider_y+tic_margin), tic_w, slider_h-2.0*tic_margin, facecolor=\"lightgray\", edgecolor=\"lightgray\", alpha=0.6, zorder=base_zorder+3))\n",
    "        ax.text(slider_x+i*step+tic_shift, slider_y-1.25*tic_margin, tic_labels[i], verticalalignment=tic_label_v_align, horizontalalignment=\"center\", color=text_color, fontsize=10, zorder=base_zorder+1)\n",
    "\n",
    "    zoom_knob_patch = Rectangle((slider_x+knob_initial_pos*step-0.5*knob_w, slider_y-0.5*(knob_h-slider_h)), knob_w, knob_h, facecolor=\"lightgray\", alpha=1.0, zorder=base_zorder+5)\n",
    "    ax.add_patch (zoom_knob_patch)\n",
    "    zoom_knob = DraggableRectangleSlider(zoom_knob_patch, func='zoom', move='x', min_pos=slider_x-0.5*knob_w, max_pos=slider_x+track_w-0.5*knob_w, step=step, tol=tol)\n",
    "    zoom_knob.connect()\n",
    "\n",
    "    \n",
    "def draw_sidenav_panel (ax, genomebrowser_mode):\n",
    "    \n",
    "    # Tree representation\n",
    "    #\n",
    "    if genomebrowser_mode == \"tree\":\n",
    "        \n",
    "        species_tree = Phylo.read(tree_data_base_path+'/'+tree_data_file, tree_data_format)\n",
    "        #print (species_tree)\n",
    "        #species_tree.ladderize()   # Flip branches so deeper clades are displayed at top\n",
    "        #Phylo.draw(species_tree, axes=ax)  # FIX ax resizing nonsense\n",
    "\n",
    "        \n",
    "    # Line representation\n",
    "    #\n",
    "    elif genomebrowser_mode == \"contigs\" \\\n",
    "        or genomebrowser_mode == \"homologs\":\n",
    "            \n",
    "            contig_label_fontsize = 12\n",
    "            contig_label_x_margin = 0.0075\n",
    "            contig_label_y_margin = 0.06 / (figure_height_scaling*(total_rows+1))\n",
    "            contig_color = \"steelblue\"\n",
    "            contig_disp_margin = 0.10\n",
    "            contig_h           = 0.17 / (figure_height_scaling*(total_rows+1))\n",
    "            clickable_y_margin = 0.75 * contig_h\n",
    "            \n",
    "\n",
    "            longest_contig_len = 0\n",
    "            for i,contig_len in enumerate(Global_State['Contig_lens']):\n",
    "                if i >= max_rows:\n",
    "                    break\n",
    "                if longest_contig_len < contig_len:\n",
    "                    longest_contig_len = contig_len\n",
    "                    \n",
    "            for i,contig_len in enumerate(Global_State['Contig_lens']):\n",
    "                if i >= max_rows:\n",
    "                    break\n",
    "                pivot_pos = Global_State['pivot_pos_list'][i]\n",
    "\n",
    "                # Contig coords\n",
    "                #\n",
    "                x0 = contig_disp_margin\n",
    "                y0 = bottom_margin + (1.0-bottom_margin-top_margin)*row_delta*(total_rows-(i+1)) - 0.5*contig_h\n",
    "                w = (1.0 - 2*contig_disp_margin) * contig_len / longest_contig_len\n",
    "                h = contig_h\n",
    "                \n",
    "                # Add clickable base\n",
    "                #\n",
    "                x0_click = 0.5*contig_disp_margin\n",
    "                y0_click = y0 - 0.75*clickable_y_margin\n",
    "                w_click = (1.0 - 2*contig_disp_margin) * contig_len / longest_contig_len + contig_disp_margin\n",
    "                h_click = contig_h + 3.0*clickable_y_margin\n",
    "                clickable_contig_rect = Rectangle((x0_click,y0_click), w_click, h_click, \\\n",
    "                                          facecolor=\"none\", edgecolor=\"none\", \\\n",
    "                                          alpha=0.0, zorder=0, \\\n",
    "                                          picker=True)         \n",
    "                clickable_contig_rect.set_label(\"linear_contig_nav\")\n",
    "                clickable_contig_rect.gid = \"%d,%f,%f\"%(i,x0,w)\n",
    "                ax.add_patch(clickable_contig_rect)\n",
    "                clickable_contig_rect.figure.canvas.mpl_connect ('pick_event', onpick_contig_nav)\n",
    "        \n",
    "                # Show contig\n",
    "                #\n",
    "                contig_rect = Rectangle((x0,y0), w, h,\n",
    "                                          facecolor=contig_color, edgecolor=contig_color, \\\n",
    "                                          alpha=1.0, zorder=1)                 \n",
    "                # drop shadow\n",
    "                contig_rect.set_path_effects([path_effects.PathPatchEffect(offset=(2, -2),\n",
    "                                                                           facecolor='magenta', edgecolor='none', alpha=0.5),\n",
    "                                              path_effects.PathPatchEffect(facecolor=contig_color, edgecolor=contig_color)]) \n",
    "                ax.add_patch (contig_rect)\n",
    "\n",
    "                # Add label\n",
    "                #\n",
    "                ax.text(x0+contig_label_x_margin, y0-contig_label_y_margin,\n",
    "                        ContigSet_names[i], verticalalignment=\"top\", horizontalalignment=\"left\",\n",
    "                        color=\"black\", fontsize=contig_label_fontsize, zorder=1)\n",
    "\n",
    "                # Mark pivot\n",
    "                #\n",
    "                if Global_State['genomebrowser_mode'] == 'homologs':\n",
    "                    mark_h = 1.25 * contig_h\n",
    "                    mark_w = 0.01\n",
    "                    mark_margin = 0.02 * mark_h\n",
    "                    disp_mark_x0 = x0 + w * (1-(contig_len-pivot_pos)/contig_len) - 0.5*mark_w\n",
    "                    disp_mark_y0 = y0 + contig_h + mark_margin\n",
    "                    pivot_feature_rect = Rectangle((disp_mark_x0,disp_mark_y0), mark_w, mark_h,\n",
    "                                                    facecolor=\"black\", edgecolor=\"none\", alpha=1.0, zorder=2) \n",
    "                    ax.add_patch (pivot_feature_rect)\n",
    "\n",
    "                # Mark search results\n",
    "                #\n",
    "                try:\n",
    "                    contig_search_results = search_results[i]\n",
    "                except:\n",
    "                    contig_search_results = []\n",
    "                for j,result_list in enumerate(contig_search_results):\n",
    "                    mark_color = search_color_names[j % len(search_color_names)]\n",
    "                    for result in contig_search_results[j]:\n",
    "                        mid_pos = 0.5 * (result['beg_pos'] + result['end_pos'])\n",
    "                        mark_h = 1.25 * contig_h\n",
    "                        mark_w = 0.01\n",
    "                        mark_margin = 0.02 * mark_h\n",
    "                        disp_mark_x0 = x0 + w * (1-(contig_len-mid_pos)/contig_len) - 0.5*mark_w\n",
    "                        disp_mark_y0 = y0 + contig_h + mark_margin\n",
    "                        search_feature_rect = Rectangle((disp_mark_x0,disp_mark_y0), mark_w, mark_h,\n",
    "                                                            facecolor=mark_color, edgecolor=\"none\", alpha=1.0, zorder=3) \n",
    "                        ax.add_patch (search_feature_rect)\n",
    "                    \n",
    "                # Indicate window viewed by track panel\n",
    "                #\n",
    "                contig_mode_xshift = 0.0\n",
    "                if Global_State['genomebrowser_mode'] == 'contigs':\n",
    "                    #contig_mode_xshift = 0.5*Global_State['genomebrowser_window_bp_width'] - 0.5*(Feature_slices[i][0]['end_pos']-Feature_slices[i][0]['beg_pos'])\n",
    "                    contig_mode_xshift = 0.5*Global_State['genomebrowser_window_bp_width']\n",
    "                    \n",
    "                #if (pivot_pos+Global_State['genomebrowser_xshift']) >= -.0000001 \\\n",
    "                #    and (pivot_pos+Global_State['genomebrowser_xshift']) <= (contig_len + .0000001):\n",
    "                track_window_h = 1.6 * contig_h\n",
    "                track_window_w = w * Global_State['genomebrowser_window_bp_width'] / contig_len\n",
    "                track_window_x0 = x0 - 0.5*track_window_w + w * ((pivot_pos+Global_State['genomebrowser_xshift']+contig_mode_xshift) / contig_len)\n",
    "                track_window_y0 = y0 + 0.5*h - 0.5*track_window_h\n",
    "                if track_window_x0 > x0+w or track_window_x0 + track_window_w < x0:\n",
    "                    continue                            \n",
    "                track_window_rect = Rectangle((track_window_x0,track_window_y0), track_window_w, track_window_h, \\\n",
    "                                                    facecolor=\"red\", edgecolor=\"red\", alpha=0.4, zorder=4) \n",
    "                ax.add_patch (track_window_rect)\n",
    "\n",
    "                \n",
    "    # Circle representation\n",
    "    #\n",
    "    elif genomebrowser_mode == \"genome\":\n",
    "\n",
    "        # set up circle genome representation\n",
    "        contig_label_fontsize = 12\n",
    "        contig_label_y_margin = 0.5 / (figure_height_scaling*(total_rows+1))\n",
    "        contig_color = \"steelblue\"\n",
    "        diameter = 0.75\n",
    "        clickable_diameter = 1.25 * diameter\n",
    "        genome_lw = 10\n",
    "        marker_lw = 15\n",
    "        track_window_lw = 20\n",
    "        marker_len = 0.1\n",
    "        GC_window_bp = 5000\n",
    "        GC_bar_high_lw = 20\n",
    "        GC_bar_high_perc = 0.20\n",
    "        GC_base_diameter = 0.55\n",
    "        GC_lw_to_coord_scale = 0.01\n",
    "        ellipse_center_x = 0.50\n",
    "        #ellipse_center_y = 0.70\n",
    "        ellipse_center_y = (total_rows+1 - 2.5) / (total_rows+1)\n",
    "        ellipse_center = (ellipse_center_x, ellipse_center_y)\n",
    "        ellipse_to_circle_scaling = 0.25*figure_width / (figure_height_scaling*(total_rows+1))        \n",
    "\n",
    "        \n",
    "        # Add clickable base\n",
    "        #\n",
    "        x_diameter = 1.0 * clickable_diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * clickable_diameter\n",
    "        #Ellipse(xy, width, height, angle=0.0, **kwargs)\n",
    "        clickable_contig_image = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                          facecolor=\"none\", edgecolor=\"none\", alpha=0.0, zorder=0,\n",
    "                                          picker=True)\n",
    "        clickable_contig_image.set_label(\"circular_contig_nav\")\n",
    "        clickable_contig_image.gid = \"%d,%f,%f,%f\"%(0,ellipse_center_x,ellipse_center_y,ellipse_to_circle_scaling)\n",
    "        ax.add_patch(clickable_contig_image)\n",
    "        clickable_contig_image.figure.canvas.mpl_connect ('pick_event', onpick_contig_nav)\n",
    "        \n",
    "        # Draw circle genome representation\n",
    "        #\n",
    "        x_diameter = 1.0 * diameter\n",
    "        y_diameter = ellipse_to_circle_scaling * diameter\n",
    "        #Ellipse(xy, width, height, angle=0.0, **kwargs)\n",
    "        contig_image = Ellipse (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                facecolor=\"white\", edgecolor=contig_color, lw=genome_lw, alpha=1.0, zorder=1,\n",
    "                               picker=False)\n",
    "        ax.add_patch(contig_image)\n",
    "\n",
    "        # Add label\n",
    "        #\n",
    "        ax.text(ellipse_center_x, ellipse_center_y - 0.5*y_diameter - contig_label_y_margin,\n",
    "                ContigSet_names[0], verticalalignment=\"top\", horizontalalignment=\"center\",\n",
    "                color=\"black\", fontsize=contig_label_fontsize)\n",
    "\n",
    "        # Mark contig edge\n",
    "        #\n",
    "        mark_width = 1.0\n",
    "        arc_end = 90 + 0.5*mark_width\n",
    "        arc_beg = 90 - 0.5*mark_width\n",
    "        contig_edge_arc = Arc (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=\"white\", lw=genome_lw, alpha=1.0, zorder=2)\n",
    "        ax.add_patch (contig_edge_arc)        \n",
    "\n",
    "        # Mark pivot\n",
    "        #\n",
    "        mark_width = 1.0\n",
    "        mark_pad = 0.125\n",
    "        mark_x_diameter = x_diameter + 1.0 * mark_pad\n",
    "        mark_y_diameter = y_diameter + ellipse_to_circle_scaling * mark_pad\n",
    "        # pivot feature\n",
    "        arc_end = 90 + 0.5*mark_width - 360 * (Global_State['PrimaryAnchor_pivot_pos'] / Global_State['PrimaryContig_len'])\n",
    "        arc_beg = 90 - 0.5*mark_width - 360 * (Global_State['PrimaryAnchor_pivot_pos'] / Global_State['PrimaryContig_len'])\n",
    "        pivot_feature_arc = Arc (ellipse_center, mark_x_diameter, mark_y_diameter, \\\n",
    "                                    theta1=arc_beg, theta2=arc_end, \\\n",
    "                                    edgecolor=\"black\", lw=marker_lw, alpha=1.0, zorder=3)\n",
    "        ax.add_patch (pivot_feature_arc)\n",
    "        \n",
    "        # Mark search features\n",
    "        #\n",
    "        try:\n",
    "            contig_search_results = search_results[0]\n",
    "        except:\n",
    "            contig_search_results = []\n",
    "        for j,result_list in enumerate(contig_search_results):\n",
    "            mark_color = search_color_names[j % len(search_color_names)]\n",
    "            for result in contig_search_results[j]:\n",
    "                mid_pos = 0.5 * (result['beg_pos'] + result['end_pos'])\n",
    "                arc_end = 90 + 0.5*mark_width - 360 * (mid_pos / Global_State['PrimaryContig_len'])\n",
    "                arc_beg = 90 - 0.5*mark_width - 360 * (mid_pos / Global_State['PrimaryContig_len'])\n",
    "                search_result_arc = Arc (ellipse_center, mark_x_diameter, mark_y_diameter, \\\n",
    "                                    theta1=arc_beg, theta2=arc_end, \\\n",
    "                                    edgecolor=mark_color, lw=marker_lw, alpha=1.0, zorder=4)\n",
    "                ax.add_patch (search_result_arc)\n",
    "        \n",
    "        # Indicate window viewed by track panel\n",
    "        #\n",
    "        track_extra = 0.5*(total_rows-1)*Global_State['genomebrowser_window_bp_width']\n",
    "        arc_end = 90 - 360 * ((Global_State['PrimaryAnchor_pivot_pos']+Global_State['genomebrowser_xshift']-track_extra-0.5*Global_State['genomebrowser_window_bp_width']) / Global_State['PrimaryContig_len'])\n",
    "        arc_beg = 90 - 360 * ((Global_State['PrimaryAnchor_pivot_pos']+Global_State['genomebrowser_xshift']+track_extra+0.5*Global_State['genomebrowser_window_bp_width']) / Global_State['PrimaryContig_len']) \n",
    "        if arc_end > 90:\n",
    "            arc_end = 90 - 0.5\n",
    "        if arc_beg < -270:\n",
    "            arc_beg = -270 + 0.5\n",
    "                \n",
    "        track_window_arc = Arc (ellipse_center, x_diameter, y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=\"red\", lw=track_window_lw, alpha=0.40, zorder=5)  # facecolor does nothing (no fill for Arc)\n",
    "        ax.add_patch (track_window_arc)\n",
    "        \n",
    "        # Add GC\n",
    "        #       \n",
    "        GC_bar_scale = GC_bar_high_lw / GC_bar_high_perc\n",
    "        GC_steps = int(round(Global_State['PrimaryContig_len'] / GC_window_bp))\n",
    "        for GC_i in range(1,GC_steps-2):\n",
    "            arc_end = 90 - 360 * (GC_i*GC_window_bp / Global_State['PrimaryContig_len'])\n",
    "            arc_beg = 90 - 360 * ((GC_i+1)*GC_window_bp / Global_State['PrimaryContig_len'])\n",
    "            GC_in_window = compute_GC (Global_Genbank_Genomes[0].seq[(GC_i-1)*GC_window_bp:(GC_i+2)*GC_window_bp])\n",
    "            GC_delta = GC_in_window - Global_State['PrimaryContig_GCavg']\n",
    "            if (GC_delta > 0):\n",
    "                GC_color = \"blue\"\n",
    "                GC_bar_lw = int(round(GC_bar_scale * GC_delta))\n",
    "                GC_bar_diameter = GC_base_diameter + 0.5*GC_bar_lw*GC_lw_to_coord_scale\n",
    "                GC_x_diameter = 1.0 * GC_bar_diameter\n",
    "                GC_y_diameter = ellipse_to_circle_scaling * GC_bar_diameter\n",
    "                GC_arc = Arc (ellipse_center, GC_x_diameter, GC_y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=GC_color, lw=GC_bar_lw, alpha=1.0, zorder=1)  # facecolor does nothing (no fill for Arc)\n",
    "                ax.add_patch (GC_arc)\n",
    "            else:\n",
    "                GC_color = \"red\"\n",
    "                GC_bar_lw = int(round(GC_bar_scale * -GC_delta))\n",
    "                GC_bar_diameter = GC_base_diameter - 0.5*GC_bar_lw*GC_lw_to_coord_scale\n",
    "                GC_x_diameter = 1.0 * GC_bar_diameter\n",
    "                GC_y_diameter = ellipse_to_circle_scaling * GC_bar_diameter\n",
    "                GC_arc = Arc (ellipse_center, GC_x_diameter, GC_y_diameter, \\\n",
    "                                theta1=arc_beg, theta2=arc_end, \\\n",
    "                                edgecolor=GC_color, lw=GC_bar_lw, alpha=1.0, zorder=1)  # facecolor does nothing (no fill for Arc)\n",
    "                ax.add_patch (GC_arc)\n",
    "\n",
    "                \n",
    "def draw_genomebrowser_panel (ax, \\\n",
    "                               ContigSet_names, \\\n",
    "                               OrthologSet_locusTags, \\\n",
    "                               genomebrowser_window_bp_width, \\\n",
    "                               genomebrowser_xshift, \\\n",
    "                               genomebrowser_mode):                  \n",
    "    \n",
    "    # Read Domains\n",
    "    #\n",
    "    if Global_State['genomebrowser_color_namespace'] == 'domains' \\\n",
    "        or Global_State['genomebrowser_color_namespace'] == 'cog' \\\n",
    "        or Global_State['genomebrowser_color_namespace'] == 'pfam':\n",
    "\n",
    "        # update Global_Domains, but really should be method of Domains object\n",
    "        getDomainHits (ContigSet_names, \\\n",
    "                       genomebrowser_mode=genomebrowser_mode, \\\n",
    "                       domain_data_format=domain_data_format)\n",
    "        \n",
    "    \n",
    "    # Read Features\n",
    "    #\n",
    "    Feature_slices = getFeatureSlices (ContigSet_names, \\\n",
    "                                       OrthologSet_locusTags, \\\n",
    "                                       genomebrowser_mode=genomebrowser_mode, \\\n",
    "                                       genome_data_format=genome_data_format, \\\n",
    "                                       window_size=genomebrowser_window_bp_width, \\\n",
    "                                       genomebrowser_xshift=genomebrowser_xshift)\n",
    "            \n",
    "\n",
    "    # Add features elements to figure\n",
    "    #\n",
    "    if genomebrowser_mode == \"contigs\" \\\n",
    "        or genomebrowser_mode == \"genome\" \\\n",
    "        or genomebrowser_mode == \"homologs\" \\\n",
    "        or genomebrowser_mode == \"tree\":\n",
    "\n",
    "        # Determine repeats\n",
    "        #\n",
    "        annot_repeat = set()\n",
    "        annot_seen   = set()\n",
    "        seq_repeat   = set()\n",
    "        seq_seen     = set()\n",
    "        for i in range(0,len(Feature_slices)):\n",
    "            if i >= max_rows:\n",
    "                break\n",
    "                \n",
    "            for j in range(1,len(Feature_slices[i])):  # skip [0]. it's the pivot and if in window will be repeated\n",
    "                if Feature_slices[i][j]['type'] == \"CDS\":\n",
    "                    if Global_State['genomebrowser_color_namespace'] == 'annot' \\\n",
    "                        and 'annot' in Feature_slices[i][j] \\\n",
    "                        and Feature_slices[i][j]['annot'] != '':\n",
    "                        if Feature_slices[i][j]['annot'] in annot_seen:\n",
    "                            annot_repeat.add(Feature_slices[i][j]['annot'])\n",
    "                        else:\n",
    "                            annot_seen.add(Feature_slices[i][j]['annot'])\n",
    "                            \n",
    "                    elif Global_State['genomebrowser_color_namespace'] == 'ec' \\\n",
    "                        and 'EC_number' in Feature_slices[i][j] \\\n",
    "                        and Feature_slices[i][j]['EC_number'] != '':\n",
    "                        if Feature_slices[i][j]['EC_number'] in annot_seen:\n",
    "                            annot_repeat.add(Feature_slices[i][j]['EC_number'])\n",
    "                        else:\n",
    "                            annot_seen.add(Feature_slices[i][j]['EC_number'])\n",
    "                            \n",
    "                    elif Global_State['genomebrowser_color_namespace'] == 'cog' \\\n",
    "                        or Global_State['genomebrowser_color_namespace'] == 'pfam' \\\n",
    "                        or Global_State['genomebrowser_color_namespace'] == 'domains':\n",
    "                            \n",
    "                        if Global_State['genomebrowser_mode'] == \"genome\":\n",
    "                            contig_i = 0\n",
    "                        else:\n",
    "                            contig_i = i\n",
    "                        try:\n",
    "                            domain_hits = Global_Domains[contig_i][Feature_slices[i][j]['ID']]\n",
    "                        except:\n",
    "                            domain_hits = []\n",
    "                        for domhit in domain_hits:\n",
    "                            domfam = domhit[DOMHIT_DOMFAM_I]\n",
    "                            if domfam in annot_seen:\n",
    "                                annot_repeat.add(domfam)\n",
    "                            else:\n",
    "                                annot_seen.add(domfam)\n",
    "                            \n",
    "                elif (Feature_slices[i][j]['type'] == \"CRISPR\" or Feature_slices[i][j]['type'] == \"CRISPR spacer\") \\\n",
    "                    and 'fwd_dna_seq' in Feature_slices[i][j]:\n",
    "                    if Feature_slices[i][j]['fwd_dna_seq'] in seq_seen:\n",
    "                        seq_repeat.add(Feature_slices[i][j]['fwd_dna_seq'])\n",
    "                    else:\n",
    "                        seq_seen.add(Feature_slices[i][j]['fwd_dna_seq'])\n",
    "                        \n",
    "\n",
    "        # Draw feature elements\n",
    "        #\n",
    "        for i in range(0,len(Feature_slices)):\n",
    "            if i >= max_rows:\n",
    "                break\n",
    "                \n",
    "            pivot_pos = 0.5 * (Feature_slices[i][0]['beg_pos']+Feature_slices[i][0]['end_pos'])\n",
    "            track_xshift = 0\n",
    "            if genomebrowser_mode == \"genome\":\n",
    "                #track_xshift = genomebrowser_window_bp_width * (i - (total_rows-1)/2)  # e.g. total_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                track_xshift = genomebrowser_window_bp_width * (i - (Global_State['genome_mode_n_rows']-1)/2)  # e.g. n_rows=7 -> -3,-2,-1,0,1,2,3\n",
    "                #print (\"%d %f\"%(i,track_xshift))\n",
    "            contig_mode_xshift = 0\n",
    "            if genomebrowser_mode == \"contigs\":\n",
    "                contig_mode_xshift = 0.5*genomebrowser_window_bp_width - 0.5*(Feature_slices[i][0]['end_pos']-Feature_slices[i][0]['beg_pos'])\n",
    "            prev_top_beg_pos = 0\n",
    "            prev_bot_beg_pos = 0\n",
    "            prev_top_name = ''\n",
    "            prev_bot_name = ''\n",
    "            \n",
    "            for j in range(1,len(Feature_slices[i])):  # skip [0]. it's the pivot and if in window will be repeated\n",
    "                pivot_feature_flag = False\n",
    "                if Feature_slices[i][j]['beg_pos'] == Feature_slices[i][0]['beg_pos'] and \\\n",
    "                   Feature_slices[i][j]['end_pos'] == Feature_slices[i][0]['end_pos']:\n",
    "                    pivot_feature_flag = True\n",
    "                \n",
    "                # show label on top or bottom, if it fits\n",
    "                this_label_show_top = True\n",
    "                this_label_show_bot = False\n",
    "                this_beg_pos = disp_coord_transform (Feature_slices[i][j]['beg_pos'], \\\n",
    "                                                     pivot_pos, \\\n",
    "                                                     genomebrowser_window_bp_width, \\\n",
    "                                                     genomebrowser_xshift, \\\n",
    "                                                     track_xshift, \\\n",
    "                                                     contig_mode_xshift)\n",
    "                #if j == 1:\n",
    "                if this_beg_pos < left_margin:\n",
    "                    this_beg_pos = left_margin\n",
    "                elif len(prev_top_name) > arrow_label_scaling*(this_beg_pos-prev_top_beg_pos):\n",
    "                    this_label_show_top = False\n",
    "                if this_label_show_top == False and len(prev_bot_name) <= arrow_label_scaling*(this_beg_pos-prev_bot_beg_pos):\n",
    "                    this_label_show_bot = True\n",
    "\n",
    "                # FIX\n",
    "                max_row_n = total_rows  \n",
    "                if num_genomes < total_rows:\n",
    "                    max_row_n = num_genomes\n",
    "                    \n",
    "                # finally draw arrow!  (or other shape)\n",
    "                draw_feature_element (ax, \\\n",
    "                                 i+1, \\\n",
    "                                 max_row_n, \\\n",
    "                                 i, \\\n",
    "                                 j, \\\n",
    "                                 Feature_slices[i][j], \\\n",
    "                                 pivot_pos, \\\n",
    "                                 track_xshift, \\\n",
    "                                 contig_mode_xshift, \\\n",
    "                                 pivot_feature_flag, \\\n",
    "                                 annot_repeat, \\\n",
    "                                 seq_repeat, \\\n",
    "                                 this_label_show_top, this_label_show_bot)\n",
    "\n",
    "                if this_label_show_top:\n",
    "                    prev_top_beg_pos = this_beg_pos\n",
    "                    prev_top_name = Feature_slices[i][j]['name']\n",
    "                if this_label_show_bot:\n",
    "                    prev_bot_beg_pos = this_beg_pos\n",
    "                    prev_bot_name = Feature_slices[i][j]['name']\n",
    "\n",
    "                    \n",
    "        \n",
    "# Panel updates\n",
    "#\n",
    "def update_mode_panel (ax):\n",
    "    ax.cla()\n",
    "    draw_mode_panel (ax, \\\n",
    "                     Global_State['ContigSet_names'][0], \\\n",
    "                     Global_State['genomebrowser_mode'], \\\n",
    "                     Global_State['Dataset_names_list'][0])    \n",
    "    pyplot.show()\n",
    "\n",
    "def update_control_panel (ax):\n",
    "    ax.cla()\n",
    "    draw_control_panel (ax, \\\n",
    "                        Global_State['genomebrowser_zoom'], \\\n",
    "                        Global_State['genomebrowser_xshift'], \\\n",
    "                        Global_State['genomebrowser_color_namespace'])    \n",
    "    pyplot.show()\n",
    "\n",
    "def update_sidenav_panel (ax):\n",
    "    ax.cla()\n",
    "    draw_sidenav_panel (ax, \\\n",
    "                        Global_State['genomebrowser_mode'])\n",
    "    pyplot.show()\n",
    "    \n",
    "def update_genomebrowser_panel (ax):\n",
    "\n",
    "    # clear panel\n",
    "    ax.cla()\n",
    "    \n",
    "    # Update contig order\n",
    "    #   might be used elsewhere later, likely when track ordering changes\n",
    "    #\n",
    "    Contig_order = []\n",
    "    Contig_order_lookup = {}\n",
    "    for i,genome_name in enumerate(ContigSet_names):\n",
    "        (genome_id,scaffold_id) = genome_name.split(\".\")        \n",
    "        try:\n",
    "            col = Contig_order_lookup['genome_id']\n",
    "        except:\n",
    "            Contig_order_lookup['genome_id'] = {}\n",
    "        Contig_order_lookup['genome_id']['scaffold_id'] = i\n",
    "        Contig_order.append(genome_name)\n",
    "        \n",
    "    # Reset feature and popup_box storage\n",
    "    #\n",
    "    feature_artist_label_to_feature = {}\n",
    "    feature_artist_label_to_popup_box = {}\n",
    "    Global_State['popup_zorder'] = def_popup_zorder\n",
    "    \n",
    "    # Add Genome tracks\n",
    "    #\n",
    "    draw_genomebrowser_panel (ax, \\\n",
    "                              ContigSet_names = Global_State['ContigSet_names'], \\\n",
    "                              OrthologSet_locusTags = Global_State['OrthologSet_locusTags'], \\\n",
    "                              genomebrowser_window_bp_width = Global_State['genomebrowser_window_bp_width'], \\\n",
    "                              genomebrowser_xshift = Global_State['genomebrowser_xshift'], \\\n",
    "                              genomebrowser_mode = Global_State['genomebrowser_mode'])\n",
    "    pyplot.show()\n",
    "\n",
    "\n",
    "###############################################################################################################                \n",
    "# Main\n",
    "###############################################################################################################\n",
    "\n",
    "\n",
    "# Instatiate fig_FGB\n",
    "#   \n",
    "#fig_FGB.suptitle(\"Flexible Genome Browser\", fontsize=22)\n",
    "fig_FGB = pyplot.figure(1, (figure_width, top_nav_height + figure_height_scaling*(total_rows+1)))\n",
    "#fig_FGB.suptitle(\"Flexible Genome Browser\", fontsize=22)\n",
    "\n",
    "# want to maintain constant top nav height, so adjust proportion according to number of rows\n",
    "# e.g. with 6 tracks, top nav should be 2 units, so top rowspan=2 when bottom rowspan = 6 and grid is 8\n",
    "top_nav_rows = int(round(1000*top_nav_height/(top_nav_height+figure_height_scaling*(total_rows+1))))\n",
    "\n",
    "# subplot2grid(shape, loc, rowspan=1, colspan=1)\n",
    "ax_top_left   = pyplot.subplot2grid((1000,4),            (0,0), rowspan=top_nav_rows,        colspan=1)\n",
    "ax_top_center = pyplot.subplot2grid((1000,4),            (0,1), rowspan=top_nav_rows,        colspan=3)\n",
    "ax_left       = pyplot.subplot2grid((1000,4), (top_nav_rows,0), rowspan=(1000-top_nav_rows), colspan=1)\n",
    "ax_center     = pyplot.subplot2grid((1000,4), (top_nav_rows,1), rowspan=(1000-top_nav_rows), colspan=3)\n",
    "\n",
    "\n",
    "# Let's turn off visibility of all tick labels and boxes here\n",
    "for ax in fig_FGB.axes:\n",
    "    ax.xaxis.set_visible(False)  # remove axis labels and ticks\n",
    "    ax.yaxis.set_visible(False)\n",
    "    for t in ax.get_xticklabels()+ax.get_yticklabels():  # remove tick labels\n",
    "        t.set_visible(False)\n",
    "#for ax in [ax_top_left, ax_top_center]:\n",
    "#    ax.set_axis_bgcolor(nav_bg_color)\n",
    "for ax in [ax_top_left, ax_top_center, ax_left, ax_center]:\n",
    "    ax.spines['top'].set_visible(False) # Get rid of top axis line\n",
    "    ax.spines['bottom'].set_visible(False) #  bottom axis line\n",
    "    ax.spines['left'].set_visible(False) #  Get rid of bottom axis line\n",
    "    ax.spines['right'].set_visible(False) #  Get rid of bottom axis line\n",
    "\n",
    "fig_FGB.tight_layout()  # left justify and space subplots reasonably.  Must follow axis creation\n",
    "\n",
    "\n",
    "# Load domain family descriptions\n",
    "#\n",
    "#domain_family_desc = read_domain_family_desc ()  # these need to be global (kludge)\n",
    "read_domain_family_desc ()  # these need to be global (kludge)\n",
    "\n",
    "\n",
    "# Draw genome browser panel (must happen first to load Global_State)\n",
    "#\n",
    "feature_artist_label_to_feature = {}     # these need to be global (kludge)\n",
    "feature_artist_label_to_popup_box = {}   # these need to be global (kludge)\n",
    "pos_knob_patch = None                    # these need to be global (kludge)\n",
    "update_genomebrowser_panel (ax_center)\n",
    "    \n",
    "    \n",
    "# Draw sidenav (tree, circles, etc.)\n",
    "#\n",
    "update_sidenav_panel (ax_left)\n",
    "    \n",
    "\n",
    "# Draw mode panel\n",
    "#\n",
    "update_mode_panel (ax_top_left)\n",
    "\n",
    "\n",
    "# Draw control panel (currently, must occur after primary genome read in update_genomebrowser_panel)\n",
    "#\n",
    "update_control_panel (ax_top_center)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
